<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2022.01.06 documentation</title>

<link href="../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
<link rel="stylesheet" href="../../_static/css/docs.css" type="text/css" />
<link rel="stylesheet" href="../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../_static/sg_gallery.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />

<link rel="stylesheet" href="../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../.././_static/pygments.css">
<link rel="stylesheet" href="../.././_static/css/common.css">
<link rel="stylesheet" href="../.././_static/css/link.css">
<link rel="stylesheet" href="../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
<script src="../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../.././_static/js/initializeHighlight.js"></script>
<script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
<script src="../../_static/jquery.js"></script>
<script src="../../_static/underscore.js"></script>
<script src="../../_static/doctools.js"></script>
<script src="../../_static/thebelab-helper.js"></script>
<script src="../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
<script src="../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3/discussions" class="nav-link" title="Forum">Forum</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-using-conda">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../install.html#manual-creation-of-the-conda-environment">Manual creation of the <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#overview">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/warning.html">Warning – apps are alpha!</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#define-the-apps">Define the apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#using-apps-like-functions">Using apps like functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-log-the-settings-and-data-analysed">You can log the settings and data analysed</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/dstore.html">Specifying data for analysis</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#a-read-only-data-store">A read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#data-store-members">Data store “members”</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#showing-the-last-few-members">Showing the last few members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#filtering-a-data-store-for-specific-members">Filtering a data store for specific members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#tinydb-data-stores-are-special">TinyDB data stores are special</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/available-apps.html">What apps are there?</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-progressive-align-app">The <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-model-app">The <code class="docutils literal notranslate"><span class="pre">model</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-hypothesis-app">The <code class="docutils literal notranslate"><span class="pre">hypothesis</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-natsel-apps">The <code class="docutils literal notranslate"><span class="pre">natsel</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-user-function-apps">The <code class="docutils literal notranslate"><span class="pre">user_function</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/user_function.html">Turn your functions into composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-user-app-with-a-different-output-type">A user app with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../cookbook/index.html">Cookbook</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/manipulating_biological_data.html">Manipulating biological data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/loading_sequences.html">Loading an alignment from a file</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/loading_sequences.html#making-a-sequence-collection-from-standard-python-objects">Making a sequence collection from standard python objects</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/alignments.html">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alignments.html#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/DNA_and_RNA_sequences.html"><code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/annotations.html">Annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/annotations.html#annotations-with-coordinates">Annotations with coordinates</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#automated-introduction-from-reading-genbank-files">Automated introduction from reading genbank files</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#customising-annotation-construction-from-reading-a-genbank-file">Customising annotation construction from reading a genbank file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#creating-directly-on-a-sequence">Creating directly on a sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#via">Via</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#adding-as-a-series-or-item-wise">Adding as a series or item-wise</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#taking-the-union-of-annotations">Taking the union of annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#getting-annotation-coordinates">Getting annotation coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#annotations-have-shadows">Annotations have shadows</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#adding-to-a-sequence-member-of-an-alignment">Adding to a sequence member of an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#adding-to-an-alignment">Adding to an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#slicing-sequences-and-alignments-by-annotations">Slicing sequences and alignments by annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#slicing-by-pseudo-feature-or-feature-series">Slicing by pseudo-feature or feature series</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#slice-series-must-not-be-overlapping">Slice series must not be overlapping</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#you-can-slice-an-annotation-itself">You can slice an annotation itself</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#sequence-vs-alignment-slicing">Sequence vs Alignment slicing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#copying-annotations">Copying annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#querying">Querying</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#querying-produces-objects-only-valid-for-their-source">Querying produces objects only valid for their source</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#querying-for-absent-annotation">Querying for absent annotation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#querying-features-that-span-gaps-in-alignments">Querying features that span gaps in alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#as-one-span-unifies-features-with-discontinuous-alignment-coordinates"><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#behaviour-of-annotations-on-nucleic-acid-sequences">Behaviour of annotations on nucleic acid sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#masking-annotated-regions">Masking annotated regions</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#masking-annotated-regions-on-alignments">Masking annotated regions on alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#you-can-take-mask-of-the-shadow">You can take mask of the shadow</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#what-features-of-a-certain-type-are-available">What features of a certain type are available?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#getting-all-features-of-a-type-or-everything-but-that-type">Getting all features of a type, or everything but that type</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#getting-sequence-features-when-you-have-an-alignment-object">Getting sequence features when you have an alignment object</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/annotations.html#annotation-display-on-sequences">Annotation display on sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html">Translate DNA sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/genetic_code.html#translating-a-dna-sequence-with-a-terminating-stop-codon">Translating a DNA sequence with a terminating stop codon</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/genetic_code.html#by-removing-the-trailing-stop-codon-first">By removing the trailing stop codon first</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/genetic_code.html#by-slicing-the-dnasequence-first">By slicing the <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> first</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#loading-delimited-data-without-a-header-line">Loading delimited data without a header line</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/tables.html#using-load-delimited">Using <code class="docutils literal notranslate"><span class="pre">load_delimited()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/tables.html#using-filteringparser">Using <code class="docutils literal notranslate"><span class="pre">FilteringParser</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/tables.html#loading-only-some-rows">Loading only some rows</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../cookbook/evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../cookbook/useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../cookbook/useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="../cookbook/useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="seq_features.html">Advanced sequence handling</a></li>
<li class="toctree-l4"><a class="reference internal" href="complete_seq_features.html">Annotating alignments and sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3 current"><a class="reference internal" href="index.html#modelling-evolution">Modelling Evolution</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="index.html#utilities">Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="parallel.html">Parallel computations</a><ul>
<li class="toctree-l5"><a class="reference internal" href="parallel.html#parallel-computation-on-a-single-computer">Parallel computation on a single computer</a><ul>
<li class="toctree-l6"><a class="reference internal" href="parallel.html#using-app-apply-to">Using <code class="docutils literal notranslate"><span class="pre">app.apply_to()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="parallel.html#directly-using-cogent3-util-parallel-map">Directly using <code class="docutils literal notranslate"><span class="pre">cogent3.util.parallel.map()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="parallel.html#parallel-computation-on-multiple-computers">Parallel computation on multiple computers</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_model.html">get_model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html">Sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html">Dotplot basics</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html">Coevolution analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-support.html">Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-square.html">Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-radial.html">Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-circular.html">Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-angular.html">Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../genindex.html">Index</a></li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="allowing-substitution-model-parameters-to-differ-between-branches">
<span id="scope-params-on-trees"></span><h1>Allowing substitution model parameters to differ between branches<a class="headerlink" href="#allowing-substitution-model-parameters-to-differ-between-branches" title="Permalink to this headline">¶</a></h1>
<p><em>Section author: Gavin Huttley</em></p>
<p>A common task concerns assessing how substitution model exchangeability parameters differ between evolutionary lineages. This is most commonly of interest for the case of testing for natural selection. Here I’ll demonstrate the different ways of scoping parameters across trees for the codon model case and how these can be used for evolutionary modelling.</p>
<p>We start with the standard imports, plus using a canned codon substitution model and then load the sample data set.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">load_tree</span>
<span class="kn">from</span> <span class="nn">cogent3.evolve.models</span> <span class="kn">import</span> <span class="n">MG94HKY</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">load_tree</span><span class="p">(</span><span class="s2">&quot;data/test.tree&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>We construct the substitution model and likelihood function and set the alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sm</span> <span class="o">=</span> <span class="n">MG94HKY</span><span class="p">()</span>
<span class="n">lf</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">make_likelihood_function</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">digits</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">space</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">lf</span><span class="o">.</span><span class="n">set_alignment</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>At this point we have a likelihood function with two exchangeability parameters from the substitution model (<code class="docutils literal notranslate"><span class="pre">kappa</span></code> the transition/transversion ratio; <code class="docutils literal notranslate"><span class="pre">omega</span></code> the nonsynonymous/synonymous ratio) plus branch lengths for all tree edges. To facilitate subsequent discussion I now display the tree</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">ascii_art</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>                              /-Human
                    /edge.0--|
          /edge.1--|          \-HowlerMon
         |         |
         |          \-Mouse
-root----|
         |--NineBande
         |
          \-DogFaced
</pre></div>
</div>
</div>
</div>
<p>In order to scope a parameter on a tree (meaning specifying a subset of edges for which the parameter is to be treated differently to the remainder of the tree) requires uniquely identifying the edges. We do this using the following arguments to the likelihood function <code class="docutils literal notranslate"><span class="pre">set_param_rule</span></code> method:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">tip_names</span></code>: the name of two tips</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">outgroup_name</span></code>: the name of a tip that is not part of the clade of interest</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clade</span></code>: if <code class="docutils literal notranslate"><span class="pre">True</span></code>, all lineages descended from the tree node identified by the <code class="docutils literal notranslate"><span class="pre">tip_names</span></code> and <code class="docutils literal notranslate"><span class="pre">outgroup_name</span></code> argument are affected by the other arguments. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, then the <code class="docutils literal notranslate"><span class="pre">stem</span></code> argument must apply.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">stem</span></code>: Whether the edge leading to the node is included.</p></li>
</ul>
<p>The next concepts include exactly what can be scoped and how. In the case of testing for distinctive periods of natural selection it is common to specify distinct values for <code class="docutils literal notranslate"><span class="pre">omega</span></code> for an edge. I’ll first illustrate some possible uses for the arguments above by setting <code class="docutils literal notranslate"><span class="pre">omega</span></code> to be distinctive for specific edges. I will set a value for omega so that printing the likelihood function illustrates what edges have been effected, but I won’t actually do any model fitting.</p>
<section id="specifying-a-clade">
<h2>Specifying a clade<a class="headerlink" href="#specifying-a-clade" title="Permalink to this headline">¶</a></h2>
<p>I’m going to cause <code class="docutils literal notranslate"><span class="pre">omega</span></code> to attain a different value for all branches aside from the primate clade and stem (<code class="docutils literal notranslate"><span class="pre">HowlerMon</span></code>, <code class="docutils literal notranslate"><span class="pre">Human</span></code>, <code class="docutils literal notranslate"><span class="pre">edge.0</span></code>).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;DogFaced&quot;</span><span class="p">,</span> <span class="s2">&quot;Mouse&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span>
    <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -9489.9506</p>
<p>number of free parameters = 10</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">1.00</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.03</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.04</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.04</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.02</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.11</span></td><td><span class="c3col_right">2.00</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>As you can see <code class="docutils literal notranslate"><span class="pre">omega</span></code> for the primate edges I listed above have the default parameter value (1.0), while the others have what I’ve assigned. In fact, you could omit the <code class="docutils literal notranslate"><span class="pre">clade</span></code> argument as this is the default, but I think for readability of scripts it’s best to be explicit.</p>
</section>
<section id="specifying-a-stem">
<h2>Specifying a stem<a class="headerlink" href="#specifying-a-stem" title="Permalink to this headline">¶</a></h2>
<p>This time I’ll specify the stem leading to the primates as the edge of interest.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>I need to reset the <code class="docutils literal notranslate"><span class="pre">lf</span></code> so all edges have the default value again. I’ll show this only for this example, but rest assured I’m doing it for all others too.</p>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s2">&quot;omega&quot;</span><span class="p">,</span> <span class="n">init</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
    <span class="n">stem</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -9424.8896</p>
<p>number of free parameters = 10</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">1.00</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.03</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.04</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.04</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.02</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.11</span></td><td><span class="c3col_right">1.00</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
</section>
<section id="specifying-clade-and-stem">
<h2>Specifying clade and stem<a class="headerlink" href="#specifying-clade-and-stem" title="Permalink to this headline">¶</a></h2>
<p>I’ll specify that both the primates and their stem are to be considered.</p>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">init</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span>
    <span class="n">stem</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -9442.4271</p>
<p>number of free parameters = 10</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">1.00</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.03</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.04</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.04</span></td><td><span class="c3col_right">2.00</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.02</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_right">1.00</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.11</span></td><td><span class="c3col_right">1.00</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
</section>
<section id="alternate-arguments-for-specifying-edges">
<h2>Alternate arguments for specifying edges<a class="headerlink" href="#alternate-arguments-for-specifying-edges" title="Permalink to this headline">¶</a></h2>
<p>The likelihood function <code class="docutils literal notranslate"><span class="pre">set_param_rule</span></code> method also has the arguments of <code class="docutils literal notranslate"><span class="pre">edge</span></code> and <code class="docutils literal notranslate"><span class="pre">edges</span></code>. These allow specific naming of the tree edge(s) to be affected by a rule. In general, however, the <code class="docutils literal notranslate"><span class="pre">tip_names</span></code> + <code class="docutils literal notranslate"><span class="pre">outgroup_name</span></code> combo is more robust.</p>
</section>
<section id="applications-of-scoped-parameters">
<h2>Applications of scoped parameters<a class="headerlink" href="#applications-of-scoped-parameters" title="Permalink to this headline">¶</a></h2>
<p>The general use-cases for which a tree scope can be applied are:</p>
<ol class="arabic simple">
<li><p>constraining all edges identified by a rule to have a specific value which is constant and not modifiable</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">is_constant</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>all edges identified by a rule have the same but different value to the rest of the tree</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span> <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span> <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span> <span class="n">clade</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p>allowing all edges identified by a rule to have different values of the parameter with the remaining tree edges having the same value</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<ol class="arabic simple" start="4">
<li><p>allowing all edges to have a different value</p></li>
</ol>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s2">&quot;omega&quot;</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>I’ll demonstrate these cases sequentially as they involve gradually increasing the degrees of freedom in the model. First we’ll constrain <code class="docutils literal notranslate"><span class="pre">omega</span></code> to equal 1 on the primate edges. I’ll then optimise the model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>here I’m specifying a constant value for the parameter and so I <strong>must</strong> use the argument <code class="docutils literal notranslate"><span class="pre">value</span></code> to set it. This not to be confused with the argument <code class="docutils literal notranslate"><span class="pre">init</span></code> that is used for providing initial (starting) values for fitting.</p>
</div>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">value</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
    <span class="n">is_constant</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -8640.9290</p>
<p>number of free parameters = 9</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">3.87</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_left">1.0</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_left">1.0</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_left">0.921816033174961</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.84</span></td><td><span class="c3col_left">0.921816033174961</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.06</span></td><td><span class="c3col_left">0.921816033174961</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_left">0.921816033174961</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.34</span></td><td><span class="c3col_left">0.921816033174961</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>I’ll now free up <code class="docutils literal notranslate"><span class="pre">omega</span></code> on the primate clade, but making it a single value shared by all primate lineages.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">is_constant</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -8639.7171</p>
<p>number of free parameters = 10</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">3.85</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_right">0.77</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_right">0.77</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.84</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.06</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.34</span></td><td><span class="c3col_right">0.92</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>Finally I’ll allow all primate edges to have different values of <code class="docutils literal notranslate"><span class="pre">omega</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span>
    <span class="s2">&quot;omega&quot;</span><span class="p">,</span>
    <span class="n">tip_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">],</span>
    <span class="n">outgroup_name</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
    <span class="n">clade</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -8638.9572</p>
<p>number of free parameters = 11</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">3.85</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_right">0.59</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_right">0.95</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.84</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.06</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.34</span></td><td><span class="c3col_right">0.92</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
<p>We now allow <code class="docutils literal notranslate"><span class="pre">omega</span></code> to be different on all edges.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">lf</span><span class="o">.</span><span class="n">set_param_rule</span><span class="p">(</span><span class="s2">&quot;omega&quot;</span><span class="p">,</span> <span class="n">is_independent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">lf</span><span class="o">.</span><span class="n">optimise</span><span class="p">(</span><span class="n">local</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">lf</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><h4>MG94HKY</h4>
<p>log-likelihood = -8636.1383</p>
<p>number of free parameters = 15</p>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Global params</span>
</caption>
<thead class="head_cell">
<th>kappa</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">3.85</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Edge params</span>
</caption>
<thead class="head_cell">
<th>edge</th><th>parent</th><th>length</th><th>omega</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">Human</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.09</span></td><td><span class="c3col_right">0.59</span></td></tr>
<tr><td><span class="c3col_left">HowlerMon</span></td><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_right">0.95</span></td></tr>
<tr><td><span class="c3col_left">edge.0</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.12</span></td><td><span class="c3col_right">1.13</span></td></tr>
<tr><td><span class="c3col_left">Mouse</span></td><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_right">0.84</span></td><td><span class="c3col_right">0.92</span></td></tr>
<tr><td><span class="c3col_left">edge.1</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.06</span></td><td><span class="c3col_right">0.38</span></td></tr>
<tr><td><span class="c3col_left">NineBande</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.28</span></td><td><span class="c3col_right">1.27</span></td></tr>
<tr><td><span class="c3col_left">DogFaced</span></td><td><span class="c3col_left">root</span></td><td><span class="c3col_right">0.34</span></td><td><span class="c3col_right">0.84</span></td></tr>
</tbody>
</table>

</div>

<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<caption>
<span class="cell_title">Motif params</span>
</caption>
<thead class="head_cell">
<th>A</th><th>C</th><th>G</th><th>T</th>
</thead>
<tbody>
<tr><td><span class="c3col_right">0.37</span></td><td><span class="c3col_right">0.19</span></td><td><span class="c3col_right">0.21</span></td><td><span class="c3col_right">0.23</span></td></tr>
</tbody>
</table>

</div></div></div>
</div>
</section>
</section>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020-2022, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 4.3.2 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>