
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Alignment &#8212; cogent3 2023.12.12 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/custom.css?v=da3cec82" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script src="../../../../_static/documentation_options.js?v=d202a915"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../../_static/thebelab-helper.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../../../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'doc/api/alignment/classes/cogent3.core.alignment.Alignment';</script>
    <link rel="icon" href="../../../../_static/c3-square.svg"/>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="ArrayAlignment" href="cogent3.core.alignment.ArrayAlignment.html" />
    <link rel="prev" title="SequenceCollection" href="cogent3.core.alignment.SequenceCollection.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../../_static/c3-logo.png" class="logo__image only-light" alt="cogent3 2023.12.12 documentation - Home"/>
    <script>document.write(`<img src="../../../../_static/c3-logo.png" class="logo__image only-dark" alt="cogent3 2023.12.12 documentation - Home"/>`);</script>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../draw/index.html">
                        Image Gallery
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../install.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        User Guides
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../community.html">
                        Community
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../licenses.html">
                        License
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../history.html">
                        Project History
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../pycogent.html">
                        How does cogent3 relate to PyCogent?
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../projects.html">
                        Projects using cogent3
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../data_file_links.html">
                        Data Files Used in the Documentation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../genindex.html">
                        Index
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../draw/index.html">
                        Image Gallery
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../install.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        User Guides
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../community.html">
                        Community
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../licenses.html">
                        License
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../history.html">
                        Project History
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../pycogent.html">
                        How does cogent3 relate to PyCogent?
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../projects.html">
                        Projects using cogent3
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../data_file_links.html">
                        Data Files Used in the Documentation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../genindex.html">
                        Index
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../app/index.html">apps</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-app.html">Overview of apps</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app-overview.html">Overview of using apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app-help.html">Getting help on an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app-get.html">Getting an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/dstore.html">Data stores – collections of data records</a></li>

<li class="toctree-l3"><a class="reference internal" href="../../../app/not-completed.html">Tracking records that could not be processed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-align.html">Sequence Alignment</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/align-nucleotide.html">Using a nucleotide model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/align-codon.html">Using a codon model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/align-protein.html">Using a protein model</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-evo-model.html">Modelling sequence evolution</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-ancestral-states.html">Reconstructing ancestral states</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-evo-natsel.html">Testing for natural selection</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-custom.html">Write your own Apps</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/user_function.html">Turn your functions into composable apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html">Manipulating biological data</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/loading_sequences.html">Loading a sequence from a file</a></li>





<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html">Sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/protein_sequences.html">Protein sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/alignments.html">Sequence Collections and Alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/features.html">Features</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/annotation_db.html">Annotation Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/moltypes.html">Molecular types</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/alphabet.html">Alphabets</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/what_codes.html">Available genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/genetic_code.html">Using genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/simple_trees.html">Trees</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/tables.html">Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/building_alignments.html">Building alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html">Genetic distance calculation</a></li>


<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/building_phylogenies.html">Building phylogenies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/useful_utilities.html">Useful Utilities</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../examples/index.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../../examples/period_estimation.html">Estimating periodic signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/simple.html">The simplest script</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/codon_models.html">Using codon models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/parallel.html">Parallel computations</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../../index.html">API</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.open_.html">open</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_model.html">get_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
<li class="toctree-l2 current active has-children"><a class="reference internal" href="../alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l3 current active"><a class="current reference internal" href="#">Alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-14"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-15"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-16"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-17"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-18"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-19"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-20"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../app/composable/define_app.html">define_app</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">User Guides</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="../alignment.html" class="nav-link"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Alignment</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="alignment">
<h1>Alignment<a class="headerlink" href="#alignment" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment" title="Link to this definition">#</a></dt>
<dd><p>An annotatable alignment class</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_db</strong></dt><dd></dd>
<dt><a class="reference internal" href="#cogent3.core.alignment.Alignment.num_seqs" title="cogent3.core.alignment.Alignment.num_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_seqs</span></code></a></dt><dd><p>Returns the number of sequences in the alignment.</p>
</dd>
<dt><a class="reference internal" href="#cogent3.core.alignment.Alignment.positions" title="cogent3.core.alignment.Alignment.positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></a></dt><dd><p>Iterates over positions in the alignment, in order.</p>
</dd>
<dt><strong>seqs</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.add_feature" title="cogent3.core.alignment.Alignment.add_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_feature</span></code></a>(*, biotype, name, spans[, ...])</p></td>
<td><p>add feature on named sequence, or on the alignment itself</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.add_from_ref_aln" title="cogent3.core.alignment.Alignment.add_from_ref_aln"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code></a>(ref_aln[, before_name, ...])</p></td>
<td><p>Insert sequence(s) to self based on their alignment to a reference sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.add_seqs" title="cogent3.core.alignment.Alignment.add_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_seqs</span></code></a>(other[, before_name, after_name])</p></td>
<td><p>Returns new object of class self with sequences from other added.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.alignment_quality" title="cogent3.core.alignment.Alignment.alignment_quality"><code class="xref py py-obj docutils literal notranslate"><span class="pre">alignment_quality</span></code></a>([app_name])</p></td>
<td><p>Computes the alignment quality using the indicated app</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.annotate_from_gff" title="cogent3.core.alignment.Alignment.annotate_from_gff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_from_gff</span></code></a>(f[, seq_ids])</p></td>
<td><p>copies annotations from a gff file to a sequence in self</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.apply_pssm" title="cogent3.core.alignment.Alignment.apply_pssm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply_pssm</span></code></a>([pssm, path, background, ...])</p></td>
<td><p>scores sequences using the specified pssm</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.coevolution" title="cogent3.core.alignment.Alignment.coevolution"><code class="xref py py-obj docutils literal notranslate"><span class="pre">coevolution</span></code></a>([method, segments, drawable, ...])</p></td>
<td><p>performs pairwise coevolution measurement</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.copy" title="cogent3.core.alignment.Alignment.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Returns deep copy of self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.copy_annotations" title="cogent3.core.alignment.Alignment.copy_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_annotations</span></code></a>(seq_db)</p></td>
<td><p>copy annotations into attached annotation db</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.count_gaps_per_pos" title="cogent3.core.alignment.Alignment.count_gaps_per_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_gaps_per_pos</span></code></a>([include_ambiguity])</p></td>
<td><p>return counts of gaps per position as a DictArray</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.count_gaps_per_seq" title="cogent3.core.alignment.Alignment.count_gaps_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_gaps_per_seq</span></code></a>([induced_by, unique, ...])</p></td>
<td><p>return counts of gaps per sequence as a DictArray</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.counts" title="cogent3.core.alignment.Alignment.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>([motif_length, include_ambiguity, ...])</p></td>
<td><p>counts of motifs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.counts_per_pos" title="cogent3.core.alignment.Alignment.counts_per_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts_per_pos</span></code></a>([motif_length, ...])</p></td>
<td><p>return DictArray of counts per position</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.counts_per_seq" title="cogent3.core.alignment.Alignment.counts_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts_per_seq</span></code></a>([motif_length, ...])</p></td>
<td><p>counts of non-overlapping motifs per sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.deepcopy" title="cogent3.core.alignment.Alignment.deepcopy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deepcopy</span></code></a>([sliced])</p></td>
<td><p>returns deep copy of self.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.degap" title="cogent3.core.alignment.Alignment.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>(**kwargs)</p></td>
<td><p>Returns copy in which sequences have no gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.distance_matrix" title="cogent3.core.alignment.Alignment.distance_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance_matrix</span></code></a>([calc, show_progress, ...])</p></td>
<td><p>Returns pairwise distances between sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.dotplot" title="cogent3.core.alignment.Alignment.dotplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dotplot</span></code></a>([name1, name2, window, threshold, ...])</p></td>
<td><p>make a dotplot between specified sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.entropy_per_pos" title="cogent3.core.alignment.Alignment.entropy_per_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">entropy_per_pos</span></code></a>([motif_length, ...])</p></td>
<td><p>returns shannon entropy per position</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.entropy_per_seq" title="cogent3.core.alignment.Alignment.entropy_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">entropy_per_seq</span></code></a>([motif_length, ...])</p></td>
<td><p>returns the Shannon entropy per sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.filtered" title="cogent3.core.alignment.Alignment.filtered"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filtered</span></code></a>(predicate[, motif_length, ...])</p></td>
<td><p>The alignment positions where predicate(column) is true.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_ambiguous_positions" title="cogent3.core.alignment.Alignment.get_ambiguous_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_ambiguous_positions</span></code></a>()</p></td>
<td><p>Returns dict of seq:{position:char} for ambiguous chars.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_degapped_relative_to" title="cogent3.core.alignment.Alignment.get_degapped_relative_to"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_degapped_relative_to</span></code></a>(name)</p></td>
<td><p>Remove all columns with gaps in sequence with given name.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_drawable" title="cogent3.core.alignment.Alignment.get_drawable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawable</span></code></a>(*[, biotype, width, vertical])</p></td>
<td><p>make a figure from sequence features</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_drawables" title="cogent3.core.alignment.Alignment.get_drawables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawables</span></code></a>(*[, biotype])</p></td>
<td><p>returns a dict of drawables, keyed by type</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_features" title="cogent3.core.alignment.Alignment.get_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_features</span></code></a>(*[, seqid, biotype, name, ...])</p></td>
<td><p>yields Feature instances</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_gap_array" title="cogent3.core.alignment.Alignment.get_gap_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_gap_array</span></code></a>([include_ambiguity])</p></td>
<td><p>returns bool array with gap state True, False otherwise</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_gapped_seq" title="cogent3.core.alignment.Alignment.get_gapped_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_gapped_seq</span></code></a>(seq_name[, recode_gaps])</p></td>
<td><p>Return a gapped Sequence object for the specified seqname.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_identical_sets" title="cogent3.core.alignment.Alignment.get_identical_sets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_identical_sets</span></code></a>([mask_degen])</p></td>
<td><p>returns sets of names for sequences that are identical</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_lengths" title="cogent3.core.alignment.Alignment.get_lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_lengths</span></code></a>([include_ambiguity, allow_gap])</p></td>
<td><p>returns {name: seq length, ...}</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_motif_probs" title="cogent3.core.alignment.Alignment.get_motif_probs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_motif_probs</span></code></a>([alphabet, ...])</p></td>
<td><p>Return a dictionary of motif probs, calculated as the averaged frequency across sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_position_indices" title="cogent3.core.alignment.Alignment.get_position_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_indices</span></code></a>(f[, native, negate])</p></td>
<td><p>Returns list of column indices for which f(col) is True.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_projected_feature" title="cogent3.core.alignment.Alignment.get_projected_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_projected_feature</span></code></a>(*, seqid, feature)</p></td>
<td><p>returns an alignment feature projected onto the seqid sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_projected_features" title="cogent3.core.alignment.Alignment.get_projected_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_projected_features</span></code></a>(*, seqid, **kwargs)</p></td>
<td><p>projects all features from other sequences onto seqid</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_seq" title="cogent3.core.alignment.Alignment.get_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq</span></code></a>(seqname)</p></td>
<td><p>Return a ungapped Sequence object for the specified seqname.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_seq_indices" title="cogent3.core.alignment.Alignment.get_seq_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq_indices</span></code></a>(f[, negate])</p></td>
<td><p>Returns list of keys of seqs where f(row) is True.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_similar" title="cogent3.core.alignment.Alignment.get_similar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_similar</span></code></a>(target[, min_similarity, ...])</p></td>
<td><p>Returns new Alignment containing sequences similar to target.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.get_translation" title="cogent3.core.alignment.Alignment.get_translation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_translation</span></code></a>([gc, incomplete_ok, ...])</p></td>
<td><p>translate from nucleic acid to protein</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.has_terminal_stop" title="cogent3.core.alignment.Alignment.has_terminal_stop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_terminal_stop</span></code></a>([gc, strict, allow_partial])</p></td>
<td><p>Returns True if any sequence has a terminal stop codon.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.has_terminal_stops" title="cogent3.core.alignment.Alignment.has_terminal_stops"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_terminal_stops</span></code></a>(**kwargs)</p></td>
<td><p>deprecated</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.information_plot" title="cogent3.core.alignment.Alignment.information_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">information_plot</span></code></a>([width, height, window, ...])</p></td>
<td><p>plot information per position</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.is_ragged" title="cogent3.core.alignment.Alignment.is_ragged"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_ragged</span></code></a>()</p></td>
<td><p>Returns True if alignment has sequences of different lengths.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.iter_positions" title="cogent3.core.alignment.Alignment.iter_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_positions</span></code></a>([pos_order])</p></td>
<td><p>Iterates over positions in the alignment, in order.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.iter_selected" title="cogent3.core.alignment.Alignment.iter_selected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_selected</span></code></a>([seq_order, pos_order])</p></td>
<td><p>Iterates over elements in the alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.iter_seqs" title="cogent3.core.alignment.Alignment.iter_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_seqs</span></code></a>([seq_order])</p></td>
<td><p>Iterates over values (sequences) in the alignment, in order.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.iupac_consensus" title="cogent3.core.alignment.Alignment.iupac_consensus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iupac_consensus</span></code></a>([alphabet, allow_gaps])</p></td>
<td><p>Returns string containing IUPAC consensus sequence of the alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.majority_consensus" title="cogent3.core.alignment.Alignment.majority_consensus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">majority_consensus</span></code></a>()</p></td>
<td><p>Returns list containing most frequent item at each position.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.make_feature" title="cogent3.core.alignment.Alignment.make_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_feature</span></code></a>(*, feature[, on_alignment])</p></td>
<td><p>create a feature on named sequence, or on the alignment itself</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.matching_ref" title="cogent3.core.alignment.Alignment.matching_ref"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matching_ref</span></code></a>(ref_name, gap_fraction, gap_run)</p></td>
<td><p>Returns new alignment with seqs well aligned with a reference.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.no_degenerates" title="cogent3.core.alignment.Alignment.no_degenerates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">no_degenerates</span></code></a>([motif_length, allow_gap])</p></td>
<td><p>returns new alignment without degenerate characters</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.omit_bad_seqs" title="cogent3.core.alignment.Alignment.omit_bad_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_bad_seqs</span></code></a>([quantile])</p></td>
<td><p>Returns new alignment without sequences with a number of uniquely introduced gaps exceeding quantile</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.omit_gap_pos" title="cogent3.core.alignment.Alignment.omit_gap_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_pos</span></code></a>([allowed_gap_frac, motif_length])</p></td>
<td><p>Returns new alignment where all cols (motifs) have &lt;= allowed_gap_frac gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.omit_gap_runs" title="cogent3.core.alignment.Alignment.omit_gap_runs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_runs</span></code></a>([allowed_run])</p></td>
<td><p>Returns new alignment where all seqs have runs of gaps &lt;=allowed_run.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.omit_gap_seqs" title="cogent3.core.alignment.Alignment.omit_gap_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code></a>([allowed_gap_frac])</p></td>
<td><p>Returns new alignment with seqs that have &lt;= allowed_gap_frac.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.pad_seqs" title="cogent3.core.alignment.Alignment.pad_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pad_seqs</span></code></a>([pad_length])</p></td>
<td><p>Returns copy in which sequences are padded to same length.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.probs_per_pos" title="cogent3.core.alignment.Alignment.probs_per_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">probs_per_pos</span></code></a>([motif_length, ...])</p></td>
<td><p>returns MotifFreqsArray per position</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.probs_per_seq" title="cogent3.core.alignment.Alignment.probs_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">probs_per_seq</span></code></a>([motif_length, ...])</p></td>
<td><p>return MotifFreqsArray per sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.quick_tree" title="cogent3.core.alignment.Alignment.quick_tree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">quick_tree</span></code></a>([calc, bootstrap, drop_invalid, ...])</p></td>
<td><p>Returns pairwise distances between sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.rc" title="cogent3.core.alignment.Alignment.rc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rc</span></code></a>()</p></td>
<td><p>Returns the reverse complement alignment</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.rename_seqs" title="cogent3.core.alignment.Alignment.rename_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rename_seqs</span></code></a>(renamer)</p></td>
<td><p>returns new instance with sequences renamed</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.replace_seqs" title="cogent3.core.alignment.Alignment.replace_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_seqs</span></code></a>(seqs[, aa_to_codon])</p></td>
<td><p>Returns new alignment with same shape but with data taken from seqs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.reverse_complement" title="cogent3.core.alignment.Alignment.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_complement</span></code></a>()</p></td>
<td><p>Returns the reverse complement alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.sample" title="cogent3.core.alignment.Alignment.sample"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample</span></code></a>([n, with_replacement, motif_length, ...])</p></td>
<td><p>Returns random sample of positions from self, e.g. to bootstrap.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.seqlogo" title="cogent3.core.alignment.Alignment.seqlogo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">seqlogo</span></code></a>([width, height, wrap, vspace, colours])</p></td>
<td><p>returns Drawable sequence logo using mutual information</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.set_repr_policy" title="cogent3.core.alignment.Alignment.set_repr_policy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_repr_policy</span></code></a>([num_seqs, num_pos, ...])</p></td>
<td><p>specify policy for repr(self)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.sliding_windows" title="cogent3.core.alignment.Alignment.sliding_windows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sliding_windows</span></code></a>(window, step[, start, end])</p></td>
<td><p>Generator yielding new alignments of given length and interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.strand_symmetry" title="cogent3.core.alignment.Alignment.strand_symmetry"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strand_symmetry</span></code></a>([motif_length])</p></td>
<td><p>returns dict of strand symmetry test results per seq</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.take_positions" title="cogent3.core.alignment.Alignment.take_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_positions</span></code></a>(cols[, negate])</p></td>
<td><p>Returns new Alignment containing only specified positions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.take_positions_if" title="cogent3.core.alignment.Alignment.take_positions_if"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_positions_if</span></code></a>(f[, negate])</p></td>
<td><p>Returns new Alignment containing cols where f(col) is True.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.take_seqs" title="cogent3.core.alignment.Alignment.take_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_seqs</span></code></a>(seqs[, negate])</p></td>
<td><p>Returns new Alignment containing only specified seqs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.take_seqs_if" title="cogent3.core.alignment.Alignment.take_seqs_if"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_seqs_if</span></code></a>(f[, negate])</p></td>
<td><p>Returns new Alignment containing seqs where f(row) is True.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_dict" title="cogent3.core.alignment.Alignment.to_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dict</span></code></a>()</p></td>
<td><p>Returns the alignment as dict of names -&gt; strings.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_dna" title="cogent3.core.alignment.Alignment.to_dna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dna</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of DNA moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_fasta" title="cogent3.core.alignment.Alignment.to_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_fasta</span></code></a>()</p></td>
<td><p>Return alignment in Fasta format</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_html" title="cogent3.core.alignment.Alignment.to_html"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_html</span></code></a>([name_order, wrap, limit, ref_name, ...])</p></td>
<td><p>returns html with embedded styles for sequence colouring</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_json" title="cogent3.core.alignment.Alignment.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns json formatted string</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_moltype" title="cogent3.core.alignment.Alignment.to_moltype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_moltype</span></code></a>(moltype)</p></td>
<td><p>returns copy of self with moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_nexus" title="cogent3.core.alignment.Alignment.to_nexus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_nexus</span></code></a>(seq_type[, wrap])</p></td>
<td><p>Return alignment in NEXUS format and mapping to sequence ids</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_phylip" title="cogent3.core.alignment.Alignment.to_phylip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_phylip</span></code></a>()</p></td>
<td><p>Return alignment in PHYLIP format and mapping to sequence ids</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_pretty" title="cogent3.core.alignment.Alignment.to_pretty"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_pretty</span></code></a>([name_order, wrap])</p></td>
<td><p>returns a string representation of the alignment in pretty print format</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_protein" title="cogent3.core.alignment.Alignment.to_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_protein</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of PROTEIN moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_rich_dict" title="cogent3.core.alignment.Alignment.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>()</p></td>
<td><p>returns detailed content including info and moltype attributes</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_rna" title="cogent3.core.alignment.Alignment.to_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rna</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of RNA moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.to_type" title="cogent3.core.alignment.Alignment.to_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_type</span></code></a>([array_align, moltype, alphabet])</p></td>
<td><p>returns alignment of type indicated by array_align</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.trim_stop_codons" title="cogent3.core.alignment.Alignment.trim_stop_codons"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim_stop_codons</span></code></a>([gc, strict, allow_partial])</p></td>
<td><p>Removes any terminal stop codons from the sequences</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.variable_positions" title="cogent3.core.alignment.Alignment.variable_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">variable_positions</span></code></a>([include_gap_motif])</p></td>
<td><p>Return a list of variable position indexes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.with_gaps_from" title="cogent3.core.alignment.Alignment.with_gaps_from"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_gaps_from</span></code></a>(template)</p></td>
<td><p>Same alignment but overwritten with the gaps from 'template'</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.with_masked_annotations" title="cogent3.core.alignment.Alignment.with_masked_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_masked_annotations</span></code></a>(biotypes[, ...])</p></td>
<td><p>returns an alignment with annot_types regions replaced by mask_char if shadow is False, otherwise all other regions are masked.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.with_modified_termini" title="cogent3.core.alignment.Alignment.with_modified_termini"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_modified_termini</span></code></a>()</p></td>
<td><p>Changes the termini to include termini char instead of gapmotif.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.Alignment.write" title="cogent3.core.alignment.Alignment.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>([filename, format])</p></td>
<td><p>Write the alignment to a file, preserving order of sequences.</p></td>
</tr>
</tbody>
</table>
<table class="table">
<tbody>
<tr class="row-odd"><td><p><strong>gapped_by_map</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spans</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'+'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.add_feature" title="Link to this definition">#</a></dt>
<dd><p>add feature on named sequence, or on the alignment itself</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seqid</strong></dt><dd><p>sequence name, incompatible with on_alignment</p>
</dd>
<dt><strong>parent_id</strong></dt><dd><p>name of the parent feature</p>
</dd>
<dt><strong>biotype</strong></dt><dd><p>biological type, e.g. CDS</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>spans</strong></dt><dd><p>plus strand coordinates of feature</p>
</dd>
<dt><strong>strand</strong></dt><dd><p>‘+’ (default) or ‘-’</p>
</dd>
<dt><strong>on_alignment</strong></dt><dd><p>the feature is in alignment coordinates, incompatible with setting
seqid. Set to True if seqid not provided.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Feature</dt><dd></dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>ValueError if define a seqid not on alignment or use seqid and</dt><dd></dd>
<dt>on_alignment.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.add_from_ref_aln">
<span class="sig-name descname"><span class="pre">add_from_ref_aln</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_aln</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">after_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.add_from_ref_aln" title="Link to this definition">#</a></dt>
<dd><p>Insert sequence(s) to self based on their alignment to a reference
sequence. Assumes the first sequence in ref_aln.names[0] is the
reference.</p>
<p>By default the sequence is appended to the end of the alignment,
this can be changed by using either before_name or after_name
arguments.</p>
<p>Returns Alignment object of the same class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ref_aln</strong></dt><dd><p>reference alignment (Alignment object/series) of
reference sequence and sequences to add.
New sequences in ref_aln (ref_aln.names[1:] are sequences to add.
If series is used as ref_aln, it must have the structure
[[‘ref_name’, SEQ], [‘name’, SEQ]]</p>
</dd>
<dt><strong>before_name</strong></dt><dd><p>name of the sequence before which
sequence is added</p>
</dd>
<dt><strong>after_name</strong></dt><dd><p>name of the sequence after which sequence is added
If both before_name and after_name are specified seqs will be
inserted using before_name.</p>
</dd>
<dt><strong>Example:</strong></dt><dd></dd>
<dt><strong>Aln1:</strong></dt><dd></dd>
<dt><strong>-AC-DEFGHI (name: seq1)</strong></dt><dd></dd>
<dt><strong>XXXXXX–XX (name: seq2)</strong></dt><dd></dd>
<dt><strong>YYYY-YYYYY (name: seq3)</strong></dt><dd></dd>
<dt><strong>Aln2:</strong></dt><dd></dd>
<dt><strong>ACDEFGHI   (name: seq1)</strong></dt><dd></dd>
<dt><strong>KL–MNPR   (name: seqX)</strong></dt><dd></dd>
<dt><strong>KLACMNPR   (name: seqY)</strong></dt><dd></dd>
<dt><strong>KL–MNPR   (name: seqZ)</strong></dt><dd></dd>
<dt><strong>Out:</strong></dt><dd></dd>
<dt><strong>-AC-DEFGHI (name: seq1)</strong></dt><dd></dd>
<dt><strong>XXXXXX–XX (name: seq2)</strong></dt><dd></dd>
<dt><strong>YYYY-YYYYY (name: seq3)</strong></dt><dd></dd>
<dt><strong>-KL—MNPR (name: seqX)</strong></dt><dd></dd>
<dt><strong>-KL-ACMNPR (name: seqY)</strong></dt><dd></dd>
<dt><strong>-KL—MNPR (name: seqZ)</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.add_seqs">
<span class="sig-name descname"><span class="pre">add_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">after_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.add_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new object of class self with sequences from other added.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong></dt><dd><p>same class as self or coerceable to that class</p>
</dd>
<dt><strong>before_name</strong><span class="classifier">str</span></dt><dd><p>which sequence is added</p>
</dd>
<dt><strong>after_name</strong><span class="classifier">str</span></dt><dd><p>which sequence is added</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If both before_name and after_name are specified, the seqs will be
inserted using before_name.</p>
<p>By default the sequence is appended to the end of the alignment,
this can be changed by using either before_name or after_name arguments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.alignment_quality">
<span class="sig-name descname"><span class="pre">alignment_quality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">app_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ic_score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.alignment_quality" title="Link to this definition">#</a></dt>
<dd><p>Computes the alignment quality using the indicated app</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>app_name</strong></dt><dd><p>name of an alignment score calculating app, e.g. ‘ic_score’,
‘cogent3_score’, ‘sp_score’</p>
</dd>
<dt><strong>kwargs</strong></dt><dd><p>keyword arguments to be passed to the app. Use
<code class="docutils literal notranslate"><span class="pre">cogent3.app_help(app_name)</span></code> to see the available options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>float or a NotCompleted instance if the score could not be computed</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.annotate_from_gff">
<span class="sig-name descname"><span class="pre">annotate_from_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PathLike</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.annotate_from_gff" title="Link to this definition">#</a></dt>
<dd><p>copies annotations from a gff file to a sequence in self</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>f</strong><span class="classifier">path to gff annotation file.</span></dt><dd></dd>
<dt><strong>seq_name</strong><span class="classifier">names of seqs to be annotated.</span></dt><dd></dd>
<dt><strong>does not support setting offset, set offset directly on sequences with seq.annotation_offset = offset</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.annotation_db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation_db</span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.annotation_db" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.apply_pssm">
<span class="sig-name descname"><span class="pre">apply_pssm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pssm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ui</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.apply_pssm" title="Link to this definition">#</a></dt>
<dd><p>scores sequences using the specified pssm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pssm</strong><span class="classifier">profile.PSSM</span></dt><dd><p>if not provided, will be loaded from path</p>
</dd>
<dt><strong>path</strong></dt><dd><p>path to either a jaspar or cisbp matrix (path must end have a suffix
matching the format).</p>
</dd>
<dt><strong>pseudocount</strong></dt><dd><p>adjustment for zero in matrix</p>
</dd>
<dt><strong>names</strong></dt><dd><p>returns only scores for these sequences and in the name order</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>numpy array of log2 based scores at every position</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.coevolution">
<span class="sig-name descname"><span class="pre">coevolution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'nmi'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">segments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drawable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ui</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.coevolution" title="Link to this definition">#</a></dt>
<dd><p>performs pairwise coevolution measurement</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>method</strong><span class="classifier">str</span></dt><dd><p>coevolution metric, defaults to ‘nmi’ (Normalized Mutual
Information). Valid choices are ‘rmi’ (Resampled Mutual Information)
and ‘mi’, mutual information.</p>
</dd>
<dt><strong>segments</strong><span class="classifier">coordinate series</span></dt><dd><p>coordinates of the form [(start, end), …] where all possible
pairs of alignment positions within and between segments are
examined.</p>
</dd>
<dt><strong>drawable</strong><span class="classifier">None or str</span></dt><dd><p>Result object is capable of plotting data specified type. str value
must be one of plot type ‘box’, ‘heatmap’, ‘violin’.</p>
</dd>
<dt><strong>show_progress</strong><span class="classifier">bool</span></dt><dd><p>shows a progress bar</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>DictArray of results with lower-triangular values. Upper triangular</dt><dd></dd>
<dt>elements and estimates that could not be computed for numerical reasons</dt><dd></dd>
<dt>are set as nan</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns deep copy of self.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.copy_annotations">
<span class="sig-name descname"><span class="pre">copy_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SupportsFeatures</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.copy_annotations" title="Link to this definition">#</a></dt>
<dd><p>copy annotations into attached annotation db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_db</strong></dt><dd><p>compatible annotation db</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Only copies annotations for records with seqid in self.names</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.count_gaps_per_pos">
<span class="sig-name descname"><span class="pre">count_gaps_per_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.count_gaps_per_pos" title="Link to this definition">#</a></dt>
<dd><p>return counts of gaps per position as a DictArray</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong><span class="classifier">bool</span></dt><dd><p>if True, ambiguity characters that include the gap state are
included</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.count_gaps_per_seq">
<span class="sig-name descname"><span class="pre">count_gaps_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">induced_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drawable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.count_gaps_per_seq" title="Link to this definition">#</a></dt>
<dd><p>return counts of gaps per sequence as a DictArray</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>induced_by</strong><span class="classifier">bool</span></dt><dd><p>a gapped column is considered to be induced by a seq if the seq
has a non-gap character in that column.</p>
</dd>
<dt><strong>unique</strong><span class="classifier">bool</span></dt><dd><p>count is limited to gaps uniquely induced by each sequence</p>
</dd>
<dt><strong>include_ambiguity</strong><span class="classifier">bool</span></dt><dd><p>if True, ambiguity characters that include the gap state are
included</p>
</dd>
<dt><strong>drawable</strong><span class="classifier">bool or str</span></dt><dd><p>if True, resulting object is capable of plotting data via specified
plot type ‘bar’, ‘box’ or ‘violin’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.counts" title="Link to this definition">#</a></dt>
<dd><p>counts of motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.counts_per_pos">
<span class="sig-name descname"><span class="pre">counts_per_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.counts_per_pos" title="Link to this definition">#</a></dt>
<dd><p>return DictArray of counts per position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>alert</strong></dt><dd><p>warns if motif_length &gt; 1 and alignment trimmed to produce
motif columns</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.counts_per_seq">
<span class="sig-name descname"><span class="pre">counts_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.counts_per_seq" title="Link to this definition">#</a></dt>
<dd><p>counts of non-overlapping motifs per sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if False, all canonical states included</p>
</dd>
<dt><strong>alert</strong></dt><dd><p>warns if motif_length &gt; 1 and alignment trimmed to produce
motif columns</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>MotifCountsArray</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.deepcopy">
<span class="sig-name descname"><span class="pre">deepcopy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sliced</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.deepcopy" title="Link to this definition">#</a></dt>
<dd><p>returns deep copy of self.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sliced</strong></dt><dd><p>if True, reduces the sequence to current
interval. This also causes dropping
annotations.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.default_gap">
<span class="sig-name descname"><span class="pre">default_gap</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'-'</span></em><a class="headerlink" href="#cogent3.core.alignment.Alignment.default_gap" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.degap">
<span class="sig-name descname"><span class="pre">degap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.degap" title="Link to this definition">#</a></dt>
<dd><p>Returns copy in which sequences have no gaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>kwargs</strong></dt><dd><p>passed to class constructor</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.distance_matrix">
<span class="sig-name descname"><span class="pre">distance_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">calc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdist'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_invalid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.distance_matrix" title="Link to this definition">#</a></dt>
<dd><p>Returns pairwise distances between sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>calc</strong><span class="classifier">str</span></dt><dd><p>a pairwise distance calculator or name of one. For options see
cogent3.evolve.fast_distance.available_distances</p>
</dd>
<dt><strong>show_progress</strong><span class="classifier">bool</span></dt><dd><p>controls progress display for distance calculation</p>
</dd>
<dt><strong>drop_invalid</strong><span class="classifier">bool</span></dt><dd><p>If True, sequences for which a pairwise distance could not be
calculated are excluded. If False, an ArithmeticError is raised if
a distance could not be computed on observed data.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.dotplot">
<span class="sig-name descname"><span class="pre">dotplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.dotplot" title="Link to this definition">#</a></dt>
<dd><p>make a dotplot between specified sequences. Random sequences
chosen if names not provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name1, name2</strong><span class="classifier">str or None</span></dt><dd><p>names of sequences. If one is not provided, a random choice is made</p>
</dd>
<dt><strong>window</strong><span class="classifier">int</span></dt><dd><p>k-mer size for comparison between sequences</p>
</dd>
<dt><strong>threshold</strong><span class="classifier">int</span></dt><dd><p>windows where the sequences are identical &gt;= threshold are a match</p>
</dd>
<dt><strong>k</strong><span class="classifier">int</span></dt><dd><p>size of k-mer to break sequences into. Larger values increase
speed but reduce resolution. If not specified, is computed as the
maximum of (window-threshold), (window % k) * k &lt;= threshold.</p>
</dd>
<dt><strong>min_gap</strong><span class="classifier">int</span></dt><dd><p>permitted gap for joining adjacent line segments, default is no gap
joining</p>
</dd>
<dt><strong>width</strong><span class="classifier">int</span></dt><dd><p>figure width. Figure height is computed based on the ratio of
len(seq1) / len(seq2)</p>
</dd>
<dt><strong>title</strong></dt><dd><p>title for the plot</p>
</dd>
<dt><strong>rc</strong><span class="classifier">bool or None</span></dt><dd><p>include dotplot of reverse compliment also. Only applies to Nucleic
acids moltypes</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>a Drawable or AnnotatedDrawable</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.entropy_per_pos">
<span class="sig-name descname"><span class="pre">entropy_per_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.entropy_per_pos" title="Link to this definition">#</a></dt>
<dd><p>returns shannon entropy per position</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.entropy_per_seq">
<span class="sig-name descname"><span class="pre">entropy_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.entropy_per_seq" title="Link to this definition">#</a></dt>
<dd><p>returns the Shannon entropy per sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of characters per tuple.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>For motif_length &gt; 1, it’s advisable to specify exclude_unobserved=True,
this avoids unnecessary calculations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.filtered">
<span class="sig-name descname"><span class="pre">filtered</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">predicate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_remainder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.filtered" title="Link to this definition">#</a></dt>
<dd><p>The alignment positions where predicate(column) is true.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>predicate</strong><span class="classifier">callable</span></dt><dd><p>a callback function that takes an tuple of motifs and returns
True/False</p>
</dd>
<dt><strong>motif_length</strong><span class="classifier">int</span></dt><dd><p>length of the motifs the sequences should be split  into, eg. 3 for
filtering aligned codons.</p>
</dd>
<dt><strong>drop_remainder</strong><span class="classifier">bool</span></dt><dd><p>If length is not modulo motif_length, allow dropping the terminal
remaining columns</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.gap_chars">
<span class="sig-name descname"><span class="pre">gap_chars</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'-':</span> <span class="pre">None,</span> <span class="pre">'?':</span> <span class="pre">None}</span></em><a class="headerlink" href="#cogent3.core.alignment.Alignment.gap_chars" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.gapped_by_map">
<span class="sig-name descname"><span class="pre">gapped_by_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">keep</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.gapped_by_map" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_ambiguous_positions">
<span class="sig-name descname"><span class="pre">get_ambiguous_positions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_ambiguous_positions" title="Link to this definition">#</a></dt>
<dd><p>Returns dict of seq:{position:char} for ambiguous chars.</p>
<p>Used in likelihood calculations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_degapped_relative_to">
<span class="sig-name descname"><span class="pre">get_degapped_relative_to</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_degapped_relative_to" title="Link to this definition">#</a></dt>
<dd><p>Remove all columns with gaps in sequence with given name.</p>
<p>Returns Alignment object of the same class.
Note that the seqs in the new Alignment are always new objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong></dt><dd><p>sequence name</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_drawable">
<span class="sig-name descname"><span class="pre">get_drawable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_drawable" title="Link to this definition">#</a></dt>
<dd><p>make a figure from sequence features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>passed to get_features(biotype). Can be a single biotype or
series. Only features matching this will be included.</p>
</dd>
<dt><strong>width</strong></dt><dd><p>width in pixels</p>
</dd>
<dt><strong>vertical</strong></dt><dd><p>rotates the drawable</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>a Drawable instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_drawables">
<span class="sig-name descname"><span class="pre">get_drawables</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_drawables" title="Link to this definition">#</a></dt>
<dd><p>returns a dict of drawables, keyed by type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>passed to get_features(biotype). Can be a single biotype or
series. Only features matching this will be included.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_features">
<span class="sig-name descname"><span class="pre">get_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Iterator</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_features" title="Link to this definition">#</a></dt>
<dd><p>yields Feature instances</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seqid</strong></dt><dd><p>limit search to features on this named sequence, defaults to search all</p>
</dd>
<dt><strong>biotype</strong></dt><dd><p>biotype of the feature, e.g. CDS, gene</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>on_alignment</strong></dt><dd><p>limit query to features on Alignment, ignores sequences. Ignored on
SequenceCollection instances.</p>
</dd>
<dt><strong>allow_partial</strong></dt><dd><p>allow features partially overlaping self</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>When dealing with a nucleic acid moltype, the returned features will
yield a sequence segment that is consistently oriented irrespective
of strand of the current instance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_gap_array">
<span class="sig-name descname"><span class="pre">get_gap_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_gap_array" title="Link to this definition">#</a></dt>
<dd><p>returns bool array with gap state True, False otherwise</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong><span class="classifier">bool</span></dt><dd><p>if True, ambiguity characters that include the gap state are
included</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_gapped_seq">
<span class="sig-name descname"><span class="pre">get_gapped_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_gapped_seq" title="Link to this definition">#</a></dt>
<dd><p>Return a gapped Sequence object for the specified seqname.</p>
<p>Note: always returns Sequence object, not ArraySequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_identical_sets">
<span class="sig-name descname"><span class="pre">get_identical_sets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask_degen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_identical_sets" title="Link to this definition">#</a></dt>
<dd><p>returns sets of names for sequences that are identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>mask_degen</strong></dt><dd><p>if True, degenerate characters are ignored</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_lengths">
<span class="sig-name descname"><span class="pre">get_lengths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_lengths" title="Link to this definition">#</a></dt>
<dd><p>returns {name: seq length, …}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_motif_probs">
<span class="sig-name descname"><span class="pre">get_motif_probs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_motif_probs" title="Link to this definition">#</a></dt>
<dd><p>Return a dictionary of motif probs, calculated as the averaged
frequency across sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong></dt><dd><p>if True resolved ambiguous codes are
included in estimation of frequencies, default is False.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, motifs that are not present in
the alignment are excluded from the returned dictionary,
default is False.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>allow gap motif</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_position_indices">
<span class="sig-name descname"><span class="pre">get_position_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">native</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_position_indices" title="Link to this definition">#</a></dt>
<dd><p>Returns list of column indices for which f(col) is True.</p>
<dl class="simple">
<dt>f<span class="classifier">callable</span></dt><dd><p>function that returns true/false given an alignment position</p>
</dd>
<dt>native<span class="classifier">boolean</span></dt><dd><p>if True, and ArrayAlignment, f is provided with slice of array
otherwise the string is used</p>
</dd>
<dt>negate<span class="classifier">boolean</span></dt><dd><p>if True, not f() is used</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_projected_feature">
<span class="sig-name descname"><span class="pre">get_projected_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_projected_feature" title="Link to this definition">#</a></dt>
<dd><p>returns an alignment feature projected onto the seqid sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seqid</strong></dt><dd><p>name of the sequence to project the feature onto</p>
</dd>
<dt><strong>feature</strong></dt><dd><p>a Feature, bound to self, that will be projected</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>a new Feature bound to seqid</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The alignment coordinates of feature are converted into the seqid
sequence coordinates and the object is bound to that sequence.</p>
<p>The feature is added to the annotation_db.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_projected_features">
<span class="sig-name descname"><span class="pre">get_projected_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_projected_features" title="Link to this definition">#</a></dt>
<dd><p>projects all features from other sequences onto seqid</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_seq">
<span class="sig-name descname"><span class="pre">get_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_seq" title="Link to this definition">#</a></dt>
<dd><p>Return a ungapped Sequence object for the specified seqname.</p>
<p>Note: always returns Sequence object, not ArraySequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_seq_indices">
<span class="sig-name descname"><span class="pre">get_seq_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_seq_indices" title="Link to this definition">#</a></dt>
<dd><p>Returns list of keys of seqs where f(row) is True.</p>
<p>List will be in the same order as self.names, if present.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_similar">
<span class="sig-name descname"><span class="pre">get_similar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_similarity=0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_similarity=1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric=&lt;cogent3.util.transform.for_seq</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_similar" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing sequences similar to target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>target</strong></dt><dd><p>sequence object to compare to. Can be in the alignment.</p>
</dd>
<dt><strong>min_similarity</strong></dt><dd><p>minimum similarity that will be kept. Default 0.0.</p>
</dd>
<dt><strong>max_similarity</strong></dt><dd><p>maximum similarity that will be kept. Default 1.0.
(Note that both min_similarity and max_similarity are inclusive.)
metric
similarity function to use. Must be f(first_seq, second_seq).</p>
</dd>
<dt><strong>The default metric is fraction similarity, ranging from 0.0 (0%</strong></dt><dd></dd>
<dt><strong>identical) to 1.0 (100% identical). The Sequence classes have lots</strong></dt><dd></dd>
<dt><strong>of methods that can be passed in as unbound methods to act as the</strong></dt><dd></dd>
<dt><strong>metric, e.g. frac_same_gaps.</strong></dt><dd></dd>
<dt><strong>transform</strong></dt><dd><p>transformation function to use on the sequences before
the metric is calculated. If None, uses the whole sequences in each
case. A frequent transformation is a function that returns a specified
range of a sequence, e.g. eliminating the ends. Note that the
transform applies to both the real sequence and the target sequence.</p>
</dd>
<dt><strong>WARNING: if the transformation changes the type of the sequence (e.g.</strong></dt><dd></dd>
<dt><strong>extracting a string from an RnaSequence object), distance metrics that</strong></dt><dd></dd>
<dt><strong>depend on instance data of the original class may fail.</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.get_translation">
<span class="sig-name descname"><span class="pre">get_translation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incomplete_ok</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.get_translation" title="Link to this definition">#</a></dt>
<dd><p>translate from nucleic acid to protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>genetic code, either the number or name
(use cogent3.core.genetic_code.available_codes)</p>
</dd>
<dt><strong>incomplete_ok</strong></dt><dd><p>codons that are mixes of nucleotide and gaps converted to ‘?’.
raises a ValueError if False</p>
</dd>
<dt><strong>include_stop</strong></dt><dd><p>whether to allow a stops in the translated sequence</p>
</dd>
<dt><strong>trim_stop</strong></dt><dd><p>exclude terminal stop codons if they exist</p>
</dd>
<dt><strong>kwargs</strong></dt><dd><p>related to construction of the resulting object</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>A new instance of self translated into protein</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.has_terminal_stop">
<span class="sig-name descname"><span class="pre">has_terminal_stop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.has_terminal_stop" title="Link to this definition">#</a></dt>
<dd><p>Returns True if any sequence has a terminal stop codon.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>strict</strong></dt><dd><p>If True, raises an exception if a seq length not divisible by 3</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.has_terminal_stops">
<span class="sig-name descname"><span class="pre">has_terminal_stops</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.has_terminal_stops" title="Link to this definition">#</a></dt>
<dd><p>deprecated</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.information_plot">
<span class="sig-name descname"><span class="pre">information_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.information_plot" title="Link to this definition">#</a></dt>
<dd><p>plot information per position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>width</strong><span class="classifier">int</span></dt><dd><p>figure width in pixels</p>
</dd>
<dt><strong>height</strong><span class="classifier">int</span></dt><dd><p>figure height in pixels</p>
</dd>
<dt><strong>window</strong><span class="classifier">int or None</span></dt><dd><p>used for smoothing line, defaults to sqrt(length)</p>
</dd>
<dt><strong>stat</strong><span class="classifier">str</span></dt><dd><p>‘mean’ or ‘median, used as the summary statistic for each window</p>
</dd>
<dt><strong>include_gap</strong></dt><dd><p>whether to include gap counts, shown on right y-axis</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.is_array">
<span class="sig-name descname"><span class="pre">is_array</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'array',</span> <span class="pre">'array_seqs'}</span></em><a class="headerlink" href="#cogent3.core.alignment.Alignment.is_array" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.is_ragged">
<span class="sig-name descname"><span class="pre">is_ragged</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.is_ragged" title="Link to this definition">#</a></dt>
<dd><p>Returns True if alignment has sequences of different lengths.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.iter_positions">
<span class="sig-name descname"><span class="pre">iter_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.iter_positions" title="Link to this definition">#</a></dt>
<dd><p>Iterates over positions in the alignment, in order.</p>
<p>pos_order refers to a list of indices (ints) specifying the column
order. This lets you rearrange positions if you want to (e.g. to pull
out individual codon positions).</p>
<p>Note that self.iter_positions() always returns new objects, by default
lists of elements. Use map(f, self.iter_positions) to apply the
constructor or function f to the resulting lists (f must take a single
list as a parameter).</p>
<p>Will raise IndexError if one of the indices in order exceeds the
sequence length. This will always happen on ragged alignments:
assign to self.seq_len to set all sequences to the same length.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.iter_selected">
<span class="sig-name descname"><span class="pre">iter_selected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.iter_selected" title="Link to this definition">#</a></dt>
<dd><p>Iterates over elements in the alignment.</p>
<p>seq_order (names) can be used to select a subset of seqs.
pos_order (positions) can be used to select a subset of positions.</p>
<p>Always iterates along a seq first, then down a position (transposes
normal order of a[i][j]; possibly, this should change)..</p>
<p>WARNING: Alignment.iter_selected() is not the same as alignment.iteritems()
(which is the built-in dict iteritems that iterates over key-value
pairs).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.iter_seqs">
<span class="sig-name descname"><span class="pre">iter_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.iter_seqs" title="Link to this definition">#</a></dt>
<dd><p>Iterates over values (sequences) in the alignment, in order.</p>
<p>seq_order: list of keys giving the order in which seqs will be returned.
Defaults to self.Names. Note that only these sequences will be
returned, and that KeyError will be raised if there are sequences
in order that have been deleted from the Alignment. If self.Names
is None, returns the sequences in the same order as
self.named_seqs.values().</p>
<p>Use map(f, self.seqs()) to apply the constructor f to each seq. f must
accept a single list as an argument.</p>
<p>Always returns references to the same objects that are values of the
alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.iupac_consensus">
<span class="sig-name descname"><span class="pre">iupac_consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.iupac_consensus" title="Link to this definition">#</a></dt>
<dd><p>Returns string containing IUPAC consensus sequence of the alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.majority_consensus">
<span class="sig-name descname"><span class="pre">majority_consensus</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.majority_consensus" title="Link to this definition">#</a></dt>
<dd><p>Returns list containing most frequent item at each position.</p>
<p>Optional parameter transform gives constructor for type to which result
will be converted (useful when consensus should be same type as
originals).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.make_feature">
<span class="sig-name descname"><span class="pre">make_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">FeatureDataType</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.make_feature" title="Link to this definition">#</a></dt>
<dd><p>create a feature on named sequence, or on the alignment itself</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>feature</strong></dt><dd><p>a dict with all the necessary data rto construct a feature</p>
</dd>
<dt><strong>on_alignment</strong></dt><dd><p>the feature is in alignment coordinates, incompatible with setting
‘seqid’. Set to True if ‘seqid’ not provided.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Feature</dt><dd></dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>ValueError if define a ‘seqid’ not on alignment or use ‘seqid’ and</dt><dd></dd>
<dt>on_alignment.</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To get a feature AND add it to annotation_db, use add_feature().</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.matching_ref">
<span class="sig-name descname"><span class="pre">matching_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_fraction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_run</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.matching_ref" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment with seqs well aligned with a reference.</p>
<dl class="simple">
<dt>gap_fraction = fraction of positions that either have a gap in the</dt><dd><p>template but not in the seq or in the seq but not in the template</p>
</dd>
<dt>gap_run = number of consecutive gaps tolerated in query relative to</dt><dd><p>sequence or sequence relative to query</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.moltype">
<span class="sig-name descname"><span class="pre">moltype</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">MolType(('a',</span> <span class="pre">'b',</span> <span class="pre">'c',</span> <span class="pre">'d',</span> <span class="pre">'e',</span> <span class="pre">'f',</span> <span class="pre">'g',</span> <span class="pre">'h',</span> <span class="pre">'i',</span> <span class="pre">'j',</span> <span class="pre">'k',</span> <span class="pre">'l',</span> <span class="pre">'m',</span> <span class="pre">'n',</span> <span class="pre">'o',</span> <span class="pre">'p',</span> <span class="pre">'q',</span> <span class="pre">'r',</span> <span class="pre">'s',</span> <span class="pre">'t',</span> <span class="pre">'u',</span> <span class="pre">'v',</span> <span class="pre">'w',</span> <span class="pre">'x',</span> <span class="pre">'y',</span> <span class="pre">'z',</span> <span class="pre">'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'J',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'O',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'X',</span> <span class="pre">'Y',</span> <span class="pre">'Z'))</span></em><a class="headerlink" href="#cogent3.core.alignment.Alignment.moltype" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.no_degenerates">
<span class="sig-name descname"><span class="pre">no_degenerates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.no_degenerates" title="Link to this definition">#</a></dt>
<dd><p>returns new alignment without degenerate characters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>sequences are segmented into units of this size</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>whether gaps are to be treated as a degenerate
character (default, most evolutionary modelling treats gaps as
N) or not.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.num_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_seqs</span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.num_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of sequences in the alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.omit_bad_seqs">
<span class="sig-name descname"><span class="pre">omit_bad_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.omit_bad_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment without sequences with a number of uniquely
introduced gaps exceeding quantile</p>
<p>Uses count_gaps_per_seq(unique=True) to obtain the counts of gaps
uniquely introduced by a sequence. The cutoff is the quantile of
this distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>quantile</strong><span class="classifier">float or None</span></dt><dd><p>sequences whose unique gap count is in a quantile larger than this
cutoff are excluded. The default quantile is (num_seqs - 1) / num_seqs</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.omit_gap_pos">
<span class="sig-name descname"><span class="pre">omit_gap_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">allowed_gap_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.999999</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.omit_gap_pos" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment where all cols (motifs) have &lt;= allowed_gap_frac gaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allowed_gap_frac</strong></dt><dd><p>specifies proportion of gaps is allowed in each
column (default is just &lt; 1, i.e. only cols with at least one gap
character are preserved). Set to 1 - e-6 to exclude strictly gapped
columns.</p>
</dd>
<dt><strong>motif_length</strong></dt><dd><p>set’s the “column” width, e.g. setting to 3
corresponds to codons. A motif that includes a gap at any position
is included in the counting. Default is 1.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.omit_gap_runs">
<span class="sig-name descname"><span class="pre">omit_gap_runs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">allowed_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.omit_gap_runs" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment where all seqs have runs of gaps &lt;=allowed_run.</p>
<p>Note that seqs with exactly allowed_run gaps are not deleted.
Default is for allowed_run to be 1 (i.e. no consecutive gaps allowed).</p>
<p>Because the test for whether the current gap run exceeds the maximum
allowed gap run is only triggered when there is at least one gap, even
negative values for allowed_run will still let sequences with no gaps
through.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.omit_gap_seqs">
<span class="sig-name descname"><span class="pre">omit_gap_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">allowed_gap_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.omit_gap_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment with seqs that have &lt;= allowed_gap_frac.</p>
<p>allowed_gap_frac should be a fraction between 0 and 1 inclusive.
Default is 0.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.pad_seqs">
<span class="sig-name descname"><span class="pre">pad_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pad_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.pad_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns copy in which sequences are padded to same length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pad_length</strong></dt><dd><p>Length all sequences are to be padded to.  Will pad
to max sequence length if pad_length is None or less than max
length.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.positions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">positions</span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.positions" title="Link to this definition">#</a></dt>
<dd><p>Iterates over positions in the alignment, in order.</p>
<p>pos_order refers to a list of indices (ints) specifying the column
order. This lets you rearrange positions if you want to (e.g. to pull
out individual codon positions).</p>
<p>Note that self.iter_positions() always returns new objects, by default
lists of elements. Use map(f, self.iter_positions) to apply the
constructor or function f to the resulting lists (f must take a single
list as a parameter).</p>
<p>Will raise IndexError if one of the indices in order exceeds the
sequence length. This will always happen on ragged alignments:
assign to self.seq_len to set all sequences to the same length.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.probs_per_pos">
<span class="sig-name descname"><span class="pre">probs_per_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.probs_per_pos" title="Link to this definition">#</a></dt>
<dd><p>returns MotifFreqsArray per position</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.probs_per_seq">
<span class="sig-name descname"><span class="pre">probs_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.probs_per_seq" title="Link to this definition">#</a></dt>
<dd><p>return MotifFreqsArray per sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of characters per tuple.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.quick_tree">
<span class="sig-name descname"><span class="pre">quick_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">calc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'percent'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bootstrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_invalid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ui</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.quick_tree" title="Link to this definition">#</a></dt>
<dd><p>Returns pairwise distances between sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>calc</strong><span class="classifier">str</span></dt><dd><p>a pairwise distance calculator or name of one. For options see
cogent3.evolve.fast_distance.available_distances</p>
</dd>
<dt><strong>show_progress</strong><span class="classifier">bool</span></dt><dd><p>controls progress display for distance calculation</p>
</dd>
<dt><strong>drop_invalid</strong><span class="classifier">bool</span></dt><dd><p>If True, sequences for which a pairwise distance could not be
calculated are excluded. If False, an ArithmeticError is raised if
a distance could not be computed on observed data.</p>
</dd>
<dt><strong>bootstrap</strong><span class="classifier">int or None</span></dt><dd><p>Number of non-parametric bootstrap replicates. Resamples alignment
columns with replacement and builds a phylogeny for each such
resampling.</p>
</dd>
<dt><strong>drop_invalid</strong><span class="classifier">bool</span></dt><dd><p>If True, sequences for which a pairwise distance could not be
calculated are excluded. If False, an ArithmeticError is raised if
a distance could not be computed on observed data.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>a phylogenetic tree. If bootstrap specified, returns the weighted</dt><dd></dd>
<dt>majority consensus. Support for each node is stored as</dt><dd></dd>
<dt>edge.params[‘params’].</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Sequences in the observed alignment for which distances could not be
computed are omitted. Bootstrap replicates are required to have
distances for all seqs present in the observed data distance matrix.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.rc" title="Link to this definition">#</a></dt>
<dd><p>Returns the reverse complement alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.rename_seqs">
<span class="sig-name descname"><span class="pre">rename_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">renamer</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.rename_seqs" title="Link to this definition">#</a></dt>
<dd><p>returns new instance with sequences renamed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>renamer</strong><span class="classifier">callable</span></dt><dd><p>function that will take current sequences and return the new one</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.replace_seqs">
<span class="sig-name descname"><span class="pre">replace_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aa_to_codon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.replace_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment with same shape but with data taken from seqs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>aa_to_codon</strong></dt><dd><p>If True (default) aligns codons from protein
alignment, or, more generally, substituting in codons from a set
of protein sequences (not necessarily aligned). For this reason,
it takes characters from seqs three at a time rather than one at
a time (i.e. 3 characters in seqs are put in place of 1 character
in self). If False, seqs must be the same lengths.</p>
</dd>
<dt><strong>If seqs is an alignment, any gaps in it will be ignored.</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Returns the reverse complement alignment. A synonym for rc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.sample">
<span class="sig-name descname"><span class="pre">sample</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_replacement=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_length=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">randint=&lt;built-in</span> <span class="pre">method</span> <span class="pre">randint</span> <span class="pre">of</span> <span class="pre">numpy.random.mtrand.RandomState</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">permutation=&lt;built-in</span> <span class="pre">method</span> <span class="pre">permutation</span> <span class="pre">of</span> <span class="pre">numpy.random.mtrand.RandomState</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.sample" title="Link to this definition">#</a></dt>
<dd><p>Returns random sample of positions from self, e.g. to bootstrap.</p>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl>
<dt><strong>n</strong></dt><dd><p>the number of positions to sample from the alignment.
Default is alignment length</p>
</dd>
<dt><strong>with_replacement</strong></dt><dd><p>boolean flag for determining if sampled
positions</p>
</dd>
<dt><strong>random_series</strong></dt><dd><p>a random number generator with
.randint(min,max) .random() methods</p>
</dd>
<dt><strong>Notes:</strong></dt><dd><p>By default (resampling all positions without replacement), generates
a permutation of the positions of the alignment.</p>
<p>Setting with_replacement to True and otherwise leaving parameters
as defaults generates a standard bootstrap resampling of the
alignment.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.seqlogo">
<span class="sig-name descname"><span class="pre">seqlogo</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">700</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vspace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.005</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colours</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.seqlogo" title="Link to this definition">#</a></dt>
<dd><p>returns Drawable sequence logo using mutual information</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>width, height</strong><span class="classifier">float</span></dt><dd><p>plot dimensions in pixels</p>
</dd>
<dt><strong>wrap</strong><span class="classifier">int</span></dt><dd><p>number of alignment columns per row</p>
</dd>
<dt><strong>vspace</strong><span class="classifier">float</span></dt><dd><p>vertical separation between rows, as a proportion of total plot</p>
</dd>
<dt><strong>colours</strong><span class="classifier">dict</span></dt><dd><p>mapping of characters to colours. If note provided, defaults to
custom for everything ecept protein, which uses protein moltype
colours.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Computes MI based on log2 and includes the gap state, so the maximum
possible value is -log2(1/num_states)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqs</span></span><a class="headerlink" href="#cogent3.core.alignment.Alignment.seqs" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.set_repr_policy">
<span class="sig-name descname"><span class="pre">set_repr_policy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.set_repr_policy" title="Link to this definition">#</a></dt>
<dd><p>specify policy for repr(self)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>num_seqs</strong><span class="classifier">int or None</span></dt><dd><p>number of sequences to include in represented display.</p>
</dd>
<dt><strong>num_pos</strong><span class="classifier">int or None</span></dt><dd><p>length of sequences to include in represented display.</p>
</dd>
<dt><strong>ref_name</strong><span class="classifier">str or None</span></dt><dd><p>name of sequence to be placed first, or “longest” (default).
If latter, indicates longest sequence will be chosen.</p>
</dd>
<dt><strong>wrap</strong><span class="classifier">int or None</span></dt><dd><p>number of printed bases per row</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.sliding_windows">
<span class="sig-name descname"><span class="pre">sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.sliding_windows" title="Link to this definition">#</a></dt>
<dd><p>Generator yielding new alignments of given length and interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>window</strong></dt><dd><p>The length of each returned alignment.</p>
</dd>
<dt><strong>step</strong></dt><dd><p>The interval between the start of the successive
alignment objects returned.</p>
</dd>
<dt><strong>start</strong></dt><dd><p>first window start position</p>
</dd>
<dt><strong>end</strong></dt><dd><p>last window start position</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.strand_symmetry">
<span class="sig-name descname"><span class="pre">strand_symmetry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.strand_symmetry" title="Link to this definition">#</a></dt>
<dd><p>returns dict of strand symmetry test results per seq</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.take_positions">
<span class="sig-name descname"><span class="pre">take_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cols</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.take_positions" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing only specified positions.</p>
<p>By default, the seqs will be lists, but an alternative constructor
can be specified.</p>
<p>Note that take_positions will fail on ragged positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.take_positions_if">
<span class="sig-name descname"><span class="pre">take_positions_if</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.take_positions_if" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing cols where f(col) is True.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.take_seqs">
<span class="sig-name descname"><span class="pre">take_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.take_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing only specified seqs.</p>
<p>Note that the seqs in the new alignment will be references to the
same objects as the seqs in the old alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.take_seqs_if">
<span class="sig-name descname"><span class="pre">take_seqs_if</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.take_seqs_if" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing seqs where f(row) is True.</p>
<p>Note that the seqs in the new Alignment are the same objects as the
seqs in the old Alignment, not copies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_dict">
<span class="sig-name descname"><span class="pre">to_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_dict" title="Link to this definition">#</a></dt>
<dd><p>Returns the alignment as dict of names -&gt; strings.</p>
<p>Note: returns strings, NOT Sequence objects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_dna">
<span class="sig-name descname"><span class="pre">to_dna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_dna" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self as an alignment of DNA moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_fasta">
<span class="sig-name descname"><span class="pre">to_fasta</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_fasta" title="Link to this definition">#</a></dt>
<dd><p>Return alignment in Fasta format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>make_seqlabel</strong></dt><dd><p>callback function that takes the seq object and
returns a label str</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_html">
<span class="sig-name descname"><span class="pre">to_html</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'longest'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_family</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Lucida</span> <span class="pre">Console'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_html" title="Link to this definition">#</a></dt>
<dd><p>returns html with embedded styles for sequence colouring</p>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl>
<dt><strong>name_order</strong></dt><dd><p>order of names for display.</p>
</dd>
<dt><strong>wrap</strong></dt><dd><p>number of alignment columns per row</p>
</dd>
<dt><strong>limit</strong></dt><dd><p>truncate alignment to this length</p>
</dd>
<dt><strong>ref_name</strong></dt><dd><p>Name of an existing sequence or ‘longest’. If the latter, the
longest sequence (excluding gaps and ambiguities) is selected as the
reference.</p>
</dd>
<dt><strong>colors</strong></dt><dd><p>{character
moltype.</p>
</dd>
<dt><strong>font_size</strong></dt><dd><p>in points. Affects labels and sequence and line spacing
(proportional to value)</p>
</dd>
<dt><strong>font_family</strong></dt><dd><p>string denoting font family</p>
</dd>
<dt><strong>To display in jupyter notebook:</strong></dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">IPython.core.display</span> <span class="kn">import</span> <span class="n">HTML</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HTML</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">to_html</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_json">
<span class="sig-name descname"><span class="pre">to_json</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_json" title="Link to this definition">#</a></dt>
<dd><p>returns json formatted string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_moltype">
<span class="sig-name descname"><span class="pre">to_moltype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">moltype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_moltype" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self with moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_nexus">
<span class="sig-name descname"><span class="pre">to_nexus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_nexus" title="Link to this definition">#</a></dt>
<dd><p>Return alignment in NEXUS format and mapping to sequence ids</p>
<dl class="simple">
<dt><strong>NOTE</strong> Not that every sequence in the alignment MUST come from</dt><dd><p>a different species!! (You can concatenate multiple sequences from
same species together before building tree)</p>
</dd>
</dl>
<p>seq_type: dna, rna, or protein</p>
<p>Raises exception if invalid alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_phylip">
<span class="sig-name descname"><span class="pre">to_phylip</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_phylip" title="Link to this definition">#</a></dt>
<dd><p>Return alignment in PHYLIP format and mapping to sequence ids</p>
<p>raises exception if invalid alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_pretty">
<span class="sig-name descname"><span class="pre">to_pretty</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_pretty" title="Link to this definition">#</a></dt>
<dd><p>returns a string representation of the alignment in pretty print format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name_order</strong></dt><dd><p>order of names for display.</p>
</dd>
<dt><strong>wrap</strong></dt><dd><p>maximum number of printed bases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_protein">
<span class="sig-name descname"><span class="pre">to_protein</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_protein" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self as an alignment of PROTEIN moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_rich_dict">
<span class="sig-name descname"><span class="pre">to_rich_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_rich_dict" title="Link to this definition">#</a></dt>
<dd><p>returns detailed content including info and moltype attributes</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_rna">
<span class="sig-name descname"><span class="pre">to_rna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_rna" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self as an alignment of RNA moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.to_type">
<span class="sig-name descname"><span class="pre">to_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array_align</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">moltype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.to_type" title="Link to this definition">#</a></dt>
<dd><p>returns alignment of type indicated by array_align</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>array_align: bool</strong></dt><dd><p>if True, returns as ArrayAlignment. Otherwise as “standard”
Alignment class. Conversion to ArrayAlignment loses annotations.</p>
</dd>
<dt><strong>moltype</strong><span class="classifier">MolType instance</span></dt><dd><p>overrides self.moltype</p>
</dd>
<dt><strong>alphabet</strong><span class="classifier">Alphabet instance</span></dt><dd><p>overrides self.alphabet</p>
</dd>
<dt><strong>If array_align would result in no change (class is same as self),</strong></dt><dd></dd>
<dt><strong>returns self</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.trim_stop_codons">
<span class="sig-name descname"><span class="pre">trim_stop_codons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.trim_stop_codons" title="Link to this definition">#</a></dt>
<dd><p>Removes any terminal stop codons from the sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>strict</strong></dt><dd><p>If True, raises an exception if a seq length not divisible by 3</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.variable_positions">
<span class="sig-name descname"><span class="pre">variable_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_gap_motif</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.variable_positions" title="Link to this definition">#</a></dt>
<dd><p>Return a list of variable position indexes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_gap_motif</strong></dt><dd><p>if False, sequences with a gap motif in a
column are ignored.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.with_gaps_from">
<span class="sig-name descname"><span class="pre">with_gaps_from</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">template</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.with_gaps_from" title="Link to this definition">#</a></dt>
<dd><p>Same alignment but overwritten with the gaps from ‘template’</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.with_masked_annotations">
<span class="sig-name descname"><span class="pre">with_masked_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">biotypes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shadow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.with_masked_annotations" title="Link to this definition">#</a></dt>
<dd><p>returns an alignment with annot_types regions replaced by mask_char
if shadow is False, otherwise all other regions are masked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotypes</strong></dt><dd><p>annotation type(s)</p>
</dd>
<dt><strong>mask_char</strong></dt><dd><p>must be a character valid for the seq moltype. The
default value is the most ambiguous character, eg. ‘?’ for DNA</p>
</dd>
<dt><strong>shadow</strong></dt><dd><p>whether to mask the annotated regions, or everything but
the annotated regions</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.with_modified_termini">
<span class="sig-name descname"><span class="pre">with_modified_termini</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.with_modified_termini" title="Link to this definition">#</a></dt>
<dd><p>Changes the termini to include termini char instead of gapmotif.</p>
<p>Useful to correct the standard gap char output by most
alignment programs when aligned sequences have different ends.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.Alignment.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.Alignment.write" title="Link to this definition">#</a></dt>
<dd><p>Write the alignment to a file, preserving order of sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>filename</strong></dt><dd><p>name of the sequence file</p>
</dd>
<dt><strong>format</strong></dt><dd><p>format of the sequence file</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If format is None, will attempt to infer format from the filename
suffix.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="cogent3.core.alignment.SequenceCollection.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">SequenceCollection</p>
      </div>
    </a>
    <a class="right-next"
       href="cogent3.core.alignment.ArrayAlignment.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">ArrayAlignment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">

  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment"><code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.add_feature"><code class="docutils literal notranslate"><span class="pre">Alignment.add_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.add_from_ref_aln"><code class="docutils literal notranslate"><span class="pre">Alignment.add_from_ref_aln()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.add_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.add_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.alignment_quality"><code class="docutils literal notranslate"><span class="pre">Alignment.alignment_quality()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.annotate_from_gff"><code class="docutils literal notranslate"><span class="pre">Alignment.annotate_from_gff()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.annotation_db"><code class="docutils literal notranslate"><span class="pre">Alignment.annotation_db</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.apply_pssm"><code class="docutils literal notranslate"><span class="pre">Alignment.apply_pssm()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.coevolution"><code class="docutils literal notranslate"><span class="pre">Alignment.coevolution()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.copy"><code class="docutils literal notranslate"><span class="pre">Alignment.copy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.copy_annotations"><code class="docutils literal notranslate"><span class="pre">Alignment.copy_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.count_gaps_per_pos"><code class="docutils literal notranslate"><span class="pre">Alignment.count_gaps_per_pos()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.count_gaps_per_seq"><code class="docutils literal notranslate"><span class="pre">Alignment.count_gaps_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.counts"><code class="docutils literal notranslate"><span class="pre">Alignment.counts()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.counts_per_pos"><code class="docutils literal notranslate"><span class="pre">Alignment.counts_per_pos()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.counts_per_seq"><code class="docutils literal notranslate"><span class="pre">Alignment.counts_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.deepcopy"><code class="docutils literal notranslate"><span class="pre">Alignment.deepcopy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.default_gap"><code class="docutils literal notranslate"><span class="pre">Alignment.default_gap</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.degap"><code class="docutils literal notranslate"><span class="pre">Alignment.degap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.distance_matrix"><code class="docutils literal notranslate"><span class="pre">Alignment.distance_matrix()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.dotplot"><code class="docutils literal notranslate"><span class="pre">Alignment.dotplot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.entropy_per_pos"><code class="docutils literal notranslate"><span class="pre">Alignment.entropy_per_pos()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.entropy_per_seq"><code class="docutils literal notranslate"><span class="pre">Alignment.entropy_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.filtered"><code class="docutils literal notranslate"><span class="pre">Alignment.filtered()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.gap_chars"><code class="docutils literal notranslate"><span class="pre">Alignment.gap_chars</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.gapped_by_map"><code class="docutils literal notranslate"><span class="pre">Alignment.gapped_by_map()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_ambiguous_positions"><code class="docutils literal notranslate"><span class="pre">Alignment.get_ambiguous_positions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_degapped_relative_to"><code class="docutils literal notranslate"><span class="pre">Alignment.get_degapped_relative_to()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_drawable"><code class="docutils literal notranslate"><span class="pre">Alignment.get_drawable()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_drawables"><code class="docutils literal notranslate"><span class="pre">Alignment.get_drawables()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_features"><code class="docutils literal notranslate"><span class="pre">Alignment.get_features()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_gap_array"><code class="docutils literal notranslate"><span class="pre">Alignment.get_gap_array()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_gapped_seq"><code class="docutils literal notranslate"><span class="pre">Alignment.get_gapped_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_identical_sets"><code class="docutils literal notranslate"><span class="pre">Alignment.get_identical_sets()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_lengths"><code class="docutils literal notranslate"><span class="pre">Alignment.get_lengths()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_motif_probs"><code class="docutils literal notranslate"><span class="pre">Alignment.get_motif_probs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_position_indices"><code class="docutils literal notranslate"><span class="pre">Alignment.get_position_indices()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_projected_feature"><code class="docutils literal notranslate"><span class="pre">Alignment.get_projected_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_projected_features"><code class="docutils literal notranslate"><span class="pre">Alignment.get_projected_features()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_seq"><code class="docutils literal notranslate"><span class="pre">Alignment.get_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_seq_indices"><code class="docutils literal notranslate"><span class="pre">Alignment.get_seq_indices()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_similar"><code class="docutils literal notranslate"><span class="pre">Alignment.get_similar()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.get_translation"><code class="docutils literal notranslate"><span class="pre">Alignment.get_translation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.has_terminal_stop"><code class="docutils literal notranslate"><span class="pre">Alignment.has_terminal_stop()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.has_terminal_stops"><code class="docutils literal notranslate"><span class="pre">Alignment.has_terminal_stops()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.information_plot"><code class="docutils literal notranslate"><span class="pre">Alignment.information_plot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.is_array"><code class="docutils literal notranslate"><span class="pre">Alignment.is_array</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.is_ragged"><code class="docutils literal notranslate"><span class="pre">Alignment.is_ragged()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.iter_positions"><code class="docutils literal notranslate"><span class="pre">Alignment.iter_positions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.iter_selected"><code class="docutils literal notranslate"><span class="pre">Alignment.iter_selected()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.iter_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.iter_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.iupac_consensus"><code class="docutils literal notranslate"><span class="pre">Alignment.iupac_consensus()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.majority_consensus"><code class="docutils literal notranslate"><span class="pre">Alignment.majority_consensus()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.make_feature"><code class="docutils literal notranslate"><span class="pre">Alignment.make_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.matching_ref"><code class="docutils literal notranslate"><span class="pre">Alignment.matching_ref()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.moltype"><code class="docutils literal notranslate"><span class="pre">Alignment.moltype</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.no_degenerates"><code class="docutils literal notranslate"><span class="pre">Alignment.no_degenerates()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.num_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.num_seqs</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.omit_bad_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.omit_bad_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.omit_gap_pos"><code class="docutils literal notranslate"><span class="pre">Alignment.omit_gap_pos()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.omit_gap_runs"><code class="docutils literal notranslate"><span class="pre">Alignment.omit_gap_runs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.omit_gap_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.omit_gap_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.pad_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.pad_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.positions"><code class="docutils literal notranslate"><span class="pre">Alignment.positions</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.probs_per_pos"><code class="docutils literal notranslate"><span class="pre">Alignment.probs_per_pos()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.probs_per_seq"><code class="docutils literal notranslate"><span class="pre">Alignment.probs_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.quick_tree"><code class="docutils literal notranslate"><span class="pre">Alignment.quick_tree()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.rc"><code class="docutils literal notranslate"><span class="pre">Alignment.rc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.rename_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.rename_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.replace_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.replace_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.reverse_complement"><code class="docutils literal notranslate"><span class="pre">Alignment.reverse_complement()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.sample"><code class="docutils literal notranslate"><span class="pre">Alignment.sample()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.seqlogo"><code class="docutils literal notranslate"><span class="pre">Alignment.seqlogo()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.seqs</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.set_repr_policy"><code class="docutils literal notranslate"><span class="pre">Alignment.set_repr_policy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.sliding_windows"><code class="docutils literal notranslate"><span class="pre">Alignment.sliding_windows()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.strand_symmetry"><code class="docutils literal notranslate"><span class="pre">Alignment.strand_symmetry()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.take_positions"><code class="docutils literal notranslate"><span class="pre">Alignment.take_positions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.take_positions_if"><code class="docutils literal notranslate"><span class="pre">Alignment.take_positions_if()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.take_seqs"><code class="docutils literal notranslate"><span class="pre">Alignment.take_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.take_seqs_if"><code class="docutils literal notranslate"><span class="pre">Alignment.take_seqs_if()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_dict"><code class="docutils literal notranslate"><span class="pre">Alignment.to_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_dna"><code class="docutils literal notranslate"><span class="pre">Alignment.to_dna()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_fasta"><code class="docutils literal notranslate"><span class="pre">Alignment.to_fasta()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_html"><code class="docutils literal notranslate"><span class="pre">Alignment.to_html()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_json"><code class="docutils literal notranslate"><span class="pre">Alignment.to_json()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_moltype"><code class="docutils literal notranslate"><span class="pre">Alignment.to_moltype()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_nexus"><code class="docutils literal notranslate"><span class="pre">Alignment.to_nexus()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_phylip"><code class="docutils literal notranslate"><span class="pre">Alignment.to_phylip()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_pretty"><code class="docutils literal notranslate"><span class="pre">Alignment.to_pretty()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_protein"><code class="docutils literal notranslate"><span class="pre">Alignment.to_protein()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_rich_dict"><code class="docutils literal notranslate"><span class="pre">Alignment.to_rich_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_rna"><code class="docutils literal notranslate"><span class="pre">Alignment.to_rna()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.to_type"><code class="docutils literal notranslate"><span class="pre">Alignment.to_type()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.trim_stop_codons"><code class="docutils literal notranslate"><span class="pre">Alignment.trim_stop_codons()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.variable_positions"><code class="docutils literal notranslate"><span class="pre">Alignment.variable_positions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.with_gaps_from"><code class="docutils literal notranslate"><span class="pre">Alignment.with_gaps_from()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.with_masked_annotations"><code class="docutils literal notranslate"><span class="pre">Alignment.with_masked_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.with_modified_termini"><code class="docutils literal notranslate"><span class="pre">Alignment.with_modified_termini()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.Alignment.write"><code class="docutils literal notranslate"><span class="pre">Alignment.write()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020-2023, cogent3 Team.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.14.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>