
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>SequenceCollection &#8212; cogent3 2025.01.17 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/custom.css?v=da3cec82" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../../_static/documentation_options.js?v=21afde67"></script>
    <script src="../../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../../_static/thebelab-helper.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../../../_static/design-tabs.js?v=f930bc37"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'doc/api/alignment/classes/cogent3.core.alignment.SequenceCollection';</script>
    <link rel="icon" href="../../../../_static/c3-square.svg"/>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="Alignment" href="cogent3.core.alignment.Alignment.html" />
    <link rel="prev" title="alignment" href="../alignment.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../../_static/c3-logo.png" class="logo__image only-light" alt="cogent3 2025.01.17 documentation - Home"/>
    <img src="../../../../_static/c3-logo.png" class="logo__image only-dark pst-js-only" alt="cogent3 2025.01.17 documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../draw/index.html">
    Image Gallery
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../../../index.html">
    User Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../community.html">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../licenses.html">
    License
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../../history.html">
    Project History
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../pycogent.html">
    How does cogent3 relate to PyCogent?
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../../projects.html">
    Projects using cogent3
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../data_file_links.html">
    Data Files Used in the Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../../../genindex.html">
    Index
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../draw/index.html">
    Image Gallery
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../../../index.html">
    User Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../community.html">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../licenses.html">
    License
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../../history.html">
    Project History
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../pycogent.html">
    How does cogent3 relate to PyCogent?
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../../projects.html">
    Projects using cogent3
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../data_file_links.html">
    Data Files Used in the Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../../../genindex.html">
    Index
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../app/index.html">apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-app.html">Overview of apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app_overview/app-overview.html">Overview of using apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app_overview/app-help.html">Getting help on an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app_overview/app-get.html">Getting an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app_overview/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app_overview/dstore.html">Data stores – collections of data records</a></li>

<li class="toctree-l3"><a class="reference internal" href="../../../app/app_overview/not-completed.html">Tracking records that could not be processed</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-cookbook.html">App Cookbook</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_cookbook/index-loaders.html">Loading data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/load-aligned.html">Loading aligned sequence data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/load-unaligned.html">Loading unaligned sequence data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/load-tabular.html">Loading tabular data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/load-json.html">Loading JSON serialised objects</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_cookbook/index-align.html">Sequence Alignment</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/align-nucleotide.html">Using a nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/align-codon.html">Using a codon model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/align-protein.html">Using a protein model</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_cookbook/index-sample.html">Filtering data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-omit-degen.html">Removing degenerate characters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-omit-gap.html">Removing highly gapped positions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-omit-bad.html">Remove problem sequences from an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-omit-duplicated.html">Remove duplicated sequences from an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-min-len.html">Filter sequence collections and alignments by length</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-fixed-len.html">Sample an alignment to a fixed length</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-trim-stops.html">Remove trailing stop codons from sequences in an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-take-pos.html">Sample nucleotides from a given codon position</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-take-seqs.html">Select named sequences from a collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-take-n-seqs.html">Select <cite>n</cite> sequences from a collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/sample-concat.html">Concatenating alignments</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_cookbook/index-writers.html">Writing Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/write-seqs.html">Writing sequences and sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/write-tabular.html">Writing tabular data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/write-json.html">Writing JSON Serialised Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_cookbook/write-db.html">Writing a database to file</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-tutorials.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_tutorials/index-evo-model.html">Modelling sequence evolution</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-ancestral-states.html">Reconstructing ancestral states</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-tr-codon-model.html">Applying a time-reversible codon model</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_tutorials/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../app/app_tutorials/index-evo-natsel.html">Testing for natural selection</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app_tutorials/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-custom.html">Custom Apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/user_function.html">Why write apps?</a></li>

</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html">Manipulating biological data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/loading_sequences.html">Loading a sequence from a file</a></li>





<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html">Sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/protein_sequences.html">Protein sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/alignments.html">Sequence Collections and Alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/features.html">Features</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/annotation_db.html">Annotation Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/moltypes.html">Molecular types</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/alphabet.html">Alphabets</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/what_codes.html">Available genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/genetic_code.html">Using genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/simple_trees.html">Trees</a></li>
</ul>
</details></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/tables.html">Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/building_alignments.html">Building alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html">Genetic distance calculation</a></li>


<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/building_phylogenies.html">Building phylogenies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/useful_utilities.html">Useful Utilities</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../examples/index.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../../examples/period_estimation.html">Estimating periodic signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/simple.html">The simplest script</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/codon_models.html">Using codon models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/nonstationary_inference.html">Inferring Amino Acid frequencies with a nonstationary model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/parallel.html">Parallel computations</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../../index.html">API</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.open_.html">open</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_model.html">get_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
<li class="toctree-l2 current active has-children"><a class="reference internal" href="../alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l3 current active"><a class="current reference internal" href="#">SequenceCollection</a></li>
<li class="toctree-l3"><a class="reference internal" href="cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../new_alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">new_alignment</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.SeqsData.html">SeqsData</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.SeqDataView.html">SeqDataView</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.AlignedSeqsData.html">AlignedSeqsData</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.AlignedDataView.html">AlignedDataView</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.Aligned.html">Aligned</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alignment/classes/cogent3.core.new_alignment.Alignment.html">Alignment</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../new_sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">new_sequence</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../new_sequence/classes/cogent3.core.new_sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_sequence/classes/cogent3.core.new_sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_sequence/classes/cogent3.core.new_sequence.ProteinSequence.html">ProteinSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_sequence/classes/cogent3.core.new_sequence.SeqView.html">SeqView</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_sequence/classes/cogent3.core.new_sequence.SliceRecord.html">SliceRecord</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../new_genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">new_genetic_code</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../new_genetic_code/classes/cogent3.core.new_genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../new_moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">new_moltype</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../new_moltype/classes/cogent3.core.new_moltype.MolType.html">MolType</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../new_alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">new_alphabet</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../new_alphabet/classes/cogent3.core.new_alphabet.CharAlphabet.html">CharAlphabet</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../new_alphabet/classes/cogent3.core.new_alphabet.KmerAlphabet.html">KmerAlphabet</a></li>
</ul>
</details></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/composable/define_app.html">define_app</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">User Guides</a></li>
    
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">API</a></li>
    
    
    <li class="breadcrumb-item"><a href="../alignment.html" class="nav-link"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">SequenceCollection</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="sequencecollection">
<h1>SequenceCollection<a class="headerlink" href="#sequencecollection" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">SequenceCollection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection" title="Link to this definition">#</a></dt>
<dd><p>Container for unaligned sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_db</strong></dt><dd></dd>
<dt><strong>named_seqs</strong></dt><dd></dd>
<dt><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.num_seqs" title="cogent3.core.alignment.SequenceCollection.num_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_seqs</span></code></a></dt><dd><p>Returns the number of sequences in the alignment.</p>
</dd>
<dt><strong>seqs</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.add_feature" title="cogent3.core.alignment.SequenceCollection.add_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_feature</span></code></a>(*, seqid, biotype, name, spans)</p></td>
<td><p>add feature on named sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.add_seqs" title="cogent3.core.alignment.SequenceCollection.add_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_seqs</span></code></a>(other[, before_name, after_name])</p></td>
<td><p>Returns new object of class self with sequences from other added.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.annotate_from_gff" title="cogent3.core.alignment.SequenceCollection.annotate_from_gff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_from_gff</span></code></a>(f[, seq_ids])</p></td>
<td><p>copies annotations from a gff file to a sequence in self</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.apply_pssm" title="cogent3.core.alignment.SequenceCollection.apply_pssm"><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply_pssm</span></code></a>([pssm, path, background, ...])</p></td>
<td><p>scores sequences using the specified pssm</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.copy" title="cogent3.core.alignment.SequenceCollection.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>()</p></td>
<td><p>Returns deep copy of self.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.copy_annotations" title="cogent3.core.alignment.SequenceCollection.copy_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_annotations</span></code></a>(seq_db)</p></td>
<td><p>copy annotations into attached annotation db</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.counts" title="cogent3.core.alignment.SequenceCollection.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>([motif_length, include_ambiguity, ...])</p></td>
<td><p>counts of motifs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.counts_per_seq" title="cogent3.core.alignment.SequenceCollection.counts_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts_per_seq</span></code></a>([motif_length, ...])</p></td>
<td><p>counts of motifs per sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.deepcopy" title="cogent3.core.alignment.SequenceCollection.deepcopy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deepcopy</span></code></a>([sliced])</p></td>
<td><p>returns deep copy of self.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.degap" title="cogent3.core.alignment.SequenceCollection.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>(**kwargs)</p></td>
<td><p>Returns copy in which sequences have no gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.distance_matrix" title="cogent3.core.alignment.SequenceCollection.distance_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance_matrix</span></code></a>([calc])</p></td>
<td><p>Estimated pairwise distance between sequences</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.dotplot" title="cogent3.core.alignment.SequenceCollection.dotplot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">dotplot</span></code></a>([name1, name2, window, threshold, ...])</p></td>
<td><p>make a dotplot between specified sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.entropy_per_seq" title="cogent3.core.alignment.SequenceCollection.entropy_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">entropy_per_seq</span></code></a>([motif_length, ...])</p></td>
<td><p>Returns the Shannon entropy per sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_ambiguous_positions" title="cogent3.core.alignment.SequenceCollection.get_ambiguous_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_ambiguous_positions</span></code></a>()</p></td>
<td><p>Returns dict of seq:{position:char} for ambiguous chars.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_features" title="cogent3.core.alignment.SequenceCollection.get_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_features</span></code></a>(*[, seqid, biotype, name, ...])</p></td>
<td><p>yields Feature instances</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_identical_sets" title="cogent3.core.alignment.SequenceCollection.get_identical_sets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_identical_sets</span></code></a>([mask_degen])</p></td>
<td><p>returns sets of names for sequences that are identical</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_lengths" title="cogent3.core.alignment.SequenceCollection.get_lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_lengths</span></code></a>([include_ambiguity, allow_gap])</p></td>
<td><p>returns {name: seq length, ...}</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_motif_probs" title="cogent3.core.alignment.SequenceCollection.get_motif_probs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_motif_probs</span></code></a>([alphabet, ...])</p></td>
<td><p>Return a dictionary of motif probs, calculated as the averaged frequency across sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_seq" title="cogent3.core.alignment.SequenceCollection.get_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq</span></code></a>(seqname)</p></td>
<td><p>Return a sequence object for the specified seqname.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_seq_indices" title="cogent3.core.alignment.SequenceCollection.get_seq_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_seq_indices</span></code></a>(f[, negate])</p></td>
<td><p>Returns list of keys of seqs where f(row) is True.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_similar" title="cogent3.core.alignment.SequenceCollection.get_similar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_similar</span></code></a>(target[, min_similarity, ...])</p></td>
<td><p>Returns new Alignment containing sequences similar to target.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.get_translation" title="cogent3.core.alignment.SequenceCollection.get_translation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_translation</span></code></a>([gc, incomplete_ok, ...])</p></td>
<td><p>translate from nucleic acid to protein</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.has_terminal_stop" title="cogent3.core.alignment.SequenceCollection.has_terminal_stop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_terminal_stop</span></code></a>([gc, strict])</p></td>
<td><p>Returns True if any sequence has a terminal stop codon.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.is_ragged" title="cogent3.core.alignment.SequenceCollection.is_ragged"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_ragged</span></code></a>()</p></td>
<td><p>Returns True if alignment has sequences of different lengths.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.iter_selected" title="cogent3.core.alignment.SequenceCollection.iter_selected"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_selected</span></code></a>([seq_order, pos_order])</p></td>
<td><p>Iterates over elements in the alignment.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.iter_seqs" title="cogent3.core.alignment.SequenceCollection.iter_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_seqs</span></code></a>([seq_order])</p></td>
<td><p>Iterates over values (sequences) in the alignment, in order.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.make_feature" title="cogent3.core.alignment.SequenceCollection.make_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_feature</span></code></a>(*, feature)</p></td>
<td><p>create a feature on named sequence, or on the alignment itself</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.omit_gap_runs" title="cogent3.core.alignment.SequenceCollection.omit_gap_runs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_runs</span></code></a>([allowed_run])</p></td>
<td><p>Returns new alignment where all seqs have runs of gaps &lt;=allowed_run.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.omit_gap_seqs" title="cogent3.core.alignment.SequenceCollection.omit_gap_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code></a>([allowed_gap_frac])</p></td>
<td><p>Returns new alignment with seqs that have &lt;= allowed_gap_frac.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.pad_seqs" title="cogent3.core.alignment.SequenceCollection.pad_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pad_seqs</span></code></a>([pad_length])</p></td>
<td><p>Returns copy in which sequences are padded to same length.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.probs_per_seq" title="cogent3.core.alignment.SequenceCollection.probs_per_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">probs_per_seq</span></code></a>([motif_length, ...])</p></td>
<td><p>return MotifFreqsArray per sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.rc" title="cogent3.core.alignment.SequenceCollection.rc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rc</span></code></a>()</p></td>
<td><p>Returns the reverse complement alignment</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.rename_seqs" title="cogent3.core.alignment.SequenceCollection.rename_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rename_seqs</span></code></a>(renamer)</p></td>
<td><p>returns new instance with sequences renamed</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.reverse_complement" title="cogent3.core.alignment.SequenceCollection.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_complement</span></code></a>()</p></td>
<td><p>Returns the reverse complement alignment.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.set_repr_policy" title="cogent3.core.alignment.SequenceCollection.set_repr_policy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_repr_policy</span></code></a>([num_seqs, num_pos, ...])</p></td>
<td><p>specify policy for repr(self)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.strand_symmetry" title="cogent3.core.alignment.SequenceCollection.strand_symmetry"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strand_symmetry</span></code></a>([motif_length])</p></td>
<td><p>returns dict of strand symmetry test results per seq</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.take_seqs" title="cogent3.core.alignment.SequenceCollection.take_seqs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_seqs</span></code></a>(seqs[, negate])</p></td>
<td><p>Returns new Alignment containing only specified seqs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.take_seqs_if" title="cogent3.core.alignment.SequenceCollection.take_seqs_if"><code class="xref py py-obj docutils literal notranslate"><span class="pre">take_seqs_if</span></code></a>(f[, negate])</p></td>
<td><p>Returns new Alignment containing seqs where f(row) is True.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_dict" title="cogent3.core.alignment.SequenceCollection.to_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dict</span></code></a>()</p></td>
<td><p>Returns the alignment as a dict of sequence names -&gt; strings.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_dna" title="cogent3.core.alignment.SequenceCollection.to_dna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dna</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of DNA moltype seqs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_fasta" title="cogent3.core.alignment.SequenceCollection.to_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_fasta</span></code></a>([block_size])</p></td>
<td><p>Return alignment in Fasta format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_html" title="cogent3.core.alignment.SequenceCollection.to_html"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_html</span></code></a>([name_order, wrap, limit, colors, ...])</p></td>
<td><p>returns html with embedded styles for sequence colouring</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_json" title="cogent3.core.alignment.SequenceCollection.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns json formatted string</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_moltype" title="cogent3.core.alignment.SequenceCollection.to_moltype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_moltype</span></code></a>(moltype)</p></td>
<td><p>returns copy of self with moltype seqs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_nexus" title="cogent3.core.alignment.SequenceCollection.to_nexus"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_nexus</span></code></a>(seq_type[, wrap])</p></td>
<td><p>Return alignment in NEXUS format and mapping to sequence ids</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_phylip" title="cogent3.core.alignment.SequenceCollection.to_phylip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_phylip</span></code></a>()</p></td>
<td><p>Return alignment in PHYLIP format and mapping to sequence ids</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_protein" title="cogent3.core.alignment.SequenceCollection.to_protein"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_protein</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of PROTEIN moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_rich_dict" title="cogent3.core.alignment.SequenceCollection.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>()</p></td>
<td><p>returns detailed content including info and moltype attributes</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.to_rna" title="cogent3.core.alignment.SequenceCollection.to_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rna</span></code></a>()</p></td>
<td><p>returns copy of self as an alignment of RNA moltype seqs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.trim_stop_codons" title="cogent3.core.alignment.SequenceCollection.trim_stop_codons"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim_stop_codons</span></code></a>([gc, strict])</p></td>
<td><p>Removes any terminal stop codons from the sequences</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.with_modified_termini" title="cogent3.core.alignment.SequenceCollection.with_modified_termini"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_modified_termini</span></code></a>()</p></td>
<td><p>Changes the termini to include termini char instead of gapmotif.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.alignment.SequenceCollection.write" title="cogent3.core.alignment.SequenceCollection.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>([filename, format])</p></td>
<td><p>Write the alignment to a file, preserving order of sequences.</p></td>
</tr>
</tbody>
</table>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spans</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'+'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.add_feature" title="Link to this definition">#</a></dt>
<dd><p>add feature on named sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seqid</strong></dt><dd><p>seq name to associate with</p>
</dd>
<dt><strong>parent_id</strong></dt><dd><p>name of the parent feature</p>
</dd>
<dt><strong>biotype</strong></dt><dd><p>biological type</p>
</dd>
<dt><strong>name</strong></dt><dd><p>feature name</p>
</dd>
<dt><strong>spans</strong></dt><dd><p>plus strand coordinates</p>
</dd>
<dt><strong>strand</strong></dt><dd><p>either ‘+’ or ‘-’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Feature</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.add_seqs">
<span class="sig-name descname"><span class="pre">add_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">before_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">after_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.add_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new object of class self with sequences from other added.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>other</strong></dt><dd><p>same class as self or coerceable to that class</p>
</dd>
<dt><strong>before_name</strong><span class="classifier">str</span></dt><dd><p>which sequence is added</p>
</dd>
<dt><strong>after_name</strong><span class="classifier">str</span></dt><dd><p>which sequence is added</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If both before_name and after_name are specified, the seqs will be
inserted using before_name.</p>
<p>By default the sequence is appended to the end of the alignment,
this can be changed by using either before_name or after_name arguments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.annotate_from_gff">
<span class="sig-name descname"><span class="pre">annotate_from_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PathLike</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.annotate_from_gff" title="Link to this definition">#</a></dt>
<dd><p>copies annotations from a gff file to a sequence in self</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>f</strong></dt><dd><p>path to gff annotation file.</p>
</dd>
<dt><strong>seq_name</strong></dt><dd><p>names of seqs to be annotated. Does not support setting offset,
set offset directly on sequences with seq.annotation_offset = offset</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.annotation_db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation_db</span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.annotation_db" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.apply_pssm">
<span class="sig-name descname"><span class="pre">apply_pssm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pssm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ui</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.apply_pssm" title="Link to this definition">#</a></dt>
<dd><p>scores sequences using the specified pssm</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pssm</strong><span class="classifier">profile.PSSM</span></dt><dd><p>if not provided, will be loaded from path</p>
</dd>
<dt><strong>path</strong></dt><dd><p>path to either a jaspar or cisbp matrix (path must end have a suffix
matching the format).</p>
</dd>
<dt><strong>pseudocount</strong></dt><dd><p>adjustment for zero in matrix</p>
</dd>
<dt><strong>names</strong></dt><dd><p>returns only scores for these sequences and in the name order</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>numpy array of log2 based scores at every position</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.copy" title="Link to this definition">#</a></dt>
<dd><p>Returns deep copy of self.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.copy_annotations">
<span class="sig-name descname"><span class="pre">copy_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SupportsFeatures</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.copy_annotations" title="Link to this definition">#</a></dt>
<dd><p>copy annotations into attached annotation db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_db</strong></dt><dd><p>compatible annotation db</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Only copies annotations for records with seqid in self.names</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.counts" title="Link to this definition">#</a></dt>
<dd><p>counts of motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.counts_per_seq">
<span class="sig-name descname"><span class="pre">counts_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.counts_per_seq" title="Link to this definition">#</a></dt>
<dd><p>counts of motifs per sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of characters per tuple.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>warn</strong></dt><dd><p>warns if motif_length &gt; 1 and alignment trimmed to produce
motif columns</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>MotifCountsArray</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.deepcopy">
<span class="sig-name descname"><span class="pre">deepcopy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sliced</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.deepcopy" title="Link to this definition">#</a></dt>
<dd><p>returns deep copy of self.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sliced</strong></dt><dd><p>if True, reduces the sequence to current
interval. This also causes dropping
annotations.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.degap">
<span class="sig-name descname"><span class="pre">degap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.degap" title="Link to this definition">#</a></dt>
<dd><p>Returns copy in which sequences have no gaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>kwargs</strong></dt><dd><p>passed to class constructor</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.distance_matrix">
<span class="sig-name descname"><span class="pre">distance_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">calc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdist'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.distance_matrix" title="Link to this definition">#</a></dt>
<dd><p>Estimated pairwise distance between sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>calc</strong><span class="classifier">str</span></dt><dd><p>The distance calculation method to use, either “pdist” or “jc69”
“pdist” is an approximation of the proportion sites different
“jc69” is an approximation of the Jukes Cantor distance</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>DistanceMatrix</dt><dd><p>Estimated pairwise distances between sequences in the collection</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>pdist approximates the proportion sites different from the Jaccard
distance. Coefficients for the approximation were derived from a
polynomial fit between Jaccard distance of kmers with k=10 and the
proportion of sites different using mammalian 106 protein coding
gene DNA sequence alignments.</p>
<p>jc69 approximates the Jukes Cantor distance using the approximated
proportion sites different, i.e., a transformation of the above.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.dotplot">
<span class="sig-name descname"><span class="pre">dotplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_gap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_progress</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.dotplot" title="Link to this definition">#</a></dt>
<dd><p>make a dotplot between specified sequences. Random sequences
chosen if names not provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name1, name2</strong></dt><dd><p>names of sequences. If not provided, a random choice is made</p>
</dd>
<dt><strong>window</strong></dt><dd><p>segment size for comparison between sequences</p>
</dd>
<dt><strong>threshold</strong></dt><dd><p>windows where the sequences are identical &gt;= threshold are a match</p>
</dd>
<dt><strong>k</strong></dt><dd><p>size of k-mer to break sequences into. Larger values increase
speed but reduce resolution. If not specified, and
window == threshold, then k is set to window. Otherwise, it is
computed as the maximum of {threshold // (window - threshold), 5}.</p>
</dd>
<dt><strong>min_gap</strong></dt><dd><p>permitted gap for joining adjacent line segments, default is no gap
joining</p>
</dd>
<dt><strong>width</strong></dt><dd><p>figure width. Figure height is computed based on the ratio of
len(seq1) / len(seq2)</p>
</dd>
<dt><strong>title</strong></dt><dd><p>title for the plot</p>
</dd>
<dt><strong>rc</strong></dt><dd><p>include dotplot of reverse compliment also. Only applies to Nucleic
acids moltypes</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>a Drawable or AnnotatedDrawable</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.entropy_per_seq">
<span class="sig-name descname"><span class="pre">entropy_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.entropy_per_seq" title="Link to this definition">#</a></dt>
<dd><p>Returns the Shannon entropy per sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length: int</strong></dt><dd><p>number of characters per tuple.</p>
</dd>
<dt><strong>include_ambiguity: bool</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap: bool</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved: bool</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
<dt><strong>warn</strong></dt><dd><p>warns if motif_length &gt; 1 and alignment trimmed to produce
motif columns</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>For motif_length &gt; 1, it’s advisable to specify exclude_unobserved=True,
this avoids unnecessary calculations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_ambiguous_positions">
<span class="sig-name descname"><span class="pre">get_ambiguous_positions</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_ambiguous_positions" title="Link to this definition">#</a></dt>
<dd><p>Returns dict of seq:{position:char} for ambiguous chars.</p>
<p>Used in likelihood calculations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_features">
<span class="sig-name descname"><span class="pre">get_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Iterator</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_features" title="Link to this definition">#</a></dt>
<dd><p>yields Feature instances</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seqid</strong></dt><dd><p>limit search to features on this named sequence, defaults to search all</p>
</dd>
<dt><strong>biotype</strong></dt><dd><p>biotype of the feature, e.g. CDS, gene</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>allow_partial</strong></dt><dd><p>allow features partially overlaping self</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>When dealing with a nucleic acid moltype, the returned features will
yield a sequence segment that is consistently oriented irrespective
of strand of the current instance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_identical_sets">
<span class="sig-name descname"><span class="pre">get_identical_sets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mask_degen</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_identical_sets" title="Link to this definition">#</a></dt>
<dd><p>returns sets of names for sequences that are identical</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>mask_degen</strong></dt><dd><p>if True, degenerate characters are ignored</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_lengths">
<span class="sig-name descname"><span class="pre">get_lengths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_lengths" title="Link to this definition">#</a></dt>
<dd><p>returns {name: seq length, …}</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_motif_probs">
<span class="sig-name descname"><span class="pre">get_motif_probs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudocount</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_motif_probs" title="Link to this definition">#</a></dt>
<dd><p>Return a dictionary of motif probs, calculated as the averaged
frequency across sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_ambiguity</strong></dt><dd><p>if True resolved ambiguous codes are
included in estimation of frequencies, default is False.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, motifs that are not present in
the alignment are excluded from the returned dictionary,
default is False.</p>
</dd>
<dt><strong>allow_gap</strong></dt><dd><p>allow gap motif</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>only non-overlapping motifs are counted</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_seq">
<span class="sig-name descname"><span class="pre">get_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_seq" title="Link to this definition">#</a></dt>
<dd><p>Return a sequence object for the specified seqname.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_seq_indices">
<span class="sig-name descname"><span class="pre">get_seq_indices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_seq_indices" title="Link to this definition">#</a></dt>
<dd><p>Returns list of keys of seqs where f(row) is True.</p>
<p>List will be in the same order as self.names, if present.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_similar">
<span class="sig-name descname"><span class="pre">get_similar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_similarity=0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_similarity=1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric=&lt;cogent3.util.transform.for_seq</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform=None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_similar" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing sequences similar to target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>target</strong></dt><dd><p>sequence object to compare to. Can be in the alignment.</p>
</dd>
<dt><strong>min_similarity</strong></dt><dd><p>minimum similarity that will be kept. Default 0.0.</p>
</dd>
<dt><strong>max_similarity</strong></dt><dd><p>maximum similarity that will be kept. Default 1.0.
(Note that both min_similarity and max_similarity are inclusive.)
metric
similarity function to use. Must be f(first_seq, second_seq).</p>
</dd>
<dt><strong>The default metric is fraction similarity, ranging from 0.0 (0%</strong></dt><dd></dd>
<dt><strong>identical) to 1.0 (100% identical). The Sequence classes have lots</strong></dt><dd></dd>
<dt><strong>of methods that can be passed in as unbound methods to act as the</strong></dt><dd></dd>
<dt><strong>metric, e.g. frac_same_gaps.</strong></dt><dd></dd>
<dt><strong>transform</strong></dt><dd><p>transformation function to use on the sequences before
the metric is calculated. If None, uses the whole sequences in each
case. A frequent transformation is a function that returns a specified
range of a sequence, e.g. eliminating the ends. Note that the
transform applies to both the real sequence and the target sequence.</p>
</dd>
<dt><strong>WARNING: if the transformation changes the type of the sequence (e.g.</strong></dt><dd></dd>
<dt><strong>extracting a string from an RnaSequence object), distance metrics that</strong></dt><dd></dd>
<dt><strong>depend on instance data of the original class may fail.</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.get_translation">
<span class="sig-name descname"><span class="pre">get_translation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incomplete_ok</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim_stop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.get_translation" title="Link to this definition">#</a></dt>
<dd><p>translate from nucleic acid to protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>genetic code, either the number or name
(use cogent3.core.genetic_code.available_codes)</p>
</dd>
<dt><strong>incomplete_ok</strong></dt><dd><p>codons that are mixes of nucleotide and gaps converted to ‘?’.
raises a ValueError if False</p>
</dd>
<dt><strong>include_stop</strong></dt><dd><p>whether to allow a stops in the translated sequence</p>
</dd>
<dt><strong>trim_stop</strong></dt><dd><p>exclude terminal stop codons if they exist</p>
</dd>
<dt><strong>kwargs</strong></dt><dd><p>related to construction of the resulting object</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>A new instance of self translated into protein</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.has_terminal_stop">
<span class="sig-name descname"><span class="pre">has_terminal_stop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.has_terminal_stop" title="Link to this definition">#</a></dt>
<dd><p>Returns True if any sequence has a terminal stop codon.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>strict</strong></dt><dd><p>If True, raises an exception if a seq length not divisible by 3</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.is_array">
<span class="sig-name descname"><span class="pre">is_array</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'array',</span> <span class="pre">'array_seqs'}</span></em><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.is_array" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.is_ragged">
<span class="sig-name descname"><span class="pre">is_ragged</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.is_ragged" title="Link to this definition">#</a></dt>
<dd><p>Returns True if alignment has sequences of different lengths.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.iter_selected">
<span class="sig-name descname"><span class="pre">iter_selected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.iter_selected" title="Link to this definition">#</a></dt>
<dd><p>Iterates over elements in the alignment.</p>
<p>seq_order (names) can be used to select a subset of seqs.
pos_order (positions) can be used to select a subset of positions.</p>
<p>Always iterates along a seq first, then down a position (transposes
normal order of a[i][j]; possibly, this should change)..</p>
<p>WARNING: Alignment.iter_selected() is not the same as alignment.iteritems()
(which is the built-in dict iteritems that iterates over key-value
pairs).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.iter_seqs">
<span class="sig-name descname"><span class="pre">iter_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.iter_seqs" title="Link to this definition">#</a></dt>
<dd><p>Iterates over values (sequences) in the alignment, in order.</p>
<p>seq_order: list of keys giving the order in which seqs will be returned.
Defaults to self.Names. Note that only these sequences will be
returned, and that KeyError will be raised if there are sequences
in order that have been deleted from the Alignment. If self.Names
is None, returns the sequences in the same order as
self.named_seqs.values().</p>
<p>Use map(f, self.seqs()) to apply the constructor f to each seq. f must
accept a single list as an argument.</p>
<p>Always returns references to the same objects that are values of the
alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.make_feature">
<span class="sig-name descname"><span class="pre">make_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">FeatureDataType</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.make_feature" title="Link to this definition">#</a></dt>
<dd><p>create a feature on named sequence, or on the alignment itself</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>feature</strong></dt><dd><p>a dict with all the necessary data rto construct a feature</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Feature</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>To get a feature AND add it to annotation_db, use add_feature().</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.moltype">
<span class="sig-name descname"><span class="pre">moltype</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">MolType(('a',</span> <span class="pre">'b',</span> <span class="pre">'c',</span> <span class="pre">'d',</span> <span class="pre">'e',</span> <span class="pre">'f',</span> <span class="pre">'g',</span> <span class="pre">'h',</span> <span class="pre">'i',</span> <span class="pre">'j',</span> <span class="pre">'k',</span> <span class="pre">'l',</span> <span class="pre">'m',</span> <span class="pre">'n',</span> <span class="pre">'o',</span> <span class="pre">'p',</span> <span class="pre">'q',</span> <span class="pre">'r',</span> <span class="pre">'s',</span> <span class="pre">'t',</span> <span class="pre">'u',</span> <span class="pre">'v',</span> <span class="pre">'w',</span> <span class="pre">'x',</span> <span class="pre">'y',</span> <span class="pre">'z',</span> <span class="pre">'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'J',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'O',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'X',</span> <span class="pre">'Y',</span> <span class="pre">'Z'))</span></em><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.moltype" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.named_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">named_seqs</span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.named_seqs" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.num_seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_seqs</span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.num_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns the number of sequences in the alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.omit_gap_runs">
<span class="sig-name descname"><span class="pre">omit_gap_runs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">allowed_run</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.omit_gap_runs" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment where all seqs have runs of gaps &lt;=allowed_run.</p>
<p>Note that seqs with exactly allowed_run gaps are not deleted.
Default is for allowed_run to be 1 (i.e. no consecutive gaps allowed).</p>
<p>Because the test for whether the current gap run exceeds the maximum
allowed gap run is only triggered when there is at least one gap, even
negative values for allowed_run will still let sequences with no gaps
through.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.omit_gap_seqs">
<span class="sig-name descname"><span class="pre">omit_gap_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">allowed_gap_frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.omit_gap_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new alignment with seqs that have &lt;= allowed_gap_frac.</p>
<p>allowed_gap_frac should be a fraction between 0 and 1 inclusive.
Default is 0.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.pad_seqs">
<span class="sig-name descname"><span class="pre">pad_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pad_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.pad_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns copy in which sequences are padded to same length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pad_length</strong></dt><dd><p>Length all sequences are to be padded to.  Will pad
to max sequence length if pad_length is None or less than max
length.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.probs_per_seq">
<span class="sig-name descname"><span class="pre">probs_per_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">warn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.probs_per_seq" title="Link to this definition">#</a></dt>
<dd><p>return MotifFreqsArray per sequence</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.rc" title="Link to this definition">#</a></dt>
<dd><p>Returns the reverse complement alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.rename_seqs">
<span class="sig-name descname"><span class="pre">rename_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">renamer</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.rename_seqs" title="Link to this definition">#</a></dt>
<dd><p>returns new instance with sequences renamed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>renamer</strong><span class="classifier">callable</span></dt><dd><p>function that will take current sequences and return the new one</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Returns the reverse complement alignment. A synonym for rc.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.seqs">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seqs</span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.seqs" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.set_repr_policy">
<span class="sig-name descname"><span class="pre">set_repr_policy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_seqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.set_repr_policy" title="Link to this definition">#</a></dt>
<dd><p>specify policy for repr(self)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>num_seqs</strong><span class="classifier">int or None</span></dt><dd><p>number of sequences to include in represented display.</p>
</dd>
<dt><strong>num_pos</strong><span class="classifier">int or None</span></dt><dd><p>length of sequences to include in represented display.</p>
</dd>
<dt><strong>ref_name</strong><span class="classifier">str or None</span></dt><dd><p>name of sequence to be placed first, or “longest” (default).
If latter, indicates longest sequence will be chosen.</p>
</dd>
<dt><strong>wrap</strong><span class="classifier">int or None</span></dt><dd><p>number of printed bases per row</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.strand_symmetry">
<span class="sig-name descname"><span class="pre">strand_symmetry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.strand_symmetry" title="Link to this definition">#</a></dt>
<dd><p>returns dict of strand symmetry test results per seq</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.take_seqs">
<span class="sig-name descname"><span class="pre">take_seqs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seqs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.take_seqs" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing only specified seqs.</p>
<p>Note that the seqs in the new alignment will be references to the
same objects as the seqs in the old alignment.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.take_seqs_if">
<span class="sig-name descname"><span class="pre">take_seqs_if</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.take_seqs_if" title="Link to this definition">#</a></dt>
<dd><p>Returns new Alignment containing seqs where f(row) is True.</p>
<p>Note that the seqs in the new Alignment are the same objects as the
seqs in the old Alignment, not copies.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_dict">
<span class="sig-name descname"><span class="pre">to_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_dict" title="Link to this definition">#</a></dt>
<dd><p>Returns the alignment as a dict of sequence names -&gt; strings.</p>
<p>Note the mapping goes to strings, not Sequence objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>a dict mapping sequence names to a string representation of</dt><dd></dd>
<dt>their sequences.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_dna">
<span class="sig-name descname"><span class="pre">to_dna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_dna" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self as an alignment of DNA moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_fasta">
<span class="sig-name descname"><span class="pre">to_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">block_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">60</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_fasta" title="Link to this definition">#</a></dt>
<dd><p>Return alignment in Fasta format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>block_size</strong></dt><dd><p>the sequence length to write to each line,
by default 60</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>The Fasta formatted alignment.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_html">
<span class="sig-name descname"><span class="pre">to_html</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name_order</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_family</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Lucida</span> <span class="pre">Console'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_html" title="Link to this definition">#</a></dt>
<dd><p>returns html with embedded styles for sequence colouring</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name_order</strong></dt><dd><p>order of names for display.</p>
</dd>
<dt><strong>wrap</strong></dt><dd><p>number of alignment columns per row</p>
</dd>
<dt><strong>limit</strong></dt><dd><p>truncate alignment to this length</p>
</dd>
<dt><strong>colors</strong></dt><dd><p>{character
moltype.</p>
</dd>
<dt><strong>font_size</strong></dt><dd><p>in points. Affects labels and sequence and line spacing
(proportional to value)</p>
</dd>
<dt><strong>font_family</strong></dt><dd><p>string denoting font family</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>In a jupyter notebook, this code is used to provide the representation.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">seq_col</span>  <span class="c1"># is rendered by jupyter</span>
</pre></div>
</div>
<p>You can directly use the result for display in a notebook as</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">IPython.core.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">HTML</span>

<span class="n">HTML</span><span class="p">(</span><span class="n">seq_col</span><span class="o">.</span><span class="n">to_html</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_json">
<span class="sig-name descname"><span class="pre">to_json</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_json" title="Link to this definition">#</a></dt>
<dd><p>returns json formatted string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_moltype">
<span class="sig-name descname"><span class="pre">to_moltype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">moltype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_moltype" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self with moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_nexus">
<span class="sig-name descname"><span class="pre">to_nexus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_nexus" title="Link to this definition">#</a></dt>
<dd><p>Return alignment in NEXUS format and mapping to sequence ids</p>
<dl class="simple">
<dt><strong>NOTE</strong> Not that every sequence in the alignment MUST come from</dt><dd><p>a different species!! (You can concatenate multiple sequences from
same species together before building tree)</p>
</dd>
</dl>
<p>seq_type: dna, rna, or protein</p>
<p>Raises exception if invalid alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_phylip">
<span class="sig-name descname"><span class="pre">to_phylip</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_phylip" title="Link to this definition">#</a></dt>
<dd><p>Return alignment in PHYLIP format and mapping to sequence ids</p>
<p>raises exception if invalid alignment</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_protein">
<span class="sig-name descname"><span class="pre">to_protein</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_protein" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self as an alignment of PROTEIN moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_rich_dict">
<span class="sig-name descname"><span class="pre">to_rich_dict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_rich_dict" title="Link to this definition">#</a></dt>
<dd><p>returns detailed content including info and moltype attributes</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.to_rna">
<span class="sig-name descname"><span class="pre">to_rna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.to_rna" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self as an alignment of RNA moltype seqs</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.trim_stop_codons">
<span class="sig-name descname"><span class="pre">trim_stop_codons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.trim_stop_codons" title="Link to this definition">#</a></dt>
<dd><p>Removes any terminal stop codons from the sequences</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>strict</strong></dt><dd><p>If True, raises an exception if a seq length not divisible by 3</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.with_modified_termini">
<span class="sig-name descname"><span class="pre">with_modified_termini</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.with_modified_termini" title="Link to this definition">#</a></dt>
<dd><p>Changes the termini to include termini char instead of gapmotif.</p>
<p>Useful to correct the standard gap char output by most
alignment programs when aligned sequences have different ends.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.alignment.SequenceCollection.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.alignment.SequenceCollection.write" title="Link to this definition">#</a></dt>
<dd><p>Write the alignment to a file, preserving order of sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>filename</strong></dt><dd><p>name of the sequence file</p>
</dd>
<dt><strong>format</strong></dt><dd><p>format of the sequence file</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If format is None, will attempt to infer format from the filename
suffix.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../alignment.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></p>
      </div>
    </a>
    <a class="right-next"
       href="cogent3.core.alignment.Alignment.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Alignment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection"><code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.add_feature"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.add_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.add_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.add_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.annotate_from_gff"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.annotate_from_gff()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.annotation_db"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.annotation_db</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.apply_pssm"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.apply_pssm()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.copy"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.copy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.copy_annotations"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.copy_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.counts"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.counts()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.counts_per_seq"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.counts_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.deepcopy"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.deepcopy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.degap"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.degap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.distance_matrix"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.distance_matrix()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.dotplot"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.dotplot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.entropy_per_seq"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.entropy_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_ambiguous_positions"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_ambiguous_positions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_features"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_features()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_identical_sets"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_identical_sets()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_lengths"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_lengths()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_motif_probs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_motif_probs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_seq"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_seq_indices"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_seq_indices()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_similar"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_similar()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.get_translation"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.get_translation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.has_terminal_stop"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.has_terminal_stop()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.is_array"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.is_array</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.is_ragged"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.is_ragged()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.iter_selected"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.iter_selected()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.iter_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.iter_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.make_feature"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.make_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.moltype"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.moltype</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.named_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.named_seqs</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.num_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.num_seqs</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.omit_gap_runs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.omit_gap_runs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.omit_gap_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.omit_gap_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.pad_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.pad_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.probs_per_seq"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.probs_per_seq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.rc"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.rc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.rename_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.rename_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.reverse_complement"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.reverse_complement()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.seqs</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.set_repr_policy"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.set_repr_policy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.strand_symmetry"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.strand_symmetry()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.take_seqs"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.take_seqs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.take_seqs_if"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.take_seqs_if()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_dict"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_dna"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_dna()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_fasta"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_fasta()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_html"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_html()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_json"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_json()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_moltype"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_moltype()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_nexus"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_nexus()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_phylip"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_phylip()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_protein"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_protein()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_rich_dict"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_rich_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.to_rna"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.to_rna()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.trim_stop_codons"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.trim_stop_codons()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.with_modified_termini"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.with_modified_termini()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.alignment.SequenceCollection.write"><code class="docutils literal notranslate"><span class="pre">SequenceCollection.write()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020-2025, cogent3 Team.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.1.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>