<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2020.4 documentation</title>

<link href="../../../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../../../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/sphinx-dropdown.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/panels-bootstrap.min.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/gallery.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/gallery-binder.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/gallery-dataframe.css" type="text/css" />

<link rel="stylesheet" href="../../../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../../../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../../../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../../../.././_static/pygments.css">
<link rel="stylesheet" href="../../../.././_static/css/common.css">
<link rel="stylesheet" href="../../../.././_static/css/link.css">
<link rel="stylesheet" href="../../../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
<script src="../../../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../../../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../../../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../../../.././_static/js/initializeHighlight.js"></script>
<script src="../../../../_static/jquery.js"></script>
<script src="../../../../_static/underscore.js"></script>
<script src="../../../../_static/doctools.js"></script>
<script src="../../../../_static/language_data.js"></script>
<script src="../../../../_static/thebelab-helper.js"></script>
<script src="../../../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../../../_static/require.js"></script>
<script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
<script src="../../../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../../../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../../../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../../../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../../../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../../../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#installing-using-conda">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../install.html#manual-creation-of-the-conda-environment">Manual creation of the <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../index.html">Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#overview">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/warning.html">Warning – apps are alpha!</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#define-the-apps">Define the apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#using-apps-like-functions">Using apps like functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-log-the-settings-and-data-analysed">You can log the settings and data analysed</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/dstore.html">Specifying data for analysis</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#a-read-only-data-store">A read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#data-store-members">Data store “members”</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#showing-the-last-few-members">Showing the last few members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#filtering-a-data-store-for-specific-members">Filtering a data store for specific members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#tinydb-data-stores-are-special">TinyDB data stores are special</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/available-apps.html">What apps are there?</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-progressive-align-app">The <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-model-app">The <code class="docutils literal notranslate"><span class="pre">model</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-hypothesis-app">The <code class="docutils literal notranslate"><span class="pre">hypothesis</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-natsel-apps">The <code class="docutils literal notranslate"><span class="pre">natsel</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index.html#the-user-function-apps">The <code class="docutils literal notranslate"><span class="pre">user_function</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/user_function.html">Custom composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#a-user-function-for-with-a-different-output-type">A user function for with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html">Manipulating biological data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html">Loading an alignment from a file</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/alignments.html">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html"><code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/annotations.html">Annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/annotations.html#annotations-with-coordinates">Annotations with coordinates</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#automated-introduction-from-reading-genbank-files">Automated introduction from reading genbank files</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#customising-annotation-construction-from-reading-a-genbank-file">Customising annotation construction from reading a genbank file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#creating-directly-on-a-sequence">Creating directly on a sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#via">Via</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#adding-as-a-series-or-item-wise">Adding as a series or item-wise</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#taking-the-union-of-annotations">Taking the union of annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#getting-annotation-coordinates">Getting annotation coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#annotations-have-shadows">Annotations have shadows</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#adding-to-a-sequence-member-of-an-alignment">Adding to a sequence member of an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#adding-to-an-alignment">Adding to an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#slicing-sequences-and-alignments-by-annotations">Slicing sequences and alignments by annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#slicing-by-pseudo-feature-or-feature-series">Slicing by pseudo-feature or feature series</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#slice-series-must-not-be-overlapping">Slice series must not be overlapping</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#you-can-slice-an-annotation-itself">You can slice an annotation itself</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#sequence-vs-alignment-slicing">Sequence vs Alignment slicing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#copying-annotations">Copying annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying">Querying</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying-produces-objects-only-valid-for-their-source">Querying produces objects only valid for their source</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying-for-absent-annotation">Querying for absent annotation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#querying-features-that-span-gaps-in-alignments">Querying features that span gaps in alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#as-one-span-unifies-features-with-discontinuous-alignment-coordinates"><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#behaviour-of-annotations-on-nucleic-acid-sequences">Behaviour of annotations on nucleic acid sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#masking-annotated-regions">Masking annotated regions</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#masking-annotated-regions-on-alignments">Masking annotated regions on alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#you-can-take-mask-of-the-shadow">You can take mask of the shadow</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#what-features-of-a-certain-type-are-available">What features of a certain type are available?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#getting-all-features-of-a-type-or-everything-but-that-type">Getting all features of a type, or everything but that type</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#getting-sequence-features-when-you-have-an-alignment-object">Getting sequence features when you have an alignment object</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotations.html#annotation-display-on-sequences">Annotation display on sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html">Translate DNA sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html#translating-a-dna-sequence-with-a-terminating-stop-codon">Translating a DNA sequence with a terminating stop codon</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#by-removing-the-trailing-stop-codon-first">By removing the trailing stop codon first</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#by-slicing-the-dnasequence-first">By slicing the <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> first</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/index.html">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/seq_features.html">Advanced sequence handling</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/complete_seq_features.html">Annotating alignments and sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#modelling-evolution">Modelling Evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="../../index.html">API</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3 current"><a class="reference internal" href="../../index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../../alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.util.table.Columns.html">Columns</a></li>
</ul>
</li>
<li class="toctree-l4 current"><a class="reference internal" href="../tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul class="current">
<li class="toctree-l5 current"><a class="current reference internal" href="#">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html">Evaluating coevolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html">Dotplot</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#all-options">All options</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html">Draw sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-angular.html">Display a Phylogenetic Tree with a Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html">Display a Phylogenetic Tree with a Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-radial.html">Display a Phylogenetic Tree with a Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html">Display a Phylogenetic Tree with a Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#changing-scale-bar-placement">Changing scale bar placement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-support.html">Display a Phylogenetic Tree Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <div class="section" id="phylonode">
<h1>PhyloNode<a class="headerlink" href="#phylonode" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt id="cogent3.core.tree.PhyloNode">
<em class="property">class </em><code class="sig-name descname">PhyloNode</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>length</strong></dt><dd></dd>
<dt><a class="reference internal" href="#cogent3.core.tree.PhyloNode.parent" title="cogent3.core.tree.PhyloNode.parent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parent</span></code></a></dt><dd><p>Accessor for parent.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.ancestors" title="cogent3.core.tree.PhyloNode.ancestors"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ancestors</span></code></a>()</p></td>
<td><p>Returns all ancestors back to the root.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.append" title="cogent3.core.tree.PhyloNode.append"><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code></a>(i)</p></td>
<td><p>Appends i to self.children, in-place, cleaning up refs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.ascii_art" title="cogent3.core.tree.PhyloNode.ascii_art"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ascii_art</span></code></a>([show_internal, compact])</p></td>
<td><p>Returns a string containing an ascii drawing of the tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.balanced" title="cogent3.core.tree.PhyloNode.balanced"><code class="xref py py-obj docutils literal notranslate"><span class="pre">balanced</span></code></a>()</p></td>
<td><p>Tree ‘rooted’ here with no neighbour having &gt; 50% of the edges.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.bifurcating" title="cogent3.core.tree.PhyloNode.bifurcating"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bifurcating</span></code></a>([eps, constructor, name_unnamed])</p></td>
<td><p>Wrap multifurcating with a num of 2</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.child_groups" title="cogent3.core.tree.PhyloNode.child_groups"><code class="xref py py-obj docutils literal notranslate"><span class="pre">child_groups</span></code></a>()</p></td>
<td><p>Returns list containing lists of children sharing a state.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.compare_by_names" title="cogent3.core.tree.PhyloNode.compare_by_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compare_by_names</span></code></a>(other)</p></td>
<td><p>Equality test for trees by name</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.compare_by_subsets" title="cogent3.core.tree.PhyloNode.compare_by_subsets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compare_by_subsets</span></code></a>(other[, exclude_absent_taxa])</p></td>
<td><p>Returns fraction of overlapping subsets where self and other differ.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.compare_by_tip_distances" title="cogent3.core.tree.PhyloNode.compare_by_tip_distances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compare_by_tip_distances</span></code></a>(other[, sample, …])</p></td>
<td><p>Compares self to other using tip-to-tip distance matrices.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.compare_name" title="cogent3.core.tree.PhyloNode.compare_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compare_name</span></code></a>(other)</p></td>
<td><p>Compares TreeNode by name</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.copy" title="cogent3.core.tree.PhyloNode.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>([memo, _nil, constructor])</p></td>
<td><p>Returns a copy of self using an iterative approach</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.copy_recursive" title="cogent3.core.tree.PhyloNode.copy_recursive"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_recursive</span></code></a>([memo, _nil, constructor])</p></td>
<td><p>Returns copy of self’s structure, including shallow copy of attrs.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.copy_topology" title="cogent3.core.tree.PhyloNode.copy_topology"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_topology</span></code></a>([constructor])</p></td>
<td><p>Copies only the topology and labels of a tree, not any extra data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.deepcopy" title="cogent3.core.tree.PhyloNode.deepcopy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">deepcopy</span></code></a>([memo, _nil, constructor])</p></td>
<td><p>Returns a copy of self using an iterative approach</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.descendant_array" title="cogent3.core.tree.PhyloNode.descendant_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">descendant_array</span></code></a>([tip_list])</p></td>
<td><p>Returns numpy array with nodes in rows and descendants in columns.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.distance" title="cogent3.core.tree.PhyloNode.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a>(other)</p></td>
<td><p>Returns branch length between self and other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.extend" title="cogent3.core.tree.PhyloNode.extend"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend</span></code></a>(items)</p></td>
<td><p>Extends self.children by items, in-place, cleaning up refs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_connecting_edges" title="cogent3.core.tree.PhyloNode.get_connecting_edges"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_connecting_edges</span></code></a>(name1, name2)</p></td>
<td><p>returns a list of edges connecting two nodes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_connecting_node" title="cogent3.core.tree.PhyloNode.get_connecting_node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_connecting_node</span></code></a>(name1, name2)</p></td>
<td><p>Finds the last common ancestor of the two named edges.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_distances" title="cogent3.core.tree.PhyloNode.get_distances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_distances</span></code></a>([endpoints])</p></td>
<td><p>The distance matrix as a dictionary.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_edge_names" title="cogent3.core.tree.PhyloNode.get_edge_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_edge_names</span></code></a>(tip1name, tip2name[, clade, …])</p></td>
<td><p>Return the list of stem and/or sub tree (clade) edge name(s).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_edge_vector" title="cogent3.core.tree.PhyloNode.get_edge_vector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_edge_vector</span></code></a>([include_root])</p></td>
<td><p>Collect the list of edges in postfix order</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_figure" title="cogent3.core.tree.PhyloNode.get_figure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_figure</span></code></a>([style])</p></td>
<td><p>gets Dendrogram for plotting the phylogeny</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_max_tip_tip_distance" title="cogent3.core.tree.PhyloNode.get_max_tip_tip_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_max_tip_tip_distance</span></code></a>()</p></td>
<td><p>Returns the max tip tip distance between any pair of tips</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_newick" title="cogent3.core.tree.PhyloNode.get_newick"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_newick</span></code></a>([with_distances, semicolon, …])</p></td>
<td><p>Return the newick string for this tree.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_newick_recursive" title="cogent3.core.tree.PhyloNode.get_newick_recursive"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_newick_recursive</span></code></a>([with_distances, …])</p></td>
<td><p>Return the newick string for this edge.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_node_names" title="cogent3.core.tree.PhyloNode.get_node_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_node_names</span></code></a>([includeself, tipsonly])</p></td>
<td><p>Return a list of edges from this edge - may or may not include self.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_nodes_dict" title="cogent3.core.tree.PhyloNode.get_nodes_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_nodes_dict</span></code></a>()</p></td>
<td><p>Returns a dict keyed by node name, value is node</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_param_value" title="cogent3.core.tree.PhyloNode.get_param_value"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_param_value</span></code></a>(param, edge)</p></td>
<td><p>returns the parameter value for named edge</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_sub_tree" title="cogent3.core.tree.PhyloNode.get_sub_tree"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sub_tree</span></code></a>(name_list[, ignore_missing, …])</p></td>
<td><p>A new instance of a sub tree that contains all the otus that are listed in name_list.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_tip_names" title="cogent3.core.tree.PhyloNode.get_tip_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_tip_names</span></code></a>([includeself])</p></td>
<td><p>return the list of the names of all tips contained by this edge</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.get_xml" title="cogent3.core.tree.PhyloNode.get_xml"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_xml</span></code></a>()</p></td>
<td><p>Return XML formatted tree string.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.index_in_parent" title="cogent3.core.tree.PhyloNode.index_in_parent"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index_in_parent</span></code></a>()</p></td>
<td><p>Returns index of self in parent.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.insert" title="cogent3.core.tree.PhyloNode.insert"><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code></a>(index, i)</p></td>
<td><p>Inserts an item at specified position in self.children.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.is_root" title="cogent3.core.tree.PhyloNode.is_root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_root</span></code></a>()</p></td>
<td><p>Returns True if the current is a root, i.e. has no parent.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.is_tip" title="cogent3.core.tree.PhyloNode.is_tip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_tip</span></code></a>()</p></td>
<td><p>Returns True if the current node is a tip, i.e. has no children.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.isroot" title="cogent3.core.tree.PhyloNode.isroot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">isroot</span></code></a>()</p></td>
<td><p>Returns True if root of a tree, i.e. no parent.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.istip" title="cogent3.core.tree.PhyloNode.istip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">istip</span></code></a>()</p></td>
<td><p>Returns True if is tip, i.e. no children.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.iter_nontips" title="cogent3.core.tree.PhyloNode.iter_nontips"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_nontips</span></code></a>([include_self])</p></td>
<td><p>Iterates over nontips descended from self, [] if none.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.iter_tips" title="cogent3.core.tree.PhyloNode.iter_tips"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_tips</span></code></a>([include_self])</p></td>
<td><p>Iterates over tips descended from self, [] if self is a tip.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.last_common_ancestor" title="cogent3.core.tree.PhyloNode.last_common_ancestor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">last_common_ancestor</span></code></a>(other)</p></td>
<td><p>Finds last common ancestor of self and other, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.lca" title="cogent3.core.tree.PhyloNode.lca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lca</span></code></a>(other)</p></td>
<td><p>Finds last common ancestor of self and other, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.levelorder" title="cogent3.core.tree.PhyloNode.levelorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">levelorder</span></code></a>([include_self])</p></td>
<td><p>Performs levelorder iteration over tree</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.lowest_common_ancestor" title="cogent3.core.tree.PhyloNode.lowest_common_ancestor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">lowest_common_ancestor</span></code></a>(tipnames)</p></td>
<td><p>Lowest common ancestor for a list of tipnames</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.make_tree_array" title="cogent3.core.tree.PhyloNode.make_tree_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_tree_array</span></code></a>([dec_list])</p></td>
<td><p>Makes an array with nodes in rows and descendants in columns.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.max_tip_tip_distance" title="cogent3.core.tree.PhyloNode.max_tip_tip_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_tip_tip_distance</span></code></a>()</p></td>
<td><p>returns the max distance between any pair of tips</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.multifurcating" title="cogent3.core.tree.PhyloNode.multifurcating"><code class="xref py py-obj docutils literal notranslate"><span class="pre">multifurcating</span></code></a>(num[, eps, constructor, …])</p></td>
<td><p>return a new tree with every node having num or few children</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.name_unnamed_nodes" title="cogent3.core.tree.PhyloNode.name_unnamed_nodes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">name_unnamed_nodes</span></code></a>()</p></td>
<td><p>sets the Data property of unnamed nodes to an arbitrary value</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.non_tip_children" title="cogent3.core.tree.PhyloNode.non_tip_children"><code class="xref py py-obj docutils literal notranslate"><span class="pre">non_tip_children</span></code></a>()</p></td>
<td><p>Returns direct children in self that have descendants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.nontips" title="cogent3.core.tree.PhyloNode.nontips"><code class="xref py py-obj docutils literal notranslate"><span class="pre">nontips</span></code></a>([include_self])</p></td>
<td><p>Returns nontips descended from self.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.pop" title="cogent3.core.tree.PhyloNode.pop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code></a>([index])</p></td>
<td><p>Returns and deletes child of self at index (default: -1)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.postorder" title="cogent3.core.tree.PhyloNode.postorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">postorder</span></code></a>([include_self])</p></td>
<td><p>performs postorder iteration over tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.pre_and_postorder" title="cogent3.core.tree.PhyloNode.pre_and_postorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pre_and_postorder</span></code></a>([include_self])</p></td>
<td><p>Performs iteration over tree, visiting node before and after.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.preorder" title="cogent3.core.tree.PhyloNode.preorder"><code class="xref py py-obj docutils literal notranslate"><span class="pre">preorder</span></code></a>([include_self])</p></td>
<td><p>Performs preorder iteration over tree.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.prune" title="cogent3.core.tree.PhyloNode.prune"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prune</span></code></a>()</p></td>
<td><p>Reconstructs correct tree after nodes have been removed.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.reassign_names" title="cogent3.core.tree.PhyloNode.reassign_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reassign_names</span></code></a>(mapping[, nodes])</p></td>
<td><p>Reassigns node names based on a mapping dict</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.remove" title="cogent3.core.tree.PhyloNode.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code></a>(target)</p></td>
<td><p>Removes node by name instead of identity.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.remove_deleted" title="cogent3.core.tree.PhyloNode.remove_deleted"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_deleted</span></code></a>(is_deleted)</p></td>
<td><p>Removes all nodes where is_deleted tests true.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.remove_node" title="cogent3.core.tree.PhyloNode.remove_node"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove_node</span></code></a>(target)</p></td>
<td><p>Removes node by identity instead of value.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.root" title="cogent3.core.tree.PhyloNode.root"><code class="xref py py-obj docutils literal notranslate"><span class="pre">root</span></code></a>()</p></td>
<td><p>Returns root of the tree self is in.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.root_at_midpoint" title="cogent3.core.tree.PhyloNode.root_at_midpoint"><code class="xref py py-obj docutils literal notranslate"><span class="pre">root_at_midpoint</span></code></a>()</p></td>
<td><p>return a new tree rooted at midpoint of the two tips farthest apart</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.rooted_at" title="cogent3.core.tree.PhyloNode.rooted_at"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rooted_at</span></code></a>(edge_name)</p></td>
<td><p>Return a new tree rooted at the provided node.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.rooted_with_tip" title="cogent3.core.tree.PhyloNode.rooted_with_tip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rooted_with_tip</span></code></a>(outgroup_name)</p></td>
<td><p>A new tree with the named tip as one of the root’s children</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.same_shape" title="cogent3.core.tree.PhyloNode.same_shape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">same_shape</span></code></a>(other)</p></td>
<td><p>Ignores lengths and order, so trees should be sorted first</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.same_topology" title="cogent3.core.tree.PhyloNode.same_topology"><code class="xref py py-obj docutils literal notranslate"><span class="pre">same_topology</span></code></a>(other)</p></td>
<td><p>Tests whether two trees have the same topology.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.scale_branch_lengths" title="cogent3.core.tree.PhyloNode.scale_branch_lengths"><code class="xref py py-obj docutils literal notranslate"><span class="pre">scale_branch_lengths</span></code></a>([max_length, ultrametric])</p></td>
<td><p>Scales BranchLengths in place to integers for ascii output.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.separation" title="cogent3.core.tree.PhyloNode.separation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">separation</span></code></a>(other)</p></td>
<td><p>Returns number of edges separating self and other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.set_max_tip_tip_distance" title="cogent3.core.tree.PhyloNode.set_max_tip_tip_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_max_tip_tip_distance</span></code></a>()</p></td>
<td><p>Propagate tip distance information up the tree</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.set_param_value" title="cogent3.core.tree.PhyloNode.set_param_value"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_param_value</span></code></a>(param, edge, value)</p></td>
<td><p>set’s the value for param at named edge</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.set_tip_distances" title="cogent3.core.tree.PhyloNode.set_tip_distances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_tip_distances</span></code></a>()</p></td>
<td><p>Sets distance from each node to the most distant tip.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.siblings" title="cogent3.core.tree.PhyloNode.siblings"><code class="xref py py-obj docutils literal notranslate"><span class="pre">siblings</span></code></a>()</p></td>
<td><p>Returns all nodes that are children of the same parent as self.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.sorted" title="cogent3.core.tree.PhyloNode.sorted"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sorted</span></code></a>([sort_order])</p></td>
<td><p>An equivalent tree sorted into a standard order.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.subset" title="cogent3.core.tree.PhyloNode.subset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subset</span></code></a>()</p></td>
<td><p>Returns set of names that descend from specified node</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.subsets" title="cogent3.core.tree.PhyloNode.subsets"><code class="xref py py-obj docutils literal notranslate"><span class="pre">subsets</span></code></a>()</p></td>
<td><p>Returns all sets of names that come from specified node and its kids</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.tip_children" title="cogent3.core.tree.PhyloNode.tip_children"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tip_children</span></code></a>()</p></td>
<td><p>Returns direct children of self that are tips.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.tip_to_tip_distances" title="cogent3.core.tree.PhyloNode.tip_to_tip_distances"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tip_to_tip_distances</span></code></a>([endpoints, default_length])</p></td>
<td><p>Returns distance matrix between all pairs of tips, and a tip order.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.tips" title="cogent3.core.tree.PhyloNode.tips"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tips</span></code></a>([include_self])</p></td>
<td><p>Returns tips descended from self, [] if self is a tip.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.tips_within_distance" title="cogent3.core.tree.PhyloNode.tips_within_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tips_within_distance</span></code></a>(distance)</p></td>
<td><p>Returns tips within specified distance from self</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.to_json" title="cogent3.core.tree.PhyloNode.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns json formatted string {‘newick’: with edges and distances, ‘edge_attributes’: }</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.to_rich_dict" title="cogent3.core.tree.PhyloNode.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>()</p></td>
<td><p>returns {‘newick’: with node names, ‘edge_attributes’: {‘tip1’: {‘length’: …}, …}}</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.total_descending_branch_length" title="cogent3.core.tree.PhyloNode.total_descending_branch_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">total_descending_branch_length</span></code></a>()</p></td>
<td><p>Returns total descending branch length from self</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.total_length" title="cogent3.core.tree.PhyloNode.total_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">total_length</span></code></a>()</p></td>
<td><p>returns the sum of all branch lengths in tree</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.traverse" title="cogent3.core.tree.PhyloNode.traverse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">traverse</span></code></a>([self_before, self_after, include_self])</p></td>
<td><p>Returns iterator over descendants.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.traverse_recursive" title="cogent3.core.tree.PhyloNode.traverse_recursive"><code class="xref py py-obj docutils literal notranslate"><span class="pre">traverse_recursive</span></code></a>([self_before, …])</p></td>
<td><p>Returns iterator over descendants.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.unrooted" title="cogent3.core.tree.PhyloNode.unrooted"><code class="xref py py-obj docutils literal notranslate"><span class="pre">unrooted</span></code></a>()</p></td>
<td><p>A tree with at least 3 children at the root.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.tree.PhyloNode.write" title="cogent3.core.tree.PhyloNode.write"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write</span></code></a>(filename[, with_distances, format])</p></td>
<td><p>Save the tree to filename</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 72%" />
<col style="width: 28%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>get_node_matching_name</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>unrooted_deepcopy</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.ancestors">
<code class="sig-name descname">ancestors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.ancestors" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all ancestors back to the root. Dynamically calculated.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.append">
<code class="sig-name descname">append</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">i</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Appends i to self.children, in-place, cleaning up refs.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.ascii_art">
<code class="sig-name descname">ascii_art</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">show_internal</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">compact</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.ascii_art" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string containing an ascii drawing of the tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>show_internal</strong></dt><dd><p>includes internal edge names.</p>
</dd>
<dt><strong>compact</strong></dt><dd><p>use exactly one line per tip.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.balanced">
<code class="sig-name descname">balanced</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.balanced" title="Permalink to this definition">¶</a></dt>
<dd><p>Tree ‘rooted’ here with no neighbour having &gt; 50% of the edges.</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>Using a balanced tree can substantially improve performance of
the likelihood calculations. Note that the resulting tree has a
different orientation with the effect that specifying clades or
stems for model parameterisation should be done using the
‘outgroup_name’ argument.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.bifurcating">
<code class="sig-name descname">bifurcating</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">eps</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name_unnamed</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.bifurcating" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrap multifurcating with a num of 2</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.child_groups">
<code class="sig-name descname">child_groups</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.child_groups" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list containing lists of children sharing a state.</p>
<p>In other words, returns runs of tip and nontip children.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.compare_by_names">
<code class="sig-name descname">compare_by_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.compare_by_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality test for trees by name</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.compare_by_subsets">
<code class="sig-name descname">compare_by_subsets</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">exclude_absent_taxa</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.compare_by_subsets" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of overlapping subsets where self and other differ.</p>
<p>Other is expected to be a tree object compatible with PhyloNode.</p>
<p>Note: names present in only one of the two trees will count as
mismatches: if you don’t want this behavior, strip out the non-matching
tips first.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.compare_by_tip_distances">
<code class="sig-name descname">compare_by_tip_distances</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">sample=None</em>, <em class="sig-param">dist_f=&lt;function distance_from_r&gt;</em>, <em class="sig-param">shuffle_f=&lt;bound method Random.shuffle of &lt;random.Random object&gt;&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.compare_by_tip_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares self to other using tip-to-tip distance matrices.</p>
<p>Value returned is dist_f(m1, m2) for the two matrices. Default is
to use the Pearson correlation coefficient, with +1 giving a distance
of 0 and -1 giving a distance of +1 (the madimum possible value).
Depending on the application, you might instead want to use
distance_from_r_squared, which counts correlations of both +1 and -1
as identical (0 distance).</p>
<p>Note: automatically strips out the names that don’t match (this is
necessary for this method because the distance between non-matching
names and matching names is undefined in the tree where they don’t
match, and because we need to reorder the names in the two trees to
match up the distance matrices).</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.compare_name">
<code class="sig-name descname">compare_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.compare_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Compares TreeNode by name</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">memo</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">_nil</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">'ignored'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of self using an iterative approach</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.copy_recursive">
<code class="sig-name descname">copy_recursive</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">memo</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">_nil</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">'ignored'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.copy_recursive" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns copy of self’s structure, including shallow copy of attrs.</p>
<p>constructor is ignored; required to support old tree unit tests.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.copy_topology">
<code class="sig-name descname">copy_topology</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.copy_topology" title="Permalink to this definition">¶</a></dt>
<dd><p>Copies only the topology and labels of a tree, not any extra data.</p>
<p>Useful when you want another copy of the tree with the same structure
and labels, but want to e.g. assign different branch lengths and
environments. Does not use deepcopy from the copy module, so _much_
faster than the copy() method.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.deepcopy">
<code class="sig-name descname">deepcopy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">memo</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">_nil</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">'ignored'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.deepcopy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of self using an iterative approach</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.descendant_array">
<code class="sig-name descname">descendant_array</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tip_list</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.descendant_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns numpy array with nodes in rows and descendants in columns.</p>
<p>A value of 1 indicates that the decendant is a descendant of that node/
A value of 0 indicates that it is not</p>
<p>Also returns a list of nodes in the same order as they are listed
in the array.</p>
<p>tip_list is a list of the names of the tips that will be considered,
in the order they will appear as columns in the final array. Internal
nodes will appear as rows in preorder traversal order.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.distance">
<code class="sig-name descname">distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns branch length between self and other.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.extend">
<code class="sig-name descname">extend</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">items</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Extends self.children by items, in-place, cleaning up refs.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_connecting_edges">
<code class="sig-name descname">get_connecting_edges</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name1</span></em>, <em class="sig-param"><span class="n">name2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_connecting_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a list of edges connecting two nodes.</p>
<p>If both are tips, the LCA is excluded from the result.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_connecting_node">
<code class="sig-name descname">get_connecting_node</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name1</span></em>, <em class="sig-param"><span class="n">name2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_connecting_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the last common ancestor of the two named edges.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_distances">
<code class="sig-name descname">get_distances</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">endpoints</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>The distance matrix as a dictionary.</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>Grabs the branch lengths (evolutionary distances) as
a complete matrix (i.e. a,b and b,a).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_edge_names">
<code class="sig-name descname">get_edge_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tip1name</span></em>, <em class="sig-param"><span class="n">tip2name</span></em>, <em class="sig-param"><span class="n">clade</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">stem</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">outgroup_name</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_edge_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the list of stem and/or sub tree (clade) edge name(s).
This is done by finding the common intersection, and then getting
the list of names. If the clade traverses the root, then use the
outgroup_name argument to ensure valid specification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>tip1/2name</strong></dt><dd><p>edge 1/2 names</p>
</dd>
<dt><strong>stem</strong></dt><dd><p>whether the name of the clade stem edge is returned.</p>
</dd>
<dt><strong>clade</strong></dt><dd><p>whether the names of the edges within the clade are
returned</p>
</dd>
<dt><strong>outgroup_name</strong></dt><dd><p>if provided the calculation is done on a version of
the tree re-rooted relative to the provided tip.</p>
</dd>
<dt><strong>Usage:</strong></dt><dd><p>The returned list can be used to specify subtrees for special
parameterisation. For instance, say you want to allow the primates
to have a different value of a particular parameter. In this case,
provide the results of this method to the parameter controller
method <cite>set_param_rule()</cite> along with the parameter name etc..</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_edge_vector">
<code class="sig-name descname">get_edge_vector</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_root</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_edge_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Collect the list of edges in postfix order</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>include_root</strong></dt><dd><p>specifies whether root edge included</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_figure">
<code class="sig-name descname">get_figure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">style</span><span class="o">=</span><span class="default_value">'square'</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_figure" title="Permalink to this definition">¶</a></dt>
<dd><p>gets Dendrogram for plotting the phylogeny</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>style</strong><span class="classifier">string</span></dt><dd><p>‘square’, ‘angular’, ‘radial’ or ‘circular’</p>
</dd>
<dt><strong>kwargs</strong></dt><dd><p>arguments passed to Dendrogram constructor</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_max_tip_tip_distance">
<code class="sig-name descname">get_max_tip_tip_distance</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_max_tip_tip_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the max tip tip distance between any pair of tips</p>
<p>Returns (dist, tip_names, internal_node)</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_newick">
<code class="sig-name descname">get_newick</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">with_distances</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">semicolon</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">escape_name</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">with_node_names</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_newick" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the newick string for this tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>with_distances</strong></dt><dd><p>whether branch lengths are included.</p>
</dd>
<dt><strong>semicolon</strong></dt><dd><p>end tree string with a semicolon</p>
</dd>
<dt><strong>escape_name</strong></dt><dd><p>if any of these characters []’”(),
nodes name, wrap the name in single quotes</p>
</dd>
<dt><strong>with_node_names</strong></dt><dd><p>includes internal node names (except ‘root’)</p>
</dd>
<dt><strong>NOTE: This method returns the Newick representation of this node</strong></dt><dd></dd>
<dt><strong>and its descendents.</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_newick_recursive">
<code class="sig-name descname">get_newick_recursive</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">with_distances</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">semicolon</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">escape_name</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">with_node_names</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_newick_recursive" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the newick string for this edge.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>with_distances</strong></dt><dd><p>whether branch lengths are included.</p>
</dd>
<dt><strong>semicolon</strong></dt><dd><p>end tree string with a semicolon</p>
</dd>
<dt><strong>escape_name</strong></dt><dd><p>if any of these characters []’”(),
nodes name, wrap the name in single quotes</p>
</dd>
<dt><strong>with_node_names</strong></dt><dd><p>includes internal node names (except ‘root’)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_node_matching_name">
<code class="sig-name descname">get_node_matching_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_node_matching_name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_node_names">
<code class="sig-name descname">get_node_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">includeself</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">tipsonly</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_node_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of edges from this edge - may or may not include self.
This node (or first connection) will be the first, and then they will
be listed in the natural traverse order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>includeself</strong><span class="classifier">bool</span></dt><dd><p>excludes self.name from the result</p>
</dd>
<dt><strong>tipsonly</strong><span class="classifier">bool</span></dt><dd><p>only tips returned</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_nodes_dict">
<code class="sig-name descname">get_nodes_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_nodes_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dict keyed by node name, value is node</p>
<p>Will raise TreeError if non-unique names are encountered</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_param_value">
<code class="sig-name descname">get_param_value</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">edge</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_param_value" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the parameter value for named edge</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_sub_tree">
<code class="sig-name descname">get_sub_tree</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name_list</span></em>, <em class="sig-param"><span class="n">ignore_missing</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">keep_root</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">tipsonly</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_sub_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>A new instance of a sub tree that contains all the otus that are
listed in name_list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ignore_missing</strong></dt><dd><p>if False, get_sub_tree will raise a ValueError if
name_list contains names that aren’t nodes in the tree</p>
</dd>
<dt><strong>keep_root</strong></dt><dd><p>if False, the root of the subtree will be the last common
ancestor of all nodes kept in the subtree. Root to tip distance is
then (possibly) different from the original tree. If True, the root to
tip distance remains constant, but root may only have one child node.</p>
</dd>
<dt><strong>tipsonly</strong></dt><dd><p>only tip names matching name_list are allowed</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_tip_names">
<code class="sig-name descname">get_tip_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">includeself</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_tip_names" title="Permalink to this definition">¶</a></dt>
<dd><p>return the list of the names of all tips contained by this edge</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.get_xml">
<code class="sig-name descname">get_xml</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.get_xml" title="Permalink to this definition">¶</a></dt>
<dd><p>Return XML formatted tree string.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.index_in_parent">
<code class="sig-name descname">index_in_parent</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.index_in_parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns index of self in parent.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.insert">
<code class="sig-name descname">insert</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span></em>, <em class="sig-param"><span class="n">i</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.insert" title="Permalink to this definition">¶</a></dt>
<dd><p>Inserts an item at specified position in self.children.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.is_root">
<code class="sig-name descname">is_root</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.is_root" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the current is a root, i.e. has no parent.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.is_tip">
<code class="sig-name descname">is_tip</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.is_tip" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if the current node is a tip, i.e. has no children.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.isroot">
<code class="sig-name descname">isroot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.isroot" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if root of a tree, i.e. no parent.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.istip">
<code class="sig-name descname">istip</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.istip" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if is tip, i.e. no children.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.iter_nontips">
<code class="sig-name descname">iter_nontips</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.iter_nontips" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over nontips descended from self, [] if none.</p>
<p>include_self, if True (default is False), will return the current
node as part of the list of nontips if it is a nontip.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.iter_tips">
<code class="sig-name descname">iter_tips</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.iter_tips" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterates over tips descended from self, [] if self is a tip.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.last_common_ancestor">
<code class="sig-name descname">last_common_ancestor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.last_common_ancestor" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds last common ancestor of self and other, or None.</p>
<p>Always tests by identity.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.lca">
<code class="sig-name descname">lca</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.lca" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds last common ancestor of self and other, or None.</p>
<p>Always tests by identity.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.length">
<em class="property">property </em><code class="sig-name descname">length</code><a class="headerlink" href="#cogent3.core.tree.PhyloNode.length" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.levelorder">
<code class="sig-name descname">levelorder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.levelorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs levelorder iteration over tree</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.lowest_common_ancestor">
<code class="sig-name descname">lowest_common_ancestor</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tipnames</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.lowest_common_ancestor" title="Permalink to this definition">¶</a></dt>
<dd><p>Lowest common ancestor for a list of tipnames</p>
<p>This should be around O(H sqrt(n)), where H is height and n is the
number of tips passed in.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.make_tree_array">
<code class="sig-name descname">make_tree_array</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dec_list</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.make_tree_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes an array with nodes in rows and descendants in columns.</p>
<p>A value of 1 indicates that the decendant is a descendant of that node/
A value of 0 indicates that it is not</p>
<p>also returns a list of nodes in the same order as they are listed
in the array</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.max_tip_tip_distance">
<code class="sig-name descname">max_tip_tip_distance</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.max_tip_tip_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the max distance between any pair of tips</p>
<p>Also returns the tip names  that it is between as a tuple</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.multifurcating">
<code class="sig-name descname">multifurcating</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num</span></em>, <em class="sig-param"><span class="n">eps</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name_unnamed</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.multifurcating" title="Permalink to this definition">¶</a></dt>
<dd><p>return a new tree with every node having num or few children</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>num</strong><span class="classifier">int</span></dt><dd><p>the number of children a node can have max</p>
</dd>
<dt><strong>eps</strong><span class="classifier">float</span></dt><dd><p>default branch length to set if self or constructor is of
PhyloNode type</p>
</dd>
<dt><strong>constructor</strong></dt><dd><p>a TreeNode or subclass constructor. If None, uses self</p>
</dd>
<dt><strong>name_unnamed</strong><span class="classifier">bool</span></dt><dd><p>names unnamed nodes</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.name_unnamed_nodes">
<code class="sig-name descname">name_unnamed_nodes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.name_unnamed_nodes" title="Permalink to this definition">¶</a></dt>
<dd><p>sets the Data property of unnamed nodes to an arbitrary value</p>
<p>Internal nodes are often unnamed and so this function assigns a
value for referencing.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.non_tip_children">
<code class="sig-name descname">non_tip_children</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.non_tip_children" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns direct children in self that have descendants.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.nontips">
<code class="sig-name descname">nontips</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.nontips" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns nontips descended from self.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.parent">
<em class="property">property </em><code class="sig-name descname">parent</code><a class="headerlink" href="#cogent3.core.tree.PhyloNode.parent" title="Permalink to this definition">¶</a></dt>
<dd><p>Accessor for parent.</p>
<p>If using an algorithm that accesses parent a lot, it will be much
faster to access self._parent directly, but don’t do it if mutating
self._parent! (or, if you must, remember to clean up the refs).</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.pop">
<code class="sig-name descname">pop</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">index</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.pop" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns and deletes child of self at index (default: -1)</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.postorder">
<code class="sig-name descname">postorder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.postorder" title="Permalink to this definition">¶</a></dt>
<dd><p>performs postorder iteration over tree.</p>
<p class="rubric">Notes</p>
<p>This is somewhat inelegant compared to saving the node and its index
on the stack, but is 30% faster in the average case and 3x faster in
the worst case (for a comb tree).</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.pre_and_postorder">
<code class="sig-name descname">pre_and_postorder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.pre_and_postorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs iteration over tree, visiting node before and after.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.preorder">
<code class="sig-name descname">preorder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.preorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs preorder iteration over tree.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.prune">
<code class="sig-name descname">prune</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.prune" title="Permalink to this definition">¶</a></dt>
<dd><p>Reconstructs correct tree after nodes have been removed.</p>
<p>Internal nodes with only one child will be removed and new connections
and Branch lengths will be made to reflect change.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.reassign_names">
<code class="sig-name descname">reassign_names</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapping</span></em>, <em class="sig-param"><span class="n">nodes</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.reassign_names" title="Permalink to this definition">¶</a></dt>
<dd><p>Reassigns node names based on a mapping dict</p>
<p>mapping : dict, old_name -&gt; new_name
nodes : specific nodes for renaming (such as just tips, etc…)</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.remove">
<code class="sig-name descname">remove</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.remove" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes node by name instead of identity.</p>
<p>Returns True if node was present, False otherwise.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.remove_deleted">
<code class="sig-name descname">remove_deleted</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">is_deleted</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.remove_deleted" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes all nodes where is_deleted tests true.</p>
<p>Internal nodes that have no children as a result of removing deleted
are also removed.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.remove_node">
<code class="sig-name descname">remove_node</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.remove_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes node by identity instead of value.</p>
<p>Returns True if node was present, False otherwise.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.root">
<code class="sig-name descname">root</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.root" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns root of the tree self is in. Dynamically calculated.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.root_at_midpoint">
<code class="sig-name descname">root_at_midpoint</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.root_at_midpoint" title="Permalink to this definition">¶</a></dt>
<dd><p>return a new tree rooted at midpoint of the two tips farthest apart</p>
<p>this fn doesn’t preserve the internal node naming or structure,
but does keep tip to tip distances correct.  uses unrooted_deepcopy()</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.rooted_at">
<code class="sig-name descname">rooted_at</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">edge_name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.rooted_at" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new tree rooted at the provided node.</p>
<dl class="simple">
<dt>Usage:</dt><dd><p>This can be useful for drawing unrooted trees with an orientation
that reflects knowledge of the true root location.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.rooted_with_tip">
<code class="sig-name descname">rooted_with_tip</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">outgroup_name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.rooted_with_tip" title="Permalink to this definition">¶</a></dt>
<dd><p>A new tree with the named tip as one of the root’s children</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.same_shape">
<code class="sig-name descname">same_shape</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.same_shape" title="Permalink to this definition">¶</a></dt>
<dd><p>Ignores lengths and order, so trees should be sorted first</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.same_topology">
<code class="sig-name descname">same_topology</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.same_topology" title="Permalink to this definition">¶</a></dt>
<dd><p>Tests whether two trees have the same topology.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.scale_branch_lengths">
<code class="sig-name descname">scale_branch_lengths</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">max_length</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">ultrametric</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.scale_branch_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Scales BranchLengths in place to integers for ascii output.</p>
<p>Warning: tree might not be exactly the length you specify.</p>
<p>Set ultrametric=True if you want all the root-tip distances to end
up precisely the same.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.separation">
<code class="sig-name descname">separation</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.separation" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of edges separating self and other.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.set_max_tip_tip_distance">
<code class="sig-name descname">set_max_tip_tip_distance</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.set_max_tip_tip_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Propagate tip distance information up the tree</p>
<p>This method was originally implemented by Julia Goodrich with the intent
of being able to determine max tip to tip distances between nodes on
large trees efficiently. The code has been modified to track the
specific tips the distance is between</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.set_param_value">
<code class="sig-name descname">set_param_value</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">param</span></em>, <em class="sig-param"><span class="n">edge</span></em>, <em class="sig-param"><span class="n">value</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.set_param_value" title="Permalink to this definition">¶</a></dt>
<dd><p>set’s the value for param at named edge</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.set_tip_distances">
<code class="sig-name descname">set_tip_distances</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.set_tip_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets distance from each node to the most distant tip.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.siblings">
<code class="sig-name descname">siblings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.siblings" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all nodes that are children of the same parent as self.</p>
<p>Note: excludes self from the list. Dynamically calculated.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.sorted">
<code class="sig-name descname">sorted</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sort_order</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.sorted" title="Permalink to this definition">¶</a></dt>
<dd><p>An equivalent tree sorted into a standard order. If this is not
specified then alphabetical order is used.  At each node starting from
root, the algorithm will try to put the descendant which contains the
lowest scoring tip on the left.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns set of names that descend from specified node</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.subsets">
<code class="sig-name descname">subsets</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.subsets" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns all sets of names that come from specified node and its kids</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.tip_children">
<code class="sig-name descname">tip_children</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.tip_children" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns direct children of self that are tips.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.tip_to_tip_distances">
<code class="sig-name descname">tip_to_tip_distances</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">endpoints</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">default_length</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.tip_to_tip_distances" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns distance matrix between all pairs of tips, and a tip order.</p>
<p>Warning: .__start and .__stop added to self and its descendants.</p>
<p>tip_order contains the actual node objects, not their names (may be
confusing in some cases).</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.tips">
<code class="sig-name descname">tips</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.tips" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns tips descended from self, [] if self is a tip.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.tips_within_distance">
<code class="sig-name descname">tips_within_distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">distance</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.tips_within_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns tips within specified distance from self</p>
<p>Branch lengths of None will be interpreted as 0</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.to_json">
<code class="sig-name descname">to_json</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>returns json formatted string {‘newick’: with edges and distances, ‘edge_attributes’: }</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.to_rich_dict">
<code class="sig-name descname">to_rich_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.to_rich_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>returns {‘newick’: with node names,
‘edge_attributes’: {‘tip1’: {‘length’: …}, …}}</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.total_descending_branch_length">
<code class="sig-name descname">total_descending_branch_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.total_descending_branch_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns total descending branch length from self</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.total_length">
<code class="sig-name descname">total_length</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.total_length" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the sum of all branch lengths in tree</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.traverse">
<code class="sig-name descname">traverse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self_before</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">self_after</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.traverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns iterator over descendants. Iterative: safe for large trees.</p>
<p>self_before includes each node before its descendants if True.
self_after includes each node after its descendants if True.
include_self includes the initial node if True.</p>
<p>self_before and self_after are independent. If neither is True, only
terminal nodes will be returned.</p>
<p>Note that if self is terminal, it will only be included once even if
self_before and self_after are both True.</p>
<p>This is a depth-first traversal. Since the trees are not binary,
preorder and postorder traversals are possible, but inorder traversals
would depend on the data in the tree and are not handled here.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.traverse_recursive">
<code class="sig-name descname">traverse_recursive</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">self_before</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">self_after</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">include_self</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.traverse_recursive" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns iterator over descendants. IMPORTANT: read notes below.</p>
<p>traverse_recursive is slower than traverse, and can lead to stack
errors. However, you _must_ use traverse_recursive if you plan to
modify the tree topology as you walk over it (e.g. in post-order),
because the iterative methods use their own stack that is not updated
if you alter the tree.</p>
<p>self_before includes each node before its descendants if True.
self_after includes each node after its descendants if True.
include_self includes the initial node if True.</p>
<p>self_before and self_after are independent. If neither is True, only
terminal nodes will be returned.</p>
<p>Note that if self is terminal, it will only be included once even if
self_before and self_after are both True.</p>
<p>This is a depth-first traversal. Since the trees are not binary,
preorder and postorder traversals are possible, but inorder traversals
would depend on the data in the tree and are not handled here.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.unrooted">
<code class="sig-name descname">unrooted</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.unrooted" title="Permalink to this definition">¶</a></dt>
<dd><p>A tree with at least 3 children at the root.</p>
</dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.unrooted_deepcopy">
<code class="sig-name descname">unrooted_deepcopy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">constructor</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">parent</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.unrooted_deepcopy" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="cogent3.core.tree.PhyloNode.write">
<code class="sig-name descname">write</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span></em>, <em class="sig-param"><span class="n">with_distances</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">format</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.tree.PhyloNode.write" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the tree to filename</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>filename</strong></dt><dd><p>self</p>
</dd>
<dt><strong>with_distances</strong></dt><dd><p>whether branch lengths are included in string.</p>
</dd>
<dt><strong>format</strong></dt><dd><p>default is newick, xml and json are alternate. Argument overrides
the filename suffix. All attributes are saved in the xml format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 3.2.1 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>