<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2022.01.06 documentation</title>

<link href="../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
<link rel="stylesheet" href="../../_static/css/docs.css" type="text/css" />
<link rel="stylesheet" href="../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../_static/sg_gallery.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />

<link rel="stylesheet" href="../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../.././_static/pygments.css">
<link rel="stylesheet" href="../.././_static/css/common.css">
<link rel="stylesheet" href="../.././_static/css/link.css">
<link rel="stylesheet" href="../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
<script src="../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../.././_static/js/initializeHighlight.js"></script>
<script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
<script src="../../_static/jquery.js"></script>
<script src="../../_static/underscore.js"></script>
<script src="../../_static/doctools.js"></script>
<script src="../../_static/thebelab-helper.js"></script>
<script src="../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.20.0/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
<script src="../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3/discussions" class="nav-link" title="Forum">Forum</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-using-conda">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../install.html#manual-creation-of-the-conda-environment">Manual creation of the <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#overview">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/warning.html">Warning – apps are alpha!</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#define-the-apps">Define the apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#using-apps-like-functions">Using apps like functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-log-the-settings-and-data-analysed">You can log the settings and data analysed</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/dstore.html">Specifying data for analysis</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#a-read-only-data-store">A read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#data-store-members">Data store “members”</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#showing-the-last-few-members">Showing the last few members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#filtering-a-data-store-for-specific-members">Filtering a data store for specific members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#tinydb-data-stores-are-special">TinyDB data stores are special</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/available-apps.html">What apps are there?</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-progressive-align-app">The <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-model-app">The <code class="docutils literal notranslate"><span class="pre">model</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-hypothesis-app">The <code class="docutils literal notranslate"><span class="pre">hypothesis</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-natsel-apps">The <code class="docutils literal notranslate"><span class="pre">natsel</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-user-function-apps">The <code class="docutils literal notranslate"><span class="pre">user_function</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/user_function.html">Turn your functions into composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-user-app-with-a-different-output-type">A user app with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Cookbook</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="manipulating_biological_data.html">Manipulating biological data</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul class="current">
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html">Loading an alignment from a file</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#making-a-sequence-collection-from-standard-python-objects">Making a sequence collection from standard python objects</a></li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5 current"><a class="current reference internal" href="#">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="DNA_and_RNA_sequences.html"><code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a><ul>
<li class="toctree-l6"><a class="reference internal" href="DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="annotations.html">Annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="annotations.html#annotations-with-coordinates">Annotations with coordinates</a><ul>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#automated-introduction-from-reading-genbank-files">Automated introduction from reading genbank files</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#customising-annotation-construction-from-reading-a-genbank-file">Customising annotation construction from reading a genbank file</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#creating-directly-on-a-sequence">Creating directly on a sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#via">Via</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#adding-as-a-series-or-item-wise">Adding as a series or item-wise</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#taking-the-union-of-annotations">Taking the union of annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#getting-annotation-coordinates">Getting annotation coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#annotations-have-shadows">Annotations have shadows</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#adding-to-a-sequence-member-of-an-alignment">Adding to a sequence member of an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#adding-to-an-alignment">Adding to an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#slicing-sequences-and-alignments-by-annotations">Slicing sequences and alignments by annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#slicing-by-pseudo-feature-or-feature-series">Slicing by pseudo-feature or feature series</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#slice-series-must-not-be-overlapping">Slice series must not be overlapping</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#you-can-slice-an-annotation-itself">You can slice an annotation itself</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#sequence-vs-alignment-slicing">Sequence vs Alignment slicing</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#copying-annotations">Copying annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#querying">Querying</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#querying-produces-objects-only-valid-for-their-source">Querying produces objects only valid for their source</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#querying-for-absent-annotation">Querying for absent annotation</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#querying-features-that-span-gaps-in-alignments">Querying features that span gaps in alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#as-one-span-unifies-features-with-discontinuous-alignment-coordinates"><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#behaviour-of-annotations-on-nucleic-acid-sequences">Behaviour of annotations on nucleic acid sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#masking-annotated-regions">Masking annotated regions</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#masking-annotated-regions-on-alignments">Masking annotated regions on alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#you-can-take-mask-of-the-shadow">You can take mask of the shadow</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#what-features-of-a-certain-type-are-available">What features of a certain type are available?</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#getting-all-features-of-a-type-or-everything-but-that-type">Getting all features of a type, or everything but that type</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#getting-sequence-features-when-you-have-an-alignment-object">Getting sequence features when you have an alignment object</a></li>
<li class="toctree-l6"><a class="reference internal" href="annotations.html#annotation-display-on-sequences">Annotation display on sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html">Translate DNA sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#translating-a-dna-sequence-with-a-terminating-stop-codon">Translating a DNA sequence with a terminating stop codon</a><ul>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#by-removing-the-trailing-stop-codon-first">By removing the trailing stop codon first</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#by-slicing-the-dnasequence-first">By slicing the <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> first</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-delimited-data-without-a-header-line">Loading delimited data without a header line</a><ul>
<li class="toctree-l5"><a class="reference internal" href="tables.html#using-load-delimited">Using <code class="docutils literal notranslate"><span class="pre">load_delimited()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#using-filteringparser">Using <code class="docutils literal notranslate"><span class="pre">FilteringParser</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#loading-only-some-rows">Loading only some rows</a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../examples/index.html">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/seq_features.html">Advanced sequence handling</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/complete_seq_features.html">Annotating alignments and sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#modelling-evolution">Modelling Evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../examples/codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#utilities">Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/parallel.html">Parallel computations</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/parallel.html#parallel-computation-on-a-single-computer">Parallel computation on a single computer</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/parallel.html#using-app-apply-to">Using <code class="docutils literal notranslate"><span class="pre">app.apply_to()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/parallel.html#directly-using-cogent3-util-parallel-map">Directly using <code class="docutils literal notranslate"><span class="pre">cogent3.util.parallel.map()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/parallel.html#parallel-computation-on-multiple-computers">Parallel computation on multiple computers</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_model.html">get_model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html">Sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html">Dotplot basics</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html">Coevolution analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-support.html">Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-square.html">Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-radial.html">Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-circular.html">Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-angular.html">Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../genindex.html">Index</a></li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="sequence-collections-and-alignments">
<h1>Sequence Collections and Alignments<a class="headerlink" href="#sequence-collections-and-alignments" title="Permalink to this headline">¶</a></h1>
<p>For loading collections of unaligned or aligned sequences see <a class="reference internal" href="loading_sequences.html#load-seqs"><span class="std std-ref">Loading an alignment from a file</span></a>.</p>
<section id="what-s-the-difference-between-alignment-and-arrayalignment">
<h2>What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?<a class="headerlink" href="#what-s-the-difference-between-alignment-and-arrayalignment" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> class can be annotated, meaning you can add annotations to an Alignment or it’s member sequences and you can slice the alignment via those objects. This capability is achieved, under the hood, by having the individual sequences represent gaps as a “span”, rather than explicitly as a “-” character in the sequence itself. This representation is also efficient for very long sequences.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> class cannot be annotated. The class represents its sequences as a <code class="docutils literal notranslate"><span class="pre">numpy.ndarray</span></code> instance. Thus, the gaps are included as a state in the array. This class is better at handling a lot of sequences and should typically be faster. This is the default class returned by the <code class="docutils literal notranslate"><span class="pre">load_aligned_seqs</span></code> and <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code> functions. (See <a class="reference internal" href="loading_sequences.html#load-seqs"><span class="std std-ref">Loading an alignment from a file</span></a> for details.)</p>
<p>You can change alignment types using the <code class="docutils literal notranslate"><span class="pre">to_type()</span></code> method.</p>
</section>
<section id="basic-collection-objects">
<h2>Basic Collection objects<a class="headerlink" href="#basic-collection-objects" title="Permalink to this headline">¶</a></h2>
<section id="constructing-a-sequencecollection-or-alignment-object-from-strings">
<h3>Constructing a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from strings<a class="headerlink" href="#constructing-a-sequencecollection-or-alignment-object-from-strings" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span><span class="p">,</span> <span class="n">make_unaligned_seqs</span>

<span class="n">dna</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCGCC&quot;</span><span class="p">}</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">))</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">make_unaligned_seqs</span><span class="p">(</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;cogent3.core.alignment.ArrayAlignment&#39;&gt;
&lt;class &#39;cogent3.core.alignment.SequenceCollection&#39;&gt;
</pre></div>
</div>
</div>
</div>
</section>
<section id="constructing-a-arrayalignment-using-make-aligned-seqs">
<h3>Constructing a <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> using <code class="docutils literal notranslate"><span class="pre">make_aligned_seqs</span></code><a class="headerlink" href="#constructing-a-arrayalignment-using-make-aligned-seqs" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">dna</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCGCC&quot;</span><span class="p">}</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">seqs</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;cogent3.core.alignment.ArrayAlignment&#39;&gt;
&gt;seq1
ATGACC
&gt;seq2
ATCGCC

</pre></div>
</div>
</div>
</div>
</section>
<section id="converting-a-sequencecollection-to-fasta-format">
<h3>Converting a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> to FASTA format<a class="headerlink" href="#converting-a-sequencecollection-to-fasta-format" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span>

<span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;NineBande
GCAAGGCGCCAACAGAGCAGATGGGCTGAAAGTAAGGAAACATGTAATGATAGGCAGACT
&gt;Mouse
GCAGTGAGCCAGCAGAGCAGATGGGCTGCAAGTAAAGGAACATGTAACGACAGGCAGGTT
&gt;Human
GCAAGGAGCCAACATAACAGATGGGCTGGAAGTAAGGAAACATGTAATGATAGGCGGACT
&gt;HowlerMon
GCAAGGAGCCAACATAACAGATGGGCTGAAAGTGAGGAAACATGTAATGATAGGCAGACT
&gt;DogFaced
GCAAGGAGCCAGCAGAACAGATGGGTTGAAACTAAGGAAACATGTAATGATAGGCAGACT

</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="adding-new-sequences-to-an-existing-collection-or-alignment">
<h2>Adding new sequences to an existing collection or alignment<a class="headerlink" href="#adding-new-sequences-to-an-existing-collection-or-alignment" title="Permalink to this headline">¶</a></h2>
<p>New sequences can be either appended or inserted using the <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> method. More than one sequence can be added at the same time. Note that <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> does not modify the existing collection/alignment, it creates a new one.</p>
<section id="appending-the-sequences">
<h3>Appending the sequences<a class="headerlink" href="#appending-the-sequences" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> without additional parameters will append the sequences to the end of the collection/alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="n">new_seqs</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq0&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGT-AGG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq4&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGCC------&quot;</span><span class="p">)],</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span>
<span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_seqs</span><span class="p">(</span><span class="n">new_seqs</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG
&gt;seq0
ATG-AGT-AGG
&gt;seq4
ATGCC------

</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The order is not preserved if you use <code class="docutils literal notranslate"><span class="pre">to_fasta</span></code> method, which sorts sequences by name.</p>
</div>
</section>
<section id="inserting-the-sequences">
<h3>Inserting the sequences<a class="headerlink" href="#inserting-the-sequences" title="Permalink to this headline">¶</a></h3>
<p>Sequences can be inserted into an alignment at the specified position using either the <code class="docutils literal notranslate"><span class="pre">before_name</span></code> or <code class="docutils literal notranslate"><span class="pre">after_name</span></code> arguments.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_seqs</span><span class="p">(</span><span class="n">new_seqs</span><span class="p">,</span> <span class="n">before_name</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_seqs</span><span class="p">(</span><span class="n">new_seqs</span><span class="p">,</span> <span class="n">after_name</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAA------
&gt;seq0
ATG-AGT-AGG
&gt;seq4
ATGCC------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq0
ATG-AGT-AGG
&gt;seq4
ATGCC------
&gt;seq3
AT--AG-GATG

</pre></div>
</div>
</div>
</div>
</section>
<section id="inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence">
<h3>Inserting sequence(s) based on their alignment to a reference sequence<a class="headerlink" href="#inserting-sequence-s-based-on-their-alignment-to-a-reference-sequence" title="Permalink to this headline">¶</a></h3>
<p>Already aligned sequences can be added to an existing <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object and aligned at the same time using the <code class="docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code> method. The alignment is performed based on their alignment to a reference sequence (which must be present in both alignments). The method assumes the first sequence in <code class="docutils literal notranslate"><span class="pre">ref_aln.names[0]</span></code> is the reference.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ref_aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATAGGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq0&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGCG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq4&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGCTGGG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_from_ref_aln</span><span class="p">(</span><span class="n">ref_aln</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAA------
&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG
&gt;seq0
AT--G--AGCG
&gt;seq4
AT--GC-TGGG

</pre></div>
</div>
</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">add_from_ref_aln</span></code> has the same arguments as <code class="docutils literal notranslate"><span class="pre">add_seqs</span></code> so <code class="docutils literal notranslate"><span class="pre">before_name</span></code> and <code class="docutils literal notranslate"><span class="pre">after_name</span></code> can be used to insert the new sequences at the desired position.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method does not work with the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code> class.</p>
</div>
<section id="removing-all-columns-with-gaps-in-a-named-sequence">
<h4>Removing all columns with gaps in a named sequence<a class="headerlink" href="#removing-all-columns-with-gaps-in-a-named-sequence" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA---TG-&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_degapped_relative_to</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAATG
&gt;seq2
ATG-AAT
&gt;seq3
AT--AAT

</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="the-elements-of-a-collection-or-alignment">
<h2>The elements of a collection or alignment<a class="headerlink" href="#the-elements-of-a-collection-or-alignment" title="Permalink to this headline">¶</a></h2>
<section id="accessing-individual-sequences-from-a-collection-or-alignment-by-name">
<h3>Accessing individual sequences from a collection or alignment by name<a class="headerlink" href="#accessing-individual-sequences-from-a-collection-or-alignment-by-name" title="Permalink to this headline">¶</a></h3>
<p>Using the <code class="docutils literal notranslate"><span class="pre">get_seq</span></code> method allows for extracting an unaligned sequence from a collection or alignment by name.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="p">[(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">)],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
    <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">name</span>
<span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">is_gapped</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
<p>Alternatively, if you want to extract the aligned (i.e., gapped) sequence from an alignment, you can use <code class="docutils literal notranslate"><span class="pre">get_gapped_seq</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_gapped_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">is_gapped</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>ATGAA------
</pre></div>
</div>
</div>
</div>
<p>To see the names of the sequences in a sequence collection, you can use either the <code class="docutils literal notranslate"><span class="pre">Names</span></code> attribute or <code class="docutils literal notranslate"><span class="pre">get_seq_names</span></code> method.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">names</span>
<span class="n">aln</span><span class="o">.</span><span class="n">names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;seq1&#39;, &#39;seq2&#39;, &#39;seq3&#39;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="slice-the-sequences-from-an-alignment-like-a-list">
<h3>Slice the sequences from an alignment like a list<a class="headerlink" href="#slice-the-sequences-from-an-alignment-like-a-list" title="Permalink to this headline">¶</a></h3>
<p>The usual approach is to access a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object as a dictionary, obtaining the individual sequences using the titles as “keys” (above).  However, one can also iterate through the collection like a list.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span><span class="p">,</span> <span class="n">load_aligned_seqs</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">my_seq</span> <span class="o">=</span> <span class="n">seqs</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">my_seq</span><span class="p">[:</span><span class="mi">24</span><span class="p">]</span>
<span class="nb">str</span><span class="p">(</span><span class="n">my_seq</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>
<span class="nb">type</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">24</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">seqs</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="mi">24</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>TGTGGCACAAATACTCATGCCAGC
</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-a-subset-of-sequences-from-the-alignment">
<h3>Getting a subset of sequences from the alignment<a class="headerlink" href="#getting-a-subset-of-sequences-from-the-alignment" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">names</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">])</span>
<span class="n">new</span><span class="o">.</span><span class="n">names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[&#39;Human&#39;, &#39;HowlerMon&#39;]
</pre></div>
</div>
</div>
</div>
<p>Note, if you set <code class="docutils literal notranslate"><span class="pre">array_align=False</span></code>, then the subset contain references to the original sequences, not copies.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">,</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">)</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;HowlerMon&quot;</span><span class="p">])</span>
<span class="nb">id</span><span class="p">(</span><span class="n">new</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">))</span> <span class="o">==</span> <span class="nb">id</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="alignments">
<h2>Alignments<a class="headerlink" href="#alignments" title="Permalink to this headline">¶</a></h2>
<section id="creating-an-alignment-object-from-a-sequencecollection">
<h3>Creating an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object from a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code><a class="headerlink" href="#creating-an-alignment-object-from-a-sequencecollection" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">fasta_1</span> <span class="o">=</span> <span class="n">seq</span>
<span class="n">fasta_2</span> <span class="o">=</span> <span class="n">aln</span>
<span class="k">assert</span> <span class="n">fasta_1</span> <span class="o">==</span> <span class="n">fasta_2</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="convert-alignment-to-dna-rna-or-protein-moltypes">
<h3>Convert alignment to DNA, RNA or PROTEIN moltypes<a class="headerlink" href="#convert-alignment-to-dna-rna-or-protein-moltypes" title="Permalink to this headline">¶</a></h3>
<p>This is useful if you’ve loaded a sequence alignment without specifying the moltype and later need to convert it.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG---&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;CCTGGG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="n">dna</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_dna</span><span class="p">()</span>
<span class="n">dna</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">b</td><td><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span></td></tr>
<tr><td class="label">a</td><td><span class="A_dna">A</span><span class="C_dna">.</span><span class="G_dna">G</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 6 dna alignment</i></p>
</div></div></div>
</div>
<p>To RNA</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;ACG---&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;CCUGGG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="n">rna</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_rna</span><span class="p">()</span>
<span class="n">rna</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .U_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_rna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">b</td><td><span class="C_rna">C</span><span class="C_rna">C</span><span class="U_rna">U</span><span class="G_rna">G</span><span class="G_rna">G</span><span class="G_rna">G</span></td></tr>
<tr><td class="label">a</td><td><span class="A_rna">A</span><span class="C_rna">.</span><span class="G_rna">G</span><span class="ambig_rna">-</span><span class="ambig_rna">-</span><span class="ambig_rna">-</span></td></tr>
</table>
<p><i>2 x 6 rna alignment</i></p>
</div></div></div>
</div>
<p>To PROTEIN</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">data</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;TYV&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TE-&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="n">prot</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_protein</span><span class="p">()</span>
<span class="n">prot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .A_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.c3align .C_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: orange; }
.c3align .D_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .E_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .F_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #ff6600; }
.c3align .G_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.c3align .H_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .I_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.c3align .K_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .L_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.c3align .M_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: orange; }
.c3align .N_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #993300; }
.c3align .P_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #cc0099; }
.c3align .Q_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #993300; }
.c3align .R_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .S_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009900; }
.c3align .T_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009900; }
.c3align .U_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .V_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #009999; }
.c3align .W_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #ff6600; }
.c3align .Y_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #ff6600; }
.c3align .terminal_ambig_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_protein{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="T_protein">T</span><span class="Y_protein">Y</span><span class="V_protein">V</span></td></tr>
<tr><td class="label">y</td><td><span class="T_protein">.</span><span class="E_protein">E</span><span class="ambig_protein">-</span></td></tr>
</table>
<p><i>2 x 3 protein alignment</i></p>
</div></div></div>
</div>
</section>
<section id="handling-gaps">
<h3>Handling gaps<a class="headerlink" href="#handling-gaps" title="Permalink to this headline">¶</a></h3>
<section id="remove-all-gaps-from-an-alignment-in-fasta-format">
<h4>Remove all gaps from an alignment in FASTA format<a class="headerlink" href="#remove-all-gaps-from-an-alignment-in-fasta-format" title="Permalink to this headline">¶</a></h4>
<p>This necessarily returns a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="n">degapped</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">degap</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">degapped</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&lt;class &#39;cogent3.core.alignment.SequenceCollection&#39;&gt;
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="writing-sequences-to-file">
<h3>Writing sequences to file<a class="headerlink" href="#writing-sequences-to-file" title="Permalink to this headline">¶</a></h3>
<p>Both collection and alignment objects have a <code class="docutils literal notranslate"><span class="pre">write</span></code> method. The output format is inferred from the filename suffix,</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">dna</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCGCC&quot;</span><span class="p">}</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">dna</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;sample.fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>or by the <code class="docutils literal notranslate"><span class="pre">format</span></code> argument.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.util.io</span> <span class="kn">import</span> <span class="n">remove_files</span>

<span class="n">remove_files</span><span class="p">([</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="s2">&quot;sample.fasta&quot;</span><span class="p">],</span> <span class="n">error_on_missing</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="converting-an-alignment-to-fasta-format">
<h3>Converting an alignment to FASTA format<a class="headerlink" href="#converting-an-alignment-to-fasta-format" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">fasta_align</span> <span class="o">=</span> <span class="n">aln</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="converting-an-alignment-into-phylip-format">
<h3>Converting an alignment into Phylip format<a class="headerlink" href="#converting-an-alignment-into-phylip-format" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">got</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_phylip</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">got</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>5  60
NineBande GCAAGGCGCCAACAGAGCAGATGGGCTGAAAGTAAGGAAACATGTAATGATAGGCAGACT
Mouse     GCAGTGAGCCAGCAGAGCAGATGGGCTGCAAGTAAAGGAACATGTAACGACAGGCAGGTT
Human     GCAAGGAGCCAACATAACAGATGGGCTGGAAGTAAGGAAACATGTAATGATAGGCGGACT
HowlerMon GCAAGGAGCCAACATAACAGATGGGCTGAAAGTGAGGAAACATGTAATGATAGGCAGACT
DogFaced  GCAAGGAGCCAGCAGAACAGATGGGTTGAAACTAAGGAAACATGTAATGATAGGCAGACT

</pre></div>
</div>
</div>
</div>
</section>
<section id="converting-an-alignment-to-a-list-of-strings">
<h3>Converting an alignment to a list of strings<a class="headerlink" href="#converting-an-alignment-to-a-list-of-strings" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>
<span class="kn">from</span> <span class="nn">cogent3.core.alignment</span> <span class="kn">import</span> <span class="n">Alignment</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">Alignment</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">string_list</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
</section>
<section id="slicing-an-alignment">
<h2>Slicing an alignment<a class="headerlink" href="#slicing-an-alignment" title="Permalink to this headline">¶</a></h2>
<section id="by-rows-sequences">
<h3>By rows (sequences)<a class="headerlink" href="#by-rows-sequences" title="Permalink to this headline">¶</a></h3>
<p>An <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> can be sliced</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
TGTGGCACAAATACTCATGCCAGC
&gt;HowlerMon
TGTGGCACAAATACTCATGCCAGC
&gt;Mouse
TGTGGCACAGATGCTCATGCCAGC
&gt;NineBande
TGTGGCACAAATACTCATGCCAAC
&gt;DogFaced
TGTGGCACAAATACTCATGCCAAC

</pre></div>
</div>
</div>
</div>
<p>but a <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code> cannot be sliced</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_unaligned_seqs</span>

<span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">TypeError</span><span class="g g-Whitespace">                                 </span>Traceback (most recent call last)
<span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">folders</span><span class="o">/</span><span class="mi">24</span><span class="o">/</span><span class="mi">8</span><span class="n">k48jl6d249_n_qfxwsl6xvm0000gn</span><span class="o">/</span><span class="n">T</span><span class="o">/</span><span class="n">ipykernel_7639</span><span class="o">/</span><span class="mf">450684702.</span><span class="n">py</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="n">fn</span> <span class="o">=</span> <span class="s2">&quot;data/long_testseqs.fasta&quot;</span>
<span class="g g-Whitespace">      </span><span class="mi">4</span> <span class="n">seqs</span> <span class="o">=</span> <span class="n">load_unaligned_seqs</span><span class="p">(</span><span class="n">fn</span><span class="p">)</span>
<span class="ne">----&gt; </span><span class="mi">5</span> <span class="nb">print</span><span class="p">(</span><span class="n">seqs</span><span class="p">[:</span><span class="mi">24</span><span class="p">])</span>

<span class="ne">TypeError</span>: &#39;SequenceCollection&#39; object is not subscriptable
</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-a-single-column-from-an-alignment">
<h3>Getting a single column from an alignment<a class="headerlink" href="#getting-a-single-column-from-an-alignment" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/test.paml&quot;</span><span class="p">)</span>
<span class="n">column_four</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="getting-a-region-of-contiguous-columns">
<h3>Getting a region of contiguous columns<a class="headerlink" href="#getting-a-region-of-contiguous-columns" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">region</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">50</span><span class="p">:</span><span class="mi">70</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="iterating-over-alignment-positions">
<h3>Iterating over alignment positions<a class="headerlink" href="#iterating-over-alignment-positions" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="n">col</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">113</span><span class="p">:</span><span class="mi">115</span><span class="p">]</span><span class="o">.</span><span class="n">iter_positions</span><span class="p">()</span>
<span class="nb">type</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
<span class="nb">list</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[[ByteSequence(A), ByteSequence(A), ByteSequence(A)],
 [ByteSequence(T), ByteSequence(-), ByteSequence(-)]]
</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-codon-3rd-positions-from-alignment">
<h3>Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#getting-codon-3rd-positions-from-alignment" title="Permalink to this headline">¶</a></h3>
<p>We’ll do this by specifying the position indices of interest, creating a sequence <code class="docutils literal notranslate"><span class="pre">Feature</span></code> and using that to extract the positions.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATG---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATGATG&quot;</span><span class="p">},</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="p">))[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">])</span>
<span class="n">indices</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln</span><span class="p">))[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">]]</span>
<span class="n">indices</span>
<span class="n">pos3</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;pos3&quot;</span><span class="p">,</span> <span class="s2">&quot;pos3&quot;</span><span class="p">,</span> <span class="n">indices</span><span class="p">)</span>
<span class="n">pos3</span> <span class="o">=</span> <span class="n">pos3</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pos3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
GGG-
&gt;seq2
GGGG

</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-codon-3rd-positions-from-arrayalignment">
<h3>Getting codon 3rd positions from <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code><a class="headerlink" href="#getting-codon-3rd-positions-from-arrayalignment" title="Permalink to this headline">¶</a></h3>
<p>We can use more conventional slice notation in this instance. Note, because Python counts from 0, the 3rd position starts at index 2.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATG---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ATGATGATGATG&quot;</span><span class="p">},</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
<span class="n">pos3</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pos3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
GGG-
&gt;seq2
GGGG

</pre></div>
</div>
</div>
</div>
</section>
<section id="filtering-positions">
<span id="filter-positions"></span><h3>Filtering positions<a class="headerlink" href="#filtering-positions" title="Permalink to this headline">¶</a></h3>
<section id="trim-terminal-stop-codons">
<h4>Trim terminal stop codons<a class="headerlink" href="#trim-terminal-stop-codons" title="Permalink to this headline">¶</a></h4>
<p>For evolutionary analyses that use codon models we need to exclude terminating stop codons. For the case where the sequences are all of length divisible by 3.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGTAA---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGACA---&quot;</span><span class="p">,</span> <span class="s2">&quot;seq3&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGCAATGA&quot;</span><span class="p">},</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">trim_stop_codons</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ACG---
&gt;seq2
ACGACA
&gt;seq3
ACGCAA

</pre></div>
</div>
</div>
</div>
<p>If the alignment contains sequences not divisible by 3, use the <code class="docutils literal notranslate"><span class="pre">allow_partial</span></code> argument.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;seq1&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGTAA---&quot;</span><span class="p">,</span>
        <span class="s2">&quot;seq2&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGAC----&quot;</span><span class="p">,</span>  <span class="c1"># terminal codon incomplete</span>
        <span class="s2">&quot;seq3&quot;</span><span class="p">:</span> <span class="s2">&quot;ACGCAATGA&quot;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">new</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">trim_stop_codons</span><span class="p">(</span><span class="n">allow_partial</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">new</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ACG---
&gt;seq2
ACGAC-
&gt;seq3
ACGCAA

</pre></div>
</div>
</div>
</div>
</section>
<section id="eliminating-columns-with-non-nucleotide-characters">
<h4>Eliminating columns with non-nucleotide characters<a class="headerlink" href="#eliminating-columns-with-non-nucleotide-characters" title="Permalink to this headline">¶</a></h4>
<p>We sometimes want to eliminate ambiguous or gap data from our alignments. We show how to exclude alignment columns by the characters they contain. In the first instance we do this just for single nucleotide columns, then for trinucleotides (equivalent for handling codons). Both are done using the <code class="docutils literal notranslate"><span class="pre">no_degenerates</span></code> method.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTG---&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGNGATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>We apply to nucleotides,</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">nucs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">no_degenerates</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">nucs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAAGGG
&gt;seq2
ATGAAGGG
&gt;seq3
ATGAAGGG

</pre></div>
</div>
</div>
</div>
<p>Applying the same filter to trinucleotides (specified by setting <code class="docutils literal notranslate"><span class="pre">motif_length=3</span></code>).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">trinucs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">no_degenerates</span><span class="p">(</span><span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">trinucs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGAAG
&gt;seq2
ATGAAG
&gt;seq3
ATGAAG

</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-all-variable-positions-from-an-alignment">
<h4>Getting all variable positions from an alignment<a class="headerlink" href="#getting-all-variable-positions-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">variable_positions</span><span class="p">()</span>
<span class="n">just_variable_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_positions</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_variable_aln</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
AAGCAAAACT
&gt;HowlerMon
AAGCAAGACT
&gt;Mouse
GGGCCCAGCT
&gt;NineBande
AAATAAAACT
&gt;DogFaced
AAACAAAATA

</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-all-constant-positions-from-an-alignment">
<h4>Getting all constant positions from an alignment<a class="headerlink" href="#getting-all-constant-positions-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">pos</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">variable_positions</span><span class="p">()</span>
<span class="n">just_constant_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">take_positions</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">negate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_constant_aln</span><span class="p">[:</span><span class="mi">10</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
TGTGGCACAA
&gt;HowlerMon
TGTGGCACAA
&gt;Mouse
TGTGGCACAA
&gt;NineBande
TGTGGCACAA
&gt;DogFaced
TGTGGCACAA

</pre></div>
</div>
</div>
</div>
</section>
<section id="getting-all-variable-codons-from-an-alignment">
<h4>Getting all variable codons from an alignment<a class="headerlink" href="#getting-all-variable-codons-from-an-alignment" title="Permalink to this headline">¶</a></h4>
<p>This is done using the <code class="docutils literal notranslate"><span class="pre">filtered</span></code> method using the <code class="docutils literal notranslate"><span class="pre">motif_length</span></code> argument. We demonstrate this first for the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">variable_codons</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">filtered</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">tuple</span><span class="p">,</span> <span class="n">x</span><span class="p">)))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_variable_aln</span><span class="p">[:</span><span class="mi">9</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
AAGCAAAAC
&gt;HowlerMon
AAGCAAGAC
&gt;Mouse
GGGCCCAGC
&gt;NineBande
AAATAAAAC
&gt;DogFaced
AAACAAAAT

</pre></div>
</div>
</div>
</div>
<p>Then for the standard <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> by first converting the <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">to_type</span><span class="p">(</span><span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">variable_codons</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">filtered</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">,</span> <span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">just_variable_aln</span><span class="p">[:</span><span class="mi">9</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;Human
AAGCAAAAC
&gt;HowlerMon
AAGCAAGAC
&gt;Mouse
GGGCCCAGC
&gt;NineBande
AAATAAAAC
&gt;DogFaced
AAACAAAAT

</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="filtering-sequences">
<h3>Filtering sequences<a class="headerlink" href="#filtering-sequences" title="Permalink to this headline">¶</a></h3>
<section id="extracting-sequences-by-sequence-identifier-into-a-new-alignment-object">
<h4>Extracting sequences by sequence identifier into a new alignment object<a class="headerlink" href="#extracting-sequences-by-sequence-identifier-into-a-new-alignment-object" title="Permalink to this headline">¶</a></h4>
<p>You can use <code class="docutils literal notranslate"><span class="pre">take_seqs</span></code> to extract some sequences by sequence identifier from an alignment to a new alignment object:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/long_testseqs.fasta&quot;</span><span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;Mouse&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .	_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .
_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align . _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .!_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ."_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .#_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .$_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .%_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .&_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .'_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .(_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .)_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .*_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .+_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .,_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .-_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ./_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .0_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .1_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .2_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .3_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .4_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .5_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .6_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .7_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .8_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .9_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .:_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .;_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .<_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .=_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .>_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .?_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .@_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .B_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .C_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .D_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .E_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .F_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .G_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .H_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .I_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .J_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .K_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .L_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .M_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .N_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .O_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .P_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .R_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .S_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .T_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .U_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .V_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .W_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .X_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .[_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .\_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .]_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .^_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .__bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .`_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .a_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .b_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .c_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .d_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .e_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .f_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .g_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .h_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .i_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .j_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .k_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .l_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .m_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .n_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .o_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .p_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .r_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .s_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .t_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .u_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .v_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .w_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .x_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .{_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .|_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .}_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .~_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align . _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¡_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¢_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .£_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¤_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¥_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¦_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .§_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¨_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .©_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ª_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .«_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¬_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .­_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .®_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¯_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .°_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .±_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .²_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .³_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .´_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .µ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¶_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .·_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¸_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¹_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .º_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .»_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¼_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .½_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¾_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¿_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .À_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .È_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .É_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .×_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ß_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .à_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .è_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .é_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .÷_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ÿ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .terminal_ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">Human</td><td><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span></td></tr>
<tr><td class="label">Mouse</td><td><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">G</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">G</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">C</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">C</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">G</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">T</span><span class="T_bytes">.</span></td></tr>
</table>
<p><i>2 x 2532 (truncated to 2 x 60) bytes alignment</i></p>
</div></div></div>
</div>
<p>Alternatively, you can extract only the sequences which are not specified by passing <code class="docutils literal notranslate"><span class="pre">negate=True</span></code>:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">take_seqs</span><span class="p">([</span><span class="s2">&quot;Human&quot;</span><span class="p">,</span> <span class="s2">&quot;Mouse&quot;</span><span class="p">],</span> <span class="n">negate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .	_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .
_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align . _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .!_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ."_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .#_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .$_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .%_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .&_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .'_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .(_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .)_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .*_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .+_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .,_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .-_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ./_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .0_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .1_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .2_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .3_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .4_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .5_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .6_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .7_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .8_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .9_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .:_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .;_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .<_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .=_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .>_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .?_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .@_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .B_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .C_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .D_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .E_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .F_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .G_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .H_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .I_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .J_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .K_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .L_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .M_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .N_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .O_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .P_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .R_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .S_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .T_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .U_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .V_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .W_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .X_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .[_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .\_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .]_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .^_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .__bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .`_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .a_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .b_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .c_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .d_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .e_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .f_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .g_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .h_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .i_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .j_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .k_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .l_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .m_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .n_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .o_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .p_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .q_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .r_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .s_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .t_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .u_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .v_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .w_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .x_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .y_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .z_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .{_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .|_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .}_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .~_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align ._bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align . _bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¡_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¢_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .£_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¤_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¥_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¦_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .§_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¨_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .©_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ª_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .«_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¬_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .­_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .®_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¯_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .°_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .±_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .²_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .³_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .´_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .µ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¶_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .·_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¸_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¹_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .º_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .»_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¼_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .½_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¾_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .¿_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .À_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .È_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .É_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .×_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ß_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .à_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .á_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .â_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ã_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ä_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .å_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .æ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ç_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .è_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .é_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ê_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ë_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ì_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .í_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .î_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ï_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ð_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ñ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ò_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ó_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ô_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .õ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ö_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .÷_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ø_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ù_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ú_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .û_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ü_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ý_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .þ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ÿ_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .terminal_ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ambig_bytes{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">DogFaced</td><td><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="A_bytes">A</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="G_bytes">G</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="T_bytes">T</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="A_bytes">A</span><span class="C_bytes">C</span><span class="T_bytes">T</span></td></tr>
<tr><td class="label">HowlerMon</td><td><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">G</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">G</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span></td></tr>
<tr><td class="label">NineBande</td><td><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">T</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="A_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="G_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="T_bytes">.</span><span class="T_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">C</span><span class="T_bytes">T</span><span class="C_bytes">.</span><span class="A_bytes">.</span><span class="C_bytes">.</span><span class="T_bytes">.</span></td></tr>
</table>
<p><i>3 x 2532 (truncated to 3 x 60) bytes alignment</i></p>
</div></div></div>
</div>
</section>
<section id="extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object">
<h4>Extracting sequences using an arbitrary function into a new alignment object<a class="headerlink" href="#extracting-sequences-using-an-arbitrary-function-into-a-new-alignment-object" title="Permalink to this headline">¶</a></h4>
<p>You can use <code class="docutils literal notranslate"><span class="pre">take_seqs_if</span></code> to extract sequences into a new alignment object based on whether an arbitrary function applied to the sequence evaluates to True. For example, to extract sequences which don’t contain any N bases you could do the following:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTG---&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGNGATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="k">def</span> <span class="nf">no_N_chars</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
    <span class="k">return</span> <span class="s2">&quot;N&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">s</span>

<span class="n">aln</span><span class="o">.</span><span class="n">take_seqs_if</span><span class="p">(</span><span class="n">no_N_chars</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">seq2</td><td><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span></td></tr>
<tr><td class="label">seq1</td><td><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 12 dna alignment</i></p>
</div></div></div>
</div>
<p>You can additionally get the sequences where the provided function evaluates to False:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span><span class="o">.</span><span class="n">take_seqs_if</span><span class="p">(</span><span class="n">no_N_chars</span><span class="p">,</span> <span class="n">negate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">seq3</td><td><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="ambig_dna">N</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span></td></tr>
</table>
<p><i>1 x 12 dna alignment</i></p>
</div></div></div>
</div>
</section>
</section>
</section>
<section id="computing-alignment-statistics">
<h2>Computing alignment statistics<a class="headerlink" href="#computing-alignment-statistics" title="Permalink to this headline">¶</a></h2>
<section id="getting-motif-counts">
<h3>Getting motif counts<a class="headerlink" href="#getting-motif-counts" title="Permalink to this headline">¶</a></h3>
<p>We state the motif length we want and whether to allow gap or ambiguous characters. The latter only has meaning for IPUAC character sets (the DNA, RNA or PROTEIN moltypes). We illustrate this for the DNA moltype with motif lengths of 1 and 3.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTG---&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGGNGATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">counts</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">counts</span><span class="p">(</span><span class="n">motif_length</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">counts</span><span class="p">(</span><span class="n">include_ambiguity</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>==================
 A    C     G    T
------------------
11    0    14    7
------------------
============================================================================================================================================================================================================================================================================================================================================================================================================================================================
AAA    AAC    AAG    AAT    ACA    ACC    ACG    ACT    AGA    AGC    AGG    AGT    ATA    ATC    ATG    ATT    CAA    CAC    CAG    CAT    CCA    CCC    CCG    CCT    CGA    CGC    CGG    CGT    CTA    CTC    CTG    CTT    GAA    GAC    GAG    GAT    GCA    GCC    GCG    GCT    GGA    GGC    GGG    GGT    GTA    GTC    GTG    GTT    TAA    TAC    TAG    TAT    TCA    TCC    TCG    TCT    TGA    TGC    TGG    TGT    TTA    TTC    TTG    TTT
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  0      0      3      0      0      0      0      0      0      0      0      0      0      0      5      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      2      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0      0
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
=======================
 A    C     G    N    T
-----------------------
11    0    14    1    7
-----------------------
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only the observed motifs are returned, rather than all defined by the alphabet.</p>
</div>
</section>
<section id="computing-motif-probabilities-from-an-alignment">
<h3>Computing motif probabilities from an alignment<a class="headerlink" href="#computing-motif-probabilities-from-an-alignment" title="Permalink to this headline">¶</a></h3>
<p>The method <code class="docutils literal notranslate"><span class="pre">get_motif_probs</span></code> of <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> objects returns the probabilities for all motifs of a given length. For individual nucleotides:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">motif_probs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;T&#39;: 0.25520602569782896, &#39;C&#39;: 0.25808595480726626, &#39;A&#39;: 0.24390784226849802, &#39;G&#39;: 0.24280017722640673}
</pre></div>
</div>
</div>
</div>
<p>For dinucleotides or longer, we need to pass in an <code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> with the appropriate word length. Here is an example with trinucleotides:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">DNA</span>

<span class="n">trinuc_alphabet</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">get_word_alphabet</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">(</span><span class="n">alphabet</span><span class="o">=</span><span class="n">trinuc_alphabet</span><span class="p">)</span>
<span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">motif_probs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">motif_probs</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">  </span><span class="si">%.3f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">motif_probs</span><span class="p">[</span><span class="n">m</span><span class="p">]))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CAG  0.037
CCT  0.034
CGC  0.030
TCT  0.029
GCC  0.028
AAC  0.028
AAA  0.027
AAG  0.027
GGG  0.026
CCC  0.025
TAA  0.025
CTT  0.024
TTC  0.024
GAG  0.024
TGT  0.023
CTG  0.023
CTC  0.021
GGA  0.021
TTT  0.020
GGT  0.019
CAA  0.019
AGT  0.019
TCC  0.019
GAA  0.019
ATT  0.018
TAG  0.017
AAT  0.017
TTG  0.017
TCA  0.015
TGG  0.015
TGC  0.015
GTT  0.014
GGC  0.014
CAT  0.013
GTA  0.013
TGA  0.013
ATA  0.012
AGA  0.012
CAC  0.011
TTA  0.011
CCA  0.011
GCT  0.011
GAT  0.011
GCA  0.011
CCG  0.011
TAC  0.010
GAC  0.009
AGC  0.009
ATC  0.008
GTC  0.008
CGG  0.008
AGG  0.008
ACT  0.007
ACG  0.007
CGT  0.007
GCG  0.007
GTG  0.006
TAT  0.005
ATG  0.005
TCG  0.005
ACC  0.005
CGA  0.005
CTA  0.003
ACA  0.002
</pre></div>
</div>
</div>
</div>
<p>The same holds for other arbitrary alphabets, as long as they match the alignment <code class="docutils literal notranslate"><span class="pre">MolType</span></code>.</p>
<p>Some calculations in <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> require all non-zero values in the motif probabilities, in which case we use a pseudo-count. We illustrate that here with a simple example where T is missing. Without the pseudo-count, the frequency of T is 0.0, with the pseudo-count defined as 1e-6 then the frequency of T will be slightly less than 1e-6.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;AACAAC&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;AAGAAG&quot;</span><span class="p">)],</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">()</span>
<span class="k">assert</span> <span class="n">motif_probs</span><span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mf">0.0</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">(</span><span class="n">pseudocount</span><span class="o">=</span><span class="mf">1e-6</span><span class="p">)</span>
<span class="k">assert</span> <span class="mi">0</span> <span class="o">&lt;</span> <span class="n">motif_probs</span><span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">1e-6</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>It is important to notice that motif probabilities are computed by treating sequences as non-overlapping tuples. Below is a very simple pair of identical sequences where there are clearly 2 ‘AA’ dinucleotides per sequence but only the first one is ‘in-frame’ (frame width = 2).</p>
<p>We then create a dinucleotide <code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> object and use this to get dinucleotide probabilities. These frequencies are determined by breaking each aligned sequence up into non-overlapping dinucleotides and then doing a count. The expected value for the ‘AA’ dinucleotide in this case will be 2/8 = 0.25.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;AACGTAAG&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="s2">&quot;AACGTAAG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">dinuc_alphabet</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">alphabet</span><span class="o">.</span><span class="n">get_word_alphabet</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="n">motif_probs</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_motif_probs</span><span class="p">(</span><span class="n">alphabet</span><span class="o">=</span><span class="n">dinuc_alphabet</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">motif_probs</span><span class="p">[</span><span class="s2">&quot;AA&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mf">0.25</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>What about counting the total incidence of dinucleotides including those not in-frame?  A naive application of the Python string object’s count method will not work as desired either because it “returns the number of non-overlapping occurrences”.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;my_seq&quot;</span><span class="p">,</span> <span class="s2">&quot;AAAGTAAG&quot;</span><span class="p">)]</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">seqs</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">my_seq</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;my_seq&quot;</span><span class="p">)</span>
<span class="n">my_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="s2">&quot;AAA&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
<span class="s2">&quot;AAAA&quot;</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;AA&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>2
</pre></div>
</div>
</div>
</div>
<p>To count all occurrences of a given dinucleotide in a DNA sequence, one could use a standard Python approach such as list comprehension:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span><span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span> <span class="n">seq</span><span class="o">=</span><span class="s2">&quot;AAAGTAAG&quot;</span><span class="p">)</span>
<span class="n">seq</span>
<span class="n">di_nucs</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span> <span class="p">:</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)]</span>
<span class="nb">sum</span><span class="p">([</span><span class="n">nn</span> <span class="o">==</span> <span class="s2">&quot;AA&quot;</span> <span class="k">for</span> <span class="n">nn</span> <span class="ow">in</span> <span class="n">di_nucs</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>3
</pre></div>
</div>
</div>
</div>
</section>
<section id="working-with-alignment-gaps">
<h3>Working with alignment gaps<a class="headerlink" href="#working-with-alignment-gaps" title="Permalink to this headline">¶</a></h3>
<section id="filtering-extracted-columns-for-the-gap-character">
<h4>Filtering extracted columns for the gap character<a class="headerlink" href="#filtering-extracted-columns-for-the-gap-character" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="n">col</span> <span class="o">=</span> <span class="n">aln</span><span class="p">[</span><span class="mi">113</span><span class="p">:</span><span class="mi">115</span><span class="p">]</span><span class="o">.</span><span class="n">iter_positions</span><span class="p">()</span>
<span class="n">c1</span><span class="p">,</span> <span class="n">c2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
<span class="n">c1</span><span class="p">,</span> <span class="n">c2</span>
<span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">c1</span><span class="p">))</span>
<span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">c2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[ByteSequence(-), ByteSequence(-)]
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculating-the-gap-fraction">
<h4>Calculating the gap fraction<a class="headerlink" href="#calculating-the-gap-fraction" title="Permalink to this headline">¶</a></h4>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span><span class="s2">&quot;data/primate_cdx2_promoter.fasta&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">aln</span><span class="p">[</span><span class="mi">113</span><span class="p">:</span><span class="mi">150</span><span class="p">]</span><span class="o">.</span><span class="n">iter_positions</span><span class="p">():</span>
    <span class="n">ungapped</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">column</span><span class="p">))</span>
    <span class="n">gap_fraction</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ungapped</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">column</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">gap_fraction</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>0.0
0.6666666666666666
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
</pre></div>
</div>
</div>
</div>
</section>
<section id="extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps">
<h4>Extracting maps of aligned to unaligned positions (i.e., gap maps)<a class="headerlink" href="#extracting-maps-of-aligned-to-unaligned-positions-i-e-gap-maps" title="Permalink to this headline">¶</a></h4>
<p>It’s often important to know how an alignment position relates to a position in one or more of the sequences in the alignment. The <code class="docutils literal notranslate"><span class="pre">gap_maps</span></code> method of the individual sequences is useful for this. To get a map of sequence to alignment positions for a specific sequence in your alignment, do the following:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAGG-TG--&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAAG--GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seq_to_aln_map</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_gapped_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">gap_maps</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p>It’s now possible to look up positions in the <code class="docutils literal notranslate"><span class="pre">seq1</span></code>, and find out what they map to in the alignment:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq_to_aln_map</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">seq_to_aln_map</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>9
</pre></div>
</div>
</div>
</div>
<p>This tells us that in position 3 in <code class="docutils literal notranslate"><span class="pre">seq1</span></code> corresponds to position 3 in <code class="docutils literal notranslate"><span class="pre">aln</span></code>, and that position 8 in <code class="docutils literal notranslate"><span class="pre">seq1</span></code> corresponds to position 9 in <code class="docutils literal notranslate"><span class="pre">aln</span></code>.</p>
<p>Notice that we grabbed the first result from the call to <code class="docutils literal notranslate"><span class="pre">gap_maps</span></code>. This is the sequence position to alignment position map. The second value returned is the alignment position to sequence position map, so if you want to find out what sequence positions the alignment positions correspond to (opposed to what alignment positions the sequence positions correspond to) for a given sequence, you would take the following steps:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln_to_seq_map</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_gapped_seq</span><span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">gap_maps</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>7
</pre></div>
</div>
</div>
</div>
<p>If an alignment position is a gap, and therefore has no corresponding sequence position, you’ll get a <code class="docutils literal notranslate"><span class="pre">KeyError</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">seq_pos</span> <span class="o">=</span> <span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyError</span><span class="g g-Whitespace">                                  </span>Traceback (most recent call last)
<span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">folders</span><span class="o">/</span><span class="mi">24</span><span class="o">/</span><span class="mi">8</span><span class="n">k48jl6d249_n_qfxwsl6xvm0000gn</span><span class="o">/</span><span class="n">T</span><span class="o">/</span><span class="n">ipykernel_7639</span><span class="o">/</span><span class="mf">8528689.</span><span class="n">py</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">seq_pos</span> <span class="o">=</span> <span class="n">aln_to_seq_map</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>

<span class="ne">KeyError</span>: 7
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The first position in alignments and sequences is always numbered position 0.</p>
</div>
</section>
<section id="filtering-alignments-based-on-gaps">
<h4>Filtering alignments based on gaps<a class="headerlink" href="#filtering-alignments-based-on-gaps" title="Permalink to this headline">¶</a></h4>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An alternate, computationally faster, approach to removing gaps is to use the <code class="docutils literal notranslate"><span class="pre">filtered</span></code> method as discussed in <a class="reference internal" href="#filter-positions"><span class="std std-ref">Filtering positions</span></a>.</p>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">omit_gap_runs</span></code> method can be applied to remove long stretches of gaps in an alignment. In the following example, we remove sequences that have more than two adjacent gaps anywhere in the aligned sequence.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA---TG-&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_runs</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

</pre></div>
</div>
</div>
</div>
<p>If instead, we just wanted to remove positions from the alignment which are gaps in more than a certain percentage of the sequences, we could use the <code class="docutils literal notranslate"><span class="pre">omit_gap_pos</span></code> function. For example:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA---TG-&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_pos</span><span class="p">(</span><span class="mf">0.40</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq1
ATGA--TG-
&gt;seq2
ATGAGGATG
&gt;seq3
AT-AGGATG

</pre></div>
</div>
</div>
</div>
<p>If you wanted to remove sequences which contain more than a certain percent gap characters, you could use the <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code> method. This is commonly applied to filter partial sequences from an alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;ATGAA------&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATG-AGTGATG&quot;</span><span class="p">),</span>
        <span class="p">(</span><span class="s2">&quot;seq3&quot;</span><span class="p">,</span> <span class="s2">&quot;AT--AG-GATG&quot;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">filtered_aln</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">omit_gap_seqs</span><span class="p">(</span><span class="mf">0.50</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">filtered_aln</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq2
ATG-AGTGATG
&gt;seq3
AT--AG-GATG

</pre></div>
</div>
</div>
</div>
<p>Note that following this call to <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code>, the 4th column of <code class="docutils literal notranslate"><span class="pre">filtered_aln</span></code> is 100% gaps. This is generally not desirable, so a call to <code class="docutils literal notranslate"><span class="pre">omit_gap_seqs</span></code> is frequently followed with a call to <code class="docutils literal notranslate"><span class="pre">omit_gap_pos</span></code> with no parameters – this defaults to removing positions which are all gaps:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">filtered_aln</span><span class="o">.</span><span class="n">omit_gap_pos</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;seq2
ATGAGTGATG
&gt;seq3
AT-AG-GATG

</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
</section>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020-2022, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 4.3.2 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>