
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Annotation Databases &#8212; cogent3 2024.08.09 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=da3cec82" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9db0cdfc"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/thebelab-helper.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'doc/cookbook/annotation_db';</script>
    <link rel="icon" href="../../_static/c3-square.svg"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Molecular types" href="moltypes.html" />
    <link rel="prev" title="Features" href="features.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/c3-logo.png" class="logo__image only-light" alt="cogent3 2024.08.09 documentation - Home"/>
    <script>document.write(`<img src="../../_static/c3-logo.png" class="logo__image only-dark" alt="cogent3 2024.08.09 documentation - Home"/>`);</script>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../draw/index.html">
    Image Gallery
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    User Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../community.html">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../licenses.html">
    License
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../history.html">
    Project History
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../pycogent.html">
    How does cogent3 relate to PyCogent?
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../projects.html">
    Projects using cogent3
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../data_file_links.html">
    Data Files Used in the Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../genindex.html">
    Index
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../draw/index.html">
    Image Gallery
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    User Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../community.html">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../licenses.html">
    License
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../history.html">
    Project History
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../pycogent.html">
    How does cogent3 relate to PyCogent?
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../projects.html">
    Projects using cogent3
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../data_file_links.html">
    Data Files Used in the Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../genindex.html">
    Index
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../app/index.html">apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-app.html">Overview of apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/app-overview.html">Overview of using apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/app-help.html">Getting help on an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/app-get.html">Getting an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/dstore.html">Data stores – collections of data records</a></li>

<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/not-completed.html">Tracking records that could not be processed</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-cookbook.html">App Cookbook</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-loaders.html">Loading data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-aligned.html">Loading aligned sequence data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-unaligned.html">Loading unaligned sequence data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-tabular.html">Loading tabular data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-json.html">Loading JSON serialised objects</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-align.html">Sequence Alignment</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/align-nucleotide.html">Using a nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/align-codon.html">Using a codon model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/align-protein.html">Using a protein model</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-sample.html">Filtering data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-degen.html">Removing degenerate characters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-gap.html">Removing highly gapped positions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-bad.html">Remove problem sequences from an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-duplicated.html">Remove duplicated sequences from an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-min-len.html">Filter sequence collections and alignments by length</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-fixed-len.html">Sample an alignment to a fixed length</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-take-pos.html">Sample nucleotides from a given codon position</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-take-seqs.html">Select named sequences from a collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-take-n-seqs.html">Select <cite>n</cite> sequences from a collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-concat.html">Concatenating alignments</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-writers.html">Writing Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-seqs.html">Writing sequences and sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-tabular.html">Writing tabular data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-json.html">Writing JSON Serialised Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-db.html">Writing a database to file</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-tutorials.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_tutorials/index-evo-model.html">Modelling sequence evolution</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-ancestral-states.html">Reconstructing ancestral states</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-tr-codon-model.html">Applying a time-reversible codon model</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_tutorials/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_tutorials/index-evo-natsel.html">Testing for natural selection</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-custom.html">Custom Apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/user_function.html">Why write apps?</a></li>

</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">Cookbook</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active has-children"><a class="reference internal" href="manipulating_biological_data.html">Manipulating biological data</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="loading_sequences.html">Loading a sequence from a file</a></li>





<li class="toctree-l3"><a class="reference internal" href="DNA_and_RNA_sequences.html">Sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="protein_sequences.html">Protein sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="alignments.html">Sequence Collections and Alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="features.html">Features</a></li>
<li class="toctree-l3 current active"><a class="current reference internal" href="#">Annotation Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="moltypes.html">Molecular types</a></li>
<li class="toctree-l3"><a class="reference internal" href="alphabet.html">Alphabets</a></li>
<li class="toctree-l3"><a class="reference internal" href="what_codes.html">Available genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="genetic_code.html">Using genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="simple_trees.html">Trees</a></li>
</ul>
</details></li>
<li class="toctree-l2"><a class="reference internal" href="tables.html">Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="building_alignments.html">Building alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="calc_genetic_distance.html">Genetic distance calculation</a></li>


<li class="toctree-l2"><a class="reference internal" href="building_phylogenies.html">Building phylogenies</a></li>
<li class="toctree-l2"><a class="reference internal" href="evo_modelling.html">Evolutionary Analysis Using Likelihood</a></li>
<li class="toctree-l2"><a class="reference internal" href="useful_utilities.html">Useful Utilities</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../examples/index.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>

<li class="toctree-l2"><a class="reference internal" href="../examples/period_estimation.html">Estimating periodic signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/simple.html">The simplest script</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/codon_models.html">Using codon models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/parallel.html">Parallel computations</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/index.html">API</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.open_.html">open</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_model.html">get_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</details></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/composable/define_app.html">define_app</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">User Guides</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="manipulating_biological_data.html" class="nav-link">Manipulating biological data</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Annotation Databases</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="annotation-databases">
<span id="anno-db"></span><h1>Annotation Databases<a class="headerlink" href="#annotation-databases" title="Link to this heading">#</a></h1>
<p>This guide shows you how to use <code class="docutils literal notranslate"><span class="pre">cogent3</span></code>’s annotation databases, which are in-memory SQLite databases, to store, query and manipulate the <a class="reference internal" href="features.html#howto-features"><span class="std std-ref">Features</span></a> (also known as annotations) of one or more biological sequences.</p>
<section id="what-are-the-different-types-of-annotationdb">
<h2>What are the different types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>?<a class="headerlink" href="#what-are-the-different-types-of-annotationdb" title="Link to this heading">#</a></h2>
<p>There are three types of databases that are available: <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code>, <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code>, and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code>. A <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code> provides a “user” table where custom features can be added. The latter two, as hinted in their names, are distinguished by the type of data file used to generate them. Both <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> have a “user” table for custom features, which allows them to be merged with a <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code>. However, they cannot be merged with each other!</p>
</section>
<section id="how-to-create-a-standalone-basicannotationdb">
<h2>How to create a standalone <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code><a class="headerlink" href="#how-to-create-a-standalone-basicannotationdb" title="Link to this heading">#</a></h2>
<p>Achieved by creating an <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code> instance. This is an empty database to which we can add features.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.core.annotation_db</span> <span class="kn">import</span> <span class="n">BasicAnnotationDb</span>

<span class="n">anno_db</span> <span class="o">=</span> <span class="n">BasicAnnotationDb</span><span class="p">()</span>
<span class="n">anno_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=0)
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-load-an-standalone-annotationdb-from-a-data-file">
<h2>How to load an standalone <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> from a data file<a class="headerlink" href="#how-to-load-an-standalone-annotationdb-from-a-data-file" title="Link to this heading">#</a></h2>
<p>Typically, we want to load a collection of features from a genomic annotation file, such as a GFF or Genbank file. For the following examples, we will use data from the bacterium <em>Mycoplasma genitalium</em>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the list of <a class="reference internal" href="../data_file_links.html#data-links"><span class="std std-ref">Data Files Used in the Documentation</span></a> to download the data used in the following examples.</p>
</div>
<section id="from-a-gff-file">
<h3>From a GFF file<a class="headerlink" href="#from-a-gff-file" title="Link to this heading">#</a></h3>
<p>To load features from a GFF file, we can use the <code class="docutils literal notranslate"><span class="pre">load_annotations</span></code> function and provide the path to the GFF file. The function automatically determines the file type based on the file extension, and it returns a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> object.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_annotations</span>

<span class="n">gff_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">)</span>
<span class="n">gff_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=1169)
</pre></div>
</div>
</div>
</div>
</section>
<section id="from-a-genbank-file">
<h3>From a Genbank file<a class="headerlink" href="#from-a-genbank-file" title="Link to this heading">#</a></h3>
<p>To load features from a Genbank file, we can once again use the <code class="docutils literal notranslate"><span class="pre">load_annotations</span></code> function and provide the path to the Genbank file. The function detects the file type based on the file extension, and it returns a <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> object.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_annotations</span>

<span class="n">gb_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gb&quot;</span><span class="p">)</span>
<span class="n">gb_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;NC_000908&#39;, source=&#39;:memory:&#39;, namer=None, total_records=1127)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="how-to-generate-a-summary-of-an-annotationdb">
<h2>How to generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-generate-a-summary-of-an-annotationdb" title="Link to this heading">#</a></h2>
<p>To generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>, we can access the <code class="docutils literal notranslate"><span class="pre">describe</span></code> attribute of the database. This attribute returns a <code class="docutils literal notranslate"><span class="pre">cogent3.util.table.Table</span></code> instance that shows the number of records for each seqid, the count for each biotype, and the number of rows in each table (in this example there is a “gff” table with 1,169 rows and an empty “user” table).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">summary</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">describe</span>
<span class="n">summary</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">seqid('NC_000908.2')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">biotype('CDS')</span></td><td><span class="c3col_right">521</span></td></tr>
<tr><td><span class="c3col_left">biotype('RNase_P_RNA')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('SRP_RNA')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('exon')</span></td><td><span class="c3col_right">42</span></td></tr>
<tr><td><span class="c3col_left">biotype('gene')</span></td><td><span class="c3col_right">546</span></td></tr>
<tr><td><span class="c3col_left">biotype('pseudogene')</span></td><td><span class="c3col_right">17</span></td></tr>
<tr><td><span class="c3col_left">biotype('rRNA')</span></td><td><span class="c3col_right">3</span></td></tr>
<tr><td><span class="c3col_left">biotype('region')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('tRNA')</span></td><td><span class="c3col_right">36</span></td></tr>
<tr><td><span class="c3col_left">biotype('tmRNA')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">0</span></td></tr>
</tbody>
</table>
<p>
13 rows x 2 columns</p>
</div></div></div>
</div>
</section>
<section id="how-to-add-custom-features-to-an-annotationdb">
<h2>How to add custom features to an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-add-custom-features-to-an-annotationdb" title="Link to this heading">#</a></h2>
<p>This is achieved via the <code class="docutils literal notranslate"><span class="pre">add_features</span></code> method and for all three types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> it will be added to the “user” table. The method requires information about the feature, such as its biotype, name, genomic location (spans), and the seqid. The seqid is necessary when linking an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> object, see <a class="reference internal" href="#assign-db-to-seq"><span class="std std-ref">How to assign an AnnotationDb to a sequence</span></a> for more information.</p>
<p>We can add a feature to the empty <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code> we created above. Now the database has one record!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">anno_db</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;NC_000908&quot;</span><span class="p">,</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;interesting_gene&quot;</span><span class="p">,</span>
    <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span>
    <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">anno_db</span><span class="o">.</span><span class="n">describe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">seqid('NC_000908')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('gene')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">1</span></td></tr>
</tbody>
</table>
<p>
3 rows x 2 columns</p>
</div></div></div>
</div>
<p>We can also add a feature to our <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> or <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code>. Below, the previously empty “user” table now has a row count of one, indicating that our feature has been successfully added to the database.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;interesting_gene&quot;</span><span class="p">,</span>
    <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span>
    <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">gff_db</span><span class="o">.</span><span class="n">describe</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:,</span> <span class="p">:]</span>  <span class="c1"># showing just last two rows</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">1</span></td></tr>
</tbody>
</table>
<p>
2 rows x 2 columns</p>
</div></div></div>
</div>
</section>
<section id="how-to-write-an-annotationdb-to-disk-for-efficient-re-loading">
<h2>How to write an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to disk for efficient re-loading<a class="headerlink" href="#how-to-write-an-annotationdb-to-disk-for-efficient-re-loading" title="Link to this heading">#</a></h2>
<p>In the above examples, all databases indicate that <code class="docutils literal notranslate"><span class="pre">source=&quot;:memory:&quot;</span></code>, i.e. they are in-memory databases. We can write any database to disk using the <code class="docutils literal notranslate"><span class="pre">write()</span></code> method and providing an outpath.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write to disk</span>
<span class="n">gb_db</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;data/m-genitalium-database.gbdb&quot;</span><span class="p">)</span>

<span class="c1"># do something</span>

<span class="c1"># re-load from disk</span>
<span class="n">quick_load_gb_db</span> <span class="o">=</span> <span class="n">GenbankAnnotationDb</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;data/m-genitalium-database.gbdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The suffix of the outpath (“.gbdb” in the above example) can be arbitrarily chosen, however, this behaviour may change in the future to only accept registered suffixes! 👀</p>
</div>
</section>
<section id="how-to-query-an-annotationdb">
<h2>How to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-query-an-annotationdb" title="Link to this heading">#</a></h2>
<p>Note, there are two methods with the same interface available to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code>. A generator that yields all features that matched the query. The <strong>minimal information</strong> required to create a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> <code class="docutils literal notranslate"><span class="pre">Feature</span></code> object is provided in the returned dictionary. For more information on Features see <a class="reference internal" href="features.html#howto-features"><span class="std std-ref">Features</span></a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code>. A generator that yields all features that matched the query. The <strong>complete record</strong> for each matching feature is provided in the returned dictionary.</p></li>
</ol>
<p>Put simply, a “feature” is a subset of a “record”.</p>
<section id="querying-via-feature-name">
<h3>Querying via Feature Name<a class="headerlink" href="#querying-via-feature-name" title="Link to this heading">#</a></h3>
<p>To query a database for a feature by its name, provide the name of the feature as an argument to either <code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code> or <code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code>. Since an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> can contain records for more than one sequence, it is best practice to also include the seqid of the sequence of interest.</p>
<p>For example, querying the <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> for the 16s rRNA gene:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mg_16s</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">gb_db</span><span class="o">.</span><span class="n">get_features_matching</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;MG_RS00775&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;NC_000908&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">mg_16s</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908&#39;,
  &#39;biotype&#39;: &#39;gene&#39;,
  &#39;spans&#39;: [(np.int64(170011), np.int64(171529))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;MG_RS00775&#39;,
  &#39;on_alignment&#39;: None}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-feature-biotype">
<h3>Querying via Feature Biotype<a class="headerlink" href="#querying-via-feature-biotype" title="Link to this heading">#</a></h3>
<p>Similarly, <code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code> and <code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code> can be used to query the database for all features that match a given biotype.</p>
<p>For example, querying the <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> for all pseudogenes:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pseudogenes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff_db</span><span class="o">.</span><span class="n">get_features_matching</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">))</span>
<span class="n">pseudogenes</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>  <span class="c1"># showing just the first two</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;pseudogene&#39;,
  &#39;spans&#39;: [(np.int64(85561), np.int64(86589))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;gene-MG_RS02910&#39;,
  &#39;on_alignment&#39;: None},
 {&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;pseudogene&#39;,
  &#39;spans&#39;: [(np.int64(86784), np.int64(87549))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;gene-MG_RS00385&#39;,
  &#39;on_alignment&#39;: None}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-region-of-interest">
<h3>Querying via region of interest<a class="headerlink" href="#querying-via-region-of-interest" title="Link to this heading">#</a></h3>
<p>We can provide <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> arguments to <code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code> and <code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code> and all features within the coordinates will be returned.</p>
<p>For example, the adhesin protein of <em>M. genitalium</em> is organised in an operon between positions 220600 to 229079, so we can query for genes in that region to return all operon genes:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">operon_cds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">gff_db</span><span class="o">.</span><span class="n">get_features_matching</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">220600</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">229067</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">operon_cds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(np.int64(220606), np.int64(221563))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;cds-WP_009886005.1&#39;,
  &#39;on_alignment&#39;: None},
 {&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(np.int64(221569), np.int64(225904))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;cds-WP_010869366.1&#39;,
  &#39;on_alignment&#39;: None},
 {&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(np.int64(225914), np.int64(229067))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;cds-WP_041593683.1&#39;,
  &#39;on_alignment&#39;: None}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-the-extended-attributes-field">
<h3>Querying via the extended attributes field<a class="headerlink" href="#querying-via-the-extended-attributes-field" title="Link to this heading">#</a></h3>
<p>A particularly useful functionality of a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> is the ability to search the extended attributes field. This allows querying for records that have matches to a specific string provided to the <code class="docutils literal notranslate"><span class="pre">attributes</span></code> argument within their extended attributes field.</p>
<p>For example, we can query for all CDS related to replication:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">replication_records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">gff_db</span><span class="o">.</span><span class="n">get_records_matching</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="s2">&quot;replication&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">replication_records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># showing just the first match</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
 &#39;source&#39;: &#39;Protein Homology&#39;,
 &#39;biotype&#39;: &#39;CDS&#39;,
 &#39;start&#39;: 685,
 &#39;stop&#39;: 1828,
 &#39;score&#39;: &#39;.&#39;,
 &#39;strand&#39;: &#39;+&#39;,
 &#39;phase&#39;: &#39;0&#39;,
 &#39;attributes&#39;: &#39;ID=cds-WP_009885562.1;Parent=gene-MG_RS00005;Dbxref=Genbank:WP_009885562.1;Name=WP_009885562.1;Ontology_term=GO:0006260,GO:0003887,GO:0009360;gbkey=CDS;gene=dnaN;go_component=DNA polymerase III complex|0009360||IEA;go_function=DNA-directed DNA polymerase activity|0003887||IEA;go_process=DNA replication|0006260||IEA;inference=COORDINATES: similar to AA sequence:RefSeq:WP_010874358.1;locus_tag=MG_RS00005;product=DNA polymerase III subunit beta;protein_id=WP_009885562.1;transl_table=4&#39;,
 &#39;comments&#39;: None,
 &#39;spans&#39;: array([[ 685, 1828]]),
 &#39;name&#39;: &#39;cds-WP_009885562.1&#39;,
 &#39;parent_id&#39;: &#39;gene-MG_RS00005&#39;}
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Extended attribute querying only works for GFF databases!</p>
</div>
</section>
</section>
<section id="how-to-interrogate-an-annotationdb">
<h2>How to interrogate an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-interrogate-an-annotationdb" title="Link to this heading">#</a></h2>
<p>An <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> can be interrogated to explore the properties of a sequence without needing the sequence information.</p>
<section id="how-many-unique-genes-are-in-a-given-genome">
<h3>How many unique genes are in a given genome?<a class="headerlink" href="#how-many-unique-genes-are-in-a-given-genome" title="Link to this heading">#</a></h3>
<p><em>Mycoplasma genitalium</em> has the smallest bacterial genome, so the number of genes in the loaded database represents the approximate minimal set of genes required for bacterial life! We can see the total number of genes by using the <code class="docutils literal notranslate"><span class="pre">num_matches()</span></code> method and specifying the condition we want to be matched is that the biotype is “gene”.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gb_db</span><span class="o">.</span><span class="n">num_matches</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>563
</pre></div>
</div>
</div>
</div>
<p>The count is 563, however, this may include genes with more than one copy. To determine the number of distinct genes we can use the <code class="docutils literal notranslate"><span class="pre">count_distinct()</span></code> method and specify <code class="docutils literal notranslate"><span class="pre">biotype=&quot;gene&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">name=True</span></code> to indicate we are interested in genes with distinct names.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">total_genes</span> <span class="o">=</span> <span class="n">gb_db</span><span class="o">.</span><span class="n">count_distinct</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">single_copy</span> <span class="o">=</span> <span class="n">total_genes</span><span class="p">[</span><span class="n">total_genes</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
<span class="nb">len</span><span class="p">(</span><span class="n">single_copy</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>561
</pre></div>
</div>
</div>
</div>
<p>The count of unique genes is 561. This means that almost every gene is present only once in the genome, very little redundancy here!</p>
<p>Just for fun, let’s try this with the GFF database… (downloaded from the exact same source)</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">total_genes</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">num_matches</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;total genes: &quot;</span><span class="p">,</span> <span class="n">total_genes</span><span class="p">)</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">count_distinct</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">single_copy</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="n">genes</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;single copy genes: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">single_copy</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>total genes:  547
single copy genes:  547
</pre></div>
</div>
</div>
</div>
<p>What? 🤯</p>
</section>
<section id="how-to-find-the-children-of-a-feature">
<h3>How to find the “children” of a Feature<a class="headerlink" href="#how-to-find-the-children-of-a-feature" title="Link to this heading">#</a></h3>
<p>To find the “children” of a feature, we can use the <code class="docutils literal notranslate"><span class="pre">get_feature_children()</span></code> method. A “child” refers to a feature that is nested within or contained by another “parent” feature. For example, a child feature could be an exon contained within a gene or a CDS contained within a transcript.</p>
<p>This method returns a generator that yields all the child features of the specified feature.</p>
<p>For example, let’s find the children of “gene-MG_RS00035”:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">children</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff_db</span><span class="o">.</span><span class="n">get_feature_children</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;gene-MG_RS00035&quot;</span><span class="p">))</span>
<span class="n">children</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(np.int64(9155), np.int64(9920))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;cds-WP_009885556.1&#39;,
  &#39;on_alignment&#39;: None}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-find-the-parent-of-a-feature">
<h3>How to find the “parent” of a Feature<a class="headerlink" href="#how-to-find-the-parent-of-a-feature" title="Link to this heading">#</a></h3>
<p>To find the “parent” of a feature, we can use the <code class="docutils literal notranslate"><span class="pre">get_feature_parent()</span></code> method, which achieves the inverse of the above method.</p>
<p>For example, we can use the “child” we returned above <code class="docutils literal notranslate"><span class="pre">&quot;cds-WP_009885556.1&quot;</span></code>, to find the original parent gene!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">parents</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff_db</span><span class="o">.</span><span class="n">get_feature_parent</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;cds-WP_009885556.1&quot;</span><span class="p">))</span>
<span class="n">parents</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;gene&#39;,
  &#39;spans&#39;: [(np.int64(9155), np.int64(9920))],
  &#39;strand&#39;: &#39;+&#39;,
  &#39;name&#39;: &#39;gene-MG_RS00035&#39;,
  &#39;on_alignment&#39;: None}]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="how-to-combine-two-annotationdb-instances">
<h2>How to combine two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances<a class="headerlink" href="#how-to-combine-two-annotationdb-instances" title="Link to this heading">#</a></h2>
<section id="checking-the-compatibility-of-two-annotationdb-instances">
<h3>Checking the compatibility of two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances<a class="headerlink" href="#checking-the-compatibility-of-two-annotationdb-instances" title="Link to this heading">#</a></h3>
<p>Combining data requires compatibility of the databases, this can be checked via the <code class="docutils literal notranslate"><span class="pre">compatible()</span></code> method. Below we check whether a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> is compatible with a <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">compatible</span><span class="p">(</span><span class="n">anno_db</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>The method evaluates to <code class="docutils literal notranslate"><span class="pre">True</span></code>, indicating that the data of the two databases can be merged.</p>
<p>What about merging a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code>?</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">compatible</span><span class="p">(</span><span class="n">gb_db</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
<p>The method evaluates to <code class="docutils literal notranslate"><span class="pre">False</span></code>. Merging a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> is not possible.</p>
</section>
<section id="taking-the-union-of-two-annotationdb-instances">
<h3>Taking the union of two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances<a class="headerlink" href="#taking-the-union-of-two-annotationdb-instances" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">union()</span></code> method will return a <strong>new instance</strong> with merged records.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">union_db</span> <span class="o">=</span> <span class="n">gb_db</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">anno_db</span><span class="p">)</span>
<span class="n">union_db</span><span class="o">.</span><span class="n">describe</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:,</span> <span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">num_rows('gb')</span></td><td><span class="c3col_right">1,127</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">1</span></td></tr>
</tbody>
</table>
<p>
2 rows x 2 columns</p>
</div></div></div>
</div>
<p>In the new merged database, there is now content in both the “user” and “gff” table.</p>
</section>
<section id="updating-an-annotationdb-with-the-record-from-another-database">
<h3>Updating an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> with the record from another database<a class="headerlink" href="#updating-an-annotationdb-with-the-record-from-another-database" title="Link to this heading">#</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">update()</span></code> method will update records of a given database with another and return the <strong>same instance</strong> of the database.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">anno_db</span><span class="p">)</span>
<span class="n">gff_db</span><span class="o">.</span><span class="n">describe</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:,</span> <span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">2</span></td></tr>
</tbody>
</table>
<p>
2 rows x 2 columns</p>
</div></div></div>
</div>
</section>
<section id="initialise-a-annotationdb-with-another-database">
<h3>Initialise a <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> with another database<a class="headerlink" href="#initialise-a-annotationdb-with-another-database" title="Link to this heading">#</a></h3>
<p>You can assign a compatible database to the <code class="docutils literal notranslate"><span class="pre">db</span></code> argument in the <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> constructor. If it’s the same class, its db will be bound to self and directly modified.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.core.annotation_db</span> <span class="kn">import</span> <span class="n">GenbankAnnotationDb</span>

<span class="n">new_gb_db</span> <span class="o">=</span> <span class="n">GenbankAnnotationDb</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;m-genitalium-database.gbdb&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="n">anno_db</span><span class="p">)</span>
<span class="n">new_gb_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;&lt;multiple seqids&gt;&#39;, source=&#39;m-genitalium-database.gbdb&#39;, namer=None, total_records=1)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="how-to-get-a-subset-of-an-annotationdb">
<h2>How to get a subset of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-get-a-subset-of-an-annotationdb" title="Link to this heading">#</a></h2>
<p>If you want a subset of a db, use the same arguments as you would for <code class="docutils literal notranslate"><span class="pre">db.get_records_matching()</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_annotations</span>

<span class="n">gff_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">)</span>
<span class="n">just_cds</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">subset</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="n">just_cds</span><span class="o">.</span><span class="n">describe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">seqid('NC_000908.2')</span></td><td><span class="c3col_right">521</span></td></tr>
<tr><td><span class="c3col_left">biotype('CDS')</span></td><td><span class="c3col_right">521</span></td></tr>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">521</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">0</span></td></tr>
</tbody>
</table>
<p>
4 rows x 2 columns</p>
</div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The result is an in-memory database by default. To have this written to disk, assign a path to the source argument, e.g. <code class="docutils literal notranslate"><span class="pre">gff_db.subset(source=&quot;som/path/subset.gff3db&quot;,</span> <span class="pre">biotype=&quot;CDS&quot;)</span></code>.</p>
</div>
</section>
<section id="how-to-assign-an-annotationdb-to-a-sequence">
<span id="assign-db-to-seq"></span><h2>How to assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a sequence<a class="headerlink" href="#how-to-assign-an-annotationdb-to-a-sequence" title="Link to this heading">#</a></h2>
<p>For more extensive documentation about annotating alignments and sequences see <a class="reference internal" href="features.html#howto-features"><span class="std std-ref">Features</span></a>.</p>
<section id="directly-assign-an-annotationdb-to-a-sequence">
<h3>Directly assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a Sequence<a class="headerlink" href="#directly-assign-an-annotationdb-to-a-sequence" title="Link to this heading">#</a></h3>
<p>Assign the AnnotationDb to the <code class="docutils literal notranslate"><span class="pre">annotation_db</span></code> attribute of a Sequence</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_seq</span>

<span class="n">seq1</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAATTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;NC_000908&quot;</span><span class="p">,</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">seq1</span><span class="o">.</span><span class="n">annotation_db</span> <span class="o">=</span> <span class="n">anno_db</span>
<span class="n">seq1</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=1)
</pre></div>
</div>
</div>
</div>
</section>
<section id="loading-an-annotationdb-and-sequence-using-the-load-seq-function">
<h3>Loading an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> using the <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code> function<a class="headerlink" href="#loading-an-annotationdb-and-sequence-using-the-load-seq-function" title="Link to this heading">#</a></h3>
<section id="for-a-single-sequence-genbank-file">
<h4>For a single sequence Genbank file<a class="headerlink" href="#for-a-single-sequence-genbank-file" title="Link to this heading">#</a></h4>
<p>Loading a sequence from a Genbank file will automatically create a database instance containing all features present in the file. This database instance will be bound to the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> instance via the <code class="docutils literal notranslate"><span class="pre">.annotation_db</span></code> attribute, accessing this attribute displays a representation of the bound annotations.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">gb_seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/mycoplasma-genitalium.gb&quot;</span><span class="p">)</span>
<span class="n">gb_seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;NC_000908&#39;, source=&#39;:memory:&#39;, namer=None, total_records=1127)
</pre></div>
</div>
</div>
</div>
</section>
<section id="for-a-single-sequence-fasta-file-and-an-associated-gff-annotation-file">
<h4>For a single sequence FASTA file and an associated GFF annotation file<a class="headerlink" href="#for-a-single-sequence-fasta-file-and-an-associated-gff-annotation-file" title="Link to this heading">#</a></h4>
<p>Data can be loaded by providing the path to the gff file to the <code class="docutils literal notranslate"><span class="pre">annotation_path</span></code> argument of <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/mycoplasma-genitalium.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">gff_seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=0)
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This assumes an exact match of the sequence name between files!</p>
</div>
<p>In the above example, the sequence name in the fasta file does not match any records in the gff3 file (it is <code class="docutils literal notranslate"><span class="pre">&quot;NC_000908.2</span> <span class="pre">Mycoplasmoides</span> <span class="pre">genitalium</span> <span class="pre">G37,</span> <span class="pre">complete</span> <span class="pre">sequence&quot;</span></code> in the former, and <code class="docutils literal notranslate"><span class="pre">&quot;NC_000908.2&quot;</span></code> in the latter). However, if you are confident that they are related, then you can use the <code class="docutils literal notranslate"><span class="pre">label_to_name</span></code> argument of <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code> to change the sequence name as follows:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/mycoplasma-genitalium.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">,</span>
    <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=1169)
</pre></div>
</div>
</div>
</div>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
</section>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="features.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Features</p>
      </div>
    </a>
    <a class="right-next"
       href="moltypes.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Molecular types</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#what-are-the-different-types-of-annotationdb">What are the different types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>?</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-create-a-standalone-basicannotationdb">How to create a standalone <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-load-an-standalone-annotationdb-from-a-data-file">How to load an standalone <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> from a data file</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#from-a-gff-file">From a GFF file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#from-a-genbank-file">From a Genbank file</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-generate-a-summary-of-an-annotationdb">How to generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-add-custom-features-to-an-annotationdb">How to add custom features to an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-write-an-annotationdb-to-disk-for-efficient-re-loading">How to write an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to disk for efficient re-loading</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-query-an-annotationdb">How to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-feature-name">Querying via Feature Name</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-feature-biotype">Querying via Feature Biotype</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-region-of-interest">Querying via region of interest</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-the-extended-attributes-field">Querying via the extended attributes field</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-interrogate-an-annotationdb">How to interrogate an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-many-unique-genes-are-in-a-given-genome">How many unique genes are in a given genome?</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-the-children-of-a-feature">How to find the “children” of a Feature</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-the-parent-of-a-feature">How to find the “parent” of a Feature</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-combine-two-annotationdb-instances">How to combine two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#checking-the-compatibility-of-two-annotationdb-instances">Checking the compatibility of two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#taking-the-union-of-two-annotationdb-instances">Taking the union of two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#updating-an-annotationdb-with-the-record-from-another-database">Updating an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> with the record from another database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#initialise-a-annotationdb-with-another-database">Initialise a <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> with another database</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-get-a-subset-of-an-annotationdb">How to get a subset of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-assign-an-annotationdb-to-a-sequence">How to assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a sequence</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#directly-assign-an-annotationdb-to-a-sequence">Directly assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a Sequence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#loading-an-annotationdb-and-sequence-using-the-load-seq-function">Loading an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> using the <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code> function</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#for-a-single-sequence-genbank-file">For a single sequence Genbank file</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#for-a-single-sequence-fasta-file-and-an-associated-gff-annotation-file">For a single sequence FASTA file and an associated GFF annotation file</a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020-2024, cogent3 Team.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.0.2.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.15.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>