<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2023.07.18 documentation</title>

<link href="../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
<link rel="stylesheet" href="../../_static/css/docs.css" type="text/css" />
<link rel="stylesheet" href="../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../_static/sg_gallery.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />

<link rel="stylesheet" href="../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../.././_static/pygments.css">
<link rel="stylesheet" href="../.././_static/css/common.css">
<link rel="stylesheet" href="../.././_static/css/link.css">
<link rel="stylesheet" href="../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
<script src="../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../.././_static/js/initializeHighlight.js"></script>
<script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
<script src="../../_static/jquery.js"></script>
<script src="../../_static/underscore.js"></script>
<script src="../../_static/doctools.js"></script>
<script src="../../_static/thebelab-helper.js"></script>
<script src="../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
<script src="../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3/discussions" class="nav-link" title="Forum">Forum</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-using-conda-mamba">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code> / <code class="docutils literal notranslate"><span class="pre">mamba</span></code></a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">User Guides</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../app/index.html">apps</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/index-app.html">Overview of apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#types-of-apps">Types of apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#composability">Composability</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#composability-rules">Composability rules</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#an-example">An example</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#using-apps-sequentially-like-functions">Using apps sequentially like functions</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#the-settings-and-data-analysed-will-be-logged">The settings and data analysed will be logged</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#failures-are-recorded">Failures are recorded</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/app-help.html">Getting help on an app</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app-get.html">Getting an app</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/dstore.html">Data stores – collections of data records</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#how-do-i-use-a-data-store">How do I use a data store?</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#types-of-data-store">Types of data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#the-structure-of-data-stores">The structure of data stores</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#supported-operations-on-a-data-store">Supported operations on a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#opening-a-data-store">Opening a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#opening-a-read-only-data-store">Opening a read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#summarising-the-data-store">Summarising the data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#data-store-members">Data store “members”</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/dstore.html#get-one-member">Get one member</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/dstore.html#members-can-read-their-own-data">Members can read their own data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#datastoresqlite-stores-serialised-data"><code class="docutils literal notranslate"><span class="pre">DataStoreSqlite</span></code> stores serialised data</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/dstore.html#pulling-it-all-together">Pulling it all together</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#defining-the-data-stores-for-analysis">Defining the data stores for analysis</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#create-an-app-and-apply-it">Create an app and apply it</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#inspecting-the-outcome">Inspecting the outcome</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index-align.html">Sequence Alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index-evo-model.html">Modelling sequence evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index-evo-natsel.html">Testing for natural selection</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index-custom.html">Write your own Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/user_function.html">Turn your functions into composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-user-app-with-a-different-output-type">A user app with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Cookbook</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="manipulating_biological_data.html">Manipulating biological data</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul class="current">
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html">Loading a sequence from a file</a></li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#loading-an-alignment-from-a-file-or-url">Loading an alignment from a file or url</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-from-a-url">Loading from a url</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#making-a-sequence-collection-from-standard-python-objects">Making a sequence collection from standard python objects</a></li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="DNA_and_RNA_sequences.html">Sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="DNA_and_RNA_sequences.html#generic-molecular-types">Generic molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#converting-a-nucleic-acid-sequence-object-to-protein">Converting a nucleic acid sequence object to protein</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="alignments.html">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="features.html">Features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="features.html#how-to-create-a-custom-feature">How to create a custom <code class="docutils literal notranslate"><span class="pre">Feature</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="features.html#how-to-load-bulk-features-from-a-file">How to load bulk Features from a File</a></li>
<li class="toctree-l6"><a class="reference internal" href="features.html#how-to-query-a-sequence-or-alignment-for-features">How to query a Sequence or Alignment for Features</a></li>
<li class="toctree-l6"><a class="reference internal" href="features.html#examples-using-the-methods-available-on-features">Examples using the methods available on Features</a></li>
</ul>
</li>
<li class="toctree-l5 current"><a class="current reference internal" href="#">Annotation Databases</a><ul>
<li class="toctree-l6"><a class="reference internal" href="#what-are-the-different-types-of-annotationdb">What are the different types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-create-a-standalone-basicannotationdb">How to create a standalone <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-load-an-standalone-annotationdb-from-a-data-file">How to load an standalone <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> from a data file</a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-generate-a-summary-of-an-annotationdb">How to generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-add-custom-features-to-an-annotationdb">How to add custom features to an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-write-an-annotationdb-to-disk-for-efficient-re-loading">How to write an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to disk for efficient re-loading</a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-query-an-annotationdb">How to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-interrogate-an-annotationdb">How to interrogate an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-combine-two-annotationdb-instances">How to combine two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances</a></li>
<li class="toctree-l6"><a class="reference internal" href="#how-to-assign-an-annotationdb-to-a-sequence">How to assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a sequence</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html">Using genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#selecting-codes-in-methods-that-support-them">Selecting codes in methods that support them</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#translate-dna-sequences">Translate DNA sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#translating-a-dna-sequence">Translating a DNA sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#translating-a-dna-sequence-containing-stop-codons">Translating a DNA sequence containing stop codons</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-from-a-url">Loading from a url</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-delimited-data-without-a-header-line">Loading delimited data without a header line</a><ul>
<li class="toctree-l5"><a class="reference internal" href="tables.html#using-load-delimited">Using <code class="docutils literal notranslate"><span class="pre">load_delimited()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#using-filteringparser">Using <code class="docutils literal notranslate"><span class="pre">FilteringParser</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#loading-only-some-rows">Loading only some rows</a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../examples/index.html">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#modelling-evolution">Modelling Evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../examples/codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#utilities">Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/parallel.html">Parallel computations</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/parallel.html#parallel-computation-on-a-single-computer">Parallel computation on a single computer</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/parallel.html#using-app-apply-to">Using <code class="docutils literal notranslate"><span class="pre">app.apply_to()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/parallel.html#directly-using-cogent3-util-parallel-map">Directly using <code class="docutils literal notranslate"><span class="pre">cogent3.util.parallel.map()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/parallel.html#parallel-computation-on-multiple-computers">Parallel computation on multiple computers</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.open_.html">open</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_model.html">get_model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#composable-apps">Composable Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/app/composable/define_app.html">define_app</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#datastore">Datastore</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#deserialising">Deserialising</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l4"><a class="reference internal" href="../api/util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_seq-features.html">Sequence Features</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_seq-features.html#drawing-all-features-on-a-sequence-segment">Drawing all features on a sequence segment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_seq-features.html#drawing-selected-feature-biotypes">Drawing selected feature biotypes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html">Sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html">Dotplot basics</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html">Coevolution analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-support.html">Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-square.html">Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-radial.html">Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-circular.html">Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-angular.html">Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../genindex.html">Index</a></li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="annotation-databases">
<span id="anno-db"></span><h1>Annotation Databases<a class="headerlink" href="#annotation-databases" title="Permalink to this headline">¶</a></h1>
<p>This guide shows you how to use <code class="docutils literal notranslate"><span class="pre">cogent3</span></code>’s annotation databases, which are in-memory SQLite databases, to store, query and manipulate the <a class="reference internal" href="features.html#howto-features"><span class="std std-ref">Features</span></a> (also known as annotations) of one or more biological sequences.</p>
<section id="what-are-the-different-types-of-annotationdb">
<h2>What are the different types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>?<a class="headerlink" href="#what-are-the-different-types-of-annotationdb" title="Permalink to this headline">¶</a></h2>
<p>There are three types of databases that are available: <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code>, <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code>, and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code>. A <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code> provides a “user” table where custom features can be added. The latter two, as hinted in their names, are distinguished by the type of data file used to generate them. Both <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> have a “user” table for custom features, which allows them to be merged with a <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code>. However, they cannot be merged with each other!</p>
</section>
<section id="how-to-create-a-standalone-basicannotationdb">
<h2>How to create a standalone <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code><a class="headerlink" href="#how-to-create-a-standalone-basicannotationdb" title="Permalink to this headline">¶</a></h2>
<p>Achieved by creating an <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code> instance. This is an empty database to which we can add features.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.core.annotation_db</span> <span class="kn">import</span> <span class="n">BasicAnnotationDb</span>

<span class="n">anno_db</span> <span class="o">=</span> <span class="n">BasicAnnotationDb</span><span class="p">()</span>
<span class="n">anno_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=0)
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-load-an-standalone-annotationdb-from-a-data-file">
<h2>How to load an standalone <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> from a data file<a class="headerlink" href="#how-to-load-an-standalone-annotationdb-from-a-data-file" title="Permalink to this headline">¶</a></h2>
<p>Typically, we want to load a collection of features from a genomic annotation file, such as a GFF or Genbank file. For the following examples, we will use data from the bacterium <em>Mycoplasma genitalium</em>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the list of <a class="reference internal" href="../data_file_links.html#data-links"><span class="std std-ref">Data Files Used in the Documentation</span></a> to download the data used in the following examples.</p>
</div>
<section id="from-a-gff-file">
<h3>From a GFF file<a class="headerlink" href="#from-a-gff-file" title="Permalink to this headline">¶</a></h3>
<p>To load features from a GFF file, we can use the <code class="docutils literal notranslate"><span class="pre">load_annotations</span></code> function and provide the path to the GFF file. The function automatically determines the file type based on the file extension, and it returns a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> object.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_annotations</span>

<span class="n">gff_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">)</span>
<span class="n">gff_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=1169)
</pre></div>
</div>
</div>
</div>
</section>
<section id="from-a-genbank-file">
<h3>From a Genbank file<a class="headerlink" href="#from-a-genbank-file" title="Permalink to this headline">¶</a></h3>
<p>To load features from a Genbank file, we can once again use the <code class="docutils literal notranslate"><span class="pre">load_annotations</span></code> function and provide the path to the Genbank file. The function detects the file type based on the file extension, and it returns a <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> object.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_annotations</span>

<span class="n">gb_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gb&quot;</span><span class="p">)</span>
<span class="n">gb_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;NC_000908&#39;, source=&#39;:memory:&#39;, namer=None, total_records=1127)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="how-to-generate-a-summary-of-an-annotationdb">
<h2>How to generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-generate-a-summary-of-an-annotationdb" title="Permalink to this headline">¶</a></h2>
<p>To generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>, we can access the <code class="docutils literal notranslate"><span class="pre">describe</span></code> attribute of the database. This attribute returns a <code class="docutils literal notranslate"><span class="pre">cogent3.util.table.Table</span></code> instance that shows the number of records for each seqid, the count for each biotype, and the number of rows in each table (in this example there is a “gff” table with 1,169 rows and an empty “user” table).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">summary</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">describe</span>
<span class="n">summary</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">seqid('NC_000908.2')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">biotype('CDS')</span></td><td><span class="c3col_right">521</span></td></tr>
<tr><td><span class="c3col_left">biotype('RNase_P_RNA')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('SRP_RNA')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('exon')</span></td><td><span class="c3col_right">42</span></td></tr>
<tr><td><span class="c3col_left">biotype('gene')</span></td><td><span class="c3col_right">546</span></td></tr>
<tr><td><span class="c3col_left">biotype('pseudogene')</span></td><td><span class="c3col_right">17</span></td></tr>
<tr><td><span class="c3col_left">biotype('rRNA')</span></td><td><span class="c3col_right">3</span></td></tr>
<tr><td><span class="c3col_left">biotype('region')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('tRNA')</span></td><td><span class="c3col_right">36</span></td></tr>
<tr><td><span class="c3col_left">biotype('tmRNA')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">0</span></td></tr>
</tbody>
</table>
<p>
13 rows x 2 columns</p>
</div></div></div>
</div>
</section>
<section id="how-to-add-custom-features-to-an-annotationdb">
<h2>How to add custom features to an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-add-custom-features-to-an-annotationdb" title="Permalink to this headline">¶</a></h2>
<p>This is achieved via the <code class="docutils literal notranslate"><span class="pre">add_features</span></code> method and for all three types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> it will be added to the “user” table. The method requires information about the feature, such as its biotype, name, genomic location (spans), and the seqid. The seqid is necessary when linking an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> object, see <a class="reference internal" href="#assign-db-to-seq"><span class="std std-ref">How to assign an AnnotationDb to a sequence</span></a> for more information.</p>
<p>We can add a feature to the empty <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code> we created above. Now the database has one record!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> <span class="n">anno_db</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
           <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;NC_000908&quot;</span><span class="p">,</span>
           <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
           <span class="n">name</span><span class="o">=</span><span class="s2">&quot;interesting_gene&quot;</span><span class="p">,</span>
           <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span>
           <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
            <span class="p">)</span>
<span class="n">anno_db</span><span class="o">.</span><span class="n">describe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">seqid('NC_000908')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">biotype('gene')</span></td><td><span class="c3col_right">1</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">1</span></td></tr>
</tbody>
</table>
<p>
3 rows x 2 columns</p>
</div></div></div>
</div>
<p>We can also add a feature to our <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> or <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code>. Below, the previously empty “user” table now has a row count of one, indicating that our feature has been successfully added to the database.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;interesting_gene&quot;</span><span class="p">,</span>
    <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span>
    <span class="n">strand</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">gff_db</span><span class="o">.</span><span class="n">describe</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:,</span> <span class="p">:]</span> <span class="c1"># showing just last two rows</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">1</span></td></tr>
</tbody>
</table>
<p>
2 rows x 2 columns</p>
</div></div></div>
</div>
</section>
<section id="how-to-write-an-annotationdb-to-disk-for-efficient-re-loading">
<h2>How to write an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to disk for efficient re-loading<a class="headerlink" href="#how-to-write-an-annotationdb-to-disk-for-efficient-re-loading" title="Permalink to this headline">¶</a></h2>
<p>In the above examples, all databases indicate that <code class="docutils literal notranslate"><span class="pre">source=&quot;:memory:&quot;</span></code>, i.e. they are in-memory databases. We can write any database to disk using the <code class="docutils literal notranslate"><span class="pre">write()</span></code> method and providing an outpath.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># write to disk</span>
<span class="n">gb_db</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;data/m-genitalium-database.gbdb&quot;</span><span class="p">)</span>

<span class="c1"># do something</span>

<span class="c1"># re-load from disk</span>
<span class="n">quick_load_gb_db</span> <span class="o">=</span> <span class="n">GenbankAnnotationDb</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;data/m-genitalium-database.gbdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The suffix of the outpath (“.gbdb” in the above example) can be arbitrarily chosen, however, this behaviour may change in the future to only accept registered suffixes! 👀</p>
</div>
</section>
<section id="how-to-query-an-annotationdb">
<h2>How to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-query-an-annotationdb" title="Permalink to this headline">¶</a></h2>
<p>Note, there are two methods with the same interface available to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code>. A generator that yields all features that matched the query. The <strong>minimal information</strong> required to create a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> <code class="docutils literal notranslate"><span class="pre">Feature</span></code> object is provided in the returned dictionary. For more information on Features see <a class="reference internal" href="features.html#howto-features"><span class="std std-ref">Features</span></a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code>. A generator that yields all features that matched the query. The <strong>complete record</strong> for each matching feature is provided in the returned dictionary.</p></li>
</ol>
<p>Put simply, a “feature” is a subset of a “record”.</p>
<section id="querying-via-feature-name">
<h3>Querying via Feature Name<a class="headerlink" href="#querying-via-feature-name" title="Permalink to this headline">¶</a></h3>
<p>To query a database for a feature by its name, provide the name of the feature as an argument to either <code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code> or <code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code>. Since an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> can contain records for more than one sequence, it is best practice to also include the seqid of the sequence of interest.</p>
<p>For example, querying the <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> for the 16s rRNA gene:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mg_16s</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">gb_db</span><span class="o">.</span><span class="n">get_features_matching</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;MG_RS00775&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;NC_000908&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">mg_16s</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908&#39;,
  &#39;biotype&#39;: &#39;gene&#39;,
  &#39;spans&#39;: [(170011, 171529)],
  &#39;name&#39;: &#39;MG_RS00775&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-feature-biotype">
<h3>Querying via Feature Biotype<a class="headerlink" href="#querying-via-feature-biotype" title="Permalink to this headline">¶</a></h3>
<p>Similarly, <code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code> and <code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code> can be used to query the database for all features that match a given biotype.</p>
<p>For example, querying the <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> for all pseudogenes:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pseudogenes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff_db</span><span class="o">.</span><span class="n">get_features_matching</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">))</span>
<span class="n">pseudogenes</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span> <span class="c1"># showing just the first two</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;pseudogene&#39;,
  &#39;spans&#39;: [(85561, 86589)],
  &#39;name&#39;: &#39;gene-MG_RS02910&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False},
 {&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;pseudogene&#39;,
  &#39;spans&#39;: [(86784, 87549)],
  &#39;name&#39;: &#39;gene-MG_RS00385&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-region-of-interest">
<h3>Querying via region of interest<a class="headerlink" href="#querying-via-region-of-interest" title="Permalink to this headline">¶</a></h3>
<p>We can provide <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> arguments to <code class="docutils literal notranslate"><span class="pre">get_features_matching()</span></code> and <code class="docutils literal notranslate"><span class="pre">get_records_matching()</span></code> and all features within the coordinates will be returned.</p>
<p>For example, the adhesin protein of <em>M. genitalium</em> is organised in an operon between positions 220600 to 229079, so we can query for genes in that region to return all operon genes:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">operon_cds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">gff_db</span><span class="o">.</span><span class="n">get_features_matching</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">220600</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">229067</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">operon_cds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(220606, 221563)],
  &#39;name&#39;: &#39;cds-WP_009886005.1&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False},
 {&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(221569, 225904)],
  &#39;name&#39;: &#39;cds-WP_010869366.1&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False},
 {&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(225914, 229067)],
  &#39;name&#39;: &#39;cds-WP_041593683.1&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-the-extended-attributes-field">
<h3>Querying via the extended attributes field<a class="headerlink" href="#querying-via-the-extended-attributes-field" title="Permalink to this headline">¶</a></h3>
<p>A particularly useful functionality of a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> is the ability to search the extended attributes field. This allows querying for records that have matches to a specific string provided to the <code class="docutils literal notranslate"><span class="pre">attributes</span></code> argument within their extended attributes field.</p>
<p>For example, we can query for all CDS related to replication:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">replication_records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">gff_db</span><span class="o">.</span><span class="n">get_records_matching</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="s2">&quot;replication&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">replication_records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="c1"># showing just the first match</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
 &#39;source&#39;: &#39;Protein Homology&#39;,
 &#39;biotype&#39;: &#39;CDS&#39;,
 &#39;start&#39;: 685,
 &#39;end&#39;: 1828,
 &#39;score&#39;: &#39;.&#39;,
 &#39;strand&#39;: &#39;+&#39;,
 &#39;phase&#39;: &#39;0&#39;,
 &#39;attributes&#39;: &#39;ID=cds-WP_009885562.1;Parent=gene-MG_RS00005;Dbxref=Genbank:WP_009885562.1;Name=WP_009885562.1;Ontology_term=GO:0006260,GO:0003887,GO:0009360;gbkey=CDS;gene=dnaN;go_component=DNA polymerase III complex|0009360||IEA;go_function=DNA-directed DNA polymerase activity|0003887||IEA;go_process=DNA replication|0006260||IEA;inference=COORDINATES: similar to AA sequence:RefSeq:WP_010874358.1;locus_tag=MG_RS00005;product=DNA polymerase III subunit beta;protein_id=WP_009885562.1;transl_table=4&#39;,
 &#39;comments&#39;: None,
 &#39;spans&#39;: array([[ 685, 1828]]),
 &#39;name&#39;: &#39;cds-WP_009885562.1&#39;,
 &#39;parent_id&#39;: &#39;gene-MG_RS00005&#39;}
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Extended attribute querying only works for GFF databases!</p>
</div>
</section>
</section>
<section id="how-to-interrogate-an-annotationdb">
<h2>How to interrogate an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#how-to-interrogate-an-annotationdb" title="Permalink to this headline">¶</a></h2>
<p>An <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> can be interrogated to explore the properties of a sequence without needing the sequence information.</p>
<section id="how-many-unique-genes-are-in-a-given-genome">
<h3>How many unique genes are in a given genome?<a class="headerlink" href="#how-many-unique-genes-are-in-a-given-genome" title="Permalink to this headline">¶</a></h3>
<p><em>Mycoplasma genitalium</em> has the smallest bacterial genome, so the number of genes in the loaded database represents the approximate minimal set of genes required for bacterial life! We can see the total number of genes by using the <code class="docutils literal notranslate"><span class="pre">num_matches()</span></code> method and specifying the condition we want to be matched is that the biotype is “gene”.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gb_db</span><span class="o">.</span><span class="n">num_matches</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>563
</pre></div>
</div>
</div>
</div>
<p>The count is 563, however, this may include genes with more than one copy. To determine the number of distinct genes we can use the <code class="docutils literal notranslate"><span class="pre">count_distinct()</span></code> method and specify <code class="docutils literal notranslate"><span class="pre">biotype=&quot;gene&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">name=True</span></code> to indicate we are interested in genes with distinct names.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">total_genes</span> <span class="o">=</span> <span class="n">gb_db</span><span class="o">.</span><span class="n">count_distinct</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">single_copy</span> <span class="o">=</span> <span class="n">total_genes</span><span class="p">[</span><span class="n">total_genes</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
<span class="nb">len</span><span class="p">(</span><span class="n">single_copy</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>561
</pre></div>
</div>
</div>
</div>
<p>The count of unique genes is 561. This means that almost every gene is present only once in the genome, very little redundancy here!</p>
<p>Just for fun, let’s try this with the GFF database… (downloaded from the exact same source)</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">total_genes</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">num_matches</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;total genes: &quot;</span><span class="p">,</span> <span class="n">total_genes</span><span class="p">)</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">gff_db</span><span class="o">.</span><span class="n">count_distinct</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">single_copy</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="n">genes</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="s2">&quot;count&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;single copy genes: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">single_copy</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>total genes:  547
single copy genes:  547
</pre></div>
</div>
</div>
</div>
<p>What? 🤯</p>
</section>
<section id="how-to-find-the-children-of-a-feature">
<h3>How to find the “children” of a Feature<a class="headerlink" href="#how-to-find-the-children-of-a-feature" title="Permalink to this headline">¶</a></h3>
<p>To find the “children” of a feature, we can use the <code class="docutils literal notranslate"><span class="pre">get_feature_children()</span></code> method. A “child” refers to a feature that is nested within or contained by another “parent” feature. For example, a child feature could be an exon contained within a gene or a CDS contained within a transcript.</p>
<p>This method returns a generator that yields all the child features of the specified feature.</p>
<p>For example, let’s find the children of “gene-MG_RS00035”:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">children</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff_db</span><span class="o">.</span><span class="n">get_feature_children</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;gene-MG_RS00035&quot;</span><span class="p">))</span>
<span class="n">children</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;CDS&#39;,
  &#39;spans&#39;: [(9155, 9920)],
  &#39;name&#39;: &#39;cds-WP_009885556.1&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False}]
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-find-the-parent-of-a-feature">
<h3>How to find the “parent” of a Feature<a class="headerlink" href="#how-to-find-the-parent-of-a-feature" title="Permalink to this headline">¶</a></h3>
<p>To find the “parent” of a feature, we can use the <code class="docutils literal notranslate"><span class="pre">get_feature_parent()</span></code> method, which achieves the inverse of the above method.</p>
<p>For example, we can use the “child” we returned above <code class="docutils literal notranslate"><span class="pre">&quot;cds-WP_009885556.1&quot;</span></code>, to find the original parent gene!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">parents</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gff_db</span><span class="o">.</span><span class="n">get_feature_parent</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;cds-WP_009885556.1&quot;</span><span class="p">))</span>
<span class="n">parents</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;seqid&#39;: &#39;NC_000908.2&#39;,
  &#39;biotype&#39;: &#39;gene&#39;,
  &#39;spans&#39;: [(9155, 9920)],
  &#39;name&#39;: &#39;gene-MG_RS00035&#39;,
  &#39;on_alignment&#39;: None,
  &#39;reversed&#39;: False}]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="how-to-combine-two-annotationdb-instances">
<h2>How to combine two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances<a class="headerlink" href="#how-to-combine-two-annotationdb-instances" title="Permalink to this headline">¶</a></h2>
<section id="checking-the-compatibility-of-two-annotationdb-instances">
<h3>Checking the compatibility of two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances<a class="headerlink" href="#checking-the-compatibility-of-two-annotationdb-instances" title="Permalink to this headline">¶</a></h3>
<p>Combining data requires compatibility of the databases, this can be checked via the <code class="docutils literal notranslate"><span class="pre">compatible()</span></code> method. Below we check whether a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> is compatible with a <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">compatible</span><span class="p">(</span><span class="n">anno_db</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<p>The method evaluates to <code class="docutils literal notranslate"><span class="pre">True</span></code>, indicating that the data of the two databases can be merged.</p>
<p>What about merging a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code>?</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">compatible</span><span class="p">(</span><span class="n">gb_db</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>False
</pre></div>
</div>
</div>
</div>
<p>The method evaluates to <code class="docutils literal notranslate"><span class="pre">False</span></code>. Merging a <code class="docutils literal notranslate"><span class="pre">GffAnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">GenbankAnnotationDb</span></code> is not possible.</p>
</section>
<section id="taking-the-union-of-two-annotationdb-instances">
<h3>Taking the union of two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances<a class="headerlink" href="#taking-the-union-of-two-annotationdb-instances" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">union()</span></code> method will return a <strong>new instance</strong> with merged records.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">union_db</span> <span class="o">=</span> <span class="n">gb_db</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">anno_db</span><span class="p">)</span>
<span class="n">union_db</span><span class="o">.</span><span class="n">describe</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:,</span> <span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">num_rows('gb')</span></td><td><span class="c3col_right">1,127</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">1</span></td></tr>
</tbody>
</table>
<p>
2 rows x 2 columns</p>
</div></div></div>
</div>
<p>In the new merged database, there is now content in both the “user” and “gff” table.</p>
</section>
<section id="updating-an-annotationdb-with-the-record-from-another-database">
<h3>Updating an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> with the record from another database<a class="headerlink" href="#updating-an-annotationdb-with-the-record-from-another-database" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">update()</span></code> method will update records of a given database with another and return the <strong>same instance</strong> of the database.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_db</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">anno_db</span><span class="p">)</span>
<span class="n">gff_db</span><span class="o">.</span><span class="n">describe</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">:,</span> <span class="p">:]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html">
<div class="c3table">

<style>
.c3table table {margin: 10px 0;}
.c3table tr:last-child {border-bottom: 1px solid #000;} 
.c3table tr > th {text-align: left; padding: 0 5px;}
.c3table tr > td {text-align: left; padding: 5px;}
.c3table tr:nth-child(even) {background: #f7f7f7 !important;}
.c3table .ellipsis {background: rgba(0, 0, 0, .01);}
.c3table .index {background: rgba(161, 195, 209, 0.0); margin: 10px; font-weight: 600;}
.c3table .head_cell {background: rgba(161, 195, 209, 0.75); font-weight: bold; text-align: center;}
.c3table caption {color: rgb(250, 250, 250); background: rgba(30, 140, 200, 1); padding: 3px; white-space: nowrap; caption-side: top;}
.c3table .cell_title {font-weight: bold;}
.c3col_left { text-align: left !important; display: block;}
.c3col_right { text-align: right !important; display: block;}
.c3col_center { text-align: center !important; display: block;}
</style>

<table>

<thead class="head_cell">
<th></th><th>count</th>
</thead>
<tbody>
<tr><td><span class="c3col_left">num_rows('gff')</span></td><td><span class="c3col_right">1,169</span></td></tr>
<tr><td><span class="c3col_left">num_rows('user')</span></td><td><span class="c3col_right">2</span></td></tr>
</tbody>
</table>
<p>
2 rows x 2 columns</p>
</div></div></div>
</div>
</section>
<section id="initialise-a-annotationdb-with-another-database">
<h3>Initialise a <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> with another database<a class="headerlink" href="#initialise-a-annotationdb-with-another-database" title="Permalink to this headline">¶</a></h3>
<p>You can assign a compatible database to the <code class="docutils literal notranslate"><span class="pre">db</span></code> argument in the <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> constructor. If it’s the same class, its db will be bound to self and directly modified.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.core.annotation_db</span> <span class="kn">import</span> <span class="n">GenbankAnnotationDb</span>

<span class="n">new_gb_db</span> <span class="o">=</span> <span class="n">GenbankAnnotationDb</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="s2">&quot;m-genitalium-database.gbdb&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="n">anno_db</span><span class="p">)</span>
<span class="n">new_gb_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;&lt;multiple seqids&gt;&#39;, source=&#39;m-genitalium-database.gbdb&#39;, namer=None, total_records=1)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="how-to-assign-an-annotationdb-to-a-sequence">
<span id="assign-db-to-seq"></span><h2>How to assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a sequence<a class="headerlink" href="#how-to-assign-an-annotationdb-to-a-sequence" title="Permalink to this headline">¶</a></h2>
<p>For more extensive documentation about annotating alignments and sequences see <a class="reference internal" href="features.html#howto-features"><span class="std std-ref">Features</span></a>.</p>
<section id="directly-assign-an-annotationdb-to-a-sequence">
<h3>Directly assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a Sequence<a class="headerlink" href="#directly-assign-an-annotationdb-to-a-sequence" title="Permalink to this headline">¶</a></h3>
<p>Assign the AnnotationDb to the <code class="docutils literal notranslate"><span class="pre">annotation_db</span></code> attribute of a Sequence</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_seq</span>

<span class="n">seq1</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAATTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;NC_000908&quot;</span><span class="p">,</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">seq1</span><span class="o">.</span><span class="n">annotation_db</span> <span class="o">=</span> <span class="n">anno_db</span>
<span class="n">seq1</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=1)
</pre></div>
</div>
</div>
</div>
</section>
<section id="loading-an-annotationdb-and-sequence-using-the-load-seq-function">
<h3>Loading an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> and <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> using the <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code> function<a class="headerlink" href="#loading-an-annotationdb-and-sequence-using-the-load-seq-function" title="Permalink to this headline">¶</a></h3>
<section id="for-a-single-sequence-genbank-file">
<h4>For a single sequence Genbank file<a class="headerlink" href="#for-a-single-sequence-genbank-file" title="Permalink to this headline">¶</a></h4>
<p>Loading a sequence from a Genbank file will automatically create a database instance containing all features present in the file. This database instance will be bound to the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> instance via the <code class="docutils literal notranslate"><span class="pre">.annotation_db</span></code> attribute, accessing this attribute displays a representation of the bound annotations.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">gb_seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/mycoplasma-genitalium.gb&quot;</span><span class="p">)</span>
<span class="n">gb_seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;NC_000908&#39;, source=&#39;:memory:&#39;, namer=None, total_records=1127)
</pre></div>
</div>
</div>
</div>
</section>
<section id="for-a-single-sequence-fasta-file-and-an-associated-gff-annotation-file">
<h4>For a single sequence FASTA file and an associated GFF annotation file<a class="headerlink" href="#for-a-single-sequence-fasta-file-and-an-associated-gff-annotation-file" title="Permalink to this headline">¶</a></h4>
<p>Data can be loaded by providing the path to the gff file to the <code class="docutils literal notranslate"><span class="pre">annotation_path</span></code> argument of <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gff_seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/mycoplasma-genitalium.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">gff_seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=0)
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This assumes an exact match of the sequence name between files!</p>
</div>
<p>In the above example, the sequence name in the fasta file does not match any records in the gff3 file (it is <code class="docutils literal notranslate"><span class="pre">&quot;NC_000908.2</span> <span class="pre">Mycoplasmoides</span> <span class="pre">genitalium</span> <span class="pre">G37,</span> <span class="pre">complete</span> <span class="pre">sequence&quot;</span></code> in the former, and <code class="docutils literal notranslate"><span class="pre">&quot;NC_000908.2&quot;</span></code> in the latter). However, if you are confident that they are related, then you can use the <code class="docutils literal notranslate"><span class="pre">label_to_name</span></code> argument of <code class="docutils literal notranslate"><span class="pre">load_seq()</span></code> to change the sequence name as follows:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/mycoplasma-genitalium.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/mycoplasma-genitalium.gff&quot;</span><span class="p">,</span>
    <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=1169)
</pre></div>
</div>
</div>
</div>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
</section>
</section>
</section>
</section>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020-2023, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 4.5.0 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>