<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2020.4 documentation</title>

<link href="../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../_static/sphinx-dropdown.css" type="text/css" />
<link rel="stylesheet" href="../../_static/panels-bootstrap.min.css" type="text/css" />
<link rel="stylesheet" href="../../_static/gallery.css" type="text/css" />
<link rel="stylesheet" href="../../_static/gallery-binder.css" type="text/css" />
<link rel="stylesheet" href="../../_static/gallery-dataframe.css" type="text/css" />

<link rel="stylesheet" href="../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../.././_static/pygments.css">
<link rel="stylesheet" href="../.././_static/css/common.css">
<link rel="stylesheet" href="../.././_static/css/link.css">
<link rel="stylesheet" href="../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
<script src="../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../.././_static/js/initializeHighlight.js"></script>
<script src="../../_static/jquery.js"></script>
<script src="../../_static/underscore.js"></script>
<script src="../../_static/doctools.js"></script>
<script src="../../_static/language_data.js"></script>
<script src="../../_static/thebelab-helper.js"></script>
<script src="../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../_static/require.js"></script>
<script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
<script src="../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../install.html#installing-using-conda">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../install.html#manual-creation-of-the-conda-environment">Manual creation of the <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../app/index.html">The <code class="docutils literal notranslate"><span class="pre">apps</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#overview">Overview</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/warning.html">Warning – apps are alpha!</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#define-the-apps">Define the apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#using-apps-like-functions">Using apps like functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#you-can-log-the-settings-and-data-analysed">You can log the settings and data analysed</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/dstore.html">Specifying data for analysis</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#a-read-only-data-store">A read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#data-store-members">Data store “members”</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#showing-the-last-few-members">Showing the last few members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#filtering-a-data-store-for-specific-members">Filtering a data store for specific members</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/dstore.html#tinydb-data-stores-are-special">TinyDB data stores are special</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/available-apps.html">What apps are there?</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-progressive-align-app">The <code class="docutils literal notranslate"><span class="pre">progressive_align</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-model-app">The <code class="docutils literal notranslate"><span class="pre">model</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-hypothesis-app">The <code class="docutils literal notranslate"><span class="pre">hypothesis</span></code> App</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-natsel-apps">The <code class="docutils literal notranslate"><span class="pre">natsel</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../app/index.html#the-user-function-apps">The <code class="docutils literal notranslate"><span class="pre">user_function</span></code> Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/user_function.html">Custom composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../app/user_function.html#a-user-function-for-with-a-different-output-type">A user function for with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">Cookbook</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="manipulating_biological_data.html">Manipulating biological data</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul class="current">
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html">Loading an alignment from a file</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="alignments.html">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="alignments.html#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="DNA_and_RNA_sequences.html"><code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a><ul>
<li class="toctree-l6"><a class="reference internal" href="DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5 current"><a class="current reference internal" href="#">Annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="#annotations-with-coordinates">Annotations with coordinates</a><ul>
<li class="toctree-l6"><a class="reference internal" href="#automated-introduction-from-reading-genbank-files">Automated introduction from reading genbank files</a></li>
<li class="toctree-l6"><a class="reference internal" href="#customising-annotation-construction-from-reading-a-genbank-file">Customising annotation construction from reading a genbank file</a></li>
<li class="toctree-l6"><a class="reference internal" href="#creating-directly-on-a-sequence">Creating directly on a sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="#via">Via</a></li>
<li class="toctree-l6"><a class="reference internal" href="#adding-as-a-series-or-item-wise">Adding as a series or item-wise</a></li>
<li class="toctree-l6"><a class="reference internal" href="#taking-the-union-of-annotations">Taking the union of annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="#getting-annotation-coordinates">Getting annotation coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="#annotations-have-shadows">Annotations have shadows</a></li>
<li class="toctree-l6"><a class="reference internal" href="#adding-to-a-sequence-member-of-an-alignment">Adding to a sequence member of an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="#adding-to-an-alignment">Adding to an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="#slicing-sequences-and-alignments-by-annotations">Slicing sequences and alignments by annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="#slicing-by-pseudo-feature-or-feature-series">Slicing by pseudo-feature or feature series</a></li>
<li class="toctree-l6"><a class="reference internal" href="#slice-series-must-not-be-overlapping">Slice series must not be overlapping</a></li>
<li class="toctree-l6"><a class="reference internal" href="#you-can-slice-an-annotation-itself">You can slice an annotation itself</a></li>
<li class="toctree-l6"><a class="reference internal" href="#sequence-vs-alignment-slicing">Sequence vs Alignment slicing</a></li>
<li class="toctree-l6"><a class="reference internal" href="#copying-annotations">Copying annotations</a></li>
<li class="toctree-l6"><a class="reference internal" href="#querying">Querying</a></li>
<li class="toctree-l6"><a class="reference internal" href="#querying-produces-objects-only-valid-for-their-source">Querying produces objects only valid for their source</a></li>
<li class="toctree-l6"><a class="reference internal" href="#querying-for-absent-annotation">Querying for absent annotation</a></li>
<li class="toctree-l6"><a class="reference internal" href="#querying-features-that-span-gaps-in-alignments">Querying features that span gaps in alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="#as-one-span-unifies-features-with-discontinuous-alignment-coordinates"><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates</a></li>
<li class="toctree-l6"><a class="reference internal" href="#behaviour-of-annotations-on-nucleic-acid-sequences">Behaviour of annotations on nucleic acid sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="#masking-annotated-regions">Masking annotated regions</a></li>
<li class="toctree-l6"><a class="reference internal" href="#masking-annotated-regions-on-alignments">Masking annotated regions on alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="#you-can-take-mask-of-the-shadow">You can take mask of the shadow</a></li>
<li class="toctree-l6"><a class="reference internal" href="#what-features-of-a-certain-type-are-available">What features of a certain type are available?</a></li>
<li class="toctree-l6"><a class="reference internal" href="#getting-all-features-of-a-type-or-everything-but-that-type">Getting all features of a type, or everything but that type</a></li>
<li class="toctree-l6"><a class="reference internal" href="#getting-sequence-features-when-you-have-an-alignment-object">Getting sequence features when you have an alignment object</a></li>
<li class="toctree-l6"><a class="reference internal" href="#annotation-display-on-sequences">Annotation display on sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html">Translate DNA sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="genetic_code.html#translating-a-dna-sequence-with-a-terminating-stop-codon">Translating a DNA sequence with a terminating stop codon</a><ul>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#by-removing-the-trailing-stop-codon-first">By removing the trailing stop codon first</a></li>
<li class="toctree-l6"><a class="reference internal" href="genetic_code.html#by-slicing-the-dnasequence-first">By slicing the <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> first</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../examples/index.html">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/seq_features.html">Advanced sequence handling</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/complete_seq_features.html">Annotating alignments and sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../examples/period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../examples/period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#modelling-evolution">Modelling Evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/scope_model_params_on_trees.html#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../examples/codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../examples/index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="../examples/phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api/index.html">API</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api/index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../api/alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../api/table/classes/cogent3.util.table.Columns.html">Columns</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../api/tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../api/tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html">Evaluating coevolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html">Dotplot</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-1.html#all-options">All options</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l5"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l5"><a class="reference internal" href="../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html">Draw sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-angular.html">Display a Phylogenetic Tree with a Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-circular.html">Display a Phylogenetic Tree with a Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-radial.html">Display a Phylogenetic Tree with a Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-square.html">Display a Phylogenetic Tree with a Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#changing-scale-bar-placement">Changing scale bar placement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../draw/tree/plot_tree-support.html">Display a Phylogenetic Tree Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<div class="section" id="annotations">
<h1>Annotations<a class="headerlink" href="#annotations" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="annotations-with-coordinates">
<h1>Annotations with coordinates<a class="headerlink" href="#annotations-with-coordinates" title="Permalink to this headline">¶</a></h1>
<p>For more extensive documentation about annotations see <a class="reference internal" href="../examples/seq_features.html#seq-annotations"><span class="std std-ref">Advanced sequence handling</span></a>.</p>
<div class="section" id="automated-introduction-from-reading-genbank-files">
<h2>Automated introduction from reading genbank files<a class="headerlink" href="#automated-introduction-from-reading-genbank-files" title="Permalink to this headline">¶</a></h2>
<p>We load a sample genbank file with plenty of features and grab the CDS features.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.parse.genbank</span> <span class="kn">import</span> <span class="n">RichGenbankParser</span>

<span class="n">parser</span> <span class="o">=</span> <span class="n">RichGenbankParser</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data/ST_genome_part.gb&quot;</span><span class="p">))</span>
<span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">parser</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">accession</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AE006468
</pre></div>
</div>
</div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cds</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>[CDS &quot;thrL&quot; at [189:255]/10020, CDS &quot;thrA&quot; at [336:2799]/10020, CDS &quot;thrB&quot; at [2800:3730]/10020, CDS &quot;thrC&quot; at [3733:5020]/10020, CDS &quot;yaaA&quot; at [5887:5113]/10020, CDS &quot;yaaJ&quot; at [7396:5965]/10020, CDS &quot;talB&quot; at [7664:8618]/10020, CDS &quot;mog&quot; at [8728:9319]/10020, CDS &quot;yaaH&quot; at [9942:9375]/10020]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="customising-annotation-construction-from-reading-a-genbank-file">
<h2>Customising annotation construction from reading a genbank file<a class="headerlink" href="#customising-annotation-construction-from-reading-a-genbank-file" title="Permalink to this headline">¶</a></h2>
<p>You can write your own code to construct annotation objects. One reason you might do this is some genbank files do not have a <code class="docutils literal notranslate"><span class="pre">/gene</span></code> tag on gene related features, instead only possessing a <code class="docutils literal notranslate"><span class="pre">/locus_tag</span></code>. For illustrating the approach we only create annotations for <code class="docutils literal notranslate"><span class="pre">CDS</span></code> features. We write a custom callback function that uses the <code class="docutils literal notranslate"><span class="pre">locus_tag</span></code> as the <code class="docutils literal notranslate"><span class="pre">Feature</span></code> name.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.core.annotation</span> <span class="kn">import</span> <span class="n">Feature</span>

<span class="k">def</span> <span class="nf">add_annotation</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">feature</span><span class="p">,</span> <span class="n">spans</span><span class="p">):</span>
    <span class="n">type_</span> <span class="o">=</span> <span class="n">feature</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">type_</span> <span class="o">!=</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
        <span class="k">return</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">feature</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;locus_tag&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">name</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
    <span class="n">seq</span><span class="o">.</span><span class="n">add_annotation</span><span class="p">(</span><span class="n">Feature</span><span class="p">,</span> <span class="n">type_</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">spans</span><span class="p">)</span>

<span class="n">parser</span> <span class="o">=</span> <span class="n">RichGenbankParser</span><span class="p">(</span>
    <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data/ST_genome_part.gb&quot;</span><span class="p">),</span> <span class="n">add_annotation</span><span class="o">=</span><span class="n">add_annotation</span>
<span class="p">)</span>
<span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">parser</span><span class="p">:</span>  <span class="c1"># just reading one accession,sequence</span>
    <span class="k">break</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>[CDS &quot;STM0001&quot; at [189:255]/10020, CDS &quot;STM0002&quot; at [336:2799]/10020, CDS &quot;STM0003&quot; at [2800:3730]/10020, CDS &quot;STM0004&quot; at [3733:5020]/10020, CDS &quot;STM0005&quot; at [5887:5113]/10020, CDS &quot;STM0006&quot; at [7396:5965]/10020, CDS &quot;STM0007&quot; at [7664:8618]/10020, CDS &quot;STM0008&quot; at [8728:9319]/10020, CDS &quot;STM0009&quot; at [9942:9375]/10020]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="creating-directly-on-a-sequence">
<h2>Creating directly on a sequence<a class="headerlink" href="#creating-directly-on-a-sequence" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="kn">from</span> <span class="nn">cogent3.core.annotation</span> <span class="kn">import</span> <span class="n">Feature</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAA&quot;</span> <span class="s2">&quot;TTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">15</span><span class="p">])</span>  <span class="c1"># this will be exon 1</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="mi">30</span><span class="p">:</span><span class="mi">40</span><span class="p">])</span>  <span class="c1"># this will be exon 2</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="mi">45</span><span class="p">:</span><span class="mi">48</span><span class="p">])</span>  <span class="c1"># this will be exon 3</span>
<span class="n">s2</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CGAAACGTTT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="n">s3</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CGAAACGTTT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq3&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CCCCC
TTTTTAAAAA
CCC
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="via">
<h2>Via<a class="headerlink" href="#via" title="Permalink to this headline">¶</a></h2>
<div class="section" id="add-annotation">
<h3><code class="docutils literal notranslate"><span class="pre">add_annotation</span></code><a class="headerlink" href="#add-annotation" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>
<span class="kn">from</span> <span class="nn">cogent3.core.annotation</span> <span class="kn">import</span> <span class="n">Feature</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAA&quot;</span> <span class="s2">&quot;TTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span>
<span class="p">)</span>
<span class="n">exon1</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_annotation</span><span class="p">(</span><span class="n">Feature</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">)])</span>
<span class="n">exon2</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_annotation</span><span class="p">(</span><span class="n">Feature</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="add-feature">
<h3><code class="docutils literal notranslate"><span class="pre">add_feature</span></code><a class="headerlink" href="#add-feature" title="Permalink to this headline">¶</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAA&quot;</span> <span class="s2">&quot;TTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span>
<span class="p">)</span>
<span class="n">exon3</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">45</span><span class="p">,</span> <span class="mi">48</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<p><em>There are other annotation types.</em></p>
</div>
</div>
<div class="section" id="adding-as-a-series-or-item-wise">
<h2>Adding as a series or item-wise<a class="headerlink" href="#adding-as-a-series-or-item-wise" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s2</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CGAAACGTTT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq2&quot;</span><span class="p">)</span>
<span class="n">cpgs_series</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">,</span> <span class="s2">&quot;cpg&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">)])</span>
<span class="n">s3</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CGAAACGTTT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq3&quot;</span><span class="p">)</span>
<span class="n">cpg1</span> <span class="o">=</span> <span class="n">s3</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">,</span> <span class="s2">&quot;cpg&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)])</span>
<span class="n">cpg2</span> <span class="o">=</span> <span class="n">s3</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">,</span> <span class="s2">&quot;cpg&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="taking-the-union-of-annotations">
<h2>Taking the union of annotations<a class="headerlink" href="#taking-the-union-of-annotations" title="Permalink to this headline">¶</a></h2>
<p>Construct a pseudo-feature (<code class="docutils literal notranslate"><span class="pre">cds</span></code>) that’s a union of other features (<code class="docutils literal notranslate"><span class="pre">exon1</span></code>, <code class="docutils literal notranslate"><span class="pre">exon2</span></code>, <code class="docutils literal notranslate"><span class="pre">exon3</span></code>).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAA&quot;</span> <span class="s2">&quot;TTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span>
<span class="p">)</span>
<span class="n">exon1</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">)])</span>
<span class="n">exon2</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">)])</span>
<span class="n">exon3</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">45</span><span class="p">,</span> <span class="mi">48</span><span class="p">)])</span>
<span class="n">cds</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">([</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon2</span><span class="p">,</span> <span class="n">exon3</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="getting-annotation-coordinates">
<h2>Getting annotation coordinates<a class="headerlink" href="#getting-annotation-coordinates" title="Permalink to this headline">¶</a></h2>
<p>These are useful for doing custom things, e.g. you could construct intron features using the below.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cds</span><span class="o">.</span><span class="n">get_coordinates</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[(10, 15), (30, 40), (45, 48)]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="annotations-have-shadows">
<h2>Annotations have shadows<a class="headerlink" href="#annotations-have-shadows" title="Permalink to this headline">¶</a></h2>
<p>A shadow is a span representing everything but the annotation.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">not_cds</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">get_shadow</span><span class="p">()</span>
<span class="n">not_cds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>region &quot;not exon&quot; at [0:10, 15:30, 40:45, 48:49]/49
</pre></div>
</div>
</div>
</div>
<p>Compare to the coordinates of the original.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>region &quot;exon&quot; at [10:15, 30:40, 45:48]/49
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="adding-to-a-sequence-member-of-an-alignment">
<h2>Adding to a sequence member of an alignment<a class="headerlink" href="#adding-to-a-sequence-member-of-an-alignment" title="Permalink to this headline">¶</a></h2>
<p>The following annotation is directly applied onto the sequence and so is in ungapped sequence coordinates.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln1</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAACCCCCA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">seq_exon</span> <span class="o">=</span> <span class="n">aln1</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="adding-to-an-alignment">
<h2>Adding to an alignment<a class="headerlink" href="#adding-to-an-alignment" title="Permalink to this headline">¶</a></h2>
<p>We add an annotation directly onto an alignment. In this example we add a <code class="docutils literal notranslate"><span class="pre">Variable</span></code> that can be displayed as a red line on a drawing. The resulting annotation (<code class="docutils literal notranslate"><span class="pre">red_data</span></code> here) is in <strong>alignment coordinates</strong>!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.core.annotation</span> <span class="kn">import</span> <span class="n">Variable</span>

<span class="n">red_data</span> <span class="o">=</span> <span class="n">aln1</span><span class="o">.</span><span class="n">add_annotation</span><span class="p">(</span>
    <span class="n">Variable</span><span class="p">,</span> <span class="s2">&quot;redline&quot;</span><span class="p">,</span> <span class="s2">&quot;align&quot;</span><span class="p">,</span> <span class="p">[((</span><span class="mi">0</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span> <span class="mi">1</span><span class="p">),</span> <span class="p">((</span><span class="mi">15</span><span class="p">,</span> <span class="mi">30</span><span class="p">),</span> <span class="mi">2</span><span class="p">),</span> <span class="p">((</span><span class="mi">30</span><span class="p">,</span> <span class="mi">45</span><span class="p">),</span> <span class="mi">3</span><span class="p">)]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</div>
<div class="section" id="slicing-sequences-and-alignments-by-annotations">
<h2>Slicing sequences and alignments by annotations<a class="headerlink" href="#slicing-sequences-and-alignments-by-annotations" title="Permalink to this headline">¶</a></h2>
<p>By a feature or coordinates returns same sequence span</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAA&quot;</span> <span class="s2">&quot;TTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span>
<span class="p">)</span>
<span class="n">exon1</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">)])</span>
<span class="n">exon2</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">)])</span>
<span class="n">s1</span><span class="p">[</span><span class="n">exon1</span><span class="p">]</span>
<span class="n">s1</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">15</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">seq1</td><td><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span></td></tr>
</table>
<p><i>5 DnaSequence</i></p>
</div></div></div>
</div>
<p>Using the annotation object <code class="docutils literal notranslate"><span class="pre">get_slice</span></code> method returns the same thing.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">s1</span><span class="p">[</span><span class="n">exon2</span><span class="p">]</span>
<span class="n">exon2</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">seq1</td><td><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span></td></tr>
</table>
<p><i>10 DnaSequence</i></p>
</div></div></div>
</div>
</div>
<div class="section" id="slicing-by-pseudo-feature-or-feature-series">
<h2>Slicing by pseudo-feature or feature series<a class="headerlink" href="#slicing-by-pseudo-feature-or-feature-series" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s1</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAA&quot;</span> <span class="s2">&quot;TTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span>
<span class="p">)</span>
<span class="n">exon1</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">)])</span>
<span class="n">exon2</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">40</span><span class="p">)])</span>
<span class="n">exon3</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">45</span><span class="p">,</span> <span class="mi">48</span><span class="p">)])</span>
<span class="n">cds</span> <span class="o">=</span> <span class="n">s1</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">([</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon2</span><span class="p">,</span> <span class="n">exon3</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">cds</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon2</span><span class="p">,</span> <span class="n">exon3</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CCCCCTTTTTAAAAACCC
CCCCCTTTTTAAAAACCC
</pre></div>
</div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Slices are applied in order!</p>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon2</span><span class="p">,</span> <span class="n">exon3</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">exon2</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">exon3</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon3</span><span class="p">,</span> <span class="n">exon2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AAGAAGAAGACCCCCAAAAAAAAAATTTTTTTTTTAAAAAGGGAACCCT
CCCCCTTTTTAAAAACCC
TTTTTAAAAA
CCC
CCCCCCCCTTTTTAAAAA
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="slice-series-must-not-be-overlapping">
<h2>Slice series must not be overlapping<a class="headerlink" href="#slice-series-must-not-be-overlapping" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">s1</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">:</span><span class="mi">15</span><span class="p">]</span>
<span class="n">s1</span><span class="p">[</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="o">&lt;</span><span class="n">ipython</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">19</span><span class="o">-</span><span class="mi">973</span><span class="n">ffb8e58c5</span><span class="o">&gt;</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">s1</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">:</span><span class="mi">15</span><span class="p">]</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="n">s1</span><span class="p">[</span><span class="n">exon1</span><span class="p">,</span> <span class="n">exon1</span><span class="p">]</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">__getitem__</span><span class="nt">(self, index)</span>
<span class="g g-Whitespace">     </span><span class="mi">89</span>         <span class="nb">map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_as_map</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">90</span>         <span class="n">new</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped</span><span class="p">(</span><span class="nb">map</span><span class="p">)</span>
<span class="ne">---&gt; </span><span class="mi">91</span>         <span class="n">sliced_annots</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sliced_annotations</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="nb">map</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">92</span>         <span class="n">new</span><span class="o">.</span><span class="n">attach_annotations</span><span class="p">(</span><span class="n">sliced_annots</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">93</span>         <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;_repr_policy&quot;</span><span class="p">):</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">_sliced_annotations</span><span class="nt">(self, new, slice)</span>
<span class="g g-Whitespace">     </span><span class="mi">34</span>             <span class="n">slicemap</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_as_map</span><span class="p">(</span><span class="nb">slice</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">35</span>             <span class="c1"># try:</span>
<span class="ne">---&gt; </span><span class="mi">36</span>             <span class="n">newmap</span> <span class="o">=</span> <span class="n">slicemap</span><span class="o">.</span><span class="n">inverse</span><span class="p">()</span>
<span class="g g-Whitespace">     </span><span class="mi">37</span>             <span class="c1"># except ValueError, detail:</span>
<span class="g g-Whitespace">     </span><span class="mi">38</span>             <span class="c1">#    print &quot;Annotations dropped because %s&quot; % detail</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/location.py</span> in <span class="ni">inverse</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">745</span>     <span class="k">def</span> <span class="nf">inverse</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">746</span>         <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">__inverse</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">747</span>             <span class="bp">self</span><span class="o">.</span><span class="n">__inverse</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_inverse</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">748</span>         <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">__inverse</span>
<span class="g g-Whitespace">    </span><span class="mi">749</span> 

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/location.py</span> in <span class="ni">_inverse</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">770</span>                 <span class="n">new_spans</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">LostSpan</span><span class="p">(</span><span class="n">lo</span> <span class="o">-</span> <span class="n">last_hi</span><span class="p">))</span>
<span class="g g-Whitespace">    </span><span class="mi">771</span>             <span class="k">elif</span> <span class="n">lo</span> <span class="o">&lt;</span> <span class="n">last_hi</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">772</span>                 <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Uninvertable. Overlap: </span><span class="si">%s</span><span class="s2"> &lt; </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">lo</span><span class="p">,</span> <span class="n">last_hi</span><span class="p">))</span>
<span class="g g-Whitespace">    </span><span class="mi">773</span>             <span class="n">new_spans</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Span</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="n">start</span> <span class="o">&gt;</span> <span class="n">end</span><span class="p">))</span>
<span class="g g-Whitespace">    </span><span class="mi">774</span>             <span class="n">last_hi</span> <span class="o">=</span> <span class="n">hi</span>

<span class="ne">ValueError</span>: Uninvertable. Overlap: 9 &lt; 10
</pre></div>
</div>
</div>
</div>
<p>But <code class="docutils literal notranslate"><span class="pre">get_region_covering_all</span></code> resolves this, ensuring no overlaps.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">([</span><span class="n">exon3</span><span class="p">,</span> <span class="n">exon3</span><span class="p">])</span><span class="o">.</span><span class="n">get_slice</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CCC
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="you-can-slice-an-annotation-itself">
<h2>You can slice an annotation itself<a class="headerlink" href="#you-can-slice-an-annotation-itself" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">exon2</span><span class="p">])</span>
<span class="n">ex2_start</span> <span class="o">=</span> <span class="n">exon2</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">ex2_start</span><span class="p">])</span>
<span class="n">ex2_end</span> <span class="o">=</span> <span class="n">exon2</span><span class="p">[</span><span class="o">-</span><span class="mi">3</span><span class="p">:]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s1</span><span class="p">[</span><span class="n">ex2_end</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>TTTTTAAAAA
TTT
AAA
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="sequence-vs-alignment-slicing">
<h2>Sequence vs Alignment slicing<a class="headerlink" href="#sequence-vs-alignment-slicing" title="Permalink to this headline">¶</a></h2>
<p>You can’t slice an alignment using an annotation from a sequence.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln1</span><span class="p">[</span><span class="n">seq_exon</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="o">&lt;</span><span class="n">ipython</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">22</span><span class="o">-</span><span class="n">bbe05c8843eb</span><span class="o">&gt;</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="n">aln1</span><span class="p">[</span><span class="n">seq_exon</span><span class="p">]</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">__getitem__</span><span class="nt">(self, index)</span>
<span class="g g-Whitespace">     </span><span class="mi">87</span> 
<span class="g g-Whitespace">     </span><span class="mi">88</span>     <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="ne">---&gt; </span><span class="mi">89</span>         <span class="nb">map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_as_map</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">90</span>         <span class="n">new</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped</span><span class="p">(</span><span class="nb">map</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">91</span>         <span class="n">sliced_annots</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sliced_annotations</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="nb">map</span><span class="p">)</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">_as_map</span><span class="nt">(self, index)</span>
<span class="g g-Whitespace">     </span><span class="mi">76</span>                 <span class="n">base</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">parent</span>
<span class="g g-Whitespace">     </span><span class="mi">77</span>             <span class="k">if</span> <span class="n">base</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
<span class="ne">---&gt; </span><span class="mi">78</span>                 <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">     </span><span class="mi">79</span>                     <span class="s2">&quot;Can&#39;t map </span><span class="si">%s</span><span class="s2"> onto </span><span class="si">%s</span><span class="s2"> via </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">repr</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">containers</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">80</span>                 <span class="p">)</span>

<span class="ne">ValueError</span>: Can&#39;t map exon &quot;A&quot; at [3:8]/9 onto 2 x 10 text alignment: x[-AAACCCCCA], y[TTTT--TTTT] via []
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="copying-annotations">
<h2>Copying annotations<a class="headerlink" href="#copying-annotations" title="Permalink to this headline">¶</a></h2>
<p>You can copy annotations onto sequences with the same name, even if the length differs</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln2</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAAAAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCCCCCCCCCCCCCCCC&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
<span class="n">match_exon</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">)])</span>
<span class="n">aln2</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">copy_annotations</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">copied</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln2</span><span class="o">.</span><span class="n">get_annotations_from_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">))</span>
<span class="n">copied</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[exon &quot;A&quot; at [4:9]/10]
</pre></div>
</div>
</div>
</div>
<p>but if the feature lies outside the sequence being copied to, you get a lost span</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln2</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCCCCCCCCCCCCCCCC&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
<span class="n">match_exon</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">)])</span>
<span class="n">aln2</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">copy_annotations</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">copied</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln2</span><span class="o">.</span><span class="n">get_annotations_from_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">))</span>
<span class="n">copied</span>
<span class="n">copied</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .a_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .b_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .c_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .d_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .e_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .f_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .g_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .h_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .i_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .j_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .k_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .l_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .m_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .n_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .o_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .p_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .q_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .r_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .s_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .t_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .u_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .v_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .w_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .x_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .y_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .z_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .B_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .C_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .D_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .E_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .F_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .G_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .H_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .I_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .J_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .K_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .L_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .M_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .N_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .O_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .P_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Q_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .R_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .S_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .T_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .U_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .V_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .W_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .X_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Y_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .Z_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .terminal_ambig_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .ambig_text{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">y</td><td><span class="terminal_ambig_text">-</span><span class="terminal_ambig_text">-</span><span class="terminal_ambig_text">-</span><span class="terminal_ambig_text">-</span></td></tr>
<tr><td class="label">x</td><td><span class="terminal_ambig_text">.</span><span class="terminal_ambig_text">.</span><span class="terminal_ambig_text">.</span><span class="terminal_ambig_text">.</span></td></tr>
</table>
<p><i>2 x 4 text alignment</i></p>
</div></div></div>
</div>
<p>You can copy to a sequence with a different name, in a different alignment if the feature lies within the length</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># new test</span>
<span class="n">aln2</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAAAAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCCCCCCCCCCCCCCCC&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
<span class="n">match_exon</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">8</span><span class="p">)])</span>
<span class="n">aln2</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">copy_annotations</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="n">copied</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln2</span><span class="o">.</span><span class="n">get_annotations_from_seq</span><span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">))</span>
<span class="n">copied</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[exon &quot;A&quot; at [7:10]/10]
</pre></div>
</div>
</div>
</div>
<p>If the sequence is shorter, again you get a lost span.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln2</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAAAAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">diff_len_seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCCCCCCCCCCCCCCCCCCCCCCCC&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
<span class="n">nonmatch</span> <span class="o">=</span> <span class="n">diff_len_seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;repeat&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">14</span><span class="p">)])</span>
<span class="n">aln2</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">copy_annotations</span><span class="p">(</span><span class="n">diff_len_seq</span><span class="p">)</span>
<span class="n">copied</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln2</span><span class="o">.</span><span class="n">get_annotations_from_seq</span><span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;repeat&quot;</span><span class="p">))</span>
<span class="n">copied</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[repeat &quot;A&quot; at [10:10, -6-]/10]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="querying">
<h2>Querying<a class="headerlink" href="#querying" title="Permalink to this headline">¶</a></h2>
<p>You need to get a corresponding annotation projected into alignment coordinates via a query.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln_exon</span> <span class="o">=</span> <span class="n">aln1</span><span class="o">.</span><span class="n">get_annotations_from_any_seq</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln1</span><span class="p">[</span><span class="n">aln_exon</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;x
CCCCC
&gt;y
--TTT

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="querying-produces-objects-only-valid-for-their-source">
<h2>Querying produces objects only valid for their source<a class="headerlink" href="#querying-produces-objects-only-valid-for-their-source" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cpgsite2</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s2</span><span class="p">[</span><span class="n">cpgsite2</span><span class="p">])</span>
<span class="n">cpgsite3</span> <span class="o">=</span> <span class="n">s3</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">)</span>
<span class="n">s2</span><span class="p">[</span><span class="n">cpgsite3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CGCG
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">ValueError</span><span class="g g-Whitespace">                                </span>Traceback (most recent call last)
<span class="o">&lt;</span><span class="n">ipython</span><span class="o">-</span><span class="nb">input</span><span class="o">-</span><span class="mi">28</span><span class="o">-</span><span class="mi">91</span><span class="n">f7e6a5ce69</span><span class="o">&gt;</span> <span class="ow">in</span> <span class="o">&lt;</span><span class="n">module</span><span class="o">&gt;</span>
<span class="g g-Whitespace">      </span><span class="mi">2</span> <span class="nb">print</span><span class="p">(</span><span class="n">s2</span><span class="p">[</span><span class="n">cpgsite2</span><span class="p">])</span>
<span class="g g-Whitespace">      </span><span class="mi">3</span> <span class="n">cpgsite3</span> <span class="o">=</span> <span class="n">s3</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">)</span>
<span class="ne">----&gt; </span><span class="mi">4</span> <span class="n">s2</span><span class="p">[</span><span class="n">cpgsite3</span><span class="p">]</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">__getitem__</span><span class="nt">(self, index)</span>
<span class="g g-Whitespace">     </span><span class="mi">87</span> 
<span class="g g-Whitespace">     </span><span class="mi">88</span>     <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
<span class="ne">---&gt; </span><span class="mi">89</span>         <span class="nb">map</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_as_map</span><span class="p">(</span><span class="n">index</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">90</span>         <span class="n">new</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_mapped</span><span class="p">(</span><span class="nb">map</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">91</span>         <span class="n">sliced_annots</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sliced_annotations</span><span class="p">(</span><span class="n">new</span><span class="p">,</span> <span class="nb">map</span><span class="p">)</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">_as_map</span><span class="nt">(self, index)</span>
<span class="g g-Whitespace">     </span><span class="mi">65</span>             <span class="n">spans</span> <span class="o">=</span> <span class="p">[]</span>
<span class="g g-Whitespace">     </span><span class="mi">66</span>             <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">index</span><span class="p">:</span>
<span class="ne">---&gt; </span><span class="mi">67</span>                 <span class="n">spans</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_as_map</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">spans</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">68</span>             <span class="nb">map</span> <span class="o">=</span> <span class="n">Map</span><span class="p">(</span><span class="n">spans</span><span class="o">=</span><span class="n">spans</span><span class="p">,</span> <span class="n">parent_length</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">))</span>
<span class="g g-Whitespace">     </span><span class="mi">69</span>         <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="n">_Feature</span><span class="p">):</span>

<span class="nn">/usr/local/miniconda/envs/cogent3/lib/python3.8/site-packages/cogent3/core/annotation.py</span> in <span class="ni">_as_map</span><span class="nt">(self, index)</span>
<span class="g g-Whitespace">     </span><span class="mi">76</span>                 <span class="n">base</span> <span class="o">=</span> <span class="n">base</span><span class="o">.</span><span class="n">parent</span>
<span class="g g-Whitespace">     </span><span class="mi">77</span>             <span class="k">if</span> <span class="n">base</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">self</span><span class="p">:</span>
<span class="ne">---&gt; </span><span class="mi">78</span>                 <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
<span class="g g-Whitespace">     </span><span class="mi">79</span>                     <span class="s2">&quot;Can&#39;t map </span><span class="si">%s</span><span class="s2"> onto </span><span class="si">%s</span><span class="s2"> via </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">repr</span><span class="p">(</span><span class="bp">self</span><span class="p">),</span> <span class="n">containers</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">80</span>                 <span class="p">)</span>

<span class="ne">ValueError</span>: Can&#39;t map cpgsite &quot;cpg&quot; at [0:2]/10 onto DnaSequence(CGAAACGTTT) via []
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="querying-for-absent-annotation">
<h2>Querying for absent annotation<a class="headerlink" href="#querying-for-absent-annotation" title="Permalink to this headline">¶</a></h2>
<p>You get back an empty list, and slicing with this returns an empty sequence.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># this test is new</span>
<span class="n">dont_exist</span> <span class="o">=</span> <span class="n">s2</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;dont_exist&quot;</span><span class="p">)</span>
<span class="n">dont_exist</span>
<span class="n">s2</span><span class="p">[</span><span class="n">dont_exist</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
</table>
<p><i>0 DnaSequence</i></p>
</div></div></div>
</div>
</div>
<div class="section" id="querying-features-that-span-gaps-in-alignments">
<h2>Querying features that span gaps in alignments<a class="headerlink" href="#querying-features-that-span-gaps-in-alignments" title="Permalink to this headline">¶</a></h2>
<p>If you query for a feature from a sequence, it’s alignment coordinates may be discontinuous.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aln3</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;C-CCCAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;-T----TTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">exon</span> <span class="o">=</span> <span class="n">aln3</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;ex1&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">)])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">exon</span><span class="o">.</span><span class="n">get_slice</span><span class="p">())</span>
<span class="n">aln_exons</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln3</span><span class="o">.</span><span class="n">get_annotations_from_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln_exons</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln3</span><span class="p">[</span><span class="n">aln_exons</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CCCC
[exon &quot;ex1&quot; at [0:1, 2:5]/10]
&gt;x
CCCC
&gt;y
----

</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">T</span></code> opposite the gap is missing since this approach only returns positions directly corresponding to the feature.</p>
</div>
</div>
<div class="section" id="as-one-span-unifies-features-with-discontinuous-alignment-coordinates">
<h2><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates<a class="headerlink" href="#as-one-span-unifies-features-with-discontinuous-alignment-coordinates" title="Permalink to this headline">¶</a></h2>
<p>To get positions spanned by a feature, including gaps, use <code class="docutils literal notranslate"><span class="pre">as_one_span</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">unified</span> <span class="o">=</span> <span class="n">aln_exons</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">as_one_span</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln3</span><span class="p">[</span><span class="n">unified</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;x
C-CCC
&gt;y
-T---

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="behaviour-of-annotations-on-nucleic-acid-sequences">
<h2>Behaviour of annotations on nucleic acid sequences<a class="headerlink" href="#behaviour-of-annotations-on-nucleic-acid-sequences" title="Permalink to this headline">¶</a></h2>
<p>Reverse complementing a sequence <strong>does not</strong> reverse annotations, that is they retain the reference to the frame for which they were defined.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plus</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCAAAAAAAAAATTTTTTTTTTAAAGG&quot;</span><span class="p">)</span>
<span class="n">plus_rpt</span> <span class="o">=</span> <span class="n">plus</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;blah&quot;</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span> <span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="mi">28</span><span class="p">)])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">plus</span><span class="p">[</span><span class="n">plus_rpt</span><span class="p">])</span>
<span class="n">minus</span> <span class="o">=</span> <span class="n">plus</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">minus</span><span class="p">)</span>
<span class="n">minus_rpt</span> <span class="o">=</span> <span class="n">minus</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;blah&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">minus</span><span class="p">[</span><span class="n">minus_rpt</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AAAAAAAAAAAAA
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>
CCTTTAAAAAAAAAATTTTTTTTTTGGGGG
AAAAAAAAAAAAA
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="masking-annotated-regions">
<h2>Masking annotated regions<a class="headerlink" href="#masking-annotated-regions" title="Permalink to this headline">¶</a></h2>
<p>We mask the CDS regions.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.parse.genbank</span> <span class="kn">import</span> <span class="n">RichGenbankParser</span>

<span class="n">parser</span> <span class="o">=</span> <span class="n">RichGenbankParser</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data/ST_genome_part.gb&quot;</span><span class="p">))</span>
<span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span> <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">parser</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">no_cds</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">no_cds</span><span class="p">[</span><span class="mi">150</span><span class="p">:</span><span class="mi">400</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CAAGACAGACAAATAAAAATGACAGAGTACACAACATCC??????????????????????????????????????????????????????????????????CGCGTACAGGAAACACAGAAAAAAGCCCGCACCTGAACAGTGCGGGCTTTTTTTTCGACCAGAGATCACGAGGTAACAACC????????????????????????????????????????????????????????????????
</pre></div>
</div>
</div>
</div>
<p>The above sequence could then have positions filtered so no position with the ambiguous character ‘?’ was present.</p>
</div>
<div class="section" id="masking-annotated-regions-on-alignments">
<h2>Masking annotated regions on alignments<a class="headerlink" href="#masking-annotated-regions-on-alignments" title="Permalink to this headline">¶</a></h2>
<p>We mask exon’s on an alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;C-CCCAAAAAGGGAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;-T----TTTTG-GTT&quot;</span><span class="p">]],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
    <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">exon</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;norwegian&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">)])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">mask_char</span><span class="o">=</span><span class="s2">&quot;?&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;x
?-???AAAAAGGGAA
&gt;y
-T----TTTTG-GTT

</pre></div>
</div>
</div>
</div>
<p>These also persist through reverse complement operations.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rc</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rc</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">mask_char</span><span class="o">=</span><span class="s2">&quot;?&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;x
TTCCCTTTTTGGG-G
&gt;y
AAC-CAAAA----A-

&gt;x
TTCCCTTTTT???-?
&gt;y
AAC-CAAAA----A-

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="you-can-take-mask-of-the-shadow">
<h2>You can take mask of the shadow<a class="headerlink" href="#you-can-take-mask-of-the-shadow" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCAAAAAGGGAA&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">)</span>
<span class="n">exon</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;norwegian&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">)])</span>
<span class="n">rpt</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;repeat&quot;</span><span class="p">,</span> <span class="s2">&quot;norwegian&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">12</span><span class="p">)])</span>
<span class="n">rc</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">([</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;repeat&quot;</span><span class="p">],</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">([</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;repeat&quot;</span><span class="p">],</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>CCCC??????????
??????????GGGG
CCCC?????GGG??
??CCC?????GGGG
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="what-features-of-a-certain-type-are-available">
<h2>What features of a certain type are available?<a class="headerlink" href="#what-features-of-a-certain-type-are-available" title="Permalink to this headline">¶</a></h2>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">s</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;ATGACCCTGTAAAAAATGTGTTAACCC&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;a&quot;</span><span class="p">)</span>
<span class="n">cds1</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;cds&quot;</span><span class="p">,</span> <span class="s2">&quot;cds1&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">12</span><span class="p">)])</span>
<span class="n">cds2</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;cds&quot;</span><span class="p">,</span> <span class="s2">&quot;cds2&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">24</span><span class="p">)])</span>
<span class="n">all_cds</span> <span class="o">=</span> <span class="n">s</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;cds&quot;</span><span class="p">)</span>
<span class="n">all_cds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[cds &quot;cds1&quot; at [0:12]/27, cds &quot;cds2&quot; at [15:24]/27]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="getting-all-features-of-a-type-or-everything-but-that-type">
<h2>Getting all features of a type, or everything but that type<a class="headerlink" href="#getting-all-features-of-a-type-or-everything-but-that-type" title="Permalink to this headline">¶</a></h2>
<p>The annotation methods <code class="docutils literal notranslate"><span class="pre">get_region_covering_all</span></code> and <code class="docutils literal notranslate"><span class="pre">get_shadow</span></code> can be used to grab all the coding sequences or non-coding sequences in a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3.parse.genbank</span> <span class="kn">import</span> <span class="n">RichGenbankParser</span>

<span class="n">parser</span> <span class="o">=</span> <span class="n">RichGenbankParser</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s2">&quot;data/ST_genome_part.gb&quot;</span><span class="p">))</span>
<span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="n">seq</span> <span class="k">for</span> <span class="n">accession</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">parser</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="n">all_cds</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get_annotations_matching</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="n">coding_seqs</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">get_region_covering_all</span><span class="p">(</span><span class="n">all_cds</span><span class="p">)</span>
<span class="n">coding_seqs</span>
<span class="n">coding_seqs</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
<span class="n">noncoding_seqs</span> <span class="o">=</span> <span class="n">coding_seqs</span><span class="o">.</span><span class="n">get_shadow</span><span class="p">()</span>
<span class="n">noncoding_seqs</span>
<span class="n">noncoding_seqs</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">AE006468</td><td><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="T_dna">T</span></td></tr>
</table>
<p><i>957 (truncated to 60) DnaSequence</i></p>
</div></div></div>
</div>
</div>
<div class="section" id="getting-sequence-features-when-you-have-an-alignment-object">
<h2>Getting sequence features when you have an alignment object<a class="headerlink" href="#getting-sequence-features-when-you-have-an-alignment-object" title="Permalink to this headline">¶</a></h2>
<p>Sequence features can be accessed via a containing <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">(</span>
    <span class="n">data</span><span class="o">=</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAAAAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span> <span class="n">array_align</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">aln</span><span class="p">)</span>
<span class="n">exon</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">)])</span>
<span class="n">aln_exons</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_annotations_from_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">)</span>
<span class="n">aln_exons</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">get_annotations_from_any_seq</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">)</span>
<span class="n">aln_exons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>&gt;x
-AAAAAAAAA
&gt;y
TTTT--TTTT

</pre></div>
</div>
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[exon &quot;1&quot; at [4:9]/10]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="annotation-display-on-sequences">
<h2>Annotation display on sequences<a class="headerlink" href="#annotation-display-on-sequences" title="Permalink to this headline">¶</a></h2>
<p>We can display annotations on sequences, writing to file.</p>
<p>We first make a sequence and add some annotations.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cogent3</span> <span class="kn">import</span> <span class="n">DNA</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">DNA</span><span class="o">.</span><span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;aaaccggttt&quot;</span> <span class="o">*</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">v</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">35</span><span class="p">)])</span>
<span class="n">v</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;repeat_unit&quot;</span><span class="p">,</span> <span class="s2">&quot;repeat_unit&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">39</span><span class="p">,</span> <span class="mi">49</span><span class="p">)])</span>
<span class="n">v</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="s2">&quot;repeat_unit&quot;</span><span class="p">,</span> <span class="s2">&quot;rep2&quot;</span><span class="p">,</span> <span class="p">[(</span><span class="mi">49</span><span class="p">,</span> <span class="mi">60</span><span class="p">)])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
</div>
</div>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 3.2.1 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>