
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Features &#8212; cogent3 2026.01.12 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../_static/custom.css?v=da3cec82" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=4969098b"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/thebelab-helper.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'doc/cookbook/features';</script>
    <link rel="icon" href="../../_static/c3-square.svg"/>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Annotation Databases" href="annotation_db.html" />
    <link rel="prev" title="Sequence Collections and Alignments" href="alignments.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/c3-logo.png" class="logo__image only-light" alt="cogent3 2026.01.12 documentation - Home"/>
    <img src="../../_static/c3-logo.png" class="logo__image only-dark pst-js-only" alt="cogent3 2026.01.12 documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../draw/index.html">
    Image Gallery
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    User Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../community.html">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../licenses.html">
    License
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../history.html">
    Project History
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../pycogent.html">
    How does cogent3 relate to PyCogent?
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../projects.html">
    Projects using cogent3
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../data_file_links.html">
    Data Files Used in the Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../../genindex.html">
    Index
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../draw/index.html">
    Image Gallery
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../install.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../index.html">
    User Guides
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../community.html">
    Community
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../licenses.html">
    License
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../history.html">
    Project History
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../pycogent.html">
    How does cogent3 relate to PyCogent?
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../projects.html">
    Projects using cogent3
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../data_file_links.html">
    Data Files Used in the Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../../genindex.html">
    Index
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../app/index.html">apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-app.html">Overview of apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/app-overview.html">Overview of using apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/app-help.html">Getting help on an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/app-get.html">Getting an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/dstore.html">Data stores – collections of data records</a></li>

<li class="toctree-l3"><a class="reference internal" href="../app/app_overview/not-completed.html">Tracking records that could not be processed</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-cookbook.html">App Cookbook</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-loaders.html">Loading data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-aligned.html">Loading aligned sequence data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-unaligned.html">Loading unaligned sequence data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-tabular.html">Loading tabular data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/load-json.html">Loading JSON serialised objects</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-align.html">Sequence Alignment</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/align-nucleotide.html">Using a nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/align-codon.html">Using a codon model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/align-protein.html">Using a protein model</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-sample.html">Filtering data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-degen.html">Removing degenerate characters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-gap.html">Removing highly gapped positions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-bad.html">Remove problem sequences from an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-omit-duplicated.html">Remove duplicated sequences from an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-min-len.html">Filter sequence collections and alignments by length</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-fixed-len.html">Sample an alignment to a fixed length</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-trim-stops.html">Remove trailing stop codons from sequences in an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-take-pos.html">Sample nucleotides from a given codon position</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-take-seqs.html">Select named sequences from a collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-take-n-seqs.html">Select <cite>n</cite> sequences from a collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/sample-concat.html">Concatenating alignments</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_cookbook/index-writers.html">Writing Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-seqs.html">Writing sequences and sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-tabular.html">Writing tabular data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-json.html">Writing JSON Serialised Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_cookbook/write-db.html">Writing a database to file</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-tutorials.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_tutorials/index-evo-model.html">Modelling sequence evolution</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-ancestral-states.html">Reconstructing ancestral states</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-tr-codon-model.html">Applying a time-reversible codon model</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_tutorials/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a></li>
</ul>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../app/app_tutorials/index-evo-natsel.html">Testing for natural selection</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app/app_tutorials/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a></li>
</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../app/index-custom.html">Custom Apps</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../app/user_function.html">Why write apps?</a></li>

</ul>
</details></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="index.html">Cookbook</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active has-children"><a class="reference internal" href="manipulating_biological_data.html">Manipulating biological data</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="loading_sequences.html">Loading a single sequence from a file</a></li>








<li class="toctree-l3"><a class="reference internal" href="DNA_and_RNA_sequences.html">Sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="protein_sequences.html">Protein sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="alignments.html">Sequence Collections and Alignments</a></li>
<li class="toctree-l3 current active"><a class="current reference internal" href="#">Features</a></li>
<li class="toctree-l3"><a class="reference internal" href="annotation_db.html">Annotation Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="moltypes.html">Molecular types</a></li>
<li class="toctree-l3"><a class="reference internal" href="alphabet.html">Alphabets</a></li>
<li class="toctree-l3"><a class="reference internal" href="what_codes.html">Available genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="genetic_code.html">Using genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="simple_trees.html">Trees</a></li>
</ul>
</details></li>
<li class="toctree-l2"><a class="reference internal" href="tables.html">Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="building_alignments.html">Building alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="calc_genetic_distance.html">Genetic distance calculation</a></li>


<li class="toctree-l2"><a class="reference internal" href="building_phylogenies.html">Building phylogenies</a></li>
<li class="toctree-l2"><a class="reference internal" href="evo_modelling.html">Evolutionary Analysis Using Likelihood</a></li>
<li class="toctree-l2"><a class="reference internal" href="useful_utilities.html">Useful Utilities</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../examples/index.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/period_estimation.html">Estimating periodic signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/simple.html">The simplest script</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/codon_models.html">Using codon models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/nonstationary_inference.html">Inferring Amino Acid frequencies with a nonstationary model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a></li>
<li class="toctree-l2"><a class="reference internal" href="../examples/parallel.html">Parallel computations</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../api/index.html">API</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.open_.html">open</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.get_model.html">get_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Aligned.html">Aligned</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/sequence/sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/sequence/classes/cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/seq_storage/seq_storage.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">seq_storage</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/seq_storage/classes/cogent3.core.seq_storage.SeqsData.html">SeqsData</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/seq_storage/classes/cogent3.core.seq_storage.AlignedSeqsData.html">AlignedSeqsData</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/seqview/seqview.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">seqview</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/seqview/classes/cogent3.core.seqview.SeqView.html">SeqView</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/seqview/classes/cogent3.core.seqview.SeqDataView.html">SeqDataView</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/seqview/classes/cogent3.core.seqview.AlignedDataView.html">AlignedDataView</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/slice_record/slice_record.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">slice_record</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/slice_record/classes/cogent3.core.slice_record.SliceRecord.html">SliceRecord</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.CharAlphabet.html">CharAlphabet</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.KmerAlphabet.html">KmerAlphabet</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/alphabet/classes/cogent3.core.alphabet.SenseCodonAlphabet.html">SenseCodonAlphabet</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/table/classes/cogent3.core.table.Table.html">Table</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/table/classes/cogent3.core.table.Columns.html">Columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api/table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../api/tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../api/tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</details></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/composable/define_app.html">define_app</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api/util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">User Guides</a></li>
    
    
    <li class="breadcrumb-item"><a href="index.html" class="nav-link">Cookbook</a></li>
    
    
    <li class="breadcrumb-item"><a href="manipulating_biological_data.html" class="nav-link">Manipulating biological data</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Features</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<section id="features">
<span id="howto-features"></span><h1>Features<a class="headerlink" href="#features" title="Link to this heading">#</a></h1>
<p>This guide provides instructions on creating, querying, and utilising features to manipulate biological sequence data.</p>
<section id="how-to-create-a-custom-feature">
<h2>How to create a custom <code class="docutils literal notranslate"><span class="pre">Feature</span></code><a class="headerlink" href="#how-to-create-a-custom-feature" title="Link to this heading">#</a></h2>
<section id="via-add-feature">
<h3>Via <code class="docutils literal notranslate"><span class="pre">add_feature</span></code><a class="headerlink" href="#via-add-feature" title="Link to this heading">#</a></h3>
<p>We can dynamically create a <code class="docutils literal notranslate"><span class="pre">Feature</span></code> via the <code class="docutils literal notranslate"><span class="pre">add_feature()</span></code> method, providing the key information about the feature.</p>
<p>The new feature will be added to the <code class="docutils literal notranslate"><span class="pre">annotation_db</span></code> attribute of the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> and/or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>. A <code class="docutils literal notranslate"><span class="pre">Feature</span></code> instance will be returned.</p>
<section id="on-a-sequence">
<h4>On a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code><a class="headerlink" href="#on-a-sequence" title="Link to this heading">#</a></h4>
<p>We use <code class="docutils literal notranslate"><span class="pre">add_feature</span></code> to add a <code class="docutils literal notranslate"><span class="pre">Feature</span></code> to a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code>, providing the biotype, name/id, and a list of start and stop indices.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_seq</span>

<span class="c1"># create an example sequence</span>
<span class="n">seq</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;GCCTAATACTAAGCCTAAGCCTAAGACTAAGCCTAATACTAAGCCTAAGCCTAAGACTAA&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Add a feature</span>
<span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;a_gene&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">48</span><span class="p">)],</span> <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&#39;seq1&#39;, biotype=&#39;gene&#39;, name=&#39;a_gene&#39;, map=[12:48]/60, parent=DnaSequence(GCCTAAT... 60))
</pre></div>
</div>
</div>
</div>
<p>A feature can also be added as a series of non-overlapping genomic segments:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Add a feature as a series</span>
<span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;cpgsite&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;a_cpgsite&quot;</span><span class="p">,</span>
    <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">18</span><span class="p">),</span> <span class="p">(</span><span class="mi">21</span><span class="p">,</span> <span class="mi">29</span><span class="p">),</span> <span class="p">(</span><span class="mi">35</span><span class="p">,</span> <span class="mi">48</span><span class="p">)],</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&#39;seq1&#39;, biotype=&#39;cpgsite&#39;, name=&#39;a_cpgsite&#39;, map=[12:18, 21:29, 35:48]/60, parent=DnaSequence(GCCTAAT... 60))
</pre></div>
</div>
</div>
</div>
</section>
<section id="on-a-sequence-within-an-alignment">
<h4>On a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> within an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#on-a-sequence-within-an-alignment" title="Link to this heading">#</a></h4>
<p>We use <code class="docutils literal notranslate"><span class="pre">add_feature</span></code> and provide a value for <code class="docutils literal notranslate"><span class="pre">seqid</span></code> to add a feature to a sequence within an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> or <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code>.</p>
<p>Note the difference between the value provided to <code class="docutils literal notranslate"><span class="pre">spans</span></code>, and the value of <code class="docutils literal notranslate"><span class="pre">map</span></code> in the returned <code class="docutils literal notranslate"><span class="pre">Feature</span></code>… the resulting feature is in <strong>alignment coordinates</strong>!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="c1"># make demo alignment</span>
<span class="n">aln1</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">([[</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAACCCCCA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># add feature to seq1</span>
<span class="n">aln1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">)],</span> <span class="n">on_alignment</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&quot;from &#39;seq1&#39;&quot;, biotype=&#39;exon&#39;, name=&#39;A&#39;, map=[4:9]/10, parent=2 x 10 dna alignment: seq1[-AAACCCCCA], seq2[TTTT--TTTT])
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Feature</span></code> specifies that <code class="docutils literal notranslate"><span class="pre">seqid=&quot;from</span> <span class="pre">'seq1'&quot;</span></code>, indicating that it “belongs” to seq1.</p>
</section>
<section id="on-an-alignment">
<h4>On an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#on-an-alignment" title="Link to this heading">#</a></h4>
<p>We use <code class="docutils literal notranslate"><span class="pre">add_feature</span></code> to add a <code class="docutils literal notranslate"><span class="pre">Feature</span></code> to an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="c1"># make demo alignment</span>
<span class="n">aln1</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">([[</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAACCCCCA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;seq2&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">aln1</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;aligned_exon&quot;</span><span class="p">,</span>
    <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">8</span><span class="p">)],</span>
    <span class="n">on_alignment</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=None, biotype=&#39;exon&#39;, name=&#39;aligned_exon&#39;, map=[0:8]/10, parent=2 x 10 dna alignment: seq1[-AAACCCCCA], seq2[TTTT--TTTT])
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Feature</span></code> specifies that <code class="docutils literal notranslate"><span class="pre">seqid=None</span></code>, indicating that it belongs to the alignment</p>
</section>
<section id="via-an-annotationdb">
<h4>Via an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code><a class="headerlink" href="#via-an-annotationdb" title="Link to this heading">#</a></h4>
<p>We can use <code class="docutils literal notranslate"><span class="pre">add_feature</span></code> to add a feature directly into an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>, and assign it to the <code class="docutils literal notranslate"><span class="pre">annotation_db</span></code> attribute of a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>. For extensive documentation on handling features directly via an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> see <a class="reference internal" href="annotation_db.html#anno-db"><span class="std std-ref">Annotation Databases</span></a>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_seq</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">cogent3.core.annotation_db</span><span class="w"> </span><span class="kn">import</span> <span class="n">BasicAnnotationDb</span>

<span class="c1"># init empty db and add feature</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">BasicAnnotationDb</span><span class="p">()</span>
<span class="n">db</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">45</span><span class="p">,</span> <span class="mi">48</span><span class="p">)])</span>

<span class="c1"># make demo seq</span>
<span class="n">s1</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span>
    <span class="s2">&quot;AAGAAGAAGACCCCCAAAAAAAAAATTTTTTTTTTAAAAAGGGAACCCT&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;seq1&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span>
<span class="p">)</span>

<span class="c1"># assign db to sequence</span>
<span class="n">s1</span><span class="o">.</span><span class="n">annotation_db</span> <span class="o">=</span> <span class="n">db</span>
<span class="n">s1</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=1)
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="how-to-load-bulk-features-from-a-file">
<h2>How to load bulk Features from a File<a class="headerlink" href="#how-to-load-bulk-features-from-a-file" title="Link to this heading">#</a></h2>
<p>Typically, we want to load bulk features from a genomic annotation file, such as a GFF or Genbank file. For the following examples, we will use <em>Caenorhabditis elegans</em> chromosome I.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See the list of <a class="reference internal" href="../data_file_links.html#data-links"><span class="std std-ref">Data Files Used in the Documentation</span></a> to download the data used in the following examples.</p>
</div>
<p>To load features from a genomic annotation file along with the corresponding sequence, we can use the <code class="docutils literal notranslate"><span class="pre">load_seq</span></code> function. The features are stored in a <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> and assigned to the <code class="docutils literal notranslate"><span class="pre">annotation_db</span></code> attribute of the sequence.</p>
<section id="from-a-genbank-file">
<span id="genbank-features"></span><h3>From a Genbank file<a class="headerlink" href="#from-a-genbank-file" title="Link to this heading">#</a></h3>
<section id="directly-using-the-genbank-parser-to-get-features-as-python-primitives">
<h4>Directly using the genbank parser to get features as Python primitives<a class="headerlink" href="#directly-using-the-genbank-parser-to-get-features-as-python-primitives" title="Link to this heading">#</a></h4>
<p>This approach returns features as Python primitives, except for the coordinates for features, which is a <code class="docutils literal notranslate"><span class="pre">Location</span></code> object.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3.parse.genbank</span><span class="w"> </span><span class="kn">import</span> <span class="n">iter_genbank_records</span>

<span class="n">label</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">anns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">iter_genbank_records</span><span class="p">(</span><span class="s2">&quot;data/mycoplasma-genitalium.gb&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">anns</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;locus&#39;, &#39;length&#39;, &#39;mol_type&#39;, &#39;topology&#39;, &#39;db&#39;, &#39;date&#39;, &#39;definition&#39;, &#39;accession&#39;, &#39;version&#39;, &#39;dblink&#39;, &#39;keywords&#39;, &#39;source&#39;, &#39;species&#39;, &#39;taxonomy&#39;, &#39;references&#39;, &#39;comment&#39;, &#39;features&#39;, &#39;contig&#39;])
</pre></div>
</div>
</div>
</div>
<p>As the output indicates, variable <code class="docutils literal notranslate"><span class="pre">anns</span></code> is a dictionary. The features in the GenBank feature table are available as a list under the <code class="docutils literal notranslate"><span class="pre">&quot;features&quot;</span></code> key. Listing just the first two.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">features</span> <span class="o">=</span> <span class="n">anns</span><span class="p">[</span><span class="s2">&quot;features&quot;</span><span class="p">]</span>
<span class="n">features</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[{&#39;type&#39;: &#39;source&#39;,
  &#39;raw_location&#39;: [&#39;1..580076&#39;],
  &#39;location&#39;: LocationList(1..580076),
  &#39;organism&#39;: [&#39;Mycoplasmoides genitalium G37&#39;],
  &#39;mol_type&#39;: [&#39;genomic DNA&#39;],
  &#39;strain&#39;: [&#39;G-37&#39;],
  &#39;type_material&#39;: [&#39;type strain of Mycoplasma genitalium&#39;],
  &#39;db_xref&#39;: [&#39;ATCC:33530&#39;, &#39;taxon:243273&#39;],
  &#39;note&#39;: [&#39;strain coidentity: G37 = ATCC 33530&#39;]},
 {&#39;type&#39;: &#39;gene&#39;,
  &#39;raw_location&#39;: [&#39;686..1828&#39;],
  &#39;location&#39;: LocationList(686..1828),
  &#39;gene&#39;: [&#39;dnaN&#39;],
  &#39;locus_tag&#39;: [&#39;MG_RS00005&#39;],
  &#39;old_locus_tag&#39;: [&#39;MG_001&#39;]}]
</pre></div>
</div>
</div>
</div>
<p>Illustrating how to get specific coordinate data (which are converted from the 1-based indices of GenBank to 0-based Python indices).</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="n">features</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">gene_loc</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span>
<span class="n">gene_loc</span><span class="o">.</span><span class="n">get_coordinates</span><span class="p">(),</span> <span class="n">gene_loc</span><span class="o">.</span><span class="n">strand</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>([(685, 1828)], 1)
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-load-features-and-sequence-data">
<h4>How to load features and sequence data<a class="headerlink" href="#how-to-load-features-and-sequence-data" title="Link to this heading">#</a></h4>
<p>To load the sequence and all 40,578 features from <em>C. elegans</em> Chromosome 1, we use the <code class="docutils literal notranslate"><span class="pre">load_seq</span></code> function ⚡️</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="o">%</span><span class="k">timeit</span> load_seq(&quot;data/C-elegans-chromosome-I.gb&quot;, moltype=&quot;dna&quot;)
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>2.84 s ± 7.12 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
</pre></div>
</div>
</div>
</div>
<div class="jupyter_cell docutils container">
<div class="cell_output docutils container">
</div>
</div>
<p>The features are stored in the <code class="docutils literal notranslate"><span class="pre">annotation_db</span></code> attribute.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GenbankAnnotationDb(seqid=&#39;I&#39;, source=&#39;:memory:&#39;, namer=None, total_records=40578)
</pre></div>
</div>
</div>
</div>
<p>Now that the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> is annotated, we can query it for specific features. For more details on querying, skip to <a class="reference internal" href="#query-for-features"><span class="std std-ref">Querying for Features</span></a>.</p>
</section>
</section>
<section id="from-a-gff-file">
<h3>From a GFF file<a class="headerlink" href="#from-a-gff-file" title="Link to this heading">#</a></h3>
<section id="id1">
<h4>How to load features and sequence data<a class="headerlink" href="#id1" title="Link to this heading">#</a></h4>
<p>Given a FASTA file with sequence data and a GFF file with annotations, we can use <code class="docutils literal notranslate"><span class="pre">load_seq</span></code> to load both the sequence and its corresponding features.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;data/C-elegans-chromosome-I.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/C-elegans-chromosome-I.gff&quot;</span><span class="p">,</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=0)
</pre></div>
</div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><code class="docutils literal notranslate"><span class="pre">total_records=0</span></code>? 🤔 This is because <code class="docutils literal notranslate"><span class="pre">load_seq</span></code> assumes the sequence names match exactly between files! If the names are different, you need to provide function to the <code class="docutils literal notranslate"><span class="pre">label_to_name</span></code> argument.</p>
</div>
<p>Because the names above are different, for FASTA its <code class="docutils literal notranslate"><span class="pre">&quot;I</span> <span class="pre">dna:chromosome</span> <span class="pre">chromosome:WBcel235:I:1:15072434:1</span> <span class="pre">REF&quot;</span></code> and for GFF its <code class="docutils literal notranslate"><span class="pre">&quot;I&quot;</span></code>, we need a <code class="docutils literal notranslate"><span class="pre">label_to_name</span></code> argument. We provide a lambda function.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/C-elegans-chromosome-I.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/C-elegans-chromosome-I.gff&quot;</span><span class="p">,</span>
    <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=60166)
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-load-features-and-associate-them-with-an-existing-sequence">
<h4>How to load features and associate them with an existing sequence<a class="headerlink" href="#how-to-load-features-and-associate-them-with-an-existing-sequence" title="Link to this heading">#</a></h4>
<p>We load with the <code class="docutils literal notranslate"><span class="pre">load_annotations()</span></code> function and directly assign to the <code class="docutils literal notranslate"><span class="pre">.annotation_db</span></code> attribute.</p>
<p>If we know that the features and the sequence share the same coordinate space, then we only need to provide the path to the annotation file.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span><span class="p">,</span> <span class="n">load_annotations</span>

<span class="n">loaded_seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/C-elegans-chromosome-I.fa&quot;</span><span class="p">,</span>
    <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/C-elegans-chromosome-I.gff&quot;</span><span class="p">)</span>
<span class="n">loaded_seq</span><span class="o">.</span><span class="n">annotation_db</span> <span class="o">=</span> <span class="n">db</span>
<span class="n">loaded_seq</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=60166)
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-load-features-and-associate-them-with-sequences-in-an-existing-alignment">
<h4>How to load features and associate them with sequences in an existing alignment<a class="headerlink" href="#how-to-load-features-and-associate-them-with-sequences-in-an-existing-alignment" title="Link to this heading">#</a></h4>
<p>For example, first we load an alignment of the brca1 gene in primates.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">load_annotations</span>

<span class="n">brca1_aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span>
    <span class="s2">&quot;data/primate_brca1.fasta&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span>
<span class="p">)</span>
<span class="n">brca1_aln</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">Chimpanzee</td><td><span class="T_dna">T</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span></td></tr>
<tr><td class="label">Galago</td><td><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="A_dna">A</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">G</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span></td></tr>
<tr><td class="label">HowlerMon</td><td><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="G_dna">G</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span></td></tr>
<tr><td class="label">Rhesus</td><td><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="G_dna">G</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span></td></tr>
<tr><td class="label">Orangutan</td><td><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span></td></tr>
<tr><td class="label">Gorilla</td><td><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span></td></tr>
<tr><td class="label">Human</td><td><span class="T_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="G_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="T_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="C_dna">.</span><span class="A_dna">.</span><span class="C_dna">.</span><span class="T_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span><span class="A_dna">.</span></td></tr>
</table>
<p><i>7 x 2814 (truncated to 7 x 60) dna alignment</i></p>
</div></div></div>
</div>
<p>Next, we annotate with a GFF file that contains features specific to the human gene.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ann_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/brca1_hsa_shortened.gff&quot;</span><span class="p">,</span> <span class="n">seqids</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">])</span>
<span class="n">brca1_aln</span><span class="o">.</span><span class="n">annotation_db</span> <span class="o">=</span> <span class="n">ann_db</span>
<span class="n">brca1_aln</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=11)
</pre></div>
</div>
</div>
</div>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> is accessible via the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> (above) and <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> (below) attribute.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">brca1_aln</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;Human&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>GffAnnotationDb(source=&#39;:memory:&#39;, total_records=11)
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="how-to-query-a-sequence-or-alignment-for-features">
<span id="query-for-features"></span><h2>How to query a Sequence or Alignment for Features<a class="headerlink" href="#how-to-query-a-sequence-or-alignment-for-features" title="Link to this heading">#</a></h2>
<p>The method <code class="docutils literal notranslate"><span class="pre">get_features</span></code> yields all features that match the given arguments. You can provide conditions for the name, biotype, and start/stop location of a feature.</p>
<section id="querying-a-sequence-for-features">
<h3>Querying a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> for Features<a class="headerlink" href="#querying-a-sequence-for-features" title="Link to this heading">#</a></h3>
<section id="querying-via-feature-name">
<h4>Querying via Feature Name<a class="headerlink" href="#querying-via-feature-name" title="Link to this heading">#</a></h4>
<p>We can search for a gene given its name (AKA its unique ID). For example we can search for a gene with <code class="docutils literal notranslate"><span class="pre">name=&quot;WBGene00021661&quot;</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>

<span class="c1"># note we wrap `get_features` in `list` as generator is returned</span>
<span class="n">gene</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;WBGene00021661&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">))</span>
<span class="n">gene</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&#39;I&#39;, biotype=&#39;gene&#39;, name=&#39;WBGene00021661&#39;, map=[343068:348424]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-feature-biotype">
<h4>Querying via Feature Biotype<a class="headerlink" href="#querying-via-feature-biotype" title="Link to this heading">#</a></h4>
<p>We can search for features with a certain biotype, for example, all coding sequences (CDS):</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">))</span>
<span class="n">cds</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&#39;I&#39;, biotype=&#39;CDS&#39;, name=&#39;WBGene00000001&#39;, map=[5107843:5108016, 5108063:5108323, 5108504:5108770, 5108997:5109429, 5109509:5109947]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;CDS&#39;, name=&#39;WBGene00000006&#39;, map=[2571987:2572039, 2572092:2572300, 2572355:2572519, 2572569:2572736, 2573740:2573941, 2573998:2574267, 2574312:2574444, 2574520:2574584, 2575438:2575629, 2575840:2576006]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;CDS&#39;, name=&#39;WBGene00000010&#39;, map=[11389140:11389238, 11389874:11390003, 11390088:11390188, 11390947:11391134, 11392047:11392185, 11393291:11393519, 11393721:11393875, 11394746:11394878, 11394926:11395007, 11395068:11395177, 11396253:11396432, 11396567:11396658]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))]
</pre></div>
</div>
</div>
</div>
<p>We can also provide combinations of argument to search, for example, all CDS with a given name:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;WBGene00021661&quot;</span><span class="p">))</span>
<span class="n">cds</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&#39;I&#39;, biotype=&#39;CDS&#39;, name=&#39;WBGene00021661&#39;, map=[343540:343649, 343893:344033, 345020:345830, 346374:346516, 346560:346742, 346792:347044, 348304:348364]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;CDS&#39;, name=&#39;WBGene00021661&#39;, map=[343540:343649, 343893:344033, 345020:345830, 346374:346516, 346560:346742, 346792:347044, 347092:347113]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))]
</pre></div>
</div>
</div>
</div>
</section>
<section id="querying-via-region-of-interest">
<h4>Querying via region of interest<a class="headerlink" href="#querying-via-region-of-interest" title="Link to this heading">#</a></h4>
<p>We can provide <code class="docutils literal notranslate"><span class="pre">start</span></code> and <code class="docutils literal notranslate"><span class="pre">end</span></code> arguments to <code class="docutils literal notranslate"><span class="pre">get_features()</span></code> and all features within the coordinates will be returned.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">region_features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">10148</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">26732</span><span class="p">))</span>
<span class="n">region_features</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&#39;I&#39;, biotype=&#39;gene&#39;, name=&#39;WBGene00022276&#39;, map=[11494:16837]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;mRNA&#39;, name=&#39;WBGene00022276&#39;, map=[11494:11561, 11617:11689, 14950:15160, 16472:16837]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;CDS&#39;, name=&#39;WBGene00022276&#39;, map=[11640:11689, 14950:15160, 16472:16585]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))]
</pre></div>
</div>
</div>
</div>
<p>We can again provide a combination of conditions, for example, querying for all features with <code class="docutils literal notranslate"><span class="pre">biotype=&quot;mRNA&quot;</span></code> within a certain range, and returning the first match.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mRNA</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">10148</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">29322</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">mRNA</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&#39;I&#39;, biotype=&#39;mRNA&#39;, name=&#39;WBGene00022276&#39;, map=[11494:11561, 11617:11689, 14950:15160, 16472:16837]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="querying-a-sequence-via-an-alignment-for-features">
<h3>Querying a Sequence (via an Alignment) for Features<a class="headerlink" href="#querying-a-sequence-via-an-alignment-for-features" title="Link to this heading">#</a></h3>
<p>To query for a particular <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> within an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> or <code class="docutils literal notranslate"><span class="pre">SequenceCollection</span></code>, we can use <code class="docutils literal notranslate"><span class="pre">get_features</span></code> as shown above for a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code>, but providing the seqid for the sequence of interest.</p>
<p>For example, given an alignment of primates, we can search for features that are just on the human sequence as follows:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_aligned_seqs</span><span class="p">,</span> <span class="n">load_annotations</span>

<span class="c1"># first load alignment and annotate the human seq</span>
<span class="n">aln</span> <span class="o">=</span> <span class="n">load_aligned_seqs</span><span class="p">(</span>
    <span class="s2">&quot;data/primate_brca1.fasta&quot;</span><span class="p">,</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1"># load the annotation data</span>
<span class="n">ann_db</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;data/brca1_hsa_shortened.gff&quot;</span><span class="p">,</span> <span class="n">seqids</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Human&quot;</span><span class="p">])</span>
<span class="n">aln</span><span class="o">.</span><span class="n">annotation_db</span> <span class="o">=</span> <span class="n">ann_db</span>
<span class="c1"># query alignment providing seqid of interest</span>
<span class="n">human_exons</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;Human&quot;</span><span class="p">))</span>
<span class="n">human_exons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&quot;from &#39;Human&#39;&quot;, biotype=&#39;exon&#39;, name=&#39;unknown-0&#39;, map=[1:1507]/2814, parent=7 x 2814 dna alignment: Galago[TGTGGCAAAA...], HowlerMon[TGTGGCACAA...], Rhesus[TGTGGCACAA...], ...)]
</pre></div>
</div>
</div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">seqid=&quot;from'Human'&quot;</span></code> indicated this feature belongs to this particular sequence.</p>
</section>
<section id="querying-an-alignment-for-features">
<h3>Querying an Alignment for Features<a class="headerlink" href="#querying-an-alignment-for-features" title="Link to this heading">#</a></h3>
<section id="querying-for-features-on-any-sequence-in-an-alignment">
<h4>Querying for features on any <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> in an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#querying-for-features-on-any-sequence-in-an-alignment" title="Link to this heading">#</a></h4>
<p>todo: <code class="docutils literal notranslate"><span class="pre">on_alignment=False</span></code> and dont provide seqid</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="c1"># add a feature to the alignment we created above on difference sequence</span>
<span class="n">aln</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;gene:101&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">40</span><span class="p">,</span> <span class="mi">387</span><span class="p">)],</span> <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;Rhesus&quot;</span><span class="p">)</span>

<span class="n">any_feature</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">on_alignment</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="n">any_feature</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&quot;from &#39;Rhesus&#39;&quot;, biotype=&#39;gene&#39;, name=&#39;gene:101&#39;, map=[40:387]/2814, parent=7 x 2814 dna alignment: Galago[TGTGGCAAAA...], HowlerMon[TGTGGCACAA...], Rhesus[TGTGGCACAA...], ...),
 Feature(seqid=&quot;from &#39;Human&#39;&quot;, biotype=&#39;exon&#39;, name=&#39;unknown-0&#39;, map=[1:1507]/2814, parent=7 x 2814 dna alignment: Galago[TGTGGCAAAA...], HowlerMon[TGTGGCACAA...], Rhesus[TGTGGCACAA...], ...)]
</pre></div>
</div>
</div>
</div>
<p>Note there are features from both Rhesus, which we just added, and Human, which we annotated above</p>
</section>
<section id="querying-for-features-on-an-alignment">
<h4>Querying for features on an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code><a class="headerlink" href="#querying-for-features-on-an-alignment" title="Link to this heading">#</a></h4>
<p>todo: <code class="docutils literal notranslate"><span class="pre">on_alignment=True</span></code> and dont provide seqid</p>
<p>Using <code class="docutils literal notranslate"><span class="pre">add_feature</span></code> we add a feature to the brca1 alignment we have been using above, by specifying <code class="docutils literal notranslate"><span class="pre">on_alignment=True</span></code> this feature will be on the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code>.</p>
<p>To query for features on the alignment, we use <code class="docutils literal notranslate"><span class="pre">get_features</span></code>, again specifying <code class="docutils literal notranslate"><span class="pre">on_alignment=True</span></code>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="c1"># first we add the feature to the alignment</span>
<span class="n">aln</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;pseudogene1&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">420</span><span class="p">,</span> <span class="mi">666</span><span class="p">)],</span> <span class="n">on_alignment</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># query for features on the alignment</span>
<span class="n">aln_features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">on_alignment</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="n">aln_features</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=None, biotype=&#39;pseudogene&#39;, name=&#39;pseudogene1&#39;, map=[420:666]/2814, parent=7 x 2814 dna alignment: Galago[TGTGGCAAAA...], HowlerMon[TGTGGCACAA...], Rhesus[TGTGGCACAA...], ...)]
</pre></div>
</div>
</div>
</div>
<p>Note how even though we annotated the Human and Rhesus sequences in the above examples, only the pseudogene we added to <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> is returned by this query.</p>
</section>
<section id="querying-features-that-span-gaps-in-alignments">
<h4>Querying features that span gaps in alignments<a class="headerlink" href="#querying-features-that-span-gaps-in-alignments" title="Link to this heading">#</a></h4>
<p>If you query for a <code class="docutils literal notranslate"><span class="pre">Feature</span></code> from a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> (i.e. the feature is in sequence coordinates), its alignment coordinates may be discontinuous. This will lead to an omission of data from other sequences!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln3</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">([[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;C-CCCAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;-T----TTTT&quot;</span><span class="p">]],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">exon</span> <span class="o">=</span> <span class="n">aln3</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;ex1&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span> <span class="n">on_alignment</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
<span class="n">exon</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span></td></tr>
<tr><td class="label">y</td><td><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 4 dna alignment</i></p>
</div></div></div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aln_exons</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln3</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">aln_exons</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&quot;from &#39;x&#39;&quot;, biotype=&#39;exon&#39;, name=&#39;ex1&#39;, map=[0:1, 2:5]/10, parent=2 x 10 dna alignment: x[C-CCCAAAAA], y[-T----TTTT])
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In the above, the <code class="docutils literal notranslate"><span class="pre">T</span></code> in sequence Y opposite the gap is missing since this approach only returns positions directly corresponding to the feature.</p>
</div>
<p>To include the gaps, use the <code class="docutils literal notranslate"><span class="pre">allow_gaps</span></code> argument</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">exon</span><span class="o">.</span><span class="n">get_slice</span><span class="p">(</span><span class="n">allow_gaps</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="C_dna">C</span><span class="ambig_dna">-</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span></td></tr>
<tr><td class="label">y</td><td><span class="ambig_dna">-</span><span class="T_dna">T</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 5 dna alignment</i></p>
</div></div></div>
</div>
</section>
</section>
</section>
<section id="examples-using-the-methods-available-on-features">
<h2>Examples using the methods available on Features<a class="headerlink" href="#examples-using-the-methods-available-on-features" title="Link to this heading">#</a></h2>
<p>A <code class="docutils literal notranslate"><span class="pre">Feature</span></code> has many methods to manipulate the sequence or alignment that they are bound to.</p>
<section id="how-to-slice-a-sequence-or-alignment-by-its-features">
<h3>How to slice a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> by its features<a class="headerlink" href="#how-to-slice-a-sequence-or-alignment-by-its-features" title="Link to this heading">#</a></h3>
<p>Given a <code class="docutils literal notranslate"><span class="pre">Feature</span></code>, we can directly slice its parent sequence to return its sequence information</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/C-elegans-chromosome-I.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/C-elegans-chromosome-I.gff&quot;</span><span class="p">,</span>
    <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">pseudogene</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="mi">10148</span><span class="p">,</span> <span class="n">stop</span><span class="o">=</span><span class="mi">26732</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;pseudogene&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">seq</span><span class="p">[</span><span class="n">pseudogene</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">gene:WBGene00235381</td><td><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span></td></tr>
</table>
<p><i>DnaSequence, length=719 (truncated to 60)</i></p>
</div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This only works for the <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> that the <code class="docutils literal notranslate"><span class="pre">Feature</span></code> “belongs” to.</p>
</div>
<p>We can also achieve this via <code class="docutils literal notranslate"><span class="pre">get_slice()</span></code></p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pseudogene</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">gene:WBGene00235381</td><td><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span></td></tr>
</table>
<p><i>DnaSequence, length=719 (truncated to 60)</i></p>
</div></div></div>
</div>
</section>
<section id="how-to-display-the-features-of-a-sequence">
<h3>How to display the features of a Sequence<a class="headerlink" href="#how-to-display-the-features-of-a-sequence" title="Link to this heading">#</a></h3>
<p>We can display all the features on a sequence using <code class="docutils literal notranslate"><span class="pre">.get_drawable()</span></code>, or a subset of biotypes. We do this for only the first 50,000 base pairs. The plotly figure returned, as displayed below, is interactive! 🤩 Zoom in on the dark vertical lines in the big gene and you will see small genes on the opposite strand. Hover your cursor over each block and the gene name is displayed.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">subseq</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="mi">25000</span><span class="p">:</span><span class="mi">35000</span><span class="p">]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">subseq</span><span class="o">.</span><span class="n">get_drawable</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="s2">&quot;mRNA&quot;</span><span class="p">,</span> <span class="s2">&quot;CDS&quot;</span><span class="p">,</span> <span class="s2">&quot;misc_RNA&quot;</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><script type="text/javascript">
window.PlotlyConfig = {MathJaxConfig: 'local'};
if (window.MathJax && window.MathJax.Hub && window.MathJax.Hub.Config) {window.MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}
</script>
<script type="module">import "https://cdn.plot.ly/plotly-3.3.1.min"</script>
</div><div class="output text_html"><div>            <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG"></script><script type="text/javascript">if (window.MathJax && window.MathJax.Hub && window.MathJax.Hub.Config) {window.MathJax.Hub.Config({SVG: {font: "STIX-Web"}});}</script>                <script type="text/javascript">window.PlotlyConfig = {MathJaxConfig: 'local'};</script>
        <script charset="utf-8" src="https://cdn.plot.ly/plotly-3.3.1.min.js" integrity="sha256-4rD3fugVb/nVJYUv5Ky3v+fYXoouHaBSP20WIJuEiWg=" crossorigin="anonymous"></script>                <div id="f7f7a74b-54f3-453d-acea-c91ec45050ec" class="plotly-graph-div" style="height:300px; width:600px;"></div>            <script type="text/javascript">                window.PLOTLYENV=window.PLOTLYENV || {};                                if (document.getElementById("f7f7a74b-54f3-453d-acea-c91ec45050ec")) {                    Plotly.newPlot(                        "f7f7a74b-54f3-453d-acea-c91ec45050ec",                        [{"fill":"toself","fillcolor":"rgba(161,0,0,0.75)","hoverinfo":"text","legendgroup":"gene","line":{"color":"rgba(161,0,0,0.75)"},"mode":"lines","name":"gene","showlegend":true,"text":"WBGene00022279","x":{"dtype":"f8","bdata":"AAAAAADWv0AAAAAAANa\u002fQAAAAAAAIKhAAAAAAAAgqEAAAAAAADikQAAAAAAAIKhAAAAAAAAgqEAAAAAAANa\u002fQA=="},"y":{"dtype":"f8","bdata":"mpmZmZmZ0T\u002fNzMzMzMzgP83MzMzMzOA\u002fmpmZmZmZ4T+amZmZmZnZPwAAAAAAANA\u002fmpmZmZmZ0T+amZmZmZnRPw=="},"type":"scatter"},{"fill":"toself","fillcolor":"rgba(161,0,0,0.75)","hoverinfo":"text","legendgroup":"gene","line":{"color":"rgba(161,0,0,0.75)"},"mode":"lines","name":"gene","showlegend":false,"text":"WBGene00170953","x":{"dtype":"f8","bdata":"AAAAAAB6uUAzMzMz8425QDMzMzPzjblAAAAAAACPuUAzMzMz8425QDMzMzPzjblAAAAAAAB6uUAAAAAAAHq5QA=="},"y":{"dtype":"f8","bdata":"mpmZmZmZ0T+amZmZmZnRPwAAAAAAANA\u002fmpmZmZmZ2T+amZmZmZnhP83MzMzMzOA\u002fzczMzMzM4D+amZmZmZnRPw=="},"type":"scatter"},{"fill":"toself","fillcolor":"rgba(161,0,0,0.75)","hoverinfo":"text","legendgroup":"gene","line":{"color":"rgba(161,0,0,0.75)"},"mode":"lines","name":"gene","showlegend":false,"text":"WBGene00173569","x":{"dtype":"f8","bdata":"AAAAAAD2vEAzMzMz8wm9QDMzMzPzCb1AAAAAAAALvUAzMzMz8wm9QDMzMzPzCb1AAAAAAAD2vEAAAAAAAPa8QA=="},"y":{"dtype":"f8","bdata":"mpmZmZmZ0T+amZmZmZnRPwAAAAAAANA\u002fmpmZmZmZ2T+amZmZmZnhP83MzMzMzOA\u002fzczMzMzM4D+amZmZmZnRPw=="},"type":"scatter"},{"fill":"toself","fillcolor":"rgba(161,0,0,0.75)","hoverinfo":"text","legendgroup":"gene","line":{"color":"rgba(161,0,0,0.75)"},"mode":"lines","name":"gene","showlegend":false,"text":"WBGene00022278 (incomplete)","x":{"dtype":"f8","bdata":"AAAAAADUm0AAAAAAANSbQAAAAAAAQH9AAAAAAABAf0AAAAAAAAAAAAAAAAAAQH9AAAAAAABAf0AAAAAAANSbQA=="},"y":{"dtype":"f8","bdata":"mpmZmZmZ0T\u002fNzMzMzMzgP83MzMzMzOA\u002fmpmZmZmZ4T+amZmZmZnZPwAAAAAAANA\u002fmpmZmZmZ0T+amZmZmZnRPw=="},"type":"scatter"},{"fill":"toself","fillcolor":"rgba(140,102,139,0.75)","hoverinfo":"text","legendgroup":"mRNA","line":{"color":"rgba(140,102,139,0.75)"},"mode":"lines","name":"mRNA","showlegend":true,"text":"WBGene00022278 (incomplete)","x":[90.0,90.0,4.5,4.5,0.0,4.5,4.5,90.0,null,90.0,272,null,272,272,472,472,272,null,472,1370,null,1370,1370,1524,1524,1370,null,1524,1589,null,1589,1589,1690,1690,1589,null,1690,1732,null,1732,1732,1781,1781,1732],"y":[0.8250000000000001,1.0750000000000002,1.0750000000000002,1.1,0.9500000000000001,0.8,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(140,102,139,0.75)","hoverinfo":"text","legendgroup":"mRNA","line":{"color":"rgba(140,102,139,0.75)"},"mode":"lines","name":"mRNA","showlegend":false,"text":"WBGene00022278 (incomplete)","x":[90.0,90.0,4.5,4.5,0.0,4.5,4.5,90.0,null,90.0,272,null,272,272,472,472,272,null,472,1370,null,1370,1370,1524,1524,1370,null,1524,1589,null,1589,1589,1681,1681,1589,null,1681,1732,null,1732,1732,1781,1781,1732],"y":[0.8250000000000001,1.0750000000000002,1.0750000000000002,1.1,0.9500000000000001,0.8,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(140,102,139,0.75)","hoverinfo":"text","legendgroup":"mRNA","line":{"color":"rgba(140,102,139,0.75)"},"mode":"lines","name":"mRNA","showlegend":false,"text":"WBGene00022279","x":[3405.0,3405.0,3090.0,3090.0,2590.0,3090.0,3090.0,3405.0,null,3405.0,4099,null,4099,4099,4367,4367,4099,null,4367,4768,null,4768,4768,5424,5424,4768,null,5424,6768,null,6768,6768,7072,7072,6768,null,7072,7411,null,7411,7411,7548,7548,7411],"y":[0.8250000000000001,1.0750000000000002,1.0750000000000002,1.1,0.9500000000000001,0.8,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(140,102,139,0.75)","hoverinfo":"text","legendgroup":"mRNA","line":{"color":"rgba(140,102,139,0.75)"},"mode":"lines","name":"mRNA","showlegend":false,"text":"WBGene00022279","x":[3405.0,3405.0,3088.0,3088.0,2588.0,3088.0,3088.0,3405.0,null,3405.0,4099,null,4099,4099,4367,4367,4099,null,4367,4768,null,4768,4768,5424,5424,4768,null,5424,6768,null,6768,6768,7072,7072,6768,null,7072,7411,null,7411,7411,7542,7542,7411,null,7542,8036,null,8036,8036,8150,8150,8036],"y":[0.8250000000000001,1.0750000000000002,1.0750000000000002,1.1,0.9500000000000001,0.8,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(140,102,139,0.75)","hoverinfo":"text","legendgroup":"mRNA","line":{"color":"rgba(140,102,139,0.75)"},"mode":"lines","name":"mRNA","showlegend":false,"text":"WBGene00022278 (incomplete)","x":[90.0,90.0,4.5,4.5,0.0,4.5,4.5,90.0,null,90.0,272,null,272,272,472,472,272,null,472,1370,null,1370,1370,1524,1524,1370,null,1524,1589,null,1589,1589,1643,1643,1589],"y":[0.8250000000000001,1.0750000000000002,1.0750000000000002,1.1,0.9500000000000001,0.8,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001,null,0.9500000000000001,0.9500000000000001,null,0.8250000000000001,1.0750000000000002,1.0750000000000002,0.8250000000000001,0.8250000000000001],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(0,0,150,0.75)","hoverinfo":"text","legendgroup":"CDS","line":{"color":"rgba(0,0,150,0.75)"},"mode":"lines","name":"CDS","showlegend":true,"text":"WBGene00022279","x":[3405.0,3405.0,3285.3,3285.3,3279.0,3285.3,3285.3,3405.0,null,3405.0,4099,null,4099,4099,4367,4367,4099,null,4367,4768,null,4768,4768,5424,5424,4768,null,5424,6768,null,6768,6768,7072,7072,6768,null,7072,7411,null,7411,7411,7482,7482,7411],"y":[1.375,1.625,1.625,1.65,1.5,1.35,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(0,0,150,0.75)","hoverinfo":"text","legendgroup":"CDS","line":{"color":"rgba(0,0,150,0.75)"},"mode":"lines","name":"CDS","showlegend":false,"text":"WBGene00022278 (incomplete)","x":[90.0,90.0,4.5,4.5,0.0,4.5,4.5,90.0,null,90.0,272,null,272,272,472,472,272,null,472,1370,null,1370,1370,1524,1524,1370,null,1524,1589,null,1589,1589,1690,1690,1589,null,1690,1732,null,1732,1732,1778,1778,1732],"y":[1.375,1.625,1.625,1.65,1.5,1.35,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(0,0,150,0.75)","hoverinfo":"text","legendgroup":"CDS","line":{"color":"rgba(0,0,150,0.75)"},"mode":"lines","name":"CDS","showlegend":false,"text":"WBGene00022278 (incomplete)","x":[90.0,90.0,4.5,4.5,0.0,4.5,4.5,90.0,null,90.0,272,null,272,272,472,472,272,null,472,1370,null,1370,1370,1524,1524,1370,null,1524,1589,null,1589,1589,1681,1681,1589,null,1681,1732,null,1732,1732,1778,1778,1732],"y":[1.375,1.625,1.625,1.65,1.5,1.35,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(0,0,150,0.75)","hoverinfo":"text","legendgroup":"CDS","line":{"color":"rgba(0,0,150,0.75)"},"mode":"lines","name":"CDS","showlegend":false,"text":"WBGene00022278 (incomplete)","x":[90.0,90.0,4.5,4.5,0.0,4.5,4.5,90.0,null,90.0,272,null,272,272,472,472,272,null,472,1370,null,1370,1370,1524,1524,1370,null,1524,1589,null,1589,1589,1643,1643,1589],"y":[1.375,1.625,1.625,1.65,1.5,1.35,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375,null,1.5,1.5,null,1.375,1.625,1.625,1.375,1.375],"type":"scatter"},{"fill":"toself","fillcolor":"rgba(0,0,150,0.75)","hoverinfo":"text","legendgroup":"misc_RNA","line":{"color":"rgba(0,0,150,0.75)"},"mode":"lines","name":"misc_RNA","showlegend":true,"text":"WBGene00170953","x":{"dtype":"f8","bdata":"AAAAAAB6uUAzMzMz8425QDMzMzPzjblAAAAAAACPuUAzMzMz8425QDMzMzPzjblAAAAAAAB6uUAAAAAAAHq5QA=="},"y":{"dtype":"f8","bdata":"zMzMzMzM\u002fj\u002fMzMzMzMz+P2ZmZmZmZv4\u002fZmZmZmZmAECZmZmZmZkBQGZmZmZmZgFAZmZmZmZmAUDMzMzMzMz+Pw=="},"type":"scatter"},{"fill":"toself","fillcolor":"rgba(0,0,150,0.75)","hoverinfo":"text","legendgroup":"misc_RNA","line":{"color":"rgba(0,0,150,0.75)"},"mode":"lines","name":"misc_RNA","showlegend":false,"text":"WBGene00173569","x":{"dtype":"f8","bdata":"AAAAAAD2vEAzMzMz8wm9QDMzMzPzCb1AAAAAAAALvUAzMzMz8wm9QDMzMzPzCb1AAAAAAAD2vEAAAAAAAPa8QA=="},"y":{"dtype":"f8","bdata":"zMzMzMzM\u002fj\u002fMzMzMzMz+P2ZmZmZmZv4\u002fZmZmZmZmAECZmZmZmZkBQGZmZmZmZgFAZmZmZmZmAUDMzMzMzMz+Pw=="},"type":"scatter"}],                        {"autosize":false,"font":{"family":"Balto","size":14},"height":300,"hovermode":"closest","margin":{"b":50,"l":50,"pad":4,"r":50,"t":50},"showlegend":true,"title":{"text":"I"},"width":600,"xaxis":{"range":[0,10000],"showline":true,"zeroline":false},"yaxis":{"range":[0,2.4499999999999997],"showline":true,"visible":false,"zeroline":true},"template":{"data":{"histogram2dcontour":[{"type":"histogram2dcontour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"choropleth":[{"type":"choropleth","colorbar":{"outlinewidth":0,"ticks":""}}],"histogram2d":[{"type":"histogram2d","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"heatmap":[{"type":"heatmap","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"contourcarpet":[{"type":"contourcarpet","colorbar":{"outlinewidth":0,"ticks":""}}],"contour":[{"type":"contour","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"surface":[{"type":"surface","colorbar":{"outlinewidth":0,"ticks":""},"colorscale":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]]}],"mesh3d":[{"type":"mesh3d","colorbar":{"outlinewidth":0,"ticks":""}}],"scatter":[{"fillpattern":{"fillmode":"overlay","size":10,"solidity":0.2},"type":"scatter"}],"parcoords":[{"type":"parcoords","line":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolargl":[{"type":"scatterpolargl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"bar":[{"error_x":{"color":"#2a3f5f"},"error_y":{"color":"#2a3f5f"},"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"bar"}],"scattergeo":[{"type":"scattergeo","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterpolar":[{"type":"scatterpolar","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"histogram":[{"marker":{"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"histogram"}],"scattergl":[{"type":"scattergl","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatter3d":[{"type":"scatter3d","line":{"colorbar":{"outlinewidth":0,"ticks":""}},"marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattermap":[{"type":"scattermap","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattermapbox":[{"type":"scattermapbox","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scatterternary":[{"type":"scatterternary","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"scattercarpet":[{"type":"scattercarpet","marker":{"colorbar":{"outlinewidth":0,"ticks":""}}}],"carpet":[{"aaxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"baxis":{"endlinecolor":"#2a3f5f","gridcolor":"white","linecolor":"white","minorgridcolor":"white","startlinecolor":"#2a3f5f"},"type":"carpet"}],"table":[{"cells":{"fill":{"color":"#EBF0F8"},"line":{"color":"white"}},"header":{"fill":{"color":"#C8D4E3"},"line":{"color":"white"}},"type":"table"}],"barpolar":[{"marker":{"line":{"color":"#E5ECF6","width":0.5},"pattern":{"fillmode":"overlay","size":10,"solidity":0.2}},"type":"barpolar"}],"pie":[{"automargin":true,"type":"pie"}]},"layout":{"autotypenumbers":"strict","colorway":["#636efa","#EF553B","#00cc96","#ab63fa","#FFA15A","#19d3f3","#FF6692","#B6E880","#FF97FF","#FECB52"],"font":{"color":"#2a3f5f"},"hovermode":"closest","hoverlabel":{"align":"left"},"paper_bgcolor":"white","plot_bgcolor":"#E5ECF6","polar":{"bgcolor":"#E5ECF6","angularaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"radialaxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"ternary":{"bgcolor":"#E5ECF6","aaxis":{"gridcolor":"white","linecolor":"white","ticks":""},"baxis":{"gridcolor":"white","linecolor":"white","ticks":""},"caxis":{"gridcolor":"white","linecolor":"white","ticks":""}},"coloraxis":{"colorbar":{"outlinewidth":0,"ticks":""}},"colorscale":{"sequential":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"sequentialminus":[[0.0,"#0d0887"],[0.1111111111111111,"#46039f"],[0.2222222222222222,"#7201a8"],[0.3333333333333333,"#9c179e"],[0.4444444444444444,"#bd3786"],[0.5555555555555556,"#d8576b"],[0.6666666666666666,"#ed7953"],[0.7777777777777778,"#fb9f3a"],[0.8888888888888888,"#fdca26"],[1.0,"#f0f921"]],"diverging":[[0,"#8e0152"],[0.1,"#c51b7d"],[0.2,"#de77ae"],[0.3,"#f1b6da"],[0.4,"#fde0ef"],[0.5,"#f7f7f7"],[0.6,"#e6f5d0"],[0.7,"#b8e186"],[0.8,"#7fbc41"],[0.9,"#4d9221"],[1,"#276419"]]},"xaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"yaxis":{"gridcolor":"white","linecolor":"white","ticks":"","title":{"standoff":15},"zerolinecolor":"white","automargin":true,"zerolinewidth":2},"scene":{"xaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"yaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2},"zaxis":{"backgroundcolor":"#E5ECF6","gridcolor":"white","linecolor":"white","showbackground":true,"ticks":"","zerolinecolor":"white","gridwidth":2}},"shapedefaults":{"line":{"color":"#2a3f5f"}},"annotationdefaults":{"arrowcolor":"#2a3f5f","arrowhead":0,"arrowwidth":1},"geo":{"bgcolor":"white","landcolor":"#E5ECF6","subunitcolor":"white","showland":true,"showlakes":true,"lakecolor":"white"},"title":{"x":0.05},"mapbox":{"style":"light"}}}},                        {"responsive": true}                    ).then(function(){

var gd = document.getElementById('f7f7a74b-54f3-453d-acea-c91ec45050ec');
var x = new MutationObserver(function (mutations, observer) {{
        var display = window.getComputedStyle(gd).display;
        if (!display || display === 'none') {{
            console.log([gd, 'removed!']);
            Plotly.purge(gd);
            observer.disconnect();
        }}
}});

// Listen for the removal of the full notebook cells
var notebookContainer = gd.closest('#notebook-container');
if (notebookContainer) {{
    x.observe(notebookContainer, {childList: true});
}}

// Listen for the clearing of the current output cell
var outputEl = gd.closest('.output');
if (outputEl) {{
    x.observe(outputEl, {childList: true});
}}

                        })                };            </script>        </div></div></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If a feature extends beyond the sequence region selected, its name includes the text “(incomplete)”.</p>
</div>
</section>
<section id="how-to-find-the-coordinates-of-a-feature">
<h3>How to find the coordinates of a feature<a class="headerlink" href="#how-to-find-the-coordinates-of-a-feature" title="Link to this heading">#</a></h3>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pseudogene</span><span class="o">.</span><span class="n">get_coordinates</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[(22881, 23600)]
</pre></div>
</div>
</div>
</div>
<p>These are useful for doing custom things, e.g. if the introns are not annotated for a gene, we can generate the introns from the coordinates of the exons as follows:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">exon_coords</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">get_coordinates</span><span class="p">()</span>

<span class="n">exon_coords</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[(5107843, 5108016),
 (5108063, 5108323),
 (5108504, 5108770),
 (5108997, 5109429),
 (5109509, 5109947)]
</pre></div>
</div>
</div>
</div>
<p>We generate the intron coordinates from the second element of the first tuple, and the first element of the second tuple and so on:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">intron_coords</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">exon_coords</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
    <span class="n">intron_coords</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">exon_coords</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">exon_coords</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span>

<span class="n">intron_coords</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[(5108016, 5108063),
 (5108323, 5108504),
 (5108770, 5108997),
 (5109429, 5109509)]
</pre></div>
</div>
</div>
</div>
<p>We can then add the introns as a <code class="docutils literal notranslate"><span class="pre">Feature</span></code> to the sequence!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">intron</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;intron&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;intron:Y74C9A.3.1&quot;</span><span class="p">,</span> <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;I&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="n">intron_coords</span>
<span class="p">)</span>
<span class="n">intron</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&#39;I&#39;, biotype=&#39;intron&#39;, name=&#39;intron:Y74C9A.3.1&#39;, map=[5108016:5108063, 5108323:5108504, 5108770:5108997, 5109429:5109509]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-take-the-union-of-features">
<h3>How to take the union of features<a class="headerlink" href="#how-to-take-the-union-of-features" title="Link to this heading">#</a></h3>
<p>We can create a feature that is the union of all coding sequence.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">cds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;CDS&quot;</span><span class="p">))</span>
<span class="n">union_cds</span> <span class="o">=</span> <span class="n">cds</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">cds</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
</div>
</div>
</section>
<section id="how-to-get-the-shadow-of-a-feature">
<h3>How to get the shadow of a Feature<a class="headerlink" href="#how-to-get-the-shadow-of-a-feature" title="Link to this heading">#</a></h3>
<p>The “shadow” of a feature is a new feature containing all of the sequence <strong>except the feature</strong>!</p>
<section id="how-to-use-the-shadow-of-a-feature-to-return-the-intergenic-sequence">
<h4>How to use the shadow of a Feature to return the intergenic sequence<a class="headerlink" href="#how-to-use-the-shadow-of-a-feature-to-return-the-intergenic-sequence" title="Link to this heading">#</a></h4>
<p>We first need to query our sequence for all genes. Using the <code class="docutils literal notranslate"><span class="pre">union()</span></code> method we combine all genes into a single feature.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">))</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">genes</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span>
<span class="n">genes</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&#39;I&#39;, biotype=&#39;gene&#39;, name=&#39;WBGene00000001, WBGene00000006, WBGene00000010, WBGene00000012, WBGene00000013, WBGene00000020, WBGene00000023, WBGene00000036, WBGene00000045, WBGene00000050, WBGene00000058, WBGene00000079, WBGene00000084, WBGene00000088, WBGene00000089, WBGene00000095, WBGene00000096, WBGene00000099, WBGene00000138, WBGene00000140, WBGene00000147, WBGene00000148, WBGene00000150, WBGene00000155, WBGene00000156, WBGene00000158, WBGene00000163, WBGene00000165, WBGene00000177, WBGene00000197, WBGene00000199, WBGene00000205, WBGene00000209, WBGene00000227, WBGene00000230, WBGene00000232, WBGene00000236, WBGene00000237, WBGene00000241, WBGene00000253, WBGene00000254, WBGene00000269, WBGene00000272, WBGene00000275, WBGene00000284, WBGene00000368, WBGene00000371, WBGene00000382, WBGene00000386, WBGene00000408, WBGene00000409, WBGene00000410, WBGene00000413, WBGene00000415, WBGene00000426, WBGene00000429, WBGene00000430, WBGene00000431, WBGene00000433, WBGene00000434, WBGene00000436, WBGene00000440, WBGene00000466, WBGene00000467, WBGene00000468, WBGene00000469, WBGene00000473, WBGene00000474, WBGene00000476, WBGene00000483, WBGene00000485, WBGene00000492, WBGene00000493, WBGene00000496, WBGene00000500, WBGene00000502, WBGene00000527, WBGene00000540, WBGene00000544, WBGene00000583, WBGene00000596, WBGene00000612, WBGene00000622, WBGene00000623, WBGene00000624, WBGene00000625, WBGene00000626, WBGene00000627, WBGene00000628, WBGene00000629, WBGene00000630, WBGene00000631, WBGene00000632, WBGene00000633, WBGene00000634, WBGene00000636, WBGene00000637, WBGene00000638, WBGene00000639, WBGene00000640, WBGene00000641, WBGene00000642, WBGene00000761, WBGene00000764, WBGene00000772, WBGene00000777, WBGene00000779, WBGene00000780, WBGene00000787, WBGene00000788, WBGene00000794, WBGene00000800, WBGene00000801, WBGene00000812, WBGene00000814, WBGene00000821, WBGene00000832, WBGene00000843, WBGene00000867, WBGene00000869, WBGene00000871, WBGene00000881, WBGene00000890, WBGene00000891, WBGene00000896, WBGene00000904, WBGene00000912, WBGene00000917, WBGene00000931, WBGene00000937, WBGene00000938, WBGene00000962, WBGene00000963, WBGene00000965, WBGene00000966, WBGene00000967, WBGene00000968, WBGene00001008, WBGene00001022, WBGene00001030, WBGene00001039, WBGene00001045, WBGene00001046, WBGene00001047, WBGene00001048, WBGene00001049, WBGene00001050, WBGene00001056, WBGene00001067, WBGene00001075, WBGene00001112, WBGene00001115, WBGene00001117, WBGene00001121, WBGene00001131, WBGene00001136, WBGene00001145, WBGene00001147, WBGene00001156, WBGene00001167, WBGene00001196, WBGene00001214, WBGene00001215, WBGene00001226, WBGene00001228, WBGene00001231, WBGene00001232, WBGene00001234, WBGene00001309, WBGene00001333, WBGene00001334, WBGene00001337, WBGene00001365, WBGene00001372, WBGene00001377, WBGene00001404, WBGene00001405, WBGene00001414, WBGene00001427, WBGene00001430, WBGene00001432, WBGene00001433, WBGene00001442, WBGene00001481, WBGene00001483, WBGene00001488, WBGene00001491, WBGene00001497, WBGene00001501, WBGene00001542, WBGene00001555, WBGene00001568, WBGene00001574, WBGene00001582, WBGene00001590, WBGene00001592, WBGene00001595, WBGene00001596, WBGene00001598, WBGene00001599, WBGene00001600, WBGene00001601, WBGene00001614, WBGene00001627, WBGene00001640, WBGene00001641, WBGene00001642, WBGene00001643, WBGene00001644, WBGene00001647, WBGene00001648, WBGene00001651, WBGene00001676, WBGene00001677, WBGene00001678, WBGene00001680, WBGene00001682, WBGene00001704, WBGene00001707, WBGene00001725, WBGene00001745, WBGene00001746, WBGene00001773, WBGene00001791, WBGene00001814, WBGene00001816, WBGene00001819, WBGene00001832, WBGene00001833, WBGene00001836, WBGene00001843, WBGene00001856, WBGene00001859, WBGene00001860, WBGene00001941, WBGene00001942, WBGene00001949, WBGene00001960, WBGene00001973, WBGene00001979, WBGene00001980, WBGene00001985, WBGene00001990, WBGene00002000, WBGene00002004, WBGene00002026, WBGene00002027, WBGene00002034, WBGene00002035, WBGene00002040, WBGene00002042, WBGene00002047, WBGene00002069, WBGene00002070, WBGene00002073, WBGene00002075, WBGene00002077, WBGene00002079, WBGene00002101, WBGene00002107, WBGene00002108, WBGene00002109, WBGene00002110, WBGene00002111, WBGene00002112, WBGene00002113, WBGene00002116, WBGene00002119, WBGene00002134, WBGene00002135, WBGene00002136, WBGene00002137, WBGene00002138, WBGene00002139, WBGene00002141, WBGene00002142, WBGene00002143, WBGene00002144, WBGene00002153, WBGene00002174, WBGene00002179, WBGene00002181, WBGene00002189, WBGene00002191, WBGene00002196, WBGene00002198, WBGene00002205, WBGene00002213, WBGene00002225, WBGene00002226, WBGene00002232, WBGene00002240, WBGene00002248, WBGene00002257, WBGene00002258, WBGene00002276, WBGene00002324, WBGene00002348, WBGene00002393, WBGene00002583, WBGene00002601, WBGene00002694, WBGene00002696, WBGene00002717, WBGene00002783, WBGene00002804, WBGene00002977, WBGene00002978, WBGene00002979, WBGene00002989, WBGene00002999, WBGene00003000, WBGene00003006, WBGene00003014, WBGene00003020, WBGene00003026, WBGene00003029, WBGene00003036, WBGene00003040, WBGene00003041, WBGene00003052, WBGene00003059, WBGene00003060, WBGene00003061, WBGene00003062, WBGene00003068, WBGene00003071, WBGene00003072, WBGene00003074, WBGene00003080, WBGene00003083, WBGene00003098, WBGene00003111, WBGene00003123, WBGene00003132, WBGene00003136, WBGene00003156, WBGene00003170, WBGene00003172, WBGene00003182, WBGene00003183, WBGene00003184, WBGene00003186, WBGene00003209, WBGene00003221, WBGene00003229, WBGene00003233, WBGene00003236, WBGene00003238, WBGene00003258, WBGene00003260, WBGene00003261, WBGene00003278, WBGene00003299, WBGene00003307, WBGene00003316, WBGene00003328, WBGene00003338, WBGene00003339, WBGene00003387, WBGene00003396, WBGene00003397, WBGene00003400, WBGene00003418, WBGene00003422, WBGene00003471, WBGene00003472, WBGene00003475, WBGene00003499, WBGene00003508, WBGene00003516, WBGene00003524, WBGene00003548, WBGene00003552, WBGene00003564, WBGene00003569, WBGene00003578, WBGene00003581, WBGene00003584, WBGene00003585, WBGene00003587, WBGene00003596, WBGene00003601, WBGene00003622, WBGene00003639, WBGene00003652, WBGene00003654, WBGene00003659, WBGene00003663, WBGene00003664, WBGene00003667, WBGene00003671, WBGene00003672, WBGene00003675, WBGene00003679, WBGene00003703, WBGene00003735, WBGene00003742, WBGene00003746, WBGene00003750, WBGene00003753, WBGene00003777, WBGene00003788, WBGene00003790, WBGene00003792, WBGene00003793, WBGene00003797, WBGene00003798, WBGene00003799, WBGene00003800, WBGene00003803, WBGene00003815, WBGene00003818, WBGene00003836, WBGene00003842, WBGene00003847, WBGene00003878, WBGene00003902, WBGene00003904, WBGene00003905, WBGene00003906, WBGene00003921, WBGene00003924, WBGene00003925, WBGene00003926, WBGene00003934, WBGene00003948, WBGene00003950, WBGene00003951, WBGene00003953, WBGene00003964, WBGene00003977, WBGene00003980, WBGene00003989, WBGene00003996, WBGene00004014, WBGene00004028, WBGene00004043, WBGene00004049, WBGene00004063, WBGene00004068, WBGene00004077, WBGene00004087, WBGene00004093, WBGene00004094, WBGene00004109, WBGene00004110, WBGene00004111, WBGene00004113, WBGene00004115, WBGene00004117, WBGene00004131, WBGene00004143, WBGene00004147, WBGene00004154, WBGene00004168, WBGene00004178, WBGene00004181, WBGene00004186, WBGene00004190, WBGene00004196, WBGene00004202, WBGene00004217, WBGene00004224, WBGene00004225, WBGene00004226, WBGene00004231, WBGene00004235, WBGene00004237, WBGene00004246, WBGene00004268, WBGene00004272, WBGene00004273, WBGene00004274, WBGene00004275, WBGene00004298, WBGene00004305, WBGene00004312, WBGene00004318, WBGene00004324, WBGene00004332, WBGene00004368, WBGene00004372, WBGene00004378, WBGene00004380, WBGene00004389, WBGene00004393, WBGene00004404, WBGene00004408, WBGene00004409, WBGene00004412, WBGene00004415, WBGene00004418, WBGene00004425, WBGene00004426, WBGene00004429, WBGene00004431, WBGene00004436, WBGene00004437, WBGene00004439, WBGene00004441, WBGene00004444, WBGene00004445, WBGene00004464, WBGene00004466, WBGene00004475, WBGene00004476, WBGene00004479, WBGene00004484, WBGene00004486, WBGene00004488, WBGene00004489, WBGene00004495, WBGene00004505, WBGene00004508, WBGene00004509, WBGene00004512, WBGene00004513, WBGene00004567, WBGene00004622, WBGene00004677, WBGene00004684, WBGene00004706, WBGene00004721, WBGene00004743, WBGene00004753, WBGene00004771, WBGene00004773, WBGene00004775, WBGene00004778, WBGene00004803, WBGene00004807, WBGene00004808, WBGene00004836, WBGene00004838, WBGene00004839, WBGene00004841, WBGene00004875, WBGene00004879, WBGene00004880, WBGene00004883, WBGene00004887, WBGene00004888, WBGene00004889, WBGene00004890, WBGene00004895, WBGene00004900, WBGene00004901, WBGene00004915, WBGene00004920, WBGene00004924, WBGene00004931, WBGene00004944, WBGene00004946, WBGene00004952, WBGene00004953, WBGene00004955, WBGene00004958, WBGene00004959, WBGene00004962, WBGene00004963, WBGene00004965, WBGene00004966, WBGene00004969, WBGene00004978, WBGene00005000, WBGene00005002, WBGene00005009, WBGene00005010, WBGene00005023, WBGene00005043, WBGene00005044, WBGene00005045, WBGene00005046, WBGene00005047, WBGene00005048, WBGene00005049, WBGene00005050, WBGene00005051, WBGene00005077, WBGene00005078, WBGene00005131, WBGene00005171, WBGene00005272, WBGene00005273, WBGene00005274, WBGene00005517, WBGene00005523, WBGene00005524, WBGene00005525, WBGene00005526, WBGene00005527, WBGene00005528, WBGene00005529, WBGene00005710, WBGene00005835, WBGene00005839, WBGene00005840, WBGene00005858, WBGene00006044, WBGene00006047, WBGene00006048, WBGene00006058, WBGene00006061, WBGene00006062, WBGene00006272, WBGene00006306, WBGene00006324, WBGene00006349, WBGene00006352, WBGene00006377, WBGene00006379, WBGene00006382, WBGene00006385, WBGene00006386, WBGene00006394, WBGene00006399, WBGene00006407, WBGene00006423, WBGene00006440, WBGene00006445, WBGene00006447, WBGene00006461, WBGene00006484, WBGene00006492, WBGene00006503, WBGene00006505, WBGene00006508, WBGene00006518, WBGene00006525, WBGene00006528, WBGene00006529, WBGene00006531, WBGene00006532, WBGene00006565, WBGene00006574, WBGene00006577, WBGene00006578, WBGene00006593, WBGene00006594, WBGene00006595, WBGene00006607, WBGene00006616, WBGene00006618, WBGene00006624, WBGene00006625, WBGene00006633, WBGene00006641, WBGene00006645, WBGene00006656, WBGene00006681, WBGene00006682, WBGene00006689, WBGene00006700, WBGene00006702, WBGene00006707, WBGene00006709, WBGene00006711, WBGene00006720, WBGene00006726, WBGene00006730, WBGene00006732, WBGene00006740, WBGene00006751, WBGene00006752, WBGene00006753, WBGene00006754, WBGene00006765, WBGene00006771, WBGene00006773, WBGene00006774, WBGene00006776, WBGene00006789, WBGene00006790, WBGene00006791, WBGene00006793, WBGene00006797, WBGene00006805, WBGene00006807, WBGene00006819, WBGene00006820, WBGene00006823, WBGene00006824, WBGene00006829, WBGene00006833, WBGene00006844, WBGene00006845, WBGene00006876, WBGene00006888, WBGene00006889, WBGene00006902, WBGene00006919, WBGene00006932, WBGene00006936, WBGene00006958, WBGene00006961, WBGene00006963, WBGene00006976, WBGene00006978, WBGene00006984, WBGene00007001, WBGene00007009, WBGene00007018, WBGene00007024, WBGene00007026, WBGene00007042, WBGene00007043, WBGene00007047, WBGene00007053, WBGene00007054, WBGene00007057, WBGene00007094, WBGene00007155, WBGene00007156, WBGene00007158, WBGene00007159, WBGene00007213, WBGene00007214, WBGene00007215, WBGene00007216, WBGene00007217, WBGene00007218, WBGene00007219, WBGene00007252, WBGene00007253, WBGene00007254, WBGene00007255, WBGene00007256, WBGene00007257, WBGene00007258, WBGene00007274, WBGene00007275, WBGene00007277, WBGene00007278, WBGene00007297, WBGene00007299, WBGene00007300, WBGene00007301, WBGene00007302, WBGene00007303, WBGene00007533, WBGene00007597, WBGene00007598, WBGene00007599, WBGene00007600, WBGene00007601, WBGene00007602, WBGene00007603, WBGene00007620, WBGene00007621, WBGene00007637, WBGene00007638, WBGene00007639, WBGene00007640, WBGene00007641, WBGene00007642, WBGene00007643, WBGene00007644, WBGene00007645, WBGene00007646, WBGene00007647, WBGene00007654, WBGene00007655, WBGene00007656, WBGene00007657, WBGene00007658, WBGene00007659, WBGene00007660, WBGene00007661, WBGene00007662, WBGene00007705, WBGene00007707, WBGene00007708, WBGene00007709, WBGene00007710, WBGene00007711, WBGene00007712, WBGene00007713, WBGene00007739, WBGene00007740, WBGene00007741, WBGene00007742, WBGene00007743, WBGene00007766, WBGene00007767, WBGene00007768, WBGene00007769, WBGene00007770, WBGene00007854, WBGene00007855, WBGene00007856, WBGene00007857, WBGene00007858, WBGene00007911, WBGene00007912, WBGene00007913, WBGene00007914, WBGene00007971, WBGene00007972, WBGene00007973, WBGene00007974, WBGene00007975, WBGene00007976, WBGene00007977, WBGene00007978, WBGene00007979, WBGene00007984, WBGene00007985, WBGene00007986, WBGene00007988, WBGene00007989, WBGene00007990, WBGene00007991, WBGene00007992, WBGene00008061, WBGene00008062, WBGene00008063, WBGene00008064, WBGene00008107, WBGene00008130, WBGene00008131, WBGene00008132, WBGene00008133, WBGene00008134, WBGene00008157, WBGene00008158, WBGene00008159, WBGene00008160, WBGene00008161, WBGene00008162, WBGene00008163, WBGene00008184, WBGene00008185, WBGene00008186, WBGene00008187, WBGene00008188, WBGene00008190, WBGene00008191, WBGene00008216, WBGene00008289, WBGene00008290, WBGene00008291, WBGene00008292, WBGene00008293, WBGene00008294, WBGene00008295, WBGene00008311, WBGene00008312, WBGene00008313, WBGene00008314, WBGene00008315, WBGene00008316, WBGene00008317, WBGene00008318, WBGene00008381, WBGene00008382, WBGene00008383, WBGene00008384, WBGene00008385, WBGene00008386, WBGene00008387, WBGene00008397, WBGene00008398, WBGene00008399, WBGene00008400, WBGene00008401, WBGene00008411, WBGene00008412, WBGene00008413, WBGene00008414, WBGene00008415, WBGene00008417, WBGene00008418, WBGene00008419, WBGene00008420, WBGene00008421, WBGene00008439, WBGene00008440, WBGene00008470, WBGene00008471, WBGene00008472, WBGene00008473, WBGene00008474, WBGene00008475, WBGene00008476, WBGene00008477, WBGene00008478, WBGene00008479, WBGene00008528, WBGene00008529, WBGene00008530, WBGene00008531, WBGene00008532, WBGene00008533, WBGene00008541, WBGene00008542, WBGene00008543, WBGene00008564, WBGene00008565, WBGene00008566, WBGene00008567, WBGene00008568, WBGene00008569, WBGene00008570, WBGene00008610, WBGene00008611, WBGene00008612, WBGene00008613, WBGene00008648, WBGene00008649, WBGene00008650, WBGene00008651, WBGene00008652, WBGene00008661, WBGene00008662, WBGene00008663, WBGene00008664, WBGene00008665, WBGene00008666, WBGene00008667, WBGene00008681, WBGene00008695, WBGene00008762, WBGene00008763, WBGene00008764, WBGene00008765, WBGene00008766, WBGene00008767, WBGene00008768, WBGene00008769, WBGene00008770, WBGene00008779, WBGene00008780, WBGene00008781, WBGene00008782, WBGene00008783, WBGene00008784, WBGene00008785, WBGene00008786, WBGene00008787, WBGene00008852, WBGene00008853, WBGene00008854, WBGene00008855, WBGene00008856, WBGene00008857, WBGene00008858, WBGene00008870, WBGene00008876, WBGene00008877, WBGene00008878, WBGene00008885, WBGene00008886, WBGene00008887, WBGene00008889, WBGene00008890, WBGene00008905, WBGene00008906, WBGene00008907, WBGene00008908, WBGene00008909, WBGene00008934, WBGene00008935, WBGene00008990, WBGene00008991, WBGene00009001, WBGene00009002, WBGene00009004, WBGene00009005, WBGene00009050, WBGene00009051, WBGene00009052, WBGene00009054, WBGene00009055, WBGene00009056, WBGene00009061, WBGene00009063, WBGene00009064, WBGene00009065, WBGene00009112, WBGene00009113, WBGene00009114, WBGene00009116, WBGene00009117, WBGene00009118, WBGene00009119, WBGene00009120, WBGene00009121, WBGene00009122, WBGene00009123, WBGene00009124, WBGene00009125, WBGene00009126, WBGene00009127, WBGene00009128, WBGene00009129, WBGene00009130, WBGene00009140, WBGene00009141, WBGene00009142, WBGene00009143, WBGene00009144, WBGene00009145, WBGene00009157, WBGene00009158, WBGene00009159, WBGene00009160, WBGene00009161, WBGene00009162, WBGene00009163, WBGene00009164, WBGene00009165, WBGene00009173, WBGene00009174, WBGene00009176, WBGene00009177, WBGene00009178, WBGene00009187, WBGene00009188, WBGene00009189, WBGene00009190, WBGene00009191, WBGene00009210, WBGene00009220, WBGene00009243, WBGene00009250, WBGene00009251, WBGene00009252, WBGene00009254, WBGene00009260, WBGene00009261, WBGene00009262, WBGene00009263, WBGene00009264, WBGene00009266, WBGene00009267, WBGene00009268, WBGene00009269, WBGene00009270, WBGene00009271, WBGene00009272, WBGene00009274, WBGene00009276, WBGene00009277, WBGene00009284, WBGene00009285, WBGene00009286, WBGene00009287, WBGene00009288, WBGene00009304, WBGene00009305, WBGene00009306, WBGene00009312, WBGene00009313, WBGene00009314, WBGene00009315, WBGene00009316, WBGene00009318, WBGene00009341, WBGene00009342, WBGene00009343, WBGene00009344, WBGene00009345, WBGene00009346, WBGene00009359, WBGene00009360, WBGene00009361, WBGene00009362, WBGene00009363, WBGene00009366, WBGene00009367, WBGene00009368, WBGene00009369, WBGene00009370, WBGene00009371, WBGene00009408, WBGene00009409, WBGene00009410, WBGene00009411, WBGene00009412, WBGene00009413, WBGene00009414, WBGene00009415, WBGene00009416, WBGene00009451, WBGene00009452, WBGene00009453, WBGene00009454, WBGene00009455, WBGene00009456, WBGene00009457, WBGene00009458, WBGene00009459, WBGene00009460, WBGene00009461, WBGene00009462, WBGene00009463, WBGene00009464, WBGene00009465, WBGene00009466, WBGene00009467, WBGene00009468, WBGene00009475, WBGene00009476, WBGene00009477, WBGene00009478, WBGene00009479, WBGene00009480, WBGene00009481, WBGene00009498, WBGene00009499, WBGene00009500, WBGene00009507, WBGene00009508, WBGene00009509, WBGene00009510, WBGene00009553, WBGene00009554, WBGene00009556, WBGene00009557, WBGene00009558, WBGene00009559, WBGene00009562, WBGene00009563, WBGene00009565, WBGene00009607, WBGene00009608, WBGene00009609, WBGene00009610, WBGene00009611, WBGene00009612, WBGene00009613, WBGene00009614, WBGene00009615, WBGene00009616, WBGene00009646, WBGene00009664, WBGene00009665, WBGene00009666, WBGene00009667, WBGene00009668, WBGene00009670, WBGene00009671, WBGene00009672, WBGene00009673, WBGene00009693, WBGene00009695, WBGene00009696, WBGene00009697, WBGene00009698, WBGene00009743, WBGene00009744, WBGene00009745, WBGene00009746, WBGene00009747, WBGene00009748, WBGene00009749, WBGene00009750, WBGene00009751, WBGene00009752, WBGene00009753, WBGene00009754, WBGene00009755, WBGene00009822, WBGene00009823, WBGene00009824, WBGene00009825, WBGene00009826, WBGene00009827, WBGene00009828, WBGene00009861, WBGene00009862, WBGene00009863, WBGene00009864, WBGene00009865, WBGene00009915, WBGene00009916, WBGene00009917, WBGene00009918, WBGene00009919, WBGene00009920, WBGene00009921, WBGene00009922, WBGene00009923, WBGene00009937, WBGene00009938, WBGene00009939, WBGene00009940, WBGene00009941, WBGene00009942, WBGene00009958, WBGene00009959, WBGene00009960, WBGene00009961, WBGene00009962, WBGene00009963, WBGene00009964, WBGene00010110, WBGene00010111, WBGene00010112, WBGene00010113, WBGene00010114, WBGene00010133, WBGene00010134, WBGene00010135, WBGene00010136, WBGene00010137, WBGene00010157, WBGene00010158, WBGene00010159, WBGene00010160, WBGene00010161, WBGene00010162, WBGene00010163, WBGene00010165, WBGene00010166, WBGene00010167, WBGene00010168, WBGene00010169, WBGene00010170, WBGene00010171, WBGene00010172, WBGene00010173, WBGene00010242, WBGene00010243, WBGene00010244, WBGene00010245, WBGene00010246, WBGene00010247, WBGene00010248, WBGene00010291, WBGene00010323, WBGene00010324, WBGene00010325, WBGene00010326, WBGene00010327, WBGene00010328, WBGene00010366, WBGene00010367, WBGene00010369, WBGene00010370, WBGene00010371, WBGene00010399, WBGene00010400, WBGene00010401, WBGene00010402, WBGene00010416, WBGene00010417, WBGene00010419, WBGene00010420, WBGene00010421, WBGene00010422, WBGene00010427, WBGene00010487, WBGene00010488, WBGene00010495, WBGene00010496, WBGene00010497, WBGene00010498, WBGene00010499, WBGene00010500, WBGene00010537, WBGene00010560, WBGene00010562, WBGene00010563, WBGene00010564, WBGene00010565, WBGene00010566, WBGene00010567, WBGene00010568, WBGene00010569, WBGene00010576, WBGene00010577, WBGene00010578, WBGene00010579, WBGene00010580, WBGene00010581, WBGene00010582, WBGene00010583, WBGene00010584, WBGene00010585, WBGene00010586, WBGene00010587, WBGene00010608, WBGene00010609, WBGene00010610, WBGene00010611, WBGene00010612, WBGene00010613, WBGene00010614, WBGene00010615, WBGene00010616, WBGene00010617, WBGene00010618, WBGene00010619, WBGene00010620, WBGene00010621, WBGene00010622, WBGene00010623, WBGene00010624, WBGene00010641, WBGene00010642, WBGene00010643, WBGene00010644, WBGene00010645, WBGene00010650, WBGene00010651, WBGene00010652, WBGene00010653, WBGene00010654, WBGene00010730, WBGene00010731, WBGene00010732, WBGene00010738, WBGene00010740, WBGene00010766, WBGene00010767, WBGene00010768, WBGene00010769, WBGene00010770, WBGene00010771, WBGene00010804, WBGene00010805, WBGene00010806, WBGene00010807, WBGene00010808, WBGene00010817, WBGene00010818, WBGene00010819, WBGene00010820, WBGene00010821, WBGene00010822, WBGene00010823, WBGene00010824, WBGene00010825, WBGene00010854, WBGene00010855, WBGene00010856, WBGene00010857, WBGene00010858, WBGene00010859, WBGene00010869, WBGene00010870, WBGene00010871, WBGene00010872, WBGene00010873, WBGene00011026, WBGene00011027, WBGene00011028, WBGene00011029, WBGene00011030, WBGene00011032, WBGene00011033, WBGene00011034, WBGene00011035, WBGene00011036, WBGene00011037, WBGene00011058, WBGene00011059, WBGene00011060, WBGene00011061, WBGene00011062, WBGene00011063, WBGene00011064, WBGene00011065, WBGene00011066, WBGene00011155, WBGene00011156, WBGene00011230, WBGene00011231, WBGene00011232, WBGene00011234, WBGene00011235, WBGene00011344, WBGene00011345, WBGene00011352, WBGene00011353, WBGene00011375, WBGene00011376, WBGene00011378, WBGene00011379, WBGene00011380, WBGene00011386, WBGene00011387, WBGene00011388, WBGene00011389, WBGene00011390, WBGene00011431, WBGene00011432, WBGene00011433, WBGene00011434, WBGene00011435, WBGene00011488, WBGene00011489, WBGene00011490, WBGene00011491, WBGene00011492, WBGene00011493, WBGene00011494, WBGene00011495, WBGene00011497, WBGene00011531, WBGene00011532, WBGene00011545, WBGene00011546, WBGene00011547, WBGene00011548, WBGene00011549, WBGene00011550, WBGene00011551, WBGene00011581, WBGene00011582, WBGene00011583, WBGene00011584, WBGene00011585, WBGene00011586, WBGene00011629, WBGene00011630, WBGene00011631, WBGene00011650, WBGene00011651, WBGene00011652, WBGene00011653, WBGene00011654, WBGene00011655, WBGene00011656, WBGene00011657, WBGene00011658, WBGene00011659, WBGene00011660, WBGene00011778, WBGene00011779, WBGene00011780, WBGene00011781, WBGene00011782, WBGene00011783, WBGene00011784, WBGene00011785, WBGene00011786, WBGene00011827, WBGene00011828, WBGene00011829, WBGene00011864, WBGene00011906, WBGene00011908, WBGene00011909, WBGene00011912, WBGene00011913, WBGene00011914, WBGene00011915, WBGene00011916, WBGene00011917, WBGene00011918, WBGene00011919, WBGene00011920, WBGene00011921, WBGene00011933, WBGene00011934, WBGene00011935, WBGene00011936, WBGene00011944, WBGene00011945, WBGene00011968, WBGene00011969, WBGene00011970, WBGene00011971, WBGene00011972, WBGene00011973, WBGene00011975, WBGene00011985, WBGene00011986, WBGene00011988, WBGene00012029, WBGene00012030, WBGene00012031, WBGene00012036, WBGene00012037, WBGene00012038, WBGene00012039, WBGene00012040, WBGene00012041, WBGene00012042, WBGene00012102, WBGene00012103, WBGene00012104, WBGene00012105, WBGene00012106, WBGene00012107, WBGene00012113, WBGene00012114, WBGene00012115, WBGene00012116, WBGene00012136, WBGene00012137, WBGene00012138, WBGene00012139, WBGene00012140, WBGene00012150, WBGene00012151, WBGene00012164, WBGene00012165, WBGene00012166, WBGene00012167, WBGene00012168, WBGene00012192, WBGene00012193, WBGene00012194, WBGene00012195, WBGene00012196, WBGene00012208, WBGene00012209, WBGene00012210, WBGene00012211, WBGene00012212, WBGene00012213, WBGene00012214, WBGene00012215, WBGene00012216, WBGene00012231, WBGene00012232, WBGene00012233, WBGene00012234, WBGene00012235, WBGene00012236, WBGene00012237, WBGene00012238, WBGene00012239, WBGene00012240, WBGene00012241, WBGene00012264, WBGene00012265, WBGene00012266, WBGene00012267, WBGene00012268, WBGene00012269, WBGene00012270, WBGene00012274, WBGene00012275, WBGene00012289, WBGene00012290, WBGene00012296, WBGene00012297, WBGene00012298, WBGene00012317, WBGene00012343, WBGene00012344, WBGene00012345, WBGene00012351, WBGene00012352, WBGene00012353, WBGene00012354, WBGene00012355, WBGene00012365, WBGene00012366, WBGene00012368, WBGene00012369, WBGene00012370, WBGene00012371, WBGene00012376, WBGene00012377, WBGene00012386, WBGene00012387, WBGene00012388, WBGene00012389, WBGene00012390, WBGene00012391, WBGene00012393, WBGene00012394, WBGene00012474, WBGene00012475, WBGene00012476, WBGene00012477, WBGene00012478, WBGene00012479, WBGene00012480, WBGene00012481, WBGene00012482, WBGene00012483, WBGene00012484, WBGene00012485, WBGene00012486, WBGene00012487, WBGene00012488, WBGene00012500, WBGene00012501, WBGene00012502, WBGene00012503, WBGene00012504, WBGene00012505, WBGene00012509, WBGene00012576, WBGene00012577, WBGene00012578, WBGene00012653, WBGene00012732, WBGene00012734, WBGene00012735, WBGene00012738, WBGene00012739, WBGene00012740, WBGene00012752, WBGene00012945, WBGene00012946, WBGene00012947, WBGene00012948, WBGene00012950, WBGene00012951, WBGene00012952, WBGene00012953, WBGene00012954, WBGene00012955, WBGene00012956, WBGene00012958, WBGene00012959, WBGene00012960, WBGene00012961, WBGene00013010, WBGene00013018, WBGene00013019, WBGene00013020, WBGene00013021, WBGene00013023, WBGene00013121, WBGene00013122, WBGene00013123, WBGene00013124, WBGene00013125, WBGene00013126, WBGene00013127, WBGene00013128, WBGene00013129, WBGene00013130, WBGene00013131, WBGene00013134, WBGene00013135, WBGene00013136, WBGene00013137, WBGene00013138, WBGene00013177, WBGene00013178, WBGene00013179, WBGene00013180, WBGene00013181, WBGene00013182, WBGene00013184, WBGene00013186, WBGene00013197, WBGene00013198, WBGene00013200, WBGene00013201, WBGene00013202, WBGene00013203, WBGene00013204, WBGene00013386, WBGene00013403, WBGene00013404, WBGene00013405, WBGene00013406, WBGene00013407, WBGene00013408, WBGene00013409, WBGene00013410, WBGene00013449, WBGene00013500, WBGene00013501, WBGene00013502, WBGene00013503, WBGene00013505, WBGene00013506, WBGene00013507, WBGene00013508, WBGene00013509, WBGene00013510, WBGene00013511, WBGene00013512, WBGene00013513, WBGene00013593, WBGene00013594, WBGene00013595, WBGene00013596, WBGene00013597, WBGene00013598, WBGene00013599, WBGene00013601, WBGene00013602, WBGene00013603, WBGene00013604, WBGene00013605, WBGene00013666, WBGene00013667, WBGene00013668, WBGene00013669, WBGene00013670, WBGene00013671, WBGene00013672, WBGene00013673, WBGene00013674, WBGene00013676, WBGene00013677, WBGene00013678, WBGene00013679, WBGene00013680, WBGene00013682, WBGene00013683, WBGene00013684, WBGene00013685, WBGene00013687, WBGene00013688, WBGene00013689, WBGene00013690, WBGene00013691, WBGene00013692, WBGene00013693, WBGene00013694, WBGene00013695, WBGene00013696, WBGene00013697, WBGene00013698, WBGene00013699, WBGene00013700, WBGene00013701, WBGene00013702, WBGene00013703, WBGene00013704, WBGene00013705, WBGene00013706, WBGene00013707, WBGene00013708, WBGene00013709, WBGene00013710, WBGene00013711, WBGene00013712, WBGene00013713, WBGene00013716, WBGene00013717, WBGene00013718, WBGene00013719, WBGene00013720, WBGene00013721, WBGene00013722, WBGene00013723, WBGene00013724, WBGene00013725, WBGene00013726, WBGene00013859, WBGene00013860, WBGene00013865, WBGene00013891, WBGene00013892, WBGene00013893, WBGene00013894, WBGene00013895, WBGene00013918, WBGene00013927, WBGene00013928, WBGene00013929, WBGene00013930, WBGene00013931, WBGene00013932, WBGene00013933, WBGene00013955, WBGene00013956, WBGene00013957, WBGene00013958, WBGene00013959, WBGene00013960, WBGene00013969, WBGene00013970, WBGene00013993, WBGene00013994, WBGene00014026, WBGene00014100, WBGene00014101, WBGene00014102, WBGene00014103, WBGene00014104, WBGene00014115, WBGene00014116, WBGene00014117, WBGene00014118, WBGene00014119, WBGene00014120, WBGene00014121, WBGene00014126, WBGene00014147, WBGene00014148, WBGene00014149, WBGene00014150, WBGene00014182, WBGene00014183, WBGene00014184, WBGene00014185, WBGene00014186, WBGene00014188, WBGene00014189, WBGene00014196, WBGene00014197, WBGene00014198, WBGene00014199, WBGene00014200, WBGene00014201, WBGene00014235, WBGene00014236, WBGene00014238, WBGene00014239, WBGene00014279, WBGene00014280, WBGene00014281, WBGene00014282, WBGene00014288, WBGene00014301, WBGene00014325, WBGene00014326, WBGene00014327, WBGene00014328, WBGene00014332, WBGene00014335, WBGene00014336, WBGene00014337, WBGene00014343, WBGene00014344, WBGene00014345, WBGene00014349, WBGene00014363, WBGene00014389, WBGene00014403, WBGene00014404, WBGene00014405, WBGene00014411, WBGene00014418, WBGene00014424, WBGene00014446, WBGene00014483, WBGene00014498, WBGene00014500, WBGene00014501, WBGene00014502, WBGene00014503, WBGene00014510, WBGene00014525, WBGene00014526, WBGene00014527, WBGene00014528, WBGene00014529, WBGene00014533, WBGene00014542, WBGene00014545, WBGene00014577, WBGene00014585, WBGene00014586, WBGene00014587, WBGene00014605, WBGene00014612, WBGene00014620, WBGene00014664, WBGene00014698, WBGene00014699, WBGene00014700, WBGene00014714, WBGene00014727, WBGene00014729, WBGene00014731, WBGene00014746, WBGene00014748, WBGene00014759, WBGene00014760, WBGene00014761, WBGene00014789, WBGene00014806, WBGene00014807, WBGene00014820, WBGene00014852, WBGene00014853, WBGene00014858, WBGene00014861, WBGene00014864, WBGene00014870, WBGene00014871, WBGene00014914, WBGene00014915, WBGene00014916, WBGene00014917, WBGene00014918, WBGene00014921, WBGene00014947, WBGene00014964, WBGene00014965, WBGene00014966, WBGene00014992, WBGene00014994, WBGene00015001, WBGene00015006, WBGene00015007, WBGene00015008, WBGene00015009, WBGene00015010, WBGene00015011, WBGene00015017, WBGene00015018, WBGene00015019, WBGene00015020, WBGene00015021, WBGene00015022, WBGene00015023, WBGene00015024, WBGene00015025, WBGene00015026, WBGene00015027, WBGene00015028, WBGene00015029, WBGene00015030, WBGene00015032, WBGene00015033, WBGene00015034, WBGene00015091, WBGene00015092, WBGene00015093, WBGene00015094, WBGene00015095, WBGene00015096, WBGene00015175, WBGene00015176, WBGene00015228, WBGene00015229, WBGene00015230, WBGene00015231, WBGene00015232, WBGene00015233, WBGene00015235, WBGene00015236, WBGene00015237, WBGene00015238, WBGene00015303, WBGene00015312, WBGene00015313, WBGene00015442, WBGene00015499, WBGene00015500, WBGene00015501, WBGene00015504, WBGene00015505, WBGene00015570, WBGene00015631, WBGene00015632, WBGene00015633, WBGene00015634, WBGene00015635, WBGene00015636, WBGene00015685, WBGene00015686, WBGene00015687, WBGene00015688, WBGene00015689, WBGene00015690, WBGene00015691, WBGene00015692, WBGene00015693, WBGene00015694, WBGene00015695, WBGene00015696, WBGene00015697, WBGene00015698, WBGene00015704, WBGene00015907, WBGene00015908, WBGene00015970, WBGene00015971, WBGene00015972, WBGene00015974, WBGene00015975, WBGene00015976, WBGene00016040, WBGene00016063, WBGene00016064, WBGene00016066, WBGene00016154, WBGene00016155, WBGene00016156, WBGene00016157, WBGene00016158, WBGene00016159, WBGene00016258, WBGene00016259, WBGene00016260, WBGene00016261, WBGene00016262, WBGene00016263, WBGene00016264, WBGene00016265, WBGene00016266, WBGene00016278, WBGene00016320, WBGene00016321, WBGene00016322, WBGene00016323, WBGene00016324, WBGene00016325, WBGene00016326, WBGene00016327, WBGene00016328, WBGene00016329, WBGene00016330, WBGene00016331, WBGene00016332, WBGene00016333, WBGene00016386, WBGene00016387, WBGene00016388, WBGene00016389, WBGene00016390, WBGene00016391, WBGene00016392, WBGene00016393, WBGene00016394, WBGene00016395, WBGene00016396, WBGene00016418, WBGene00016419, WBGene00016420, WBGene00016421, WBGene00016422, WBGene00016453, WBGene00016454, WBGene00016455, WBGene00016456, WBGene00016457, WBGene00016458, WBGene00016459, WBGene00016460, WBGene00016461, WBGene00016462, WBGene00016463, WBGene00016491, WBGene00016492, WBGene00016493, WBGene00016494, WBGene00016561, WBGene00016562, WBGene00016564, WBGene00016565, WBGene00016566, WBGene00016567, WBGene00016601, WBGene00016602, WBGene00016603, WBGene00016605, WBGene00016606, WBGene00016607, WBGene00016608, WBGene00016609, WBGene00016621, WBGene00016622, WBGene00016650, WBGene00016652, WBGene00016653, WBGene00016654, WBGene00016655, WBGene00016663, WBGene00016727, WBGene00016728, WBGene00016729, WBGene00016730, WBGene00016731, WBGene00016732, WBGene00016733, WBGene00016734, WBGene00016735, WBGene00016740, WBGene00016741, WBGene00016742, WBGene00016743, WBGene00016744, WBGene00016745, WBGene00016746, WBGene00016749, WBGene00016750, WBGene00016751, WBGene00016752, WBGene00016753, WBGene00016754, WBGene00016835, WBGene00016836, WBGene00016837, WBGene00016838, WBGene00016839, WBGene00016840, WBGene00016841, WBGene00016902, WBGene00016903, WBGene00016905, WBGene00016906, WBGene00016907, WBGene00016930, WBGene00016931, WBGene00016932, WBGene00016942, WBGene00016943, WBGene00016944, WBGene00016945, WBGene00016946, WBGene00016947, WBGene00016948, WBGene00016949, WBGene00016950, WBGene00017001, WBGene00017002, WBGene00017003, WBGene00017004, WBGene00017005, WBGene00017007, WBGene00017009, WBGene00017010, WBGene00017011, WBGene00017025, WBGene00017026, WBGene00017063, WBGene00017064, WBGene00017065, WBGene00017066, WBGene00017068, WBGene00017084, WBGene00017085, WBGene00017087, WBGene00017088, WBGene00017089, WBGene00017090, WBGene00017098, WBGene00017239, WBGene00017240, WBGene00017398, WBGene00017399, WBGene00017650, WBGene00017651, WBGene00017671, WBGene00017672, WBGene00017673, WBGene00017674, WBGene00017675, WBGene00017676, WBGene00017731, WBGene00017732, WBGene00017733, WBGene00017734, WBGene00017735, WBGene00017736, WBGene00017737, WBGene00017738, WBGene00017739, WBGene00017741, WBGene00017795, WBGene00017815, WBGene00017816, WBGene00017817, WBGene00017851, WBGene00017852, WBGene00017853, WBGene00017854, WBGene00017855, WBGene00017858, WBGene00017860, WBGene00017885, WBGene00017886, WBGene00017887, WBGene00017888, WBGene00017889, WBGene00017908, WBGene00017909, WBGene00017910, WBGene00017975, WBGene00017976, WBGene00017977, WBGene00017978, WBGene00017979, WBGene00017980, WBGene00018000, WBGene00018001, WBGene00018002, WBGene00018003, WBGene00018004, WBGene00018005, WBGene00018006, WBGene00018007, WBGene00018008, WBGene00018009, WBGene00018156, WBGene00018157, WBGene00018158, WBGene00018159, WBGene00018230, WBGene00018231, WBGene00018232, WBGene00018233, WBGene00018508, WBGene00018509, WBGene00018510, WBGene00018511, WBGene00018512, WBGene00018513, WBGene00018525, WBGene00018526, WBGene00018527, WBGene00018528, WBGene00018529, WBGene00018530, WBGene00018531, WBGene00018579, WBGene00018580, WBGene00018582, WBGene00018594, WBGene00018595, WBGene00018596, WBGene00018598, WBGene00018599, WBGene00018600, WBGene00018601, WBGene00018625, WBGene00018626, WBGene00018627, WBGene00018628, WBGene00018630, WBGene00018631, WBGene00018632, WBGene00018763, WBGene00018764, WBGene00018765, WBGene00018771, WBGene00018772, WBGene00018773, WBGene00018774, WBGene00018786, WBGene00018787, WBGene00018788, WBGene00018789, WBGene00018791, WBGene00018792, WBGene00018797, WBGene00018798, WBGene00018847, WBGene00018848, WBGene00018849, WBGene00018850, WBGene00018851, WBGene00018852, WBGene00018862, WBGene00018863, WBGene00018864, WBGene00018865, WBGene00018866, WBGene00018867, WBGene00018868, WBGene00018890, WBGene00018891, WBGene00018892, WBGene00018893, WBGene00018894, WBGene00018895, WBGene00018896, WBGene00018897, WBGene00018908, WBGene00018909, WBGene00018921, WBGene00018923, WBGene00018955, WBGene00018957, WBGene00018958, WBGene00018978, WBGene00018979, WBGene00018980, WBGene00018981, WBGene00018982, WBGene00018993, WBGene00018994, WBGene00018995, WBGene00019001, WBGene00019002, WBGene00019003, WBGene00019004, WBGene00019005, WBGene00019006, WBGene00019007, WBGene00019008, WBGene00019009, WBGene00019010, WBGene00019011, WBGene00019012, WBGene00019013, WBGene00019074, WBGene00019075, WBGene00019076, WBGene00019077, WBGene00019079, WBGene00019080, WBGene00019081, WBGene00019082, WBGene00019245, WBGene00019246, WBGene00019247, WBGene00019248, WBGene00019253, WBGene00019254, WBGene00019322, WBGene00019323, WBGene00019324, WBGene00019360, WBGene00019361, WBGene00019362, WBGene00019394, WBGene00019395, WBGene00019396, WBGene00019399, WBGene00019429, WBGene00019430, WBGene00019431, WBGene00019432, WBGene00019433, WBGene00019434, WBGene00019461, WBGene00019594, WBGene00019595, WBGene00019597, WBGene00019598, WBGene00019673, WBGene00019674, WBGene00019692, WBGene00019693, WBGene00019697, WBGene00019698, WBGene00019706, WBGene00019710, WBGene00019711, WBGene00019712, WBGene00019713, WBGene00019767, WBGene00019768, WBGene00019769, WBGene00019770, WBGene00019771, WBGene00019909, WBGene00019911, WBGene00019992, WBGene00019993, WBGene00020030, WBGene00020031, WBGene00020032, WBGene00020033, WBGene00020034, WBGene00020035, WBGene00020036, WBGene00020037, WBGene00020038, WBGene00020039, WBGene00020040, WBGene00020087, WBGene00020088, WBGene00020089, WBGene00020090, WBGene00020091, WBGene00020131, WBGene00020184, WBGene00020185, WBGene00020187, WBGene00020188, WBGene00020189, WBGene00020190, WBGene00020191, WBGene00020192, WBGene00020193, WBGene00020212, WBGene00020258, WBGene00020259, WBGene00020260, WBGene00020261, WBGene00020262, WBGene00020263, WBGene00020281, WBGene00020282, WBGene00020283, WBGene00020345, WBGene00020346, WBGene00020347, WBGene00020348, WBGene00020349, WBGene00020350, WBGene00020375, WBGene00020376, WBGene00020377, WBGene00020378, WBGene00020379, WBGene00020380, WBGene00020381, WBGene00020382, WBGene00020383, WBGene00020397, WBGene00020398, WBGene00020399, WBGene00020400, WBGene00020401, WBGene00020402, WBGene00020403, WBGene00020404, WBGene00020411, WBGene00020412, WBGene00020413, WBGene00020414, WBGene00020415, WBGene00020416, WBGene00020417, WBGene00020418, WBGene00020419, WBGene00020466, WBGene00020467, WBGene00020556, WBGene00020557, WBGene00020558, WBGene00020625, WBGene00020626, WBGene00020627, WBGene00020628, WBGene00020629, WBGene00020652, WBGene00020653, WBGene00020655, WBGene00020656, WBGene00020659, WBGene00020660, WBGene00020661, WBGene00020693, WBGene00020694, WBGene00020709, WBGene00020710, WBGene00020711, WBGene00020712, WBGene00020713, WBGene00020714, WBGene00020742, WBGene00020743, WBGene00020838, WBGene00020839, WBGene00020840, WBGene00020841, WBGene00020842, WBGene00020843, WBGene00020856, WBGene00020857, WBGene00020858, WBGene00020859, WBGene00020860, WBGene00020900, WBGene00020901, WBGene00020902, WBGene00020903, WBGene00020913, WBGene00020914, WBGene00020915, WBGene00020917, WBGene00020931, WBGene00020932, WBGene00020933, WBGene00020935, WBGene00020936, WBGene00020937, WBGene00020984, WBGene00020985, WBGene00020986, WBGene00020987, WBGene00020988, WBGene00020989, WBGene00020990, WBGene00020991, WBGene00021005, WBGene00021006, WBGene00021007, WBGene00021008, WBGene00021009, WBGene00021014, WBGene00021015, WBGene00021016, WBGene00021017, WBGene00021024, WBGene00021025, WBGene00021026, WBGene00021027, WBGene00021028, WBGene00021034, WBGene00021035, WBGene00021036, WBGene00021037, WBGene00021038, WBGene00021039, WBGene00021109, WBGene00021110, WBGene00021111, WBGene00021112, WBGene00021113, WBGene00021114, WBGene00021115, WBGene00021125, WBGene00021127, WBGene00021128, WBGene00021175, WBGene00021207, WBGene00021208, WBGene00021209, WBGene00021210, WBGene00021212, WBGene00021213, WBGene00021214, WBGene00021215, WBGene00021237, WBGene00021238, WBGene00021239, WBGene00021269, WBGene00021270, WBGene00021271, WBGene00021272, WBGene00021273, WBGene00021274, WBGene00021275, WBGene00021276, WBGene00021277, WBGene00021278, WBGene00021279, WBGene00021280, WBGene00021327, WBGene00021328, WBGene00021329, WBGene00021331, WBGene00021332, WBGene00021333, WBGene00021334, WBGene00021335, WBGene00021346, WBGene00021347, WBGene00021348, WBGene00021349, WBGene00021350, WBGene00021351, WBGene00021352, WBGene00021353, WBGene00021354, WBGene00021355, WBGene00021382, WBGene00021383, WBGene00021384, WBGene00021385, WBGene00021386, WBGene00021387, WBGene00021388, WBGene00021458, WBGene00021460, WBGene00021461, WBGene00021463, WBGene00021464, WBGene00021465, WBGene00021466, WBGene00021467, WBGene00021468, WBGene00021469, WBGene00021470, WBGene00021471, WBGene00021472, WBGene00021473, WBGene00021474, WBGene00021475, WBGene00021476, WBGene00021477, WBGene00021547, WBGene00021548, WBGene00021549, WBGene00021551, WBGene00021552, WBGene00021553, WBGene00021628, WBGene00021629, WBGene00021630, WBGene00021631, WBGene00021632, WBGene00021633, WBGene00021634, WBGene00021635, WBGene00021636, WBGene00021637, WBGene00021638, WBGene00021639, WBGene00021640, WBGene00021641, WBGene00021643, WBGene00021644, WBGene00021645, WBGene00021646, WBGene00021647, WBGene00021648, WBGene00021649, WBGene00021650, WBGene00021651, WBGene00021653, WBGene00021657, WBGene00021658, WBGene00021660, WBGene00021661, WBGene00021662, WBGene00021664, WBGene00021666, WBGene00021667, WBGene00021668, WBGene00021671, WBGene00021672, WBGene00021673, WBGene00021675, WBGene00021676, WBGene00021677, WBGene00021678, WBGene00021679, WBGene00021681, WBGene00021682, WBGene00021683, WBGene00021685, WBGene00021686, WBGene00021687, WBGene00021689, WBGene00021690, WBGene00021691, WBGene00021692, WBGene00021693, WBGene00021694, WBGene00021733, WBGene00021734, WBGene00021763, WBGene00021764, WBGene00021765, WBGene00021766, WBGene00021768, WBGene00021771, WBGene00021826, WBGene00021827, WBGene00021828, WBGene00021829, WBGene00021830, WBGene00021831, WBGene00021832, WBGene00021833, WBGene00021834, WBGene00021835, WBGene00021838, WBGene00021839, WBGene00021840, WBGene00021841, WBGene00021842, WBGene00021843, WBGene00021844, WBGene00021845, WBGene00021846, WBGene00022025, WBGene00022026, WBGene00022027, WBGene00022028, WBGene00022029, WBGene00022030, WBGene00022031, WBGene00022032, WBGene00022033, WBGene00022034, WBGene00022035, WBGene00022036, WBGene00022037, WBGene00022038, WBGene00022039, WBGene00022040, WBGene00022042, WBGene00022043, WBGene00022107, WBGene00022108, WBGene00022110, WBGene00022111, WBGene00022112, WBGene00022113, WBGene00022114, WBGene00022115, WBGene00022117, WBGene00022118, WBGene00022119, WBGene00022121, WBGene00022122, WBGene00022123, WBGene00022124, WBGene00022125, WBGene00022126, WBGene00022127, WBGene00022128, WBGene00022129, WBGene00022130, WBGene00022131, WBGene00022133, WBGene00022138, WBGene00022139, WBGene00022140, WBGene00022141, WBGene00022142, WBGene00022143, WBGene00022144, WBGene00022145, WBGene00022146, WBGene00022148, WBGene00022149, WBGene00022150, WBGene00022152, WBGene00022153, WBGene00022154, WBGene00022155, WBGene00022156, WBGene00022157, WBGene00022158, WBGene00022159, WBGene00022160, WBGene00022161, WBGene00022162, WBGene00022163, WBGene00022268, WBGene00022269, WBGene00022270, WBGene00022271, WBGene00022272, WBGene00022273, WBGene00022274, WBGene00022275, WBGene00022276, WBGene00022277, WBGene00022278, WBGene00022279, WBGene00022280, WBGene00022281, WBGene00022282, WBGene00022358, WBGene00022360, WBGene00022361, WBGene00022363, WBGene00022365, WBGene00022366, WBGene00022367, WBGene00022368, WBGene00022369, WBGene00022370, WBGene00022371, WBGene00022372, WBGene00022373, WBGene00022385, WBGene00022386, WBGene00022388, WBGene00022389, WBGene00022390, WBGene00022391, WBGene00022454, WBGene00022455, WBGene00022456, WBGene00022457, WBGene00022458, WBGene00022459, WBGene00022460, WBGene00022462, WBGene00022463, WBGene00022465, WBGene00022466, WBGene00022467, WBGene00022469, WBGene00022470, WBGene00022471, WBGene00022472, WBGene00022473, WBGene00022474, WBGene00022475, WBGene00022517, WBGene00022518, WBGene00022519, WBGene00022586, WBGene00022591, WBGene00022592, WBGene00022593, WBGene00022594, WBGene00022631, WBGene00022632, WBGene00022633, WBGene00022634, WBGene00022635, WBGene00022748, WBGene00022749, WBGene00022750, WBGene00022751, WBGene00022752, WBGene00022753, WBGene00022754, WBGene00022815, WBGene00022830, WBGene00022831, WBGene00022832, WBGene00022833, WBGene00022834, WBGene00022835, WBGene00022836, WBGene00022837, WBGene00022838, WBGene00022946, WBGene00022953, WBGene00022955, WBGene00023023, WBGene00023024, WBGene00023055, WBGene00023057, WBGene00023058, WBGene00023065, WBGene00023066, WBGene00023067, WBGene00023068, WBGene00023072, WBGene00023073, WBGene00023135, WBGene00023136, WBGene00023142, WBGene00023156, WBGene00023169, WBGene00023172, WBGene00023176, WBGene00023193, WBGene00023198, WBGene00023199, WBGene00023273, WBGene00023274, WBGene00023295, WBGene00023314, WBGene00023344, WBGene00023368, WBGene00023369, WBGene00023398, WBGene00023409, WBGene00023411, WBGene00023413, WBGene00023415, WBGene00023452, WBGene00023459, WBGene00023461, WBGene00023472, WBGene00023490, WBGene00043049, WBGene00043056, WBGene00043279, WBGene00043308, WBGene00044008, WBGene00044058, WBGene00044065, WBGene00044066, WBGene00044070, WBGene00044094, WBGene00044120, WBGene00044122, WBGene00044236, WBGene00044237, WBGene00044242, WBGene00044243, WBGene00044245, WBGene00044246, WBGene00044247, WBGene00044250, WBGene00044251, WBGene00044260, WBGene00044261, WBGene00044262, WBGene00044264, WBGene00044265, WBGene00044266, WBGene00044267, WBGene00044270, WBGene00044277, WBGene00044289, WBGene00044305, WBGene00044321, WBGene00044345, WBGene00044347, WBGene00044348, WBGene00044349, WBGene00044358, WBGene00044413, WBGene00044414, WBGene00044421, WBGene00044423, WBGene00044427, WBGene00044431, WBGene00044432, WBGene00044433, WBGene00044434, WBGene00044435, WBGene00044436, WBGene00044437, WBGene00044438, WBGene00044440, WBGene00044441, WBGene00044461, WBGene00044607, WBGene00044610, WBGene00044611, WBGene00044614, WBGene00044616, WBGene00044644, WBGene00044646, WBGene00044679, WBGene00044680, WBGene00044684, WBGene00044706, WBGene00044707, WBGene00044708, WBGene00044709, WBGene00044710, WBGene00044711, WBGene00044712, WBGene00044713, WBGene00044714, WBGene00044716, WBGene00044722, WBGene00044746, WBGene00044754, WBGene00044764, WBGene00044769, WBGene00044781, WBGene00044782, WBGene00044786, WBGene00044792, WBGene00044796, WBGene00044805, WBGene00044809, WBGene00044810, WBGene00044894, WBGene00044895, WBGene00044896, WBGene00044897, WBGene00044898, WBGene00044905, WBGene00044925, WBGene00044926, WBGene00044927, WBGene00044928, WBGene00044929, WBGene00044930, WBGene00044931, WBGene00044932, WBGene00044933, WBGene00044934, WBGene00044935, WBGene00044964, WBGene00044975, WBGene00044982, WBGene00044984, WBGene00044987, WBGene00044988, WBGene00044997, WBGene00045048, WBGene00045052, WBGene00045053, WBGene00045057, WBGene00045058, WBGene00045075, WBGene00045076, WBGene00045077, WBGene00045082, WBGene00045087, WBGene00045093, WBGene00045094, WBGene00045103, WBGene00045138, WBGene00045150, WBGene00045156, WBGene00045162, WBGene00045164, WBGene00045192, WBGene00045212, WBGene00045218, WBGene00045234, WBGene00045235, WBGene00045236, WBGene00045237, WBGene00045238, WBGene00045254, WBGene00045328, WBGene00045329, WBGene00045369, WBGene00045370, WBGene00045375, WBGene00045376, WBGene00045383, WBGene00045387, WBGene00045396, WBGene00045398, WBGene00045399, WBGene00045433, WBGene00045434, WBGene00045458, WBGene00045468, WBGene00045473, WBGene00045474, WBGene00045483, WBGene00045498, WBGene00045508, WBGene00045509, WBGene00050875, WBGene00050878, WBGene00050879, WBGene00050880, WBGene00050881, WBGene00050896, WBGene00050921, WBGene00077443, WBGene00077508, WBGene00077515, WBGene00077519, WBGene00077520, WBGene00077521, WBGene00077524, WBGene00077525, WBGene00077526, WBGene00077527, WBGene00077534, WBGene00077535, WBGene00077537, WBGene00077539, WBGene00077543, WBGene00077546, WBGene00077548, WBGene00077549, WBGene00077551, WBGene00077552, WBGene00077553, WBGene00077555, WBGene00077556, WBGene00077557, WBGene00077558, WBGene00077559, WBGene00077561, WBGene00077563, WBGene00077599, WBGene00077684, WBGene00077685, WBGene00077686, WBGene00077687, WBGene00077689, WBGene00077701, WBGene00077708, WBGene00077715, WBGene00077717, WBGene00077718, WBGene00077719, WBGene00077728, WBGene00077764, WBGene00077765, WBGene00077766, WBGene00077767, WBGene00077768, WBGene00077769, WBGene00077770, WBGene00077771, WBGene00077772, WBGene00077773, WBGene00077781, WBGene00077785, WBGene00077789, WBGene00086551, WBGene00086552, WBGene00086553, WBGene00086554, WBGene00086555, WBGene00086556, WBGene00086557, WBGene00086562, WBGene00086563, WBGene00086564, WBGene00138716, WBGene00165026, WBGene00165095, WBGene00165439, WBGene00165483, WBGene00165567, WBGene00165627, WBGene00166055, WBGene00166087, WBGene00166290, WBGene00166331, WBGene00166332, WBGene00166448, WBGene00166549, WBGene00166632, WBGene00166768, WBGene00166975, WBGene00167006, WBGene00167018, WBGene00167052, WBGene00167193, WBGene00167652, WBGene00167709, WBGene00167738, WBGene00167759, WBGene00167822, WBGene00168081, WBGene00168182, WBGene00168457, WBGene00168505, WBGene00168694, WBGene00168700, WBGene00168707, WBGene00168760, WBGene00168761, WBGene00168902, WBGene00169054, WBGene00169118, WBGene00169222, WBGene00169224, WBGene00169245, WBGene00169408, WBGene00169462, WBGene00169515, WBGene00169521, WBGene00169597, WBGene00169979, WBGene00170108, WBGene00170117, WBGene00170253, WBGene00170280, WBGene00170333, WBGene00170426, WBGene00170676, WBGene00170690, WBGene00170698, WBGene00170789, WBGene00170835, WBGene00170884, WBGene00170927, WBGene00170953, WBGene00170964, WBGene00171005, WBGene00171130, WBGene00171280, WBGene00171346, WBGene00171663, WBGene00171678, WBGene00171818, WBGene00171935, WBGene00172021, WBGene00172093, WBGene00172419, WBGene00172654, WBGene00172703, WBGene00172802, WBGene00173043, WBGene00173107, WBGene00173115, WBGene00173142, WBGene00173238, WBGene00173290, WBGene00173303, WBGene00173337, WBGene00173414, WBGene00173516, WBGene00173569, WBGene00173790, WBGene00173976, WBGene00174000, WBGene00174101, WBGene00174114, WBGene00174124, WBGene00174161, WBGene00174311, WBGene00174649, WBGene00174741, WBGene00174760, WBGene00174770, WBGene00175036, WBGene00185086, WBGene00185098, WBGene00189949, WBGene00189958, WBGene00189974, WBGene00189979, WBGene00189983, WBGene00194652, WBGene00194653, WBGene00194654, WBGene00194656, WBGene00194657, WBGene00194659, WBGene00194660, WBGene00194697, WBGene00194698, WBGene00194700, WBGene00194701, WBGene00194702, WBGene00194703, WBGene00194806, WBGene00194809, WBGene00194810, WBGene00194992, WBGene00194999, WBGene00195031, WBGene00195032, WBGene00195033, WBGene00195045, WBGene00195056, WBGene00195057, WBGene00195068, WBGene00195069, WBGene00195155, WBGene00195175, WBGene00195178, WBGene00195179, WBGene00195180, WBGene00195227, WBGene00195228, WBGene00195229, WBGene00195230, WBGene00195231, WBGene00195232, WBGene00195244, WBGene00195247, WBGene00195253, WBGene00195254, WBGene00195255, WBGene00195257, WBGene00195289, WBGene00195315, WBGene00195320, WBGene00195336, WBGene00195350, WBGene00195352, WBGene00195355, WBGene00195369, WBGene00195394, WBGene00195400, WBGene00195405, WBGene00195417, WBGene00195424, WBGene00195426, WBGene00195443, WBGene00195445, WBGene00195448, WBGene00195455, WBGene00195459, WBGene00195465, WBGene00195476, WBGene00195485, WBGene00195486, WBGene00195504, WBGene00195514, WBGene00195529, WBGene00195544, WBGene00195546, WBGene00195553, WBGene00195560, WBGene00195578, WBGene00195590, WBGene00195592, WBGene00195596, WBGene00195605, WBGene00195606, WBGene00195608, WBGene00195610, WBGene00195613, WBGene00195615, WBGene00195618, WBGene00195645, WBGene00195672, WBGene00195683, WBGene00195719, WBGene00195749, WBGene00195751, WBGene00195752, WBGene00195755, WBGene00195771, WBGene00195776, WBGene00195813, WBGene00195817, WBGene00195833, WBGene00195835, WBGene00195861, WBGene00195883, WBGene00195884, WBGene00195888, WBGene00195909, WBGene00195916, WBGene00195923, WBGene00195924, WBGene00195926, WBGene00195955, WBGene00195962, WBGene00195985, WBGene00195989, WBGene00195998, WBGene00195999, WBGene00196006, WBGene00196011, WBGene00196017, WBGene00196043, WBGene00196062, WBGene00196075, WBGene00196076, WBGene00196097, WBGene00196099, WBGene00196100, WBGene00196101, WBGene00196116, WBGene00196124, WBGene00196129, WBGene00196143, WBGene00196146, WBGene00196187, WBGene00196188, WBGene00196196, WBGene00196204, WBGene00196211, WBGene00196215, WBGene00196219, WBGene00196224, WBGene00196234, WBGene00196235, WBGene00196243, WBGene00196260, WBGene00196276, WBGene00196296, WBGene00196297, WBGene00196342, WBGene00196343, WBGene00196360, WBGene00196367, WBGene00196389, WBGene00196394, WBGene00196396, WBGene00196397, WBGene00196398, WBGene00196406, WBGene00196417, WBGene00196418, WBGene00196421, WBGene00196428, WBGene00196465, WBGene00196468, WBGene00196479, WBGene00196480, WBGene00196493, WBGene00196502, WBGene00196508, WBGene00196517, WBGene00196533, WBGene00196551, WBGene00196566, WBGene00196610, WBGene00196622, WBGene00196636, WBGene00196638, WBGene00196639, WBGene00196658, WBGene00196671, WBGene00196679, WBGene00196703, WBGene00196705, WBGene00196717, WBGene00196718, WBGene00196720, WBGene00196721, WBGene00196743, WBGene00196754, WBGene00196756, WBGene00196768, WBGene00196780, WBGene00196781, WBGene00196791, WBGene00196796, WBGene00196802, WBGene00196824, WBGene00196837, WBGene00196879, WBGene00196880, WBGene00196881, WBGene00196890, WBGene00196897, WBGene00196903, WBGene00196907, WBGene00196917, WBGene00196934, WBGene00196963, WBGene00196978, WBGene00196993, WBGene00197005, WBGene00197014, WBGene00197026, WBGene00197032, WBGene00197034, WBGene00197035, WBGene00197044, WBGene00197059, WBGene00197062, WBGene00197063, WBGene00197076, WBGene00197084, WBGene00197112, WBGene00197121, WBGene00197138, WBGene00197140, WBGene00197142, WBGene00197150, WBGene00197156, WBGene00197162, WBGene00197166, WBGene00197167, WBGene00197177, WBGene00197182, WBGene00197193, WBGene00197229, WBGene00197232, WBGene00197234, WBGene00197258, WBGene00197259, WBGene00197263, WBGene00197264, WBGene00197271, WBGene00197277, WBGene00197303, WBGene00197305, WBGene00197324, WBGene00197341, WBGene00197353, WBGene00197361, WBGene00197364, WBGene00197378, WBGene00197411, WBGene00197432, WBGene00197441, WBGene00197447, WBGene00197457, WBGene00197503, WBGene00197505, WBGene00197508, WBGene00197519, WBGene00197526, WBGene00197538, WBGene00197549, WBGene00197559, WBGene00197565, WBGene00197570, WBGene00197585, WBGene00197595, WBGene00197613, WBGene00197618, WBGene00197627, WBGene00197673, WBGene00197693, WBGene00197696, WBGene00197704, WBGene00197712, WBGene00197713, WBGene00197714, WBGene00197716, WBGene00197725, WBGene00197736, WBGene00197738, WBGene00197757, WBGene00197758, WBGene00197772, WBGene00197776, WBGene00197781, WBGene00197785, WBGene00197788, WBGene00197789, WBGene00197794, WBGene00197823, WBGene00197825, WBGene00197864, WBGene00197871, WBGene00197872, WBGene00197888, WBGene00197900, WBGene00197940, WBGene00197941, WBGene00197950, WBGene00197957, WBGene00197961, WBGene00197965, WBGene00197971, WBGene00197978, WBGene00197993, WBGene00197996, WBGene00198001, WBGene00198011, WBGene00198026, WBGene00198042, WBGene00198047, WBGene00198052, WBGene00198059, WBGene00198061, WBGene00198066, WBGene00198071, WBGene00198089, WBGene00198093, WBGene00198097, WBGene00198101, WBGene00198102, WBGene00198120, WBGene00198131, WBGene00198144, WBGene00198146, WBGene00198151, WBGene00198170, WBGene00198187, WBGene00198190, WBGene00198202, WBGene00198205, WBGene00198207, WBGene00198208, WBGene00198214, WBGene00198221, WBGene00198238, WBGene00198245, WBGene00198267, WBGene00198277, WBGene00198280, WBGene00198315, WBGene00198317, WBGene00198324, WBGene00198328, WBGene00198335, WBGene00198369, WBGene00198375, WBGene00198377, WBGene00198389, WBGene00198391, WBGene00198411, WBGene00198423, WBGene00198482, WBGene00198497, WBGene00198499, WBGene00198507, WBGene00198540, WBGene00198554, WBGene00198570, WBGene00198585, WBGene00198586, WBGene00198588, WBGene00198609, WBGene00198612, WBGene00198649, WBGene00198651, WBGene00198657, WBGene00198691, WBGene00198708, WBGene00198727, WBGene00198736, WBGene00198744, WBGene00198757, WBGene00198773, WBGene00198788, WBGene00198814, WBGene00198816, WBGene00198823, WBGene00198824, WBGene00198826, WBGene00198838, WBGene00198839, WBGene00198848, WBGene00198876, WBGene00198883, WBGene00198884, WBGene00198893, WBGene00198900, WBGene00198922, WBGene00198941, WBGene00198946, WBGene00198957, WBGene00198984, WBGene00199005, WBGene00199008, WBGene00199010, WBGene00199013, WBGene00199017, WBGene00199036, WBGene00199039, WBGene00199044, WBGene00199061, WBGene00199062, WBGene00199064, WBGene00199070, WBGene00199077, WBGene00199081, WBGene00199088, WBGene00199103, WBGene00199106, WBGene00199113, WBGene00199116, WBGene00199148, WBGene00199160, WBGene00199182, WBGene00199191, WBGene00199192, WBGene00199200, WBGene00199201, WBGene00199211, WBGene00199215, WBGene00199217, WBGene00199244, WBGene00199250, WBGene00199256, WBGene00199257, WBGene00199280, WBGene00199297, WBGene00199300, WBGene00199318, WBGene00199333, WBGene00199339, WBGene00199346, WBGene00199362, WBGene00199394, WBGene00199400, WBGene00199407, WBGene00199408, WBGene00199411, WBGene00199416, WBGene00199417, WBGene00199422, WBGene00199432, WBGene00199436, WBGene00199463, WBGene00199467, WBGene00199469, WBGene00199486, WBGene00199495, WBGene00199505, WBGene00199506, WBGene00199508, WBGene00199511, WBGene00199523, WBGene00199530, WBGene00199531, WBGene00199533, WBGene00199566, WBGene00199582, WBGene00199586, WBGene00199597, WBGene00199602, WBGene00199607, WBGene00199609, WBGene00199622, WBGene00199623, WBGene00199629, WBGene00199635, WBGene00199640, WBGene00199648, WBGene00199650, WBGene00199662, WBGene00199669, WBGene00199670, WBGene00199683, WBGene00199697, WBGene00199710, WBGene00199718, WBGene00199719, WBGene00199734, WBGene00199737, WBGene00199757, WBGene00199760, WBGene00199766, WBGene00199768, WBGene00199775, WBGene00199781, WBGene00199790, WBGene00199799, WBGene00199814, WBGene00199816, WBGene00199825, WBGene00199840, WBGene00199872, WBGene00199873, WBGene00199884, WBGene00199892, WBGene00199899, WBGene00199900, WBGene00199915, WBGene00199923, WBGene00199928, WBGene00199929, WBGene00199930, WBGene00199946, WBGene00199971, WBGene00199978, WBGene00199987, WBGene00199992, WBGene00199998, WBGene00200012, WBGene00200022, WBGene00200024, WBGene00200034, WBGene00200037, WBGene00200046, WBGene00200048, WBGene00200067, WBGene00200084, WBGene00200093, WBGene00200096, WBGene00200098, WBGene00200101, WBGene00200104, WBGene00200117, WBGene00200124, WBGene00200158, WBGene00200159, WBGene00200165, WBGene00200171, WBGene00200203, WBGene00200218, WBGene00200232, WBGene00200243, WBGene00200247, WBGene00200253, WBGene00200254, WBGene00200262, WBGene00200264, WBGene00200265, WBGene00200276, WBGene00200277, WBGene00200286, WBGene00200290, WBGene00200294, WBGene00200300, WBGene00200301, WBGene00200307, WBGene00200310, WBGene00200317, WBGene00200345, WBGene00200353, WBGene00200364, WBGene00200365, WBGene00200387, WBGene00200408, WBGene00200417, WBGene00200422, WBGene00200427, WBGene00200428, WBGene00200441, WBGene00200442, WBGene00200482, WBGene00200484, WBGene00200496, WBGene00200506, WBGene00200507, WBGene00200510, WBGene00200517, WBGene00200519, WBGene00200521, WBGene00200550, WBGene00200556, WBGene00200571, WBGene00200577, WBGene00200606, WBGene00200608, WBGene00200634, WBGene00200644, WBGene00200648, WBGene00200654, WBGene00200656, WBGene00200661, WBGene00200668, WBGene00200669, WBGene00200672, WBGene00200688, WBGene00200693, WBGene00200700, WBGene00200705, WBGene00200707, WBGene00200715, WBGene00200720, WBGene00200726, WBGene00200742, WBGene00200760, WBGene00200766, WBGene00200768, WBGene00200773, WBGene00200786, WBGene00200810, WBGene00200833, WBGene00200840, WBGene00200854, WBGene00200860, WBGene00200878, WBGene00200888, WBGene00200900, WBGene00200910, WBGene00200924, WBGene00200936, WBGene00200953, WBGene00200957, WBGene00200970, WBGene00200985, WBGene00200987, WBGene00200992, WBGene00200994, WBGene00201013, WBGene00201017, WBGene00201018, WBGene00201019, WBGene00201068, WBGene00201075, WBGene00201084, WBGene00201096, WBGene00201101, WBGene00201104, WBGene00201106, WBGene00201107, WBGene00201128, WBGene00201134, WBGene00201147, WBGene00201162, WBGene00201164, WBGene00201165, WBGene00201170, WBGene00201172, WBGene00201181, WBGene00201187, WBGene00201188, WBGene00201190, WBGene00201192, WBGene00201193, WBGene00201202, WBGene00201209, WBGene00201222, WBGene00201225, WBGene00201255, WBGene00201270, WBGene00201274, WBGene00201300, WBGene00201317, WBGene00201318, WBGene00201319, WBGene00201331, WBGene00201336, WBGene00201351, WBGene00201383, WBGene00201393, WBGene00201398, WBGene00201435, WBGene00201436, WBGene00201440, WBGene00201448, WBGene00201449, WBGene00201455, WBGene00201456, WBGene00201458, WBGene00201470, WBGene00201478, WBGene00201484, WBGene00201494, WBGene00201499, WBGene00201508, WBGene00201517, WBGene00201534, WBGene00201569, WBGene00201573, WBGene00201588, WBGene00201589, WBGene00201611, WBGene00201627, WBGene00201662, WBGene00201665, WBGene00201667, WBGene00201669, WBGene00201676, WBGene00201678, WBGene00201682, WBGene00201688, WBGene00201694, WBGene00201710, WBGene00201723, WBGene00201737, WBGene00201753, WBGene00201754, WBGene00201755, WBGene00201765, WBGene00201768, WBGene00201777, WBGene00201782, WBGene00201794, WBGene00201801, WBGene00201816, WBGene00201824, WBGene00201829, WBGene00201869, WBGene00201890, WBGene00201901, WBGene00201916, WBGene00201925, WBGene00201926, WBGene00201927, WBGene00201953, WBGene00201965, WBGene00201973, WBGene00201974, WBGene00201976, WBGene00201989, WBGene00202007, WBGene00202025, WBGene00202027, WBGene00202039, WBGene00202040, WBGene00202055, WBGene00202057, WBGene00202063, WBGene00202085, WBGene00202087, WBGene00202093, WBGene00202110, WBGene00202130, WBGene00202136, WBGene00202137, WBGene00202142, WBGene00202148, WBGene00202152, WBGene00202155, WBGene00202156, WBGene00202162, WBGene00202170, WBGene00202173, WBGene00202175, WBGene00202230, WBGene00202233, WBGene00202237, WBGene00202238, WBGene00202255, WBGene00202259, WBGene00202263, WBGene00202277, WBGene00202295, WBGene00202319, WBGene00202331, WBGene00202346, WBGene00202349, WBGene00202359, WBGene00202365, WBGene00202368, WBGene00202369, WBGene00202384, WBGene00202405, WBGene00202408, WBGene00202420, WBGene00202441, WBGene00202444, WBGene00202449, WBGene00202452, WBGene00202469, WBGene00202489, WBGene00202494, WBGene00206356, WBGene00206357, WBGene00206362, WBGene00206363, WBGene00206393, WBGene00206464, WBGene00206474, WBGene00206497, WBGene00206528, WBGene00206530, WBGene00206532, WBGene00219206, WBGene00219219, WBGene00219227, WBGene00219231, WBGene00219234, WBGene00219253, WBGene00219261, WBGene00219263, WBGene00219275, WBGene00219281, WBGene00219282, WBGene00219283, WBGene00219284, WBGene00219285, WBGene00219286, WBGene00219287, WBGene00219288, WBGene00219310, WBGene00219313, WBGene00219319, WBGene00219422, WBGene00219450, WBGene00219460, WBGene00219498, WBGene00219557, WBGene00219578, WBGene00219580, WBGene00219581, WBGene00219582, WBGene00219588, WBGene00219591, WBGene00219592, WBGene00219598, WBGene00219608, WBGene00219621, WBGene00219622, WBGene00219637, WBGene00219654, WBGene00219663, WBGene00219678, WBGene00219692, WBGene00219703, WBGene00219705, WBGene00219710, WBGene00219712, WBGene00219721, WBGene00219725, WBGene00219727, WBGene00219733, WBGene00219746, WBGene00219749, WBGene00219754, WBGene00219760, WBGene00219766, WBGene00219767, WBGene00219778, WBGene00219779, WBGene00219787, WBGene00219788, WBGene00219789, WBGene00219796, WBGene00219798, WBGene00219811, WBGene00219817, WBGene00219820, WBGene00219821, WBGene00219830, WBGene00219833, WBGene00219834, WBGene00219835, WBGene00219845, WBGene00219857, WBGene00219863, WBGene00219870, WBGene00219871, WBGene00219882, WBGene00219894, WBGene00219895, WBGene00219910, WBGene00219911, WBGene00219918, WBGene00219919, WBGene00219920, WBGene00219923, WBGene00219926, WBGene00219927, WBGene00219929, WBGene00219930, WBGene00219937, WBGene00219940, WBGene00219943, WBGene00219950, WBGene00219951, WBGene00219953, WBGene00219955, WBGene00219972, WBGene00219973, WBGene00219976, WBGene00219983, WBGene00219984, WBGene00219985, WBGene00219987, WBGene00219998, WBGene00220000, WBGene00220004, WBGene00220006, WBGene00220010, WBGene00220015, WBGene00220016, WBGene00220018, WBGene00220019, WBGene00220027, WBGene00220039, WBGene00220041, WBGene00220054, WBGene00220064, WBGene00220065, WBGene00220069, WBGene00220081, WBGene00220085, WBGene00220086, WBGene00220087, WBGene00220089, WBGene00220100, WBGene00220106, WBGene00220112, WBGene00220118, WBGene00220126, WBGene00220127, WBGene00220128, WBGene00220135, WBGene00220136, WBGene00220142, WBGene00220147, WBGene00220148, WBGene00220165, WBGene00220166, WBGene00220167, WBGene00220171, WBGene00220179, WBGene00220188, WBGene00220189, WBGene00220190, WBGene00220191, WBGene00220195, WBGene00220197, WBGene00220198, WBGene00220212, WBGene00220217, WBGene00220225, WBGene00220226, WBGene00220231, WBGene00220232, WBGene00220240, WBGene00220241, WBGene00220243, WBGene00220246, WBGene00220261, WBGene00235158, WBGene00235197, WBGene00235237, WBGene00235254, WBGene00235381, WBGene00235385, WBGene00235387, WBGene00235388, WBGene00235390, WBGene00236789, WBGene00236797, WBGene00236798, WBGene00236800, WBGene00236808, WBGene00236809, WBGene00249808, WBGene00255420, WBGene00255457, WBGene00255549, WBGene00255550, WBGene00255568, WBGene00255576, WBGene00255577, WBGene00255578, WBGene00255584, WBGene00255594, WBGene00255605, WBGene00255606, WBGene00255714, WBGene00268216, WBGene00269382, WBGene00269393, WBGene00269435, WBGene00271567, WBGene00271598, WBGene00271605, WBGene00271632, WBGene00271639, WBGene00271648, WBGene00271649, WBGene00271653, WBGene00271694, WBGene00271776, WBGene00271777, WBGene00271778, WBGene00271779, WBGene00271780, WBGene00271781, WBGene00284845, WBGene00284852, WBGene00284853, WBGene00284854, WBGene00284857, WBGene00284858, WBGene00302974, WBGene00303028, WBGene00303030, WBGene00303032, WBGene00303055, WBGene00303056, WBGene00303057, WBGene00303058, WBGene00303070, WBGene00303078, WBGene00303082, WBGene00303088, WBGene00303090, WBGene00303094, WBGene00303097, WBGene00303103, WBGene00303104, WBGene00304792, WBGene00304811, WBGene00304981, WBGene00304985, WBGene00304992, WBGene00304996, WBGene00305018, WBGene00305047, WBGene00305048, WBGene00305049, WBGene00305050, WBGene00306076, WBGene00306080, WBGene00306081, WBGene00306121&#39;, map=[3746:3909, 4115:10230, 11494:16837, 17483:26781, 27588:33150, 43732:44677, 47460:49860, 49920:54426, 55292:64066, 71424:81156, 81192:91296, 91403:92720, 93021:95312, 96458:108223, 108681:110057, 110550:110771, 111037:113751, 113806:114708, 115655:115676, 115686:117498, 118092:120901, 127284:134270, 134336:137282, 137844:144565, 144810:154855, 160171:169763, 171222:176040, 178231:178291, 178523:182772, 183363:208384, 209280:211738, 212255:215491, 216027:217002, 217927:219137, 219734:221575, 231577:237853, 238615:240250, 240490:245414, 248623:251580, 252106:259157, 263501:263618, 272602:272627, 280113:289185, 291214:305477, 305810:310790, 310980:314989, 316634:318066, 318548:323352, 323854:330987, 332603:337845, 343068:348424, 348459:357236, 357239:363901, 364509:373465, 373570:381902, 382825:394415, 398890:418677, 420921:423713, 444002:461605, 461994:467137, 471134:473890, 477581:483880, 484038:488392, 488606:489905, 489968:502042, 502890:506250, 506472:519646, 520108:521554, 526047:526139, 530214:530326, 530394:530683, 530968:533084, 534454:535344, 535790:536587, 536803:536931, 537121:542204, 550695:556709, 557374:561671, 562735:565184, 565283:566588, 572172:575120, 582210:583729, 585118:585376, 585806:607571, 613289:613379, 614650:629444, 629464:636345, 636803:637852, 638128:653988, 654089:663610, 664386:666902, 667539:683472, 683723:686599, 688256:691332, 692019:694948, 694975:698324, 700277:715875, 717889:736562, 740800:757010, 757297:760447, 760847:761849, 763477:770720, 770891:776369, 779156:779241, 788096:788219, 796996:797295, 801554:801698, 803367:834765, 837995:846763, 853450:882713, 882917:895099, 897285:907370, 907447:911189, 912331:926916, 938482:947710, 948568:957456, 963691:976560, 979124:979274, 989670:992178, 995780:1005640, 1006176:1014455, 1015194:1015742, 1016404:1017176, 1018842:1019157, 1019826:1029910, 1030632:1033436, 1034442:1041047, 1041471:1053826, 1053893:1066899, 1068483:1074343, 1077424:1095796, 1107577:1112303, 1112417:1115564, 1126974:1127057, 1128323:1131765, 1136365:1154388, 1154513:1156020, 1156087:1156454, 1156578:1157487, 1157570:1175809, 1176472:1190830, 1193157:1211376, 1211484:1254719, 1254769:1258169, 1258200:1268155, 1271515:1273033, 1278807:1279042, 1279435:1279484, 1282100:1285435, 1287303:1287959, 1305561:1328648, 1332697:1334077, 1335320:1336867, 1337015:1338216, 1338783:1339773, 1340677:1341265, 1341271:1344692, 1348502:1359643, 1365302:1376409, 1376941:1379341, 1383461:1385057, 1389247:1392759, 1393045:1402975, 1405934:1413100, 1413916:1415807, 1416581:1425201, 1427226:1433695, 1436278:1441334, 1442047:1443944, 1444506:1444641, 1446496:1465358, 1465596:1469593, 1471074:1479923, 1483082:1514417, 1514754:1540926, 1543524:1556196, 1556640:1556661, 1567601:1580208, 1581393:1593674, 1594648:1596216, 1600623:1601457, 1601837:1607219, 1607704:1610041, 1610389:1628486, 1629002:1633498, 1634568:1644115, 1644587:1651457, 1652875:1655459, 1655615:1656763, 1657512:1663331, 1663401:1669792, 1670032:1672742, 1672852:1676438, 1680445:1685203, 1686215:1693675, 1696198:1697676, 1698348:1701128, 1701360:1701808, 1702104:1708553, 1709586:1715564, 1721589:1751055, 1751480:1754410, 1755328:1766400, 1767886:1790157, 1791397:1791491, 1793253:1796424, 1797434:1798722, 1799716:1806996, 1808091:1811835, 1812891:1813581, 1813908:1823043, 1823446:1825967, 1828078:1831929, 1832166:1832421, 1832896:1834763, 1834882:1835438, 1835660:1841965, 1842444:1843227, 1843383:1843959, 1845199:1850523, 1851668:1853045, 1856458:1860578, 1862606:1872306, 1874007:1884196, 1885160:1894910, 1895160:1923240, 1924299:1932260, 1932345:1945220, 1945813:1949520, 1950593:1958527, 1958706:1965066, 1967771:1973882, 1974778:1974991, 1978029:1983412, 1988245:2010525, 2019431:2019523, 2026315:2033748, 2034321:2043805, 2046538:2048913, 2049139:2050177, 2061666:2063998, 2064292:2064686, 2064975:2065481, 2065691:2069575, 2069680:2070607, 2070672:2070891, 2071351:2072547, 2072679:2076925, 2078954:2093796, 2097345:2123443, 2128587:2145971, 2149824:2151826, 2160037:2209644, 2216097:2222313, 2223352:2233099, 2233743:2234957, 2235601:2241252, 2242229:2247223, 2250130:2254908, 2255920:2272866, 2279844:2280800, 2281502:2291205, 2293126:2297328, 2297391:2304397, 2304512:2306790, 2307266:2312129, 2312495:2312958, 2313270:2315992, 2316306:2321817, 2322024:2322596, 2323991:2327329, 2327988:2330154, 2330894:2332652, 2333337:2338693, 2342215:2350536, 2350840:2350945, 2352834:2356238, 2357882:2372985, 2377088:2381851, 2382952:2401492, 2402229:2409999, 2410189:2410261, 2415130:2420353, 2420833:2421816, 2422152:2425185, 2429674:2432588, 2432589:2433615, 2434001:2435691, 2436470:2437084, 2438764:2442958, 2445419:2445983, 2446739:2448369, 2449413:2451366, 2451897:2464618, 2468792:2471935, 2472059:2485391, 2485860:2489787, 2489975:2497381, 2498133:2506259, 2506890:2513710, 2514166:2514311, 2524531:2524552, 2533267:2542114, 2544748:2546195, 2546385:2548399, 2549295:2550543, 2551374:2551590, 2552733:2555408, 2556225:2557451, 2558175:2560189, 2560371:2561830, 2562667:2562997, 2568859:2571059, 2571739:2576020, 2581014:2586124, 2587526:2595294, 2599978:2602650, 2605282:2605605, 2606740:2617360, 2617528:2620430, 2620649:2621859, 2621956:2628577, 2629684:2632320, 2632986:2634561, 2636903:2637591, 2638648:2639419, 2642723:2643778, 2644964:2657245, 2666383:2666468, 2668159:2672438, 2672939:2682797, 2682865:2683011, 2684864:2684997, 2686639:2689429, 2692108:2692399, 2692756:2694084, 2697398:2699999, 2700419:2702158, 2703262:2703458, 2706882:2717549, 2718026:2721475, 2721587:2726724, 2727013:2727545, 2730003:2731453, 2731561:2742528, 2744952:2745118, 2755539:2758204, 2758297:2758405, 2759022:2776104, 2776363:2778719, 2780849:2784271, 2784619:2786077, 2786331:2787514, 2787980:2788418, 2788747:2790343, 2791686:2794976, 2796129:2797464, 2798174:2800878, 2802802:2815830, 2816414:2817689, 2818743:2822134, 2822242:2830311, 2830334:2830483, 2831304:2831674, 2835344:2835622, 2835766:2835984, 2837983:2838050, 2845858:2849903, 2851956:2853903, 2853969:2855732, 2858520:2859913, 2864010:2864061, 2866505:2875291, 2875397:2876308, 2876336:2876554, 2876697:2876843, 2876931:2878523, 2878577:2883454, 2883548:2893718, 2895328:2900860, 2901324:2903006, 2903634:2928134, 2928394:2929628, 2929776:2931361, 2931460:2947106, 2948428:2948692, 2949600:2957910, 2967615:2967740, 2973509:2973591, 2974916:2974970, 2981656:2983076, 2983202:2985178, 2986103:2995914, 2996487:3001382, 3001405:3002847, 3004048:3013448, 3015741:3021531, 3021650:3031092, 3031337:3047794, 3052047:3070150, 3070967:3084876, 3085013:3096266, 3099680:3099769, 3101850:3110781, 3115035:3117628, 3122010:3127661, 3132969:3135588, 3137184:3138447, 3138540:3140584, 3141703:3143175, 3144410:3147793, 3148013:3152003, 3153284:3154469, 3158491:3162391, 3162812:3163585, 3163687:3168057, 3168192:3171230, 3173267:3176319, 3176417:3192857, 3193527:3198838, 3199251:3203229, 3206761:3210272, 3211045:3217267, 3217287:3222955, 3223159:3228104, 3229430:3233286, 3233390:3256077, 3256688:3256764, 3258999:3259089, 3259796:3266762, 3267093:3269112, 3269396:3280883, 3287335:3293171, 3293183:3294356, 3299569:3300516, 3301367:3307935, 3317977:3318584, 3318829:3330673, 3338366:3349914, 3352433:3359487, 3362326:3363492, 3364026:3365857, 3367639:3368550, 3368851:3371444, 3371768:3373834, 3374775:3407189, 3410012:3416368, 3419599:3424211, 3427147:3438543, 3439098:3449175, 3449325:3462174, 3462257:3478517, 3479352:3481715, 3481996:3484500, 3487803:3499706, 3502735:3512733, 3516449:3529962, 3533884:3546910, 3547070:3559807, 3559933:3562851, 3564039:3600735, 3601290:3602908, 3603337:3605744, 3606406:3608345, 3608517:3611844, 3611941:3618317, 3618888:3619693, 3620536:3625570, 3628245:3636485, 3637352:3638824, 3638851:3638924, 3639647:3641347, 3641650:3647082, 3651235:3652702, 3653996:3655265, 3655412:3662183, 3666489:3671700, 3680179:3680927, 3682889:3684086, 3684780:3694021, 3694295:3697551, 3697595:3697616, 3698101:3699300, 3703541:3704956, 3706029:3709435, 3710089:3713201, 3713448:3721157, 3727026:3731179, 3731351:3740485, 3741450:3742647, 3745501:3745522, 3745588:3748145, 3748522:3753958, 3755487:3759039, 3769671:3774500, 3774633:3777948, 3778088:3781690, 3782117:3783294, 3783906:3784226, 3786587:3787496, 3788035:3788896, 3789308:3792797, 3794555:3798068, 3798395:3805739, 3807455:3816809, 3818073:3821496, 3821655:3821990, 3822882:3826514, 3827218:3828126, 3828238:3829327, 3829514:3837807, 3838236:3839673, 3839800:3841561, 3853846:3859292, 3859368:3862733, 3862832:3865938, 3867540:3869109, 3869213:3870114, 3870641:3871267, 3871473:3873402, 3873443:3876716, 3877062:3880743, 3881815:3891598, 3892331:3895271, 3895763:3897397, 3898044:3899802, 3900406:3905208, 3908746:3910544, 3910711:3912482, 3912634:3915308, 3916063:3916391, 3917796:3925162, 3925433:3927891, 3930813:3931740, 3932209:3934813, 3935271:3953991, 3956074:3956137, 3960967:3962249, 3963087:3964177, 3964369:3965314, 3965737:3967279, 3968109:3970010, 3970373:3971302, 3972161:3973703, 3974058:3977424, 3978750:3979538, 3983182:3986581, 3987939:3988699, 3995724:3996402, 3999299:4031739, 4035701:4090885, 4094159:4098330, 4103697:4108167, 4109406:4112385, 4114860:4116370, 4116850:4118164, 4120406:4123036, 4123480:4124764, 4126055:4129310, 4129564:4132224, 4132361:4137074, 4137160:4138952, 4139510:4144583, 4145314:4147121, 4147272:4148292, 4148831:4149415, 4149980:4150638, 4151102:4151761, 4153965:4154452, 4155541:4158718, 4159013:4161113, 4162034:4165025, 4165074:4167056, 4168983:4171589, 4172203:4172315, 4172437:4173624, 4173757:4174774, 4175318:4177271, 4177304:4178447, 4178811:4182129, 4182220:4184992, 4185173:4189956, 4190157:4193129, 4193240:4203303, 4203457:4210341, 4210978:4215768, 4229569:4236115, 4236953:4238619, 4239090:4242727, 4242820:4247582, 4247752:4249415, 4249492:4253170, 4253241:4254139, 4254196:4264699, 4265106:4267967, 4268069:4270287, 4270665:4273336, 4273895:4274742, 4275887:4276967, 4277829:4279622, 4295023:4296741, 4298183:4302245, 4303929:4308244, 4311815:4311888, 4318032:4359092, 4359206:4362548, 4363274:4364077, 4364902:4367907, 4368088:4372138, 4372254:4373018, 4373655:4381814, 4382564:4385218, 4386784:4401907, 4401955:4404611, 4409688:4426188, 4427023:4428042, 4428448:4444349, 4444718:4448461, 4448584:4449774, 4449878:4450927, 4451193:4456805, 4457374:4469460, 4479638:4479659, 4481197:4511126, 4511437:4516089, 4517259:4518491, 4521534:4522813, 4525960:4534805, 4537446:4539729, 4539888:4540026, 4549680:4552785, 4552884:4553121, 4554125:4555322, 4555342:4556826, 4557652:4563759, 4564034:4565982, 4566429:4568398, 4568764:4570027, 4571191:4575038, 4575100:4577204, 4578605:4579740, 4579963:4580601, 4580707:4583810, 4583912:4584505, 4584767:4584932, 4585114:4586176, 4586272:4589450, 4589599:4589899, 4597922:4602666, 4603216:4604196, 4604402:4605380, 4605654:4633347, 4634102:4635615, 4635742:4637328, 4637431:4640882, 4641002:4644069, 4644641:4645659, 4645661:4646963, 4647423:4650777, 4652144:4652906, 4653466:4660153, 4660520:4660713, 4660827:4662445, 4664507:4665348, 4665932:4671738, 4675132:4679448, 4679664:4682856, 4683181:4683865, 4684315:4684414, 4685864:4698575, 4699609:4702866, 4704603:4705885, 4706575:4708883, 4709258:4710712, 4711413:4713576, 4714319:4715483, 4716969:4719609, 4719757:4720658, 4721130:4723982, 4724064:4726771, 4727279:4730366, 4731605:4738102, 4738170:4744321, 4745314:4745455, 4746463:4746903, 4746969:4747116, 4747623:4759603, 4759819:4761321, 4761930:4762359, 4763195:4765313, 4766132:4767657, 4768330:4769063, 4770291:4773070, 4773818:4788358, 4789189:4789592, 4790291:4790642, 4790643:4792483, 4799980:4804396, 4805606:4809674, 4821928:4826639, 4827071:4829791, 4830134:4835622, 4836186:4847972, 4848424:4867825, 4868170:4868280, 4876142:4876275, 4884902:4895516, 4895920:4900974, 4901917:4903264, 4903451:4905173, 4906838:4914047, 4918371:4926486, 4926634:4932178, 4934258:4950546, 4950582:4951481, 4951570:4957667, 4957774:4958309, 4959216:4961526, 4962450:4964390, 4964666:4967089, 4967993:4972226, 4975374:4975508, 4980306:4983959, 4984087:4986739, 4986888:4989454, 4989547:4994675, 4996834:5003202, 5003449:5006057, 5006705:5008050, 5008769:5013624, 5014350:5016260, 5018782:5020358, 5020779:5023210, 5027154:5041160, 5041369:5042092, 5042101:5043064, 5044197:5045720, 5046393:5048003, 5050039:5055000, 5055685:5056039, 5058250:5068831, 5073578:5079303, 5079909:5082279, 5084897:5085028, 5086309:5093903, 5096597:5100707, 5100750:5103476, 5103580:5104911, 5104968:5105103, 5105109:5106048, 5106203:5107625, 5107842:5113752, 5114121:5118852, 5119215:5127201, 5127827:5129611, 5129744:5131065, 5131519:5134105, 5134738:5137971, 5138249:5139679, 5141672:5149226, 5154513:5158500, 5158510:5162232, 5162601:5164884, 5165147:5176654, 5176828:5186236, 5186925:5189240, 5190137:5195243, 5196072:5203504, 5204529:5219815, 5230548:5241914, 5244282:5245402, 5245742:5263424, 5263693:5268380, 5268418:5269485, 5269544:5270941, 5271264:5272041, 5272845:5274270, 5274692:5278708, 5278802:5281519, 5282082:5283944, 5284025:5291604, 5292778:5292837, 5298754:5300331, 5300851:5317392, 5317490:5318124, 5318219:5319301, 5319711:5321151, 5325548:5330230, 5330474:5330495, 5330577:5332169, 5332248:5332780, 5333778:5334919, 5335191:5338027, 5338921:5340266, 5340828:5342045, 5342179:5344724, 5344871:5345013, 5345275:5345425, 5345926:5353774, 5354180:5357029, 5357540:5357785, 5358145:5358293, 5358368:5358507, 5358511:5365174, 5365262:5366880, 5366943:5367986, 5368477:5372643, 5372850:5384508, 5388899:5388920, 5401756:5412400, 5412496:5412640, 5412840:5415629, 5416652:5420731, 5421309:5428457, 5428532:5429908, 5429931:5430054, 5430156:5431450, 5431491:5431512, 5432196:5433051, 5433904:5443026, 5443877:5446855, 5450435:5450456, 5452326:5459523, 5460530:5460551, 5463116:5469639, 5472324:5479438, 5479487:5484661, 5484778:5486354, 5486718:5487150, 5487365:5488241, 5488339:5490631, 5491423:5493263, 5494416:5494474, 5497678:5507023, 5507115:5509177, 5509466:5511705, 5512047:5514556, 5514650:5516025, 5516069:5518567, 5518809:5518898, 5520388:5529763, 5530613:5534649, 5540543:5540564, 5546117:5561689, 5563826:5566269, 5566342:5570085, 5571795:5572949, 5573213:5576749, 5577069:5579194, 5579665:5581993, 5582095:5607529, 5607825:5608841, 5609340:5610270, 5610373:5616402, 5616597:5618551, 5619568:5620669, 5620778:5624475, 5624600:5626288, 5627812:5637621, 5643309:5644057, 5644728:5649656, 5649755:5653648, 5653827:5655857, 5655949:5658712, 5659444:5661273, 5662142:5669737, 5669944:5673352, 5674303:5676609, 5676826:5677646, 5678637:5691503, 5692263:5697500, 5702155:5705553, 5707044:5707124, 5716573:5725809, 5726307:5729310, 5729792:5735253, 5735549:5739421, 5739478:5740283, 5741185:5741622, 5741867:5743571, 5744179:5745617, 5746632:5754306, 5754311:5760799, 5760895:5762435, 5764029:5768450, 5768918:5773627, 5774584:5775610, 5776092:5788611, 5788633:5791752, 5792320:5795114, 5795399:5803734, 5803955:5807291, 5807708:5818132, 5818473:5821134, 5822976:5826148, 5829390:5837392, 5838094:5838850, 5840281:5841940, 5842708:5844050, 5844835:5846466, 5847476:5849424, 5850033:5851893, 5852022:5852962, 5854198:5855040, 5857544:5862557, 5862643:5866431, 5866561:5868139, 5868265:5869954, 5871393:5871512, 5872173:5874343, 5874415:5878958, 5879138:5879554, 5883857:5883989, 5886612:5895973, 5896617:5897405, 5898302:5904695, 5906657:5918157, 5918269:5918445, 5919054:5936957, 5937194:5938881, 5942389:5947640, 5949395:5949448, 5949858:5957013, 5957440:5958457, 5959007:5959532, 5962952:5964183, 5964400:5967211, 5967246:5973326, 5981765:5985871, 5986487:5990589, 5990740:6003379, 6004656:6007595, 6008136:6009524, 6009781:6011586, 6011976:6026877, 6027193:6028370, 6028485:6033458, 6035447:6037884, 6038402:6041267, 6041671:6046018, 6046787:6050451, 6051144:6051282, 6053784:6055821, 6056130:6058424, 6058547:6060953, 6061731:6068040, 6068525:6070122, 6070231:6078896, 6081781:6086006, 6087429:6087729, 6090008:6091385, 6092685:6092811, 6093306:6094437, 6095059:6096589, 6097578:6097791, 6098926:6099731, 6100034:6100055, 6100260:6101789, 6102066:6102396, 6104804:6105754, 6105758:6105820, 6105862:6107715, 6107816:6109655, 6110133:6114549, 6123119:6126875, 6133006:6133129, 6133244:6137977, 6138500:6139211, 6139420:6140883, 6141352:6144375, 6145454:6147102, 6148141:6148836, 6149196:6150491, 6153644:6153753, 6154074:6157440, 6157447:6159264, 6159880:6161665, 6162290:6162387, 6163379:6163599, 6164190:6164264, 6164718:6166539, 6167945:6168642, 6169083:6169195, 6171594:6171658, 6172661:6172796, 6172998:6173139, 6174994:6175138, 6175194:6180840, 6181313:6183856, 6184071:6185368, 6185461:6194897, 6196173:6196194, 6199672:6200193, 6200797:6202076, 6203973:6206927, 6206954:6211155, 6211269:6211885, 6211980:6219912, 6220092:6220764, 6221244:6223115, 6224877:6226099, 6231495:6232129, 6232776:6234280, 6234707:6235391, 6236133:6250197, 6250474:6252876, 6253081:6253584, 6253700:6254947, 6255034:6257620, 6260248:6264520, 6264564:6264709, 6265677:6269069, 6269419:6271533, 6271865:6273785, 6274529:6275291, 6280455:6280583, 6283495:6283672, 6284972:6289340, 6289432:6294068, 6295066:6297636, 6297766:6299729, 6300492:6302298, 6302412:6304296, 6304915:6306384, 6307302:6308636, 6309113:6310808, 6311228:6312416, 6312749:6314624, 6315232:6317262, 6317696:6324019, 6326620:6327968, 6329535:6330181, 6342223:6358411, 6358584:6360457, 6360837:6364172, 6365030:6366224, 6366330:6368078, 6368269:6369596, 6369747:6394350, 6402648:6402806, 6404586:6405705, 6405856:6406936, 6407525:6416812, 6418564:6418840, 6420659:6421889, 6422060:6428942, 6429526:6431586, 6432138:6432212, 6436111:6441192, 6441643:6446465, 6447303:6456984, 6457281:6458479, 6459476:6459596, 6460877:6467519, 6468112:6471922, 6473206:6488977, 6490715:6496109, 6497435:6499560, 6499981:6501410, 6502166:6503909, 6507893:6513830, 6513927:6517507, 6517528:6520918, 6521784:6523442, 6525044:6528689, 6529165:6530099, 6530130:6531963, 6531982:6533071, 6534083:6541062, 6542115:6544966, 6545109:6546593, 6546614:6550436, 6550535:6552625, 6552833:6553379, 6553480:6554290, 6554503:6554709, 6554728:6558665, 6558693:6562346, 6562648:6562669, 6562698:6564451, 6564748:6565852, 6566124:6568954, 6569556:6572600, 6574003:6587902, 6592081:6592231, 6592468:6598927, 6598981:6607520, 6608340:6608517, 6612187:6615990, 6616303:6618560, 6620664:6621295, 6621623:6624918, 6625065:6626788, 6627044:6627917, 6629568:6637429, 6640650:6642636, 6642720:6643509, 6651219:6655430, 6655607:6658371, 6659271:6659356, 6660435:6662315, 6663160:6673830, 6682299:6683740, 6686723:6686776, 6687348:6691671, 6695367:6695424, 6705545:6706404, 6706756:6708919, 6710215:6710543, 6710849:6712431, 6715966:6716990, 6717818:6723059, 6724866:6729376, 6730056:6730959, 6731266:6732276, 6734027:6736161, 6736679:6741176, 6741275:6741876, 6741983:6745329, 6745990:6753031, 6753218:6756348, 6757174:6757233, 6757762:6758178, 6758180:6760179, 6760415:6767797, 6768384:6768545, 6773972:6780151, 6780798:6784397, 6784867:6787139, 6788031:6789002, 6789038:6789738, 6790697:6790838, 6790977:6798161, 6798226:6799608, 6800385:6801973, 6802199:6810571, 6812659:6814249, 6814510:6834699, 6835505:6835631, 6836468:6838140, 6838353:6843343, 6843588:6845142, 6845176:6846959, 6847133:6847308, 6847510:6849923, 6850408:6851253, 6851371:6854140, 6854477:6857536, 6860394:6861997, 6862312:6870951, 6871133:6872142, 6872606:6873757, 6874311:6877038, 6877058:6892143, 6893302:6893447, 6897734:6900619, 6900948:6913579, 6915042:6916593, 6916765:6917561, 6917699:6922973, 6923068:6923693, 6924281:6929665, 6931137:6936928, 6937463:6939672, 6940214:6943077, 6943175:6945358, 6945597:6950499, 6951134:6954252, 6954413:6961139, 6967669:6970145, 6970362:6971888, 6972156:6973696, 6974557:6977071, 6979628:6979773, 6989066:6999706, 7002869:7003481, 7005442:7015223, 7015907:7020797, 7021294:7021584, 7021971:7029076, 7029887:7030001, 7033810:7033950, 7034158:7034366, 7035350:7035371, 7038899:7048832, 7050041:7051712, 7053687:7056385, 7057191:7059351, 7059891:7061254, 7065251:7069186, 7069439:7074721, 7075917:7082040, 7088029:7090062, 7090739:7094723, 7095485:7098400, 7098648:7099866, 7100002:7101775, 7102215:7103363, 7103811:7104136, 7105301:7106926, 7110233:7112443, 7112866:7115055, 7116328:7116655, 7116734:7116830, 7117807:7121545, 7121656:7122194, 7122613:7128246, 7132746:7134370, 7134737:7147578, 7149889:7151844, 7152461:7156786, 7159713:7160917, 7160935:7166523, 7166917:7169089, 7169996:7172192, 7172462:7174124, 7175509:7176442, 7176695:7176842, 7177903:7179430, 7181233:7181382, 7183052:7185650, 7185694:7188298, 7188464:7190581, 7190787:7193920, 7201775:7201829, 7202204:7203533, 7204540:7207717, 7208021:7209866, 7210270:7212522, 7214744:7221299, 7222290:7224481, 7224655:7227765, 7227788:7233121, 7233392:7234501, 7235351:7235372, 7235383:7238792, 7238992:7243967, 7244936:7246811, 7246903:7248689, 7248826:7251282, 7251940:7255223, 7255379:7258700, 7258945:7267057, 7267094:7267226, 7269118:7273008, 7274281:7278573, 7278734:7279623, 7279665:7282924, 7282963:7283762, 7283901:7285772, 7286300:7286321, 7290711:7290859, 7290945:7291013, 7291271:7292991, 7294037:7299983, 7300060:7300913, 7300998:7302975, 7303299:7304690, 7304917:7307012, 7307116:7308626, 7308642:7311915, 7312796:7313525, 7313624:7314279, 7317645:7321085, 7322264:7324415, 7327967:7327988, 7333034:7334533, 7339231:7352235, 7352414:7353703, 7354490:7359100, 7360225:7371534, 7372295:7372529, 7376603:7383197, 7384397:7388795, 7389053:7395918, 7398709:7401023, 7401131:7402761, 7403290:7404374, 7404949:7408511, 7422784:7454116, 7454659:7456953, 7457151:7460756, 7460855:7466429, 7472619:7476946, 7480107:7483505, 7484061:7486293, 7486647:7494736, 7495435:7505642, 7505950:7506693, 7509568:7513728, 7514075:7515126, 7518821:7519001, 7519239:7519339, 7522115:7526716, 7527223:7531715, 7532046:7542159, 7544617:7544730, 7553843:7554553, 7554854:7562349, 7564583:7567719, 7567826:7570553, 7574239:7580833, 7580943:7581628, 7582105:7583333, 7583809:7588367, 7588579:7591206, 7591287:7595372, 7595775:7602221, 7602306:7608245, 7611208:7611315, 7614175:7634143, 7634491:7639785, 7640347:7641767, 7641884:7643328, 7643430:7649777, 7650340:7656338, 7656453:7656474, 7656977:7658291, 7667728:7669900, 7670154:7672714, 7674728:7676402, 7676667:7679046, 7679447:7683045, 7683340:7694210, 7694319:7695232, 7695680:7698222, 7698300:7699284, 7699681:7700958, 7701176:7701348, 7701364:7702053, 7702428:7718639, 7718881:7718958, 7722083:7723116, 7723755:7724656, 7725018:7727112, 7727256:7729092, 7729183:7729878, 7731120:7737995, 7738670:7738691, 7755777:7756353, 7766531:7772231, 7773415:7777655, 7777669:7778011, 7779418:7785798, 7785877:7793614, 7798812:7798947, 7800454:7801822, 7813761:7819413, 7819578:7820478, 7820830:7826378, 7826441:7831629, 7832662:7835453, 7835812:7836967, 7837062:7840022, 7840438:7853640, 7853749:7859417, 7859939:7860105, 7862528:7863569, 7866135:7875554, 7884188:7885769, 7887213:7887335, 7887370:7887467, 7889309:7894864, 7896124:7897725, 7898433:7903334, 7906825:7918341, 7918437:7926503, 7926622:7934577, 7936263:7938487, 7938559:7940728, 7940892:7941994, 7942718:7945551, 7946192:7951593, 7951654:7952140, 7952462:7952648, 7952884:7953125, 7954819:7955081, 7955976:7958474, 7958646:7960158, 7960266:7961474, 7962051:7972443, 7972472:7977078, 7980230:7984758, 7984888:7993161, 7993294:7993579, 7995733:8003179, 8004132:8004187, 8005126:8006448, 8006634:8009116, 8010206:8010366, 8014405:8024422, 8024879:8026525, 8026600:8028752, 8028816:8028996, 8030958:8035696, 8035797:8039831, 8039878:8040526, 8040798:8043278, 8043343:8044175, 8046556:8046749, 8047344:8049340, 8050036:8057029, 8059803:8060478, 8060788:8061967, 8062006:8081671, 8083575:8095610, 8095739:8098444, 8103595:8109228, 8109745:8119316, 8125800:8130039, 8134114:8135051, 8135839:8135955, 8137255:8137276, 8138867:8138888, 8139467:8142537, 8147426:8149651, 8150354:8151371, 8151861:8159798, 8160467:8168711, 8172720:8173491, 8174887:8210703, 8210901:8218324, 8219997:8223954, 8224473:8226297, 8226493:8240637, 8242133:8243572, 8243693:8244963, 8245504:8250948, 8253455:8254926, 8255347:8258939, 8259436:8259498, 8262038:8268990, 8270303:8288124, 8288948:8291005, 8291298:8293640, 8294873:8307566, 8307642:8307733, 8312099:8312120, 8313120:8315432, 8315902:8318250, 8318346:8323987, 8324285:8328859, 8329061:8329647, 8330449:8332784, 8333081:8336727, 8336965:8346023, 8353381:8363858, 8363953:8369562, 8375489:8392758, 8394725:8398626, 8398917:8400504, 8401371:8403939, 8404546:8405968, 8406827:8410804, 8412421:8416197, 8416815:8426092, 8427086:8429832, 8429925:8435865, 8435957:8438608, 8439970:8440488, 8441542:8442452, 8442694:8444969, 8445346:8451001, 8451035:8451254, 8452152:8452312, 8459040:8459331, 8460224:8463015, 8463128:8463149, 8463322:8467283, 8467487:8469351, 8469858:8471852, 8472463:8473481, 8474059:8474767, 8474846:8476288, 8479438:8479492, 8481151:8482538, 8482758:8484414, 8484595:8487062, 8487617:8490938, 8491290:8495833, 8498111:8502604, 8503350:8503449, 8503993:8505566, 8509040:8511315, 8511592:8512505, 8513899:8526621, 8526757:8529271, 8529375:8531342, 8531683:8534527, 8539938:8544409, 8546062:8546725, 8549512:8570998, 8571555:8573749, 8573851:8580205, 8584423:8587507, 8591214:8591978, 8592091:8592928, 8593026:8596225, 8596332:8598551, 8598657:8600875, 8601354:8602035, 8602124:8605162, 8605262:8608631, 8608953:8609101, 8609920:8612604, 8612724:8615193, 8615387:8616948, 8617056:8618936, 8623011:8631202, 8631878:8632030, 8633298:8633403, 8633896:8637943, 8638513:8640236, 8640507:8644399, 8644459:8646304, 8646450:8651671, 8658898:8659967, 8660830:8664445, 8664975:8666497, 8666594:8669797, 8669802:8673623, 8676665:8681325, 8683196:8687217, 8687633:8692622, 8692945:8696301, 8696602:8701973, 8702465:8724279, 8727217:8728487, 8728587:8728708, 8728763:8730054, 8730155:8731722, 8732218:8735717, 8736212:8737038, 8737071:8742548, 8743848:8746985, 8749615:8751358, 8752228:8755680, 8755845:8757603, 8758155:8758996, 8759205:8759788, 8763040:8765631, 8765744:8777518, 8779730:8785415, 8790295:8790456, 8791353:8791426, 8792472:8792618, 8792963:8797295, 8797455:8804697, 8804798:8808081, 8808655:8813300, 8813760:8814979, 8815571:8816193, 8816973:8821362, 8822187:8823636, 8824400:8825603, 8826166:8826670, 8828009:8838719, 8840167:8841212, 8841769:8842309, 8842551:8845481, 8845808:8851116, 8852704:8852765, 8852825:8852917, 8853588:8861278, 8864019:8865549, 8865944:8866624, 8866807:8866973, 8867939:8868296, 8869714:8871531, 8871790:8875257, 8876318:8894137, 8897631:8901435, 8901647:8901794, 8902740:8905447, 8905647:8916805, 8917851:8918001, 8919328:8920504, 8921347:8925306, 8928749:8942292, 8945532:8946244, 8946685:8947201, 8947303:8948711, 8948777:8954390, 8954558:8957394, 8957455:8957650, 8958007:8960882, 8962321:8970081, 8970290:8971031, 8971138:8974585, 8974864:8975116, 8975365:8986540, 8987094:8987231, 8989930:8993683, 8994001:8999575, 9000096:9001742, 9001844:9003963, 9004496:9008708, 9009011:9013885, 9015315:9016179, 9016754:9017994, 9018669:9020304, 9020944:9025608, 9025778:9031232, 9031234:9036695, 9039223:9039351, 9039578:9039723, 9039989:9040126, 9041319:9042007, 9042395:9042578, 9043320:9046575, 9048187:9049349, 9049459:9055457, 9056292:9057824, 9057889:9057999, 9058143:9063160, 9063277:9064403, 9065078:9077325, 9077564:9080024, 9085350:9085848, 9087742:9087883, 9089011:9089228, 9095956:9108140, 9108172:9110918, 9118114:9124472, 9124663:9127101, 9132537:9135753, 9135763:9138365, 9138405:9145632, 9149672:9153252, 9154227:9154812, 9155463:9157462, 9158800:9158986, 9159120:9159522, 9159895:9163266, 9163530:9166952, 9168431:9177488, 9177663:9178703, 9179940:9181119, 9181809:9220728, 9221817:9226390, 9226904:9233560, 9234169:9245926, 9246027:9249358, 9249541:9250495, 9250605:9251828, 9251837:9261669, 9265950:9271058, 9271478:9278689, 9280922:9286576, 9287442:9303928, 9305750:9306844, 9308441:9310221, 9310318:9314891, 9317712:9327279, 9327423:9327495, 9327636:9327708, 9327856:9330928, 9332533:9332649, 9332945:9333043, 9334799:9342496, 9344723:9349576, 9350268:9352700, 9352846:9354612, 9355024:9357079, 9359066:9389982, 9390832:9393095, 9393760:9404039, 9407173:9408282, 9408340:9410094, 9410612:9411454, 9412305:9413222, 9414549:9418885, 9419456:9420813, 9420911:9421992, 9422555:9431120, 9431248:9433828, 9435159:9435327, 9435464:9444227, 9447361:9447421, 9448541:9448669, 9454330:9454383, 9456828:9456977, 9457455:9458842, 9459286:9460550, 9463750:9468898, 9469653:9470637, 9471549:9474801, 9476596:9477181, 9477458:9478151, 9478469:9479091, 9479100:9480912, 9481658:9482275, 9482742:9483591, 9483981:9486540, 9486644:9488029, 9488487:9494555, 9500623:9500823, 9501265:9505768, 9506546:9512236, 9512306:9512416, 9512451:9514637, 9514997:9516776, 9517025:9518736, 9518890:9518911, 9519243:9521275, 9522549:9531132, 9531626:9532869, 9533420:9540966, 9541167:9541223, 9543728:9543951, 9544399:9544549, 9544732:9558099, 9558791:9558874, 9561440:9561573, 9562352:9562437, 9563842:9563927, 9565388:9574377, 9574758:9576326, 9576356:9577889, 9578525:9581267, 9581785:9582583, 9582680:9583533, 9583868:9584735, 9585416:9588153, 9588661:9589244, 9589283:9589931, 9590505:9591825, 9591938:9593103, 9595275:9598363, 9600444:9601015, 9604809:9605179, 9605750:9605844, 9605895:9611534, 9611812:9615082, 9615191:9616333, 9616515:9618107, 9618195:9619760, 9620226:9622377, 9622612:9624724, 9625048:9625796, 9626240:9630848, 9631387:9632401, 9633182:9635945, 9638855:9641736, 9642343:9643834, 9644415:9644728, 9645637:9649238, 9650498:9652946, 9653020:9658023, 9658273:9660080, 9660173:9664915, 9665250:9670726, 9673663:9678557, 9679643:9679718, 9685397:9688653, 9689196:9690557, 9692700:9693781, 9694621:9695035, 9695419:9695440, 9695706:9696762, 9697236:9699566, 9703076:9703323, 9704869:9705517, 9708610:9708753, 9710087:9710745, 9711065:9712719, 9712810:9715232, 9715320:9728034, 9731347:9731631, 9734085:9734369, 9736635:9736785, 9737815:9744146, 9744208:9747923, 9748516:9749921, 9751181:9751336, 9751342:9751469, 9752384:9752527, 9752600:9752742, 9754121:9760787, 9762363:9767846, 9768813:9772565, 9773171:9774012, 9774073:9779542, 9780348:9785724, 9785780:9787624, 9788724:9790050, 9793755:9796675, 9797852:9797912, 9806794:9806960, 9808903:9815907, 9829761:9830569, 9839597:9844809, 9847405:9851791, 9852611:9852799, 9853653:9863495, 9863583:9869064, 9869650:9874680, 9874697:9881504, 9883695:9883767, 9884710:9884731, 9886188:9888224, 9888355:9892223, 9893620:9900817, 9901075:9902950, 9904302:9910545, 9914275:9914497, 9914946:9915096, 9916864:9916947, 9919579:9919708, 9921210:9921930, 9923609:9924280, 9925694:9925715, 9925977:9930885, 9930987:9934290, 9934636:9936022, 9936124:9938301, 9940976:9941032, 9942322:9942758, 9942955:9943165, 9947803:9953391, 9953408:9955873, 9956149:9958357, 9958781:9960095, 9960426:9961977, 9962712:9962855, 9965573:9973177, 9973380:9976315, 9976425:9979662, 9980062:9984794, 9986342:9986413, 9988451:9988763, 9989131:9989515, 9989801:9999928, 10001313:10001534, 10005522:10005544, 10011504:10011836, 10014546:10015495, 10016764:10018121, 10019642:10021090, 10021701:10022872, 10023797:10025838, 10026187:10030059, 10030173:10047977, 10048084:10062026, 10062219:10062354, 10066792:10066869, 10070882:10071900, 10072577:10074399, 10074993:10077749, 10080158:10085765, 10086264:10102012, 10102229:10102353, 10103209:10103353, 10103537:10103626, 10107348:10108778, 10110981:10111916, 10112082:10117124, 10117158:10122171, 10122458:10130215, 10133589:10133671, 10134404:10134524, 10147324:10149888, 10150279:10155105, 10155160:10158884, 10158950:10161883, 10162374:10165132, 10169440:10169580, 10169638:10172603, 10174454:10175591, 10177180:10178915, 10179017:10179747, 10179844:10185520, 10185854:10190896, 10191272:10192832, 10193241:10196997, 10197320:10197596, 10198218:10200440, 10200767:10201808, 10203436:10203743, 10220454:10221252, 10221641:10222778, 10222880:10233386, 10235230:10235407, 10237461:10237682, 10244351:10244462, 10246502:10246653, 10248287:10255327, 10255545:10262250, 10264305:10279234, 10290910:10297927, 10298139:10327870, 10328983:10330400, 10330878:10331808, 10331978:10333082, 10333470:10334590, 10334660:10336130, 10336394:10345654, 10345737:10349369, 10356250:10360898, 10366620:10371697, 10373372:10373862, 10374151:10374172, 10374802:10379634, 10380096:10380986, 10381081:10382025, 10382740:10391388, 10398620:10398759, 10398907:10400177, 10400396:10400619, 10400693:10400840, 10401063:10402762, 10406128:10406272, 10407457:10407585, 10409301:10414773, 10418233:10421858, 10422894:10429034, 10429628:10432371, 10432629:10432761, 10432855:10435876, 10436253:10436888, 10436894:10437779, 10438074:10443563, 10443666:10446043, 10446656:10452075, 10452177:10452895, 10454188:10455678, 10457827:10459047, 10464288:10471056, 10471510:10471927, 10472195:10473389, 10473512:10474435, 10474821:10478129, 10479304:10484969, 10486698:10487341, 10488451:10489585, 10491195:10491339, 10493661:10493885, 10494333:10494548, 10497961:10498106, 10501121:10501264, 10501584:10505949, 10506416:10506638, 10506997:10515586, 10516101:10516428, 10516873:10516950, 10516977:10517150, 10517320:10518055, 10518378:10519258, 10519359:10520761, 10520862:10522027, 10522090:10523996, 10524446:10525509, 10526601:10530095, 10530338:10531547, 10532666:10534484, 10536114:10538088, 10543023:10546429, 10547720:10549876, 10551971:10553047, 10553115:10553913, 10554365:10557838, 10558322:10560006, 10562969:10564903, 10565265:10565443, 10566349:10567300, 10567893:10569017, 10569084:10571920, 10571953:10577147, 10579412:10580799, 10581537:10582615, 10588500:10590465, 10590702:10593968, 10594331:10598489, 10598672:10606621, 10608465:10621435, 10623172:10624361, 10624463:10628387, 10628742:10631674, 10631684:10633150, 10633278:10634078, 10634567:10636028, 10636291:10636441, 10636691:10638966, 10639075:10640233, 10640464:10642281, 10643028:10643705, 10644244:10645617, 10646477:10647841, 10648599:10654237, 10654298:10666945, 10667807:10671357, 10671417:10672614, 10672736:10674864, 10675167:10676811, 10676921:10682374, 10682471:10683048, 10683269:10685587, 10685824:10687058, 10687362:10688060, 10688331:10690770, 10691344:10696783, 10699567:10705131, 10708372:10710378, 10713504:10718201, 10720812:10723496, 10724160:10727386, 10727754:10728808, 10728897:10730233, 10730462:10732424, 10732542:10733666, 10734021:10736615, 10736696:10739914, 10740296:10740395, 10741367:10744415, 10750029:10750293, 10750497:10776703, 10781277:10781400, 10782114:10782890, 10784186:10784831, 10785127:10787863, 10789129:10793152, 10793400:10798059, 10799622:10808240, 10808589:10810489, 10810647:10812923, 10813811:10814937, 10817341:10820843, 10823362:10828421, 10830320:10830446, 10831873:10834814, 10836978:10839012, 10841097:10844666, 10844771:10845972, 10846771:10850103, 10850404:10851268, 10855508:10855799, 10855934:10857689, 10858077:10860800, 10861193:10863559, 10863879:10866495, 10866960:10867014, 10867529:10868455, 10869369:10869474, 10869955:10875050, 10876199:10877741, 10878104:10878326, 10879137:10879572, 10879768:10894237, 10899088:10902220, 10904369:10912192, 10912370:10915684, 10917049:10919139, 10919376:10919777, 10921407:10923927, 10924239:10925750, 10926880:10933706, 10933809:10935559, 10936324:10936741, 10936814:10940521, 10940677:10944203, 10944209:10945687, 10945820:10945893, 10952933:10955204, 10958296:10959906, 10964471:10970177, 10970486:11004283, 11011434:11014745, 11015482:11017537, 11018495:11024593, 11032337:11038827, 11039055:11052497, 11059298:11061435, 11063471:11065939, 11067529:11070012, 11075754:11134667, 11142109:11144135, 11145335:11145925, 11146447:11148318, 11148972:11150303, 11151223:11151975, 11152703:11155369, 11156727:11158833, 11159524:11162661, 11163787:11177728, 11177847:11182063, 11182794:11186206, 11198965:11199146, 11206921:11211905, 11212306:11215094, 11216117:11221995, 11222398:11224163, 11224835:11233332, 11234301:11235176, 11235384:11235594, 11235850:11236784, 11237816:11237980, 11238116:11238227, 11239683:11240478, 11241224:11241970, 11242736:11243524, 11243835:11245003, 11245822:11246765, 11248269:11250018, 11250138:11251602, 11252237:11252981, 11253255:11255927, 11256525:11259233, 11260921:11262120, 11263065:11263665, 11264684:11264987, 11265104:11265407, 11265522:11265714, 11266069:11266474, 11268029:11268409, 11268618:11268958, 11269924:11271289, 11271687:11272385, 11272476:11273482, 11274891:11276346, 11276447:11280018, 11282425:11286561, 11287434:11287511, 11287747:11287818, 11289040:11291977, 11292661:11294164, 11294575:11297206, 11300742:11301466, 11302285:11305218, 11306252:11307450, 11308120:11311053, 11311523:11323110, 11323526:11333154, 11334055:11338112, 11338634:11338836, 11339786:11341521, 11342084:11347288, 11355235:11356370, 11358077:11359305, 11361084:11362946, 11364325:11374334, 11374619:11380127, 11380230:11396662, 11400142:11416433, 11417007:11421269, 11422179:11423600, 11423992:11426054, 11428654:11428747, 11429013:11432156, 11433120:11433422, 11433670:11436018, 11436781:11437407, 11439288:11440709, 11443451:11446644, 11449855:11451972, 11453899:11454604, 11454779:11455539, 11456466:11459953, 11460965:11463074, 11465141:11467612, 11469461:11471584, 11473217:11475662, 11477743:11480071, 11483356:11484640, 11484777:11486598, 11486894:11488471, 11489157:11490578, 11492334:11492773, 11493035:11495080, 11495520:11496245, 11498449:11498715, 11498859:11499655, 11501617:11507172, 11510022:11513853, 11515126:11517108, 11517557:11526797, 11526917:11557851, 11557985:11558683, 11559445:11560565, 11560619:11564387, 11565188:11570871, 11570885:11571166, 11572092:11577488, 11578640:11579612, 11579705:11579778, 11579855:11581989, 11582742:11587337, 11587695:11587757, 11592245:11601712, 11601861:11602018, 11604026:11611314, 11611671:11616244, 11616815:11620263, 11621154:11621273, 11621575:11621899, 11623114:11626562, 11628148:11628355, 11628597:11628666, 11629640:11629932, 11631643:11632918, 11632982:11640414, 11640686:11642206, 11643020:11645420, 11646524:11647020, 11647752:11648428, 11648759:11649206, 11649502:11649687, 11649981:11651825, 11652788:11654313, 11655003:11655123, 11666411:11666470, 11667910:11667969, 11668081:11668145, 11668388:11668438, 11669984:11670094, 11670497:11670549, 11675940:11682441, 11686088:11686948, 11687810:11688522, 11688672:11690836, 11694576:11697943, 11699168:11701178, 11701630:11703698, 11703717:11706612, 11707807:11712197, 11714955:11717616, 11718863:11722936, 11723399:11726326, 11731057:11733478, 11733939:11735964, 11736580:11737699, 11738655:11740762, 11741602:11743966, 11744168:11804991, 11809131:11811266, 11813778:11814451, 11815907:11819280, 11819538:11820051, 11820661:11822646, 11824975:11825787, 11827548:11828459, 11830984:11833183, 11836459:11849258, 11851553:11868105, 11877802:11887278, 11888500:11890558, 11891575:11893727, 11893930:11896284, 11896673:11898521, 11898788:11901476, 11901659:11903310, 11903733:11905148, 11905956:11907484, 11907524:11909280, 11909365:11909991, 11910327:11911021, 11911319:11912677, 11912792:11917712, 11917805:11923915, 11924540:11929516, 11930764:11931757, 11933351:11933416, 11933464:11933575, 11943665:11943815, 11947000:11953154, 11953506:11963085, 11966270:11967591, 11968021:11969847, 11971178:11971802, 11977176:11977276, 11977977:11978084, 11978756:11978847, 11979400:11985619, 11996975:11997030, 11998780:12004947, 12005176:12006296, 12006910:12008040, 12008624:12009625, 12011194:12018684, 12020755:12022676, 12024795:12028298, 12029045:12029504, 12029921:12032203, 12032284:12033737, 12034071:12037079, 12041034:12046618, 12047341:12059798, 12060270:12064897, 12065864:12065937, 12065942:12069084, 12069805:12072904, 12073071:12084225, 12084294:12087856, 12088118:12090062, 12091565:12094726, 12095481:12097987, 12098663:12106112, 12107043:12108503, 12108587:12112505, 12113016:12113596, 12114063:12116290, 12116948:12117128, 12117996:12118384, 12119110:12119586, 12120140:12121333, 12121710:12122060, 12122810:12125586, 12126454:12128965, 12130069:12131381, 12132552:12133639, 12134267:12136465, 12140559:12142757, 12142910:12142982, 12144006:12146977, 12148079:12154333, 12155123:12156391, 12157132:12158733, 12160099:12162091, 12162391:12163703, 12165408:12177782, 12181053:12183254, 12183687:12189008, 12189520:12190224, 12195267:12195334, 12198599:12198685, 12202231:12203720, 12203838:12205809, 12207096:12212689, 12213105:12215198, 12215537:12217772, 12218431:12222298, 12223552:12223605, 12225425:12225497, 12229308:12229380, 12230417:12232303, 12232711:12235193, 12236017:12238252, 12239159:12241808, 12245455:12246778, 12247770:12250210, 12251905:12253636, 12254484:12257180, 12258602:12258803, 12259911:12262848, 12263701:12264276, 12264464:12266420, 12266804:12268128, 12269254:12272190, 12272948:12275342, 12276741:12276807, 12276892:12278775, 12279751:12280434, 12281220:12282615, 12283267:12285397, 12285745:12285932, 12286527:12289316, 12290448:12293936, 12294299:12296348, 12297733:12300817, 12301423:12303157, 12303598:12306820, 12307730:12310431, 12312324:12314962, 12315192:12316688, 12317320:12318735, 12319349:12320854, 12321936:12324062, 12324757:12324945, 12326342:12327025, 12327981:12329494, 12331084:12333454, 12335034:12336605, 12337967:12339239, 12341144:12342826, 12344166:12345702, 12345818:12347265, 12347699:12349819, 12350934:12352724, 12353505:12356337, 12356941:12359341, 12359941:12363416, 12364717:12365950, 12366334:12368619, 12369188:12370607, 12371349:12372647, 12374204:12376135, 12376903:12378674, 12379334:12381032, 12381807:12383259, 12384276:12385731, 12387704:12388791, 12389759:12390979, 12392008:12393312, 12394119:12396278, 12396643:12398267, 12399289:12401206, 12403143:12404489, 12405524:12406946, 12407563:12411435, 12413808:12415745, 12417385:12418800, 12421238:12426850, 12427387:12429094, 12430027:12431500, 12433247:12434944, 12435725:12437095, 12439935:12441253, 12442298:12443780, 12444581:12446512, 12447166:12448240, 12449976:12452034, 12453388:12454764, 12454852:12455826, 12456925:12457174, 12458016:12460903, 12461655:12461895, 12462175:12465816, 12470619:12473424, 12474619:12475934, 12476652:12478594, 12483808:12486235, 12486943:12494538, 12494825:12497357, 12497437:12505010, 12505113:12507569, 12508298:12513420, 12513515:12518694, 12518786:12532135, 12532741:12534404, 12535313:12537671, 12537902:12538986, 12551536:12552375, 12557163:12557543, 12558546:12558944, 12559364:12559658, 12560411:12560672, 12561039:12561560, 12562755:12564210, 12564355:12572801, 12577953:12579226, 12579896:12580957, 12581765:12585139, 12585381:12586006, 12587241:12591091, 12591115:12592351, 12594143:12594204, 12594568:12594657, 12595689:12595773, 12596529:12598916, 12601777:12601845, 12601897:12602957, 12603757:12606248, 12608618:12608708, 12615920:12625238, 12626673:12629082, 12629308:12630701, 12631184:12632555, 12635171:12636542, 12637025:12638418, 12638684:12640622, 12640719:12641501, 12641602:12642187, 12642402:12642900, 12643007:12644321, 12647379:12652087, 12652276:12654655, 12655013:12656760, 12657761:12659446, 12659728:12660647, 12661253:12661732, 12662811:12663833, 12665066:12668893, 12669379:12674449, 12674491:12674544, 12677524:12677597, 12679206:12681812, 12681844:12682702, 12683219:12684386, 12685386:12685569, 12686115:12688206, 12688726:12690572, 12691317:12691762, 12692168:12693724, 12694405:12695934, 12696557:12698214, 12698903:12699200, 12699429:12701031, 12701438:12702886, 12703434:12704727, 12705096:12706983, 12707512:12708691, 12708832:12711596, 12712078:12714730, 12715647:12716583, 12718295:12719837, 12720204:12722889, 12724750:12726206, 12727763:12727929, 12728002:12729898, 12730025:12732722, 12733286:12741275, 12741587:12749562, 12750035:12752105, 12754090:12755126, 12758185:12759577, 12759586:12764960, 12773658:12782257, 12783282:12784281, 12787942:12806581, 12807530:12811814, 12815903:12816558, 12819215:12825602, 12825663:12834163, 12835114:12843331, 12844268:12858281, 12859492:12861293, 12861950:12863544, 12864229:12865350, 12866097:12866911, 12866987:12869576, 12878638:12902806, 12903158:12903449, 12904469:12906040, 12906461:12907754, 12908303:12909436, 12915165:12918131, 12918721:12919280, 12922003:12923548, 12927894:12928456, 12929371:12931115, 12932036:12933753, 12934763:12937362, 12938554:12942227, 12942279:12946588, 12946931:12951099, 12951133:12955322, 12955324:12958051, 12958639:12959932, 12960312:12961619, 12962037:12964579, 12964673:12966795, 12967215:12968280, 12968945:12978787, 12978838:12983974, 12988230:12995475, 12995802:13023624, 13033482:13045541, 13050444:13062122, 13066623:13069046, 13069520:13071190, 13072228:13073057, 13073836:13075381, 13076269:13077020, 13077072:13077941, 13077993:13078880, 13078920:13079690, 13079741:13080633, 13081411:13082166, 13082680:13082966, 13083320:13103119, 13104250:13106465, 13107135:13107453, 13107491:13116434, 13117628:13120324, 13121091:13123513, 13124167:13126650, 13127151:13127876, 13128469:13130387, 13130812:13132466, 13133647:13136156, 13136389:13138544, 13139121:13141898, 13143194:13144623, 13144911:13146453, 13152568:13153734, 13155223:13161508, 13162491:13182726, 13183451:13183772, 13184851:13188192, 13189791:13190583, 13191088:13194645, 13197127:13198376, 13199349:13201037, 13201709:13204293, 13204496:13205403, 13205611:13207835, 13208440:13210462, 13210584:13211352, 13211825:13213806, 13214094:13216230, 13216589:13218852, 13219140:13219464, 13221858:13221978, 13224429:13225473, 13227994:13229701, 13232170:13234686, 13235347:13236663, 13237053:13238304, 13238537:13240733, 13241654:13244730, 13246852:13249778, 13250518:13252902, 13254140:13257729, 13258380:13260217, 13260865:13262733, 13263802:13265570, 13269983:13272781, 13274110:13276464, 13277615:13280175, 13280845:13286504, 13287552:13289542, 13292505:13297558, 13297658:13297746, 13297876:13297985, 13298203:13300716, 13302758:13304311, 13307050:13308685, 13309987:13310711, 13310886:13318716, 13324168:13324254, 13324994:13325066, 13325229:13336146, 13337046:13337212, 13337429:13338931, 13339396:13340224, 13340326:13343952, 13344067:13346963, 13347065:13359643, 13360052:13366888, 13369998:13370080, 13370684:13370819, 13376838:13377349, 13381391:13384371, 13384751:13387084, 13387818:13391205, 13392117:13394001, 13397269:13398631, 13402447:13404106, 13408673:13408697, 13409896:13411554, 13411659:13413172, 13414272:13414465, 13418457:13423377, 13423996:13425148, 13426450:13428471, 13428912:13434611, 13435488:13437963, 13438057:13450218, 13451462:13452845, 13455355:13470371, 13470591:13474185, 13477151:13487368, 13489586:13491120, 13491380:13507068, 13508813:13512024, 13512312:13514290, 13514549:13526247, 13527365:13527963, 13528689:13531462, 13534736:13542395, 13543067:13545188, 13548073:13559361, 13560090:13563465, 13564566:13572047, 13572378:13575805, 13576147:13581611, 13581697:13585295, 13586036:13588831, 13589177:13598505, 13598608:13603393, 13603905:13626225, 13626517:13628447, 13629226:13633492, 13635041:13635958, 13637211:13637277, 13637844:13639311, 13640304:13648405, 13652523:13652727, 13652948:13653704, 13654008:13658167, 13658170:13659128, 13659430:13662332, 13663351:13674997, 13676254:13677792, 13680414:13681366, 13682136:13684547, 13685298:13687586, 13692684:13702954, 13709778:13714128, 13714297:13716916, 13720124:13724453, 13724576:13732046, 13734957:13751124, 13751233:13756354, 13756729:13771773, 13773066:13778784, 13778894:13792249, 13792636:13797272, 13797344:13802223, 13802392:13806801, 13806916:13814322, 13814755:13827988, 13828101:13832147, 13834875:13834941, 13835339:13840047, 13840703:13843996, 13844738:13846957, 13849500:13850893, 13852385:13854047, 13854637:13859699, 13861467:13887640, 13890341:13891406, 13893132:13894249, 13899084:13914008, 13915663:13917809, 13919374:13922122, 13922584:13922659, 13923060:13923763, 13926735:13927556, 13927945:13928792, 13938809:13942016, 13942543:13951630, 13951749:13952052, 13952160:13953203, 13957198:13960612, 13961010:13969571, 13969843:13970877, 13971282:13971419, 13972616:13973412, 13978429:13980130, 13980857:13980954, 13983225:13989668, 13992993:13996221, 13997362:14002972, 14004427:14008484, 14009870:14011531, 14012368:14013629, 14015753:14018267, 14018798:14019759, 14020360:14021312, 14027235:14027588, 14027779:14029212, 14029278:14029364, 14031146:14031667, 14034953:14036057, 14037150:14037777, 14041626:14043478, 14043530:14045694, 14046131:14048022, 14048093:14051010, 14052135:14054154, 14054592:14056894, 14058838:14058910, 14059311:14062314, 14062748:14068803, 14070555:14070619, 14075066:14075163, 14075836:14082582, 14083044:14084207, 14084511:14085525, 14085687:14086685, 14088086:14089858, 14090729:14096348, 14097491:14101090, 14101321:14104470, 14104514:14113108, 14113635:14114085, 14114193:14118650, 14118750:14127202, 14128590:14142207, 14150627:14151067, 14151840:14152282, 14153227:14154195, 14154768:14155296, 14155428:14156336, 14157610:14158078, 14158507:14161664, 14162845:14162939, 14163529:14168307, 14171053:14179360, 14182594:14196295, 14198610:14199136, 14200019:14204687, 14205716:14208726, 14209147:14212501, 14213448:14213515, 14213680:14215850, 14216700:14220624, 14220692:14231360, 14233034:14235948, 14236394:14241181, 14241897:14244452, 14245499:14245555, 14246227:14247314, 14251082:14252416, 14252956:14254513, 14255683:14257016, 14263296:14271321, 14271892:14277653, 14278780:14278828, 14282074:14282140, 14283469:14285325, 14285933:14287089, 14287825:14295954, 14297286:14298325, 14298754:14299780, 14305069:14307220, 14311756:14312842, 14316649:14318522, 14319113:14319913, 14320341:14320592, 14322236:14328965, 14331084:14339839, 14340790:14343815, 14347548:14358326, 14358561:14367594, 14367615:14369773, 14371137:14376086, 14376094:14376115, 14378385:14379425, 14381359:14384737, 14390332:14400226, 14400682:14405649, 14406027:14417934, 14418196:14431913, 14432790:14433750, 14434641:14435706, 14436809:14451889, 14461037:14465145, 14465379:14468214, 14468331:14469103, 14469499:14486767, 14486876:14496560, 14496815:14498295, 14499734:14510552, 14511430:14518719, 14529528:14546870, 14552029:14565853, 14566306:14566783, 14567168:14616523, 14620182:14620637, 14621168:14631094, 14634452:14642568, 14642713:14646601, 14647781:14651288, 14652115:14657275, 14657354:14661495, 14662523:14663141, 14664448:14674841, 14675599:14682566, 14688626:14692144, 14694153:14700392, 14701121:14706522, 14707058:14707875, 14708194:14719027, 14719166:14720477, 14720587:14726131, 14726280:14730738, 14732930:14733669, 14735222:14737195, 14737701:14738796, 14738881:14742671, 14742764:14747789, 14748617:14753777, 14755153:14756412, 14756903:14759473, 14759929:14760626, 14760637:14768989, 14770688:14775007, 14775183:14776663, 14777136:14787804, 14788124:14789517, 14790243:14790934, 14791039:14797167, 14797659:14798705, 14806328:14806473, 14807461:14820235, 14820404:14821651, 14822049:14823996, 14824492:14825952, 14826557:14827779, 14828199:14829594, 14829731:14834841, 14838916:14839138, 14839855:14843128, 14846475:14855827, 14855900:14863482, 14867923:14871365, 14872111:14875436, 14875822:14900322, 14906269:14906399, 14906791:14913083, 14913765:14952226, 14953489:14955492, 14955635:14957097, 14961496:14965361, 14967300:14968540, 14975418:14980596, 14984177:14990016, 14993823:14993959, 14994580:15001691, 15001852:15013904, 15014836:15016595, 15017282:15022990, 15023167:15030708, 15031140:15033039, 15034514:15035467, 15036008:15041274, 15041432:15050700, 15050796:15052420, 15053009:15058994, 15060298:15061159, 15062082:15063836, 15064300:15064453, 15064837:15068355, 15069279:15071033, 15072265:15072426]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))
</pre></div>
</div>
</div>
</div>
<p>Taking the “shadow” of all genes will return the intergenic region as a valid <code class="docutils literal notranslate"><span class="pre">Feature</span></code></p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">intergenic</span> <span class="o">=</span> <span class="n">genes</span><span class="o">.</span><span class="n">shadow</span><span class="p">(</span><span class="s2">&quot;not gene&quot;</span><span class="p">)</span>
<span class="n">intergenic</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>Feature(seqid=&#39;I&#39;, biotype=&#39;not gene&#39;, name=&#39;not gene&#39;, map=[0:3746, 3909:4115, 10230:11494, 16837:17483, 26781:27588, 33150:43732, 44677:47460, 49860:49920, 54426:55292, 64066:71424, 81156:81192, 91296:91403, 92720:93021, 95312:96458, 108223:108681, 110057:110550, 110771:111037, 113751:113806, 114708:115655, 115676:115686, 117498:118092, 120901:127284, 134270:134336, 137282:137844, 144565:144810, 154855:160171, 169763:171222, 176040:178231, 178291:178523, 182772:183363, 208384:209280, 211738:212255, 215491:216027, 217002:217927, 219137:219734, 221575:231577, 237853:238615, 240250:240490, 245414:248623, 251580:252106, 259157:263501, 263618:272602, 272627:280113, 289185:291214, 305477:305810, 310790:310980, 314989:316634, 318066:318548, 323352:323854, 330987:332603, 337845:343068, 348424:348459, 357236:357239, 363901:364509, 373465:373570, 381902:382825, 394415:398890, 418677:420921, 423713:444002, 461605:461994, 467137:471134, 473890:477581, 483880:484038, 488392:488606, 489905:489968, 502042:502890, 506250:506472, 519646:520108, 521554:526047, 526139:530214, 530326:530394, 530683:530968, 533084:534454, 535344:535790, 536587:536803, 536931:537121, 542204:550695, 556709:557374, 561671:562735, 565184:565283, 566588:572172, 575120:582210, 583729:585118, 585376:585806, 607571:613289, 613379:614650, 629444:629464, 636345:636803, 637852:638128, 653988:654089, 663610:664386, 666902:667539, 683472:683723, 686599:688256, 691332:692019, 694948:694975, 698324:700277, 715875:717889, 736562:740800, 757010:757297, 760447:760847, 761849:763477, 770720:770891, 776369:779156, 779241:788096, 788219:796996, 797295:801554, 801698:803367, 834765:837995, 846763:853450, 882713:882917, 895099:897285, 907370:907447, 911189:912331, 926916:938482, 947710:948568, 957456:963691, 976560:979124, 979274:989670, 992178:995780, 1005640:1006176, 1014455:1015194, 1015742:1016404, 1017176:1018842, 1019157:1019826, 1029910:1030632, 1033436:1034442, 1041047:1041471, 1053826:1053893, 1066899:1068483, 1074343:1077424, 1095796:1107577, 1112303:1112417, 1115564:1126974, 1127057:1128323, 1131765:1136365, 1154388:1154513, 1156020:1156087, 1156454:1156578, 1157487:1157570, 1175809:1176472, 1190830:1193157, 1211376:1211484, 1254719:1254769, 1258169:1258200, 1268155:1271515, 1273033:1278807, 1279042:1279435, 1279484:1282100, 1285435:1287303, 1287959:1305561, 1328648:1332697, 1334077:1335320, 1336867:1337015, 1338216:1338783, 1339773:1340677, 1341265:1341271, 1344692:1348502, 1359643:1365302, 1376409:1376941, 1379341:1383461, 1385057:1389247, 1392759:1393045, 1402975:1405934, 1413100:1413916, 1415807:1416581, 1425201:1427226, 1433695:1436278, 1441334:1442047, 1443944:1444506, 1444641:1446496, 1465358:1465596, 1469593:1471074, 1479923:1483082, 1514417:1514754, 1540926:1543524, 1556196:1556640, 1556661:1567601, 1580208:1581393, 1593674:1594648, 1596216:1600623, 1601457:1601837, 1607219:1607704, 1610041:1610389, 1628486:1629002, 1633498:1634568, 1644115:1644587, 1651457:1652875, 1655459:1655615, 1656763:1657512, 1663331:1663401, 1669792:1670032, 1672742:1672852, 1676438:1680445, 1685203:1686215, 1693675:1696198, 1697676:1698348, 1701128:1701360, 1701808:1702104, 1708553:1709586, 1715564:1721589, 1751055:1751480, 1754410:1755328, 1766400:1767886, 1790157:1791397, 1791491:1793253, 1796424:1797434, 1798722:1799716, 1806996:1808091, 1811835:1812891, 1813581:1813908, 1823043:1823446, 1825967:1828078, 1831929:1832166, 1832421:1832896, 1834763:1834882, 1835438:1835660, 1841965:1842444, 1843227:1843383, 1843959:1845199, 1850523:1851668, 1853045:1856458, 1860578:1862606, 1872306:1874007, 1884196:1885160, 1894910:1895160, 1923240:1924299, 1932260:1932345, 1945220:1945813, 1949520:1950593, 1958527:1958706, 1965066:1967771, 1973882:1974778, 1974991:1978029, 1983412:1988245, 2010525:2019431, 2019523:2026315, 2033748:2034321, 2043805:2046538, 2048913:2049139, 2050177:2061666, 2063998:2064292, 2064686:2064975, 2065481:2065691, 2069575:2069680, 2070607:2070672, 2070891:2071351, 2072547:2072679, 2076925:2078954, 2093796:2097345, 2123443:2128587, 2145971:2149824, 2151826:2160037, 2209644:2216097, 2222313:2223352, 2233099:2233743, 2234957:2235601, 2241252:2242229, 2247223:2250130, 2254908:2255920, 2272866:2279844, 2280800:2281502, 2291205:2293126, 2297328:2297391, 2304397:2304512, 2306790:2307266, 2312129:2312495, 2312958:2313270, 2315992:2316306, 2321817:2322024, 2322596:2323991, 2327329:2327988, 2330154:2330894, 2332652:2333337, 2338693:2342215, 2350536:2350840, 2350945:2352834, 2356238:2357882, 2372985:2377088, 2381851:2382952, 2401492:2402229, 2409999:2410189, 2410261:2415130, 2420353:2420833, 2421816:2422152, 2425185:2429674, 2432588:2432589, 2433615:2434001, 2435691:2436470, 2437084:2438764, 2442958:2445419, 2445983:2446739, 2448369:2449413, 2451366:2451897, 2464618:2468792, 2471935:2472059, 2485391:2485860, 2489787:2489975, 2497381:2498133, 2506259:2506890, 2513710:2514166, 2514311:2524531, 2524552:2533267, 2542114:2544748, 2546195:2546385, 2548399:2549295, 2550543:2551374, 2551590:2552733, 2555408:2556225, 2557451:2558175, 2560189:2560371, 2561830:2562667, 2562997:2568859, 2571059:2571739, 2576020:2581014, 2586124:2587526, 2595294:2599978, 2602650:2605282, 2605605:2606740, 2617360:2617528, 2620430:2620649, 2621859:2621956, 2628577:2629684, 2632320:2632986, 2634561:2636903, 2637591:2638648, 2639419:2642723, 2643778:2644964, 2657245:2666383, 2666468:2668159, 2672438:2672939, 2682797:2682865, 2683011:2684864, 2684997:2686639, 2689429:2692108, 2692399:2692756, 2694084:2697398, 2699999:2700419, 2702158:2703262, 2703458:2706882, 2717549:2718026, 2721475:2721587, 2726724:2727013, 2727545:2730003, 2731453:2731561, 2742528:2744952, 2745118:2755539, 2758204:2758297, 2758405:2759022, 2776104:2776363, 2778719:2780849, 2784271:2784619, 2786077:2786331, 2787514:2787980, 2788418:2788747, 2790343:2791686, 2794976:2796129, 2797464:2798174, 2800878:2802802, 2815830:2816414, 2817689:2818743, 2822134:2822242, 2830311:2830334, 2830483:2831304, 2831674:2835344, 2835622:2835766, 2835984:2837983, 2838050:2845858, 2849903:2851956, 2853903:2853969, 2855732:2858520, 2859913:2864010, 2864061:2866505, 2875291:2875397, 2876308:2876336, 2876554:2876697, 2876843:2876931, 2878523:2878577, 2883454:2883548, 2893718:2895328, 2900860:2901324, 2903006:2903634, 2928134:2928394, 2929628:2929776, 2931361:2931460, 2947106:2948428, 2948692:2949600, 2957910:2967615, 2967740:2973509, 2973591:2974916, 2974970:2981656, 2983076:2983202, 2985178:2986103, 2995914:2996487, 3001382:3001405, 3002847:3004048, 3013448:3015741, 3021531:3021650, 3031092:3031337, 3047794:3052047, 3070150:3070967, 3084876:3085013, 3096266:3099680, 3099769:3101850, 3110781:3115035, 3117628:3122010, 3127661:3132969, 3135588:3137184, 3138447:3138540, 3140584:3141703, 3143175:3144410, 3147793:3148013, 3152003:3153284, 3154469:3158491, 3162391:3162812, 3163585:3163687, 3168057:3168192, 3171230:3173267, 3176319:3176417, 3192857:3193527, 3198838:3199251, 3203229:3206761, 3210272:3211045, 3217267:3217287, 3222955:3223159, 3228104:3229430, 3233286:3233390, 3256077:3256688, 3256764:3258999, 3259089:3259796, 3266762:3267093, 3269112:3269396, 3280883:3287335, 3293171:3293183, 3294356:3299569, 3300516:3301367, 3307935:3317977, 3318584:3318829, 3330673:3338366, 3349914:3352433, 3359487:3362326, 3363492:3364026, 3365857:3367639, 3368550:3368851, 3371444:3371768, 3373834:3374775, 3407189:3410012, 3416368:3419599, 3424211:3427147, 3438543:3439098, 3449175:3449325, 3462174:3462257, 3478517:3479352, 3481715:3481996, 3484500:3487803, 3499706:3502735, 3512733:3516449, 3529962:3533884, 3546910:3547070, 3559807:3559933, 3562851:3564039, 3600735:3601290, 3602908:3603337, 3605744:3606406, 3608345:3608517, 3611844:3611941, 3618317:3618888, 3619693:3620536, 3625570:3628245, 3636485:3637352, 3638824:3638851, 3638924:3639647, 3641347:3641650, 3647082:3651235, 3652702:3653996, 3655265:3655412, 3662183:3666489, 3671700:3680179, 3680927:3682889, 3684086:3684780, 3694021:3694295, 3697551:3697595, 3697616:3698101, 3699300:3703541, 3704956:3706029, 3709435:3710089, 3713201:3713448, 3721157:3727026, 3731179:3731351, 3740485:3741450, 3742647:3745501, 3745522:3745588, 3748145:3748522, 3753958:3755487, 3759039:3769671, 3774500:3774633, 3777948:3778088, 3781690:3782117, 3783294:3783906, 3784226:3786587, 3787496:3788035, 3788896:3789308, 3792797:3794555, 3798068:3798395, 3805739:3807455, 3816809:3818073, 3821496:3821655, 3821990:3822882, 3826514:3827218, 3828126:3828238, 3829327:3829514, 3837807:3838236, 3839673:3839800, 3841561:3853846, 3859292:3859368, 3862733:3862832, 3865938:3867540, 3869109:3869213, 3870114:3870641, 3871267:3871473, 3873402:3873443, 3876716:3877062, 3880743:3881815, 3891598:3892331, 3895271:3895763, 3897397:3898044, 3899802:3900406, 3905208:3908746, 3910544:3910711, 3912482:3912634, 3915308:3916063, 3916391:3917796, 3925162:3925433, 3927891:3930813, 3931740:3932209, 3934813:3935271, 3953991:3956074, 3956137:3960967, 3962249:3963087, 3964177:3964369, 3965314:3965737, 3967279:3968109, 3970010:3970373, 3971302:3972161, 3973703:3974058, 3977424:3978750, 3979538:3983182, 3986581:3987939, 3988699:3995724, 3996402:3999299, 4031739:4035701, 4090885:4094159, 4098330:4103697, 4108167:4109406, 4112385:4114860, 4116370:4116850, 4118164:4120406, 4123036:4123480, 4124764:4126055, 4129310:4129564, 4132224:4132361, 4137074:4137160, 4138952:4139510, 4144583:4145314, 4147121:4147272, 4148292:4148831, 4149415:4149980, 4150638:4151102, 4151761:4153965, 4154452:4155541, 4158718:4159013, 4161113:4162034, 4165025:4165074, 4167056:4168983, 4171589:4172203, 4172315:4172437, 4173624:4173757, 4174774:4175318, 4177271:4177304, 4178447:4178811, 4182129:4182220, 4184992:4185173, 4189956:4190157, 4193129:4193240, 4203303:4203457, 4210341:4210978, 4215768:4229569, 4236115:4236953, 4238619:4239090, 4242727:4242820, 4247582:4247752, 4249415:4249492, 4253170:4253241, 4254139:4254196, 4264699:4265106, 4267967:4268069, 4270287:4270665, 4273336:4273895, 4274742:4275887, 4276967:4277829, 4279622:4295023, 4296741:4298183, 4302245:4303929, 4308244:4311815, 4311888:4318032, 4359092:4359206, 4362548:4363274, 4364077:4364902, 4367907:4368088, 4372138:4372254, 4373018:4373655, 4381814:4382564, 4385218:4386784, 4401907:4401955, 4404611:4409688, 4426188:4427023, 4428042:4428448, 4444349:4444718, 4448461:4448584, 4449774:4449878, 4450927:4451193, 4456805:4457374, 4469460:4479638, 4479659:4481197, 4511126:4511437, 4516089:4517259, 4518491:4521534, 4522813:4525960, 4534805:4537446, 4539729:4539888, 4540026:4549680, 4552785:4552884, 4553121:4554125, 4555322:4555342, 4556826:4557652, 4563759:4564034, 4565982:4566429, 4568398:4568764, 4570027:4571191, 4575038:4575100, 4577204:4578605, 4579740:4579963, 4580601:4580707, 4583810:4583912, 4584505:4584767, 4584932:4585114, 4586176:4586272, 4589450:4589599, 4589899:4597922, 4602666:4603216, 4604196:4604402, 4605380:4605654, 4633347:4634102, 4635615:4635742, 4637328:4637431, 4640882:4641002, 4644069:4644641, 4645659:4645661, 4646963:4647423, 4650777:4652144, 4652906:4653466, 4660153:4660520, 4660713:4660827, 4662445:4664507, 4665348:4665932, 4671738:4675132, 4679448:4679664, 4682856:4683181, 4683865:4684315, 4684414:4685864, 4698575:4699609, 4702866:4704603, 4705885:4706575, 4708883:4709258, 4710712:4711413, 4713576:4714319, 4715483:4716969, 4719609:4719757, 4720658:4721130, 4723982:4724064, 4726771:4727279, 4730366:4731605, 4738102:4738170, 4744321:4745314, 4745455:4746463, 4746903:4746969, 4747116:4747623, 4759603:4759819, 4761321:4761930, 4762359:4763195, 4765313:4766132, 4767657:4768330, 4769063:4770291, 4773070:4773818, 4788358:4789189, 4789592:4790291, 4790642:4790643, 4792483:4799980, 4804396:4805606, 4809674:4821928, 4826639:4827071, 4829791:4830134, 4835622:4836186, 4847972:4848424, 4867825:4868170, 4868280:4876142, 4876275:4884902, 4895516:4895920, 4900974:4901917, 4903264:4903451, 4905173:4906838, 4914047:4918371, 4926486:4926634, 4932178:4934258, 4950546:4950582, 4951481:4951570, 4957667:4957774, 4958309:4959216, 4961526:4962450, 4964390:4964666, 4967089:4967993, 4972226:4975374, 4975508:4980306, 4983959:4984087, 4986739:4986888, 4989454:4989547, 4994675:4996834, 5003202:5003449, 5006057:5006705, 5008050:5008769, 5013624:5014350, 5016260:5018782, 5020358:5020779, 5023210:5027154, 5041160:5041369, 5042092:5042101, 5043064:5044197, 5045720:5046393, 5048003:5050039, 5055000:5055685, 5056039:5058250, 5068831:5073578, 5079303:5079909, 5082279:5084897, 5085028:5086309, 5093903:5096597, 5100707:5100750, 5103476:5103580, 5104911:5104968, 5105103:5105109, 5106048:5106203, 5107625:5107842, 5113752:5114121, 5118852:5119215, 5127201:5127827, 5129611:5129744, 5131065:5131519, 5134105:5134738, 5137971:5138249, 5139679:5141672, 5149226:5154513, 5158500:5158510, 5162232:5162601, 5164884:5165147, 5176654:5176828, 5186236:5186925, 5189240:5190137, 5195243:5196072, 5203504:5204529, 5219815:5230548, 5241914:5244282, 5245402:5245742, 5263424:5263693, 5268380:5268418, 5269485:5269544, 5270941:5271264, 5272041:5272845, 5274270:5274692, 5278708:5278802, 5281519:5282082, 5283944:5284025, 5291604:5292778, 5292837:5298754, 5300331:5300851, 5317392:5317490, 5318124:5318219, 5319301:5319711, 5321151:5325548, 5330230:5330474, 5330495:5330577, 5332169:5332248, 5332780:5333778, 5334919:5335191, 5338027:5338921, 5340266:5340828, 5342045:5342179, 5344724:5344871, 5345013:5345275, 5345425:5345926, 5353774:5354180, 5357029:5357540, 5357785:5358145, 5358293:5358368, 5358507:5358511, 5365174:5365262, 5366880:5366943, 5367986:5368477, 5372643:5372850, 5384508:5388899, 5388920:5401756, 5412400:5412496, 5412640:5412840, 5415629:5416652, 5420731:5421309, 5428457:5428532, 5429908:5429931, 5430054:5430156, 5431450:5431491, 5431512:5432196, 5433051:5433904, 5443026:5443877, 5446855:5450435, 5450456:5452326, 5459523:5460530, 5460551:5463116, 5469639:5472324, 5479438:5479487, 5484661:5484778, 5486354:5486718, 5487150:5487365, 5488241:5488339, 5490631:5491423, 5493263:5494416, 5494474:5497678, 5507023:5507115, 5509177:5509466, 5511705:5512047, 5514556:5514650, 5516025:5516069, 5518567:5518809, 5518898:5520388, 5529763:5530613, 5534649:5540543, 5540564:5546117, 5561689:5563826, 5566269:5566342, 5570085:5571795, 5572949:5573213, 5576749:5577069, 5579194:5579665, 5581993:5582095, 5607529:5607825, 5608841:5609340, 5610270:5610373, 5616402:5616597, 5618551:5619568, 5620669:5620778, 5624475:5624600, 5626288:5627812, 5637621:5643309, 5644057:5644728, 5649656:5649755, 5653648:5653827, 5655857:5655949, 5658712:5659444, 5661273:5662142, 5669737:5669944, 5673352:5674303, 5676609:5676826, 5677646:5678637, 5691503:5692263, 5697500:5702155, 5705553:5707044, 5707124:5716573, 5725809:5726307, 5729310:5729792, 5735253:5735549, 5739421:5739478, 5740283:5741185, 5741622:5741867, 5743571:5744179, 5745617:5746632, 5754306:5754311, 5760799:5760895, 5762435:5764029, 5768450:5768918, 5773627:5774584, 5775610:5776092, 5788611:5788633, 5791752:5792320, 5795114:5795399, 5803734:5803955, 5807291:5807708, 5818132:5818473, 5821134:5822976, 5826148:5829390, 5837392:5838094, 5838850:5840281, 5841940:5842708, 5844050:5844835, 5846466:5847476, 5849424:5850033, 5851893:5852022, 5852962:5854198, 5855040:5857544, 5862557:5862643, 5866431:5866561, 5868139:5868265, 5869954:5871393, 5871512:5872173, 5874343:5874415, 5878958:5879138, 5879554:5883857, 5883989:5886612, 5895973:5896617, 5897405:5898302, 5904695:5906657, 5918157:5918269, 5918445:5919054, 5936957:5937194, 5938881:5942389, 5947640:5949395, 5949448:5949858, 5957013:5957440, 5958457:5959007, 5959532:5962952, 5964183:5964400, 5967211:5967246, 5973326:5981765, 5985871:5986487, 5990589:5990740, 6003379:6004656, 6007595:6008136, 6009524:6009781, 6011586:6011976, 6026877:6027193, 6028370:6028485, 6033458:6035447, 6037884:6038402, 6041267:6041671, 6046018:6046787, 6050451:6051144, 6051282:6053784, 6055821:6056130, 6058424:6058547, 6060953:6061731, 6068040:6068525, 6070122:6070231, 6078896:6081781, 6086006:6087429, 6087729:6090008, 6091385:6092685, 6092811:6093306, 6094437:6095059, 6096589:6097578, 6097791:6098926, 6099731:6100034, 6100055:6100260, 6101789:6102066, 6102396:6104804, 6105754:6105758, 6105820:6105862, 6107715:6107816, 6109655:6110133, 6114549:6123119, 6126875:6133006, 6133129:6133244, 6137977:6138500, 6139211:6139420, 6140883:6141352, 6144375:6145454, 6147102:6148141, 6148836:6149196, 6150491:6153644, 6153753:6154074, 6157440:6157447, 6159264:6159880, 6161665:6162290, 6162387:6163379, 6163599:6164190, 6164264:6164718, 6166539:6167945, 6168642:6169083, 6169195:6171594, 6171658:6172661, 6172796:6172998, 6173139:6174994, 6175138:6175194, 6180840:6181313, 6183856:6184071, 6185368:6185461, 6194897:6196173, 6196194:6199672, 6200193:6200797, 6202076:6203973, 6206927:6206954, 6211155:6211269, 6211885:6211980, 6219912:6220092, 6220764:6221244, 6223115:6224877, 6226099:6231495, 6232129:6232776, 6234280:6234707, 6235391:6236133, 6250197:6250474, 6252876:6253081, 6253584:6253700, 6254947:6255034, 6257620:6260248, 6264520:6264564, 6264709:6265677, 6269069:6269419, 6271533:6271865, 6273785:6274529, 6275291:6280455, 6280583:6283495, 6283672:6284972, 6289340:6289432, 6294068:6295066, 6297636:6297766, 6299729:6300492, 6302298:6302412, 6304296:6304915, 6306384:6307302, 6308636:6309113, 6310808:6311228, 6312416:6312749, 6314624:6315232, 6317262:6317696, 6324019:6326620, 6327968:6329535, 6330181:6342223, 6358411:6358584, 6360457:6360837, 6364172:6365030, 6366224:6366330, 6368078:6368269, 6369596:6369747, 6394350:6402648, 6402806:6404586, 6405705:6405856, 6406936:6407525, 6416812:6418564, 6418840:6420659, 6421889:6422060, 6428942:6429526, 6431586:6432138, 6432212:6436111, 6441192:6441643, 6446465:6447303, 6456984:6457281, 6458479:6459476, 6459596:6460877, 6467519:6468112, 6471922:6473206, 6488977:6490715, 6496109:6497435, 6499560:6499981, 6501410:6502166, 6503909:6507893, 6513830:6513927, 6517507:6517528, 6520918:6521784, 6523442:6525044, 6528689:6529165, 6530099:6530130, 6531963:6531982, 6533071:6534083, 6541062:6542115, 6544966:6545109, 6546593:6546614, 6550436:6550535, 6552625:6552833, 6553379:6553480, 6554290:6554503, 6554709:6554728, 6558665:6558693, 6562346:6562648, 6562669:6562698, 6564451:6564748, 6565852:6566124, 6568954:6569556, 6572600:6574003, 6587902:6592081, 6592231:6592468, 6598927:6598981, 6607520:6608340, 6608517:6612187, 6615990:6616303, 6618560:6620664, 6621295:6621623, 6624918:6625065, 6626788:6627044, 6627917:6629568, 6637429:6640650, 6642636:6642720, 6643509:6651219, 6655430:6655607, 6658371:6659271, 6659356:6660435, 6662315:6663160, 6673830:6682299, 6683740:6686723, 6686776:6687348, 6691671:6695367, 6695424:6705545, 6706404:6706756, 6708919:6710215, 6710543:6710849, 6712431:6715966, 6716990:6717818, 6723059:6724866, 6729376:6730056, 6730959:6731266, 6732276:6734027, 6736161:6736679, 6741176:6741275, 6741876:6741983, 6745329:6745990, 6753031:6753218, 6756348:6757174, 6757233:6757762, 6758178:6758180, 6760179:6760415, 6767797:6768384, 6768545:6773972, 6780151:6780798, 6784397:6784867, 6787139:6788031, 6789002:6789038, 6789738:6790697, 6790838:6790977, 6798161:6798226, 6799608:6800385, 6801973:6802199, 6810571:6812659, 6814249:6814510, 6834699:6835505, 6835631:6836468, 6838140:6838353, 6843343:6843588, 6845142:6845176, 6846959:6847133, 6847308:6847510, 6849923:6850408, 6851253:6851371, 6854140:6854477, 6857536:6860394, 6861997:6862312, 6870951:6871133, 6872142:6872606, 6873757:6874311, 6877038:6877058, 6892143:6893302, 6893447:6897734, 6900619:6900948, 6913579:6915042, 6916593:6916765, 6917561:6917699, 6922973:6923068, 6923693:6924281, 6929665:6931137, 6936928:6937463, 6939672:6940214, 6943077:6943175, 6945358:6945597, 6950499:6951134, 6954252:6954413, 6961139:6967669, 6970145:6970362, 6971888:6972156, 6973696:6974557, 6977071:6979628, 6979773:6989066, 6999706:7002869, 7003481:7005442, 7015223:7015907, 7020797:7021294, 7021584:7021971, 7029076:7029887, 7030001:7033810, 7033950:7034158, 7034366:7035350, 7035371:7038899, 7048832:7050041, 7051712:7053687, 7056385:7057191, 7059351:7059891, 7061254:7065251, 7069186:7069439, 7074721:7075917, 7082040:7088029, 7090062:7090739, 7094723:7095485, 7098400:7098648, 7099866:7100002, 7101775:7102215, 7103363:7103811, 7104136:7105301, 7106926:7110233, 7112443:7112866, 7115055:7116328, 7116655:7116734, 7116830:7117807, 7121545:7121656, 7122194:7122613, 7128246:7132746, 7134370:7134737, 7147578:7149889, 7151844:7152461, 7156786:7159713, 7160917:7160935, 7166523:7166917, 7169089:7169996, 7172192:7172462, 7174124:7175509, 7176442:7176695, 7176842:7177903, 7179430:7181233, 7181382:7183052, 7185650:7185694, 7188298:7188464, 7190581:7190787, 7193920:7201775, 7201829:7202204, 7203533:7204540, 7207717:7208021, 7209866:7210270, 7212522:7214744, 7221299:7222290, 7224481:7224655, 7227765:7227788, 7233121:7233392, 7234501:7235351, 7235372:7235383, 7238792:7238992, 7243967:7244936, 7246811:7246903, 7248689:7248826, 7251282:7251940, 7255223:7255379, 7258700:7258945, 7267057:7267094, 7267226:7269118, 7273008:7274281, 7278573:7278734, 7279623:7279665, 7282924:7282963, 7283762:7283901, 7285772:7286300, 7286321:7290711, 7290859:7290945, 7291013:7291271, 7292991:7294037, 7299983:7300060, 7300913:7300998, 7302975:7303299, 7304690:7304917, 7307012:7307116, 7308626:7308642, 7311915:7312796, 7313525:7313624, 7314279:7317645, 7321085:7322264, 7324415:7327967, 7327988:7333034, 7334533:7339231, 7352235:7352414, 7353703:7354490, 7359100:7360225, 7371534:7372295, 7372529:7376603, 7383197:7384397, 7388795:7389053, 7395918:7398709, 7401023:7401131, 7402761:7403290, 7404374:7404949, 7408511:7422784, 7454116:7454659, 7456953:7457151, 7460756:7460855, 7466429:7472619, 7476946:7480107, 7483505:7484061, 7486293:7486647, 7494736:7495435, 7505642:7505950, 7506693:7509568, 7513728:7514075, 7515126:7518821, 7519001:7519239, 7519339:7522115, 7526716:7527223, 7531715:7532046, 7542159:7544617, 7544730:7553843, 7554553:7554854, 7562349:7564583, 7567719:7567826, 7570553:7574239, 7580833:7580943, 7581628:7582105, 7583333:7583809, 7588367:7588579, 7591206:7591287, 7595372:7595775, 7602221:7602306, 7608245:7611208, 7611315:7614175, 7634143:7634491, 7639785:7640347, 7641767:7641884, 7643328:7643430, 7649777:7650340, 7656338:7656453, 7656474:7656977, 7658291:7667728, 7669900:7670154, 7672714:7674728, 7676402:7676667, 7679046:7679447, 7683045:7683340, 7694210:7694319, 7695232:7695680, 7698222:7698300, 7699284:7699681, 7700958:7701176, 7701348:7701364, 7702053:7702428, 7718639:7718881, 7718958:7722083, 7723116:7723755, 7724656:7725018, 7727112:7727256, 7729092:7729183, 7729878:7731120, 7737995:7738670, 7738691:7755777, 7756353:7766531, 7772231:7773415, 7777655:7777669, 7778011:7779418, 7785798:7785877, 7793614:7798812, 7798947:7800454, 7801822:7813761, 7819413:7819578, 7820478:7820830, 7826378:7826441, 7831629:7832662, 7835453:7835812, 7836967:7837062, 7840022:7840438, 7853640:7853749, 7859417:7859939, 7860105:7862528, 7863569:7866135, 7875554:7884188, 7885769:7887213, 7887335:7887370, 7887467:7889309, 7894864:7896124, 7897725:7898433, 7903334:7906825, 7918341:7918437, 7926503:7926622, 7934577:7936263, 7938487:7938559, 7940728:7940892, 7941994:7942718, 7945551:7946192, 7951593:7951654, 7952140:7952462, 7952648:7952884, 7953125:7954819, 7955081:7955976, 7958474:7958646, 7960158:7960266, 7961474:7962051, 7972443:7972472, 7977078:7980230, 7984758:7984888, 7993161:7993294, 7993579:7995733, 8003179:8004132, 8004187:8005126, 8006448:8006634, 8009116:8010206, 8010366:8014405, 8024422:8024879, 8026525:8026600, 8028752:8028816, 8028996:8030958, 8035696:8035797, 8039831:8039878, 8040526:8040798, 8043278:8043343, 8044175:8046556, 8046749:8047344, 8049340:8050036, 8057029:8059803, 8060478:8060788, 8061967:8062006, 8081671:8083575, 8095610:8095739, 8098444:8103595, 8109228:8109745, 8119316:8125800, 8130039:8134114, 8135051:8135839, 8135955:8137255, 8137276:8138867, 8138888:8139467, 8142537:8147426, 8149651:8150354, 8151371:8151861, 8159798:8160467, 8168711:8172720, 8173491:8174887, 8210703:8210901, 8218324:8219997, 8223954:8224473, 8226297:8226493, 8240637:8242133, 8243572:8243693, 8244963:8245504, 8250948:8253455, 8254926:8255347, 8258939:8259436, 8259498:8262038, 8268990:8270303, 8288124:8288948, 8291005:8291298, 8293640:8294873, 8307566:8307642, 8307733:8312099, 8312120:8313120, 8315432:8315902, 8318250:8318346, 8323987:8324285, 8328859:8329061, 8329647:8330449, 8332784:8333081, 8336727:8336965, 8346023:8353381, 8363858:8363953, 8369562:8375489, 8392758:8394725, 8398626:8398917, 8400504:8401371, 8403939:8404546, 8405968:8406827, 8410804:8412421, 8416197:8416815, 8426092:8427086, 8429832:8429925, 8435865:8435957, 8438608:8439970, 8440488:8441542, 8442452:8442694, 8444969:8445346, 8451001:8451035, 8451254:8452152, 8452312:8459040, 8459331:8460224, 8463015:8463128, 8463149:8463322, 8467283:8467487, 8469351:8469858, 8471852:8472463, 8473481:8474059, 8474767:8474846, 8476288:8479438, 8479492:8481151, 8482538:8482758, 8484414:8484595, 8487062:8487617, 8490938:8491290, 8495833:8498111, 8502604:8503350, 8503449:8503993, 8505566:8509040, 8511315:8511592, 8512505:8513899, 8526621:8526757, 8529271:8529375, 8531342:8531683, 8534527:8539938, 8544409:8546062, 8546725:8549512, 8570998:8571555, 8573749:8573851, 8580205:8584423, 8587507:8591214, 8591978:8592091, 8592928:8593026, 8596225:8596332, 8598551:8598657, 8600875:8601354, 8602035:8602124, 8605162:8605262, 8608631:8608953, 8609101:8609920, 8612604:8612724, 8615193:8615387, 8616948:8617056, 8618936:8623011, 8631202:8631878, 8632030:8633298, 8633403:8633896, 8637943:8638513, 8640236:8640507, 8644399:8644459, 8646304:8646450, 8651671:8658898, 8659967:8660830, 8664445:8664975, 8666497:8666594, 8669797:8669802, 8673623:8676665, 8681325:8683196, 8687217:8687633, 8692622:8692945, 8696301:8696602, 8701973:8702465, 8724279:8727217, 8728487:8728587, 8728708:8728763, 8730054:8730155, 8731722:8732218, 8735717:8736212, 8737038:8737071, 8742548:8743848, 8746985:8749615, 8751358:8752228, 8755680:8755845, 8757603:8758155, 8758996:8759205, 8759788:8763040, 8765631:8765744, 8777518:8779730, 8785415:8790295, 8790456:8791353, 8791426:8792472, 8792618:8792963, 8797295:8797455, 8804697:8804798, 8808081:8808655, 8813300:8813760, 8814979:8815571, 8816193:8816973, 8821362:8822187, 8823636:8824400, 8825603:8826166, 8826670:8828009, 8838719:8840167, 8841212:8841769, 8842309:8842551, 8845481:8845808, 8851116:8852704, 8852765:8852825, 8852917:8853588, 8861278:8864019, 8865549:8865944, 8866624:8866807, 8866973:8867939, 8868296:8869714, 8871531:8871790, 8875257:8876318, 8894137:8897631, 8901435:8901647, 8901794:8902740, 8905447:8905647, 8916805:8917851, 8918001:8919328, 8920504:8921347, 8925306:8928749, 8942292:8945532, 8946244:8946685, 8947201:8947303, 8948711:8948777, 8954390:8954558, 8957394:8957455, 8957650:8958007, 8960882:8962321, 8970081:8970290, 8971031:8971138, 8974585:8974864, 8975116:8975365, 8986540:8987094, 8987231:8989930, 8993683:8994001, 8999575:9000096, 9001742:9001844, 9003963:9004496, 9008708:9009011, 9013885:9015315, 9016179:9016754, 9017994:9018669, 9020304:9020944, 9025608:9025778, 9031232:9031234, 9036695:9039223, 9039351:9039578, 9039723:9039989, 9040126:9041319, 9042007:9042395, 9042578:9043320, 9046575:9048187, 9049349:9049459, 9055457:9056292, 9057824:9057889, 9057999:9058143, 9063160:9063277, 9064403:9065078, 9077325:9077564, 9080024:9085350, 9085848:9087742, 9087883:9089011, 9089228:9095956, 9108140:9108172, 9110918:9118114, 9124472:9124663, 9127101:9132537, 9135753:9135763, 9138365:9138405, 9145632:9149672, 9153252:9154227, 9154812:9155463, 9157462:9158800, 9158986:9159120, 9159522:9159895, 9163266:9163530, 9166952:9168431, 9177488:9177663, 9178703:9179940, 9181119:9181809, 9220728:9221817, 9226390:9226904, 9233560:9234169, 9245926:9246027, 9249358:9249541, 9250495:9250605, 9251828:9251837, 9261669:9265950, 9271058:9271478, 9278689:9280922, 9286576:9287442, 9303928:9305750, 9306844:9308441, 9310221:9310318, 9314891:9317712, 9327279:9327423, 9327495:9327636, 9327708:9327856, 9330928:9332533, 9332649:9332945, 9333043:9334799, 9342496:9344723, 9349576:9350268, 9352700:9352846, 9354612:9355024, 9357079:9359066, 9389982:9390832, 9393095:9393760, 9404039:9407173, 9408282:9408340, 9410094:9410612, 9411454:9412305, 9413222:9414549, 9418885:9419456, 9420813:9420911, 9421992:9422555, 9431120:9431248, 9433828:9435159, 9435327:9435464, 9444227:9447361, 9447421:9448541, 9448669:9454330, 9454383:9456828, 9456977:9457455, 9458842:9459286, 9460550:9463750, 9468898:9469653, 9470637:9471549, 9474801:9476596, 9477181:9477458, 9478151:9478469, 9479091:9479100, 9480912:9481658, 9482275:9482742, 9483591:9483981, 9486540:9486644, 9488029:9488487, 9494555:9500623, 9500823:9501265, 9505768:9506546, 9512236:9512306, 9512416:9512451, 9514637:9514997, 9516776:9517025, 9518736:9518890, 9518911:9519243, 9521275:9522549, 9531132:9531626, 9532869:9533420, 9540966:9541167, 9541223:9543728, 9543951:9544399, 9544549:9544732, 9558099:9558791, 9558874:9561440, 9561573:9562352, 9562437:9563842, 9563927:9565388, 9574377:9574758, 9576326:9576356, 9577889:9578525, 9581267:9581785, 9582583:9582680, 9583533:9583868, 9584735:9585416, 9588153:9588661, 9589244:9589283, 9589931:9590505, 9591825:9591938, 9593103:9595275, 9598363:9600444, 9601015:9604809, 9605179:9605750, 9605844:9605895, 9611534:9611812, 9615082:9615191, 9616333:9616515, 9618107:9618195, 9619760:9620226, 9622377:9622612, 9624724:9625048, 9625796:9626240, 9630848:9631387, 9632401:9633182, 9635945:9638855, 9641736:9642343, 9643834:9644415, 9644728:9645637, 9649238:9650498, 9652946:9653020, 9658023:9658273, 9660080:9660173, 9664915:9665250, 9670726:9673663, 9678557:9679643, 9679718:9685397, 9688653:9689196, 9690557:9692700, 9693781:9694621, 9695035:9695419, 9695440:9695706, 9696762:9697236, 9699566:9703076, 9703323:9704869, 9705517:9708610, 9708753:9710087, 9710745:9711065, 9712719:9712810, 9715232:9715320, 9728034:9731347, 9731631:9734085, 9734369:9736635, 9736785:9737815, 9744146:9744208, 9747923:9748516, 9749921:9751181, 9751336:9751342, 9751469:9752384, 9752527:9752600, 9752742:9754121, 9760787:9762363, 9767846:9768813, 9772565:9773171, 9774012:9774073, 9779542:9780348, 9785724:9785780, 9787624:9788724, 9790050:9793755, 9796675:9797852, 9797912:9806794, 9806960:9808903, 9815907:9829761, 9830569:9839597, 9844809:9847405, 9851791:9852611, 9852799:9853653, 9863495:9863583, 9869064:9869650, 9874680:9874697, 9881504:9883695, 9883767:9884710, 9884731:9886188, 9888224:9888355, 9892223:9893620, 9900817:9901075, 9902950:9904302, 9910545:9914275, 9914497:9914946, 9915096:9916864, 9916947:9919579, 9919708:9921210, 9921930:9923609, 9924280:9925694, 9925715:9925977, 9930885:9930987, 9934290:9934636, 9936022:9936124, 9938301:9940976, 9941032:9942322, 9942758:9942955, 9943165:9947803, 9953391:9953408, 9955873:9956149, 9958357:9958781, 9960095:9960426, 9961977:9962712, 9962855:9965573, 9973177:9973380, 9976315:9976425, 9979662:9980062, 9984794:9986342, 9986413:9988451, 9988763:9989131, 9989515:9989801, 9999928:10001313, 10001534:10005522, 10005544:10011504, 10011836:10014546, 10015495:10016764, 10018121:10019642, 10021090:10021701, 10022872:10023797, 10025838:10026187, 10030059:10030173, 10047977:10048084, 10062026:10062219, 10062354:10066792, 10066869:10070882, 10071900:10072577, 10074399:10074993, 10077749:10080158, 10085765:10086264, 10102012:10102229, 10102353:10103209, 10103353:10103537, 10103626:10107348, 10108778:10110981, 10111916:10112082, 10117124:10117158, 10122171:10122458, 10130215:10133589, 10133671:10134404, 10134524:10147324, 10149888:10150279, 10155105:10155160, 10158884:10158950, 10161883:10162374, 10165132:10169440, 10169580:10169638, 10172603:10174454, 10175591:10177180, 10178915:10179017, 10179747:10179844, 10185520:10185854, 10190896:10191272, 10192832:10193241, 10196997:10197320, 10197596:10198218, 10200440:10200767, 10201808:10203436, 10203743:10220454, 10221252:10221641, 10222778:10222880, 10233386:10235230, 10235407:10237461, 10237682:10244351, 10244462:10246502, 10246653:10248287, 10255327:10255545, 10262250:10264305, 10279234:10290910, 10297927:10298139, 10327870:10328983, 10330400:10330878, 10331808:10331978, 10333082:10333470, 10334590:10334660, 10336130:10336394, 10345654:10345737, 10349369:10356250, 10360898:10366620, 10371697:10373372, 10373862:10374151, 10374172:10374802, 10379634:10380096, 10380986:10381081, 10382025:10382740, 10391388:10398620, 10398759:10398907, 10400177:10400396, 10400619:10400693, 10400840:10401063, 10402762:10406128, 10406272:10407457, 10407585:10409301, 10414773:10418233, 10421858:10422894, 10429034:10429628, 10432371:10432629, 10432761:10432855, 10435876:10436253, 10436888:10436894, 10437779:10438074, 10443563:10443666, 10446043:10446656, 10452075:10452177, 10452895:10454188, 10455678:10457827, 10459047:10464288, 10471056:10471510, 10471927:10472195, 10473389:10473512, 10474435:10474821, 10478129:10479304, 10484969:10486698, 10487341:10488451, 10489585:10491195, 10491339:10493661, 10493885:10494333, 10494548:10497961, 10498106:10501121, 10501264:10501584, 10505949:10506416, 10506638:10506997, 10515586:10516101, 10516428:10516873, 10516950:10516977, 10517150:10517320, 10518055:10518378, 10519258:10519359, 10520761:10520862, 10522027:10522090, 10523996:10524446, 10525509:10526601, 10530095:10530338, 10531547:10532666, 10534484:10536114, 10538088:10543023, 10546429:10547720, 10549876:10551971, 10553047:10553115, 10553913:10554365, 10557838:10558322, 10560006:10562969, 10564903:10565265, 10565443:10566349, 10567300:10567893, 10569017:10569084, 10571920:10571953, 10577147:10579412, 10580799:10581537, 10582615:10588500, 10590465:10590702, 10593968:10594331, 10598489:10598672, 10606621:10608465, 10621435:10623172, 10624361:10624463, 10628387:10628742, 10631674:10631684, 10633150:10633278, 10634078:10634567, 10636028:10636291, 10636441:10636691, 10638966:10639075, 10640233:10640464, 10642281:10643028, 10643705:10644244, 10645617:10646477, 10647841:10648599, 10654237:10654298, 10666945:10667807, 10671357:10671417, 10672614:10672736, 10674864:10675167, 10676811:10676921, 10682374:10682471, 10683048:10683269, 10685587:10685824, 10687058:10687362, 10688060:10688331, 10690770:10691344, 10696783:10699567, 10705131:10708372, 10710378:10713504, 10718201:10720812, 10723496:10724160, 10727386:10727754, 10728808:10728897, 10730233:10730462, 10732424:10732542, 10733666:10734021, 10736615:10736696, 10739914:10740296, 10740395:10741367, 10744415:10750029, 10750293:10750497, 10776703:10781277, 10781400:10782114, 10782890:10784186, 10784831:10785127, 10787863:10789129, 10793152:10793400, 10798059:10799622, 10808240:10808589, 10810489:10810647, 10812923:10813811, 10814937:10817341, 10820843:10823362, 10828421:10830320, 10830446:10831873, 10834814:10836978, 10839012:10841097, 10844666:10844771, 10845972:10846771, 10850103:10850404, 10851268:10855508, 10855799:10855934, 10857689:10858077, 10860800:10861193, 10863559:10863879, 10866495:10866960, 10867014:10867529, 10868455:10869369, 10869474:10869955, 10875050:10876199, 10877741:10878104, 10878326:10879137, 10879572:10879768, 10894237:10899088, 10902220:10904369, 10912192:10912370, 10915684:10917049, 10919139:10919376, 10919777:10921407, 10923927:10924239, 10925750:10926880, 10933706:10933809, 10935559:10936324, 10936741:10936814, 10940521:10940677, 10944203:10944209, 10945687:10945820, 10945893:10952933, 10955204:10958296, 10959906:10964471, 10970177:10970486, 11004283:11011434, 11014745:11015482, 11017537:11018495, 11024593:11032337, 11038827:11039055, 11052497:11059298, 11061435:11063471, 11065939:11067529, 11070012:11075754, 11134667:11142109, 11144135:11145335, 11145925:11146447, 11148318:11148972, 11150303:11151223, 11151975:11152703, 11155369:11156727, 11158833:11159524, 11162661:11163787, 11177728:11177847, 11182063:11182794, 11186206:11198965, 11199146:11206921, 11211905:11212306, 11215094:11216117, 11221995:11222398, 11224163:11224835, 11233332:11234301, 11235176:11235384, 11235594:11235850, 11236784:11237816, 11237980:11238116, 11238227:11239683, 11240478:11241224, 11241970:11242736, 11243524:11243835, 11245003:11245822, 11246765:11248269, 11250018:11250138, 11251602:11252237, 11252981:11253255, 11255927:11256525, 11259233:11260921, 11262120:11263065, 11263665:11264684, 11264987:11265104, 11265407:11265522, 11265714:11266069, 11266474:11268029, 11268409:11268618, 11268958:11269924, 11271289:11271687, 11272385:11272476, 11273482:11274891, 11276346:11276447, 11280018:11282425, 11286561:11287434, 11287511:11287747, 11287818:11289040, 11291977:11292661, 11294164:11294575, 11297206:11300742, 11301466:11302285, 11305218:11306252, 11307450:11308120, 11311053:11311523, 11323110:11323526, 11333154:11334055, 11338112:11338634, 11338836:11339786, 11341521:11342084, 11347288:11355235, 11356370:11358077, 11359305:11361084, 11362946:11364325, 11374334:11374619, 11380127:11380230, 11396662:11400142, 11416433:11417007, 11421269:11422179, 11423600:11423992, 11426054:11428654, 11428747:11429013, 11432156:11433120, 11433422:11433670, 11436018:11436781, 11437407:11439288, 11440709:11443451, 11446644:11449855, 11451972:11453899, 11454604:11454779, 11455539:11456466, 11459953:11460965, 11463074:11465141, 11467612:11469461, 11471584:11473217, 11475662:11477743, 11480071:11483356, 11484640:11484777, 11486598:11486894, 11488471:11489157, 11490578:11492334, 11492773:11493035, 11495080:11495520, 11496245:11498449, 11498715:11498859, 11499655:11501617, 11507172:11510022, 11513853:11515126, 11517108:11517557, 11526797:11526917, 11557851:11557985, 11558683:11559445, 11560565:11560619, 11564387:11565188, 11570871:11570885, 11571166:11572092, 11577488:11578640, 11579612:11579705, 11579778:11579855, 11581989:11582742, 11587337:11587695, 11587757:11592245, 11601712:11601861, 11602018:11604026, 11611314:11611671, 11616244:11616815, 11620263:11621154, 11621273:11621575, 11621899:11623114, 11626562:11628148, 11628355:11628597, 11628666:11629640, 11629932:11631643, 11632918:11632982, 11640414:11640686, 11642206:11643020, 11645420:11646524, 11647020:11647752, 11648428:11648759, 11649206:11649502, 11649687:11649981, 11651825:11652788, 11654313:11655003, 11655123:11666411, 11666470:11667910, 11667969:11668081, 11668145:11668388, 11668438:11669984, 11670094:11670497, 11670549:11675940, 11682441:11686088, 11686948:11687810, 11688522:11688672, 11690836:11694576, 11697943:11699168, 11701178:11701630, 11703698:11703717, 11706612:11707807, 11712197:11714955, 11717616:11718863, 11722936:11723399, 11726326:11731057, 11733478:11733939, 11735964:11736580, 11737699:11738655, 11740762:11741602, 11743966:11744168, 11804991:11809131, 11811266:11813778, 11814451:11815907, 11819280:11819538, 11820051:11820661, 11822646:11824975, 11825787:11827548, 11828459:11830984, 11833183:11836459, 11849258:11851553, 11868105:11877802, 11887278:11888500, 11890558:11891575, 11893727:11893930, 11896284:11896673, 11898521:11898788, 11901476:11901659, 11903310:11903733, 11905148:11905956, 11907484:11907524, 11909280:11909365, 11909991:11910327, 11911021:11911319, 11912677:11912792, 11917712:11917805, 11923915:11924540, 11929516:11930764, 11931757:11933351, 11933416:11933464, 11933575:11943665, 11943815:11947000, 11953154:11953506, 11963085:11966270, 11967591:11968021, 11969847:11971178, 11971802:11977176, 11977276:11977977, 11978084:11978756, 11978847:11979400, 11985619:11996975, 11997030:11998780, 12004947:12005176, 12006296:12006910, 12008040:12008624, 12009625:12011194, 12018684:12020755, 12022676:12024795, 12028298:12029045, 12029504:12029921, 12032203:12032284, 12033737:12034071, 12037079:12041034, 12046618:12047341, 12059798:12060270, 12064897:12065864, 12065937:12065942, 12069084:12069805, 12072904:12073071, 12084225:12084294, 12087856:12088118, 12090062:12091565, 12094726:12095481, 12097987:12098663, 12106112:12107043, 12108503:12108587, 12112505:12113016, 12113596:12114063, 12116290:12116948, 12117128:12117996, 12118384:12119110, 12119586:12120140, 12121333:12121710, 12122060:12122810, 12125586:12126454, 12128965:12130069, 12131381:12132552, 12133639:12134267, 12136465:12140559, 12142757:12142910, 12142982:12144006, 12146977:12148079, 12154333:12155123, 12156391:12157132, 12158733:12160099, 12162091:12162391, 12163703:12165408, 12177782:12181053, 12183254:12183687, 12189008:12189520, 12190224:12195267, 12195334:12198599, 12198685:12202231, 12203720:12203838, 12205809:12207096, 12212689:12213105, 12215198:12215537, 12217772:12218431, 12222298:12223552, 12223605:12225425, 12225497:12229308, 12229380:12230417, 12232303:12232711, 12235193:12236017, 12238252:12239159, 12241808:12245455, 12246778:12247770, 12250210:12251905, 12253636:12254484, 12257180:12258602, 12258803:12259911, 12262848:12263701, 12264276:12264464, 12266420:12266804, 12268128:12269254, 12272190:12272948, 12275342:12276741, 12276807:12276892, 12278775:12279751, 12280434:12281220, 12282615:12283267, 12285397:12285745, 12285932:12286527, 12289316:12290448, 12293936:12294299, 12296348:12297733, 12300817:12301423, 12303157:12303598, 12306820:12307730, 12310431:12312324, 12314962:12315192, 12316688:12317320, 12318735:12319349, 12320854:12321936, 12324062:12324757, 12324945:12326342, 12327025:12327981, 12329494:12331084, 12333454:12335034, 12336605:12337967, 12339239:12341144, 12342826:12344166, 12345702:12345818, 12347265:12347699, 12349819:12350934, 12352724:12353505, 12356337:12356941, 12359341:12359941, 12363416:12364717, 12365950:12366334, 12368619:12369188, 12370607:12371349, 12372647:12374204, 12376135:12376903, 12378674:12379334, 12381032:12381807, 12383259:12384276, 12385731:12387704, 12388791:12389759, 12390979:12392008, 12393312:12394119, 12396278:12396643, 12398267:12399289, 12401206:12403143, 12404489:12405524, 12406946:12407563, 12411435:12413808, 12415745:12417385, 12418800:12421238, 12426850:12427387, 12429094:12430027, 12431500:12433247, 12434944:12435725, 12437095:12439935, 12441253:12442298, 12443780:12444581, 12446512:12447166, 12448240:12449976, 12452034:12453388, 12454764:12454852, 12455826:12456925, 12457174:12458016, 12460903:12461655, 12461895:12462175, 12465816:12470619, 12473424:12474619, 12475934:12476652, 12478594:12483808, 12486235:12486943, 12494538:12494825, 12497357:12497437, 12505010:12505113, 12507569:12508298, 12513420:12513515, 12518694:12518786, 12532135:12532741, 12534404:12535313, 12537671:12537902, 12538986:12551536, 12552375:12557163, 12557543:12558546, 12558944:12559364, 12559658:12560411, 12560672:12561039, 12561560:12562755, 12564210:12564355, 12572801:12577953, 12579226:12579896, 12580957:12581765, 12585139:12585381, 12586006:12587241, 12591091:12591115, 12592351:12594143, 12594204:12594568, 12594657:12595689, 12595773:12596529, 12598916:12601777, 12601845:12601897, 12602957:12603757, 12606248:12608618, 12608708:12615920, 12625238:12626673, 12629082:12629308, 12630701:12631184, 12632555:12635171, 12636542:12637025, 12638418:12638684, 12640622:12640719, 12641501:12641602, 12642187:12642402, 12642900:12643007, 12644321:12647379, 12652087:12652276, 12654655:12655013, 12656760:12657761, 12659446:12659728, 12660647:12661253, 12661732:12662811, 12663833:12665066, 12668893:12669379, 12674449:12674491, 12674544:12677524, 12677597:12679206, 12681812:12681844, 12682702:12683219, 12684386:12685386, 12685569:12686115, 12688206:12688726, 12690572:12691317, 12691762:12692168, 12693724:12694405, 12695934:12696557, 12698214:12698903, 12699200:12699429, 12701031:12701438, 12702886:12703434, 12704727:12705096, 12706983:12707512, 12708691:12708832, 12711596:12712078, 12714730:12715647, 12716583:12718295, 12719837:12720204, 12722889:12724750, 12726206:12727763, 12727929:12728002, 12729898:12730025, 12732722:12733286, 12741275:12741587, 12749562:12750035, 12752105:12754090, 12755126:12758185, 12759577:12759586, 12764960:12773658, 12782257:12783282, 12784281:12787942, 12806581:12807530, 12811814:12815903, 12816558:12819215, 12825602:12825663, 12834163:12835114, 12843331:12844268, 12858281:12859492, 12861293:12861950, 12863544:12864229, 12865350:12866097, 12866911:12866987, 12869576:12878638, 12902806:12903158, 12903449:12904469, 12906040:12906461, 12907754:12908303, 12909436:12915165, 12918131:12918721, 12919280:12922003, 12923548:12927894, 12928456:12929371, 12931115:12932036, 12933753:12934763, 12937362:12938554, 12942227:12942279, 12946588:12946931, 12951099:12951133, 12955322:12955324, 12958051:12958639, 12959932:12960312, 12961619:12962037, 12964579:12964673, 12966795:12967215, 12968280:12968945, 12978787:12978838, 12983974:12988230, 12995475:12995802, 13023624:13033482, 13045541:13050444, 13062122:13066623, 13069046:13069520, 13071190:13072228, 13073057:13073836, 13075381:13076269, 13077020:13077072, 13077941:13077993, 13078880:13078920, 13079690:13079741, 13080633:13081411, 13082166:13082680, 13082966:13083320, 13103119:13104250, 13106465:13107135, 13107453:13107491, 13116434:13117628, 13120324:13121091, 13123513:13124167, 13126650:13127151, 13127876:13128469, 13130387:13130812, 13132466:13133647, 13136156:13136389, 13138544:13139121, 13141898:13143194, 13144623:13144911, 13146453:13152568, 13153734:13155223, 13161508:13162491, 13182726:13183451, 13183772:13184851, 13188192:13189791, 13190583:13191088, 13194645:13197127, 13198376:13199349, 13201037:13201709, 13204293:13204496, 13205403:13205611, 13207835:13208440, 13210462:13210584, 13211352:13211825, 13213806:13214094, 13216230:13216589, 13218852:13219140, 13219464:13221858, 13221978:13224429, 13225473:13227994, 13229701:13232170, 13234686:13235347, 13236663:13237053, 13238304:13238537, 13240733:13241654, 13244730:13246852, 13249778:13250518, 13252902:13254140, 13257729:13258380, 13260217:13260865, 13262733:13263802, 13265570:13269983, 13272781:13274110, 13276464:13277615, 13280175:13280845, 13286504:13287552, 13289542:13292505, 13297558:13297658, 13297746:13297876, 13297985:13298203, 13300716:13302758, 13304311:13307050, 13308685:13309987, 13310711:13310886, 13318716:13324168, 13324254:13324994, 13325066:13325229, 13336146:13337046, 13337212:13337429, 13338931:13339396, 13340224:13340326, 13343952:13344067, 13346963:13347065, 13359643:13360052, 13366888:13369998, 13370080:13370684, 13370819:13376838, 13377349:13381391, 13384371:13384751, 13387084:13387818, 13391205:13392117, 13394001:13397269, 13398631:13402447, 13404106:13408673, 13408697:13409896, 13411554:13411659, 13413172:13414272, 13414465:13418457, 13423377:13423996, 13425148:13426450, 13428471:13428912, 13434611:13435488, 13437963:13438057, 13450218:13451462, 13452845:13455355, 13470371:13470591, 13474185:13477151, 13487368:13489586, 13491120:13491380, 13507068:13508813, 13512024:13512312, 13514290:13514549, 13526247:13527365, 13527963:13528689, 13531462:13534736, 13542395:13543067, 13545188:13548073, 13559361:13560090, 13563465:13564566, 13572047:13572378, 13575805:13576147, 13581611:13581697, 13585295:13586036, 13588831:13589177, 13598505:13598608, 13603393:13603905, 13626225:13626517, 13628447:13629226, 13633492:13635041, 13635958:13637211, 13637277:13637844, 13639311:13640304, 13648405:13652523, 13652727:13652948, 13653704:13654008, 13658167:13658170, 13659128:13659430, 13662332:13663351, 13674997:13676254, 13677792:13680414, 13681366:13682136, 13684547:13685298, 13687586:13692684, 13702954:13709778, 13714128:13714297, 13716916:13720124, 13724453:13724576, 13732046:13734957, 13751124:13751233, 13756354:13756729, 13771773:13773066, 13778784:13778894, 13792249:13792636, 13797272:13797344, 13802223:13802392, 13806801:13806916, 13814322:13814755, 13827988:13828101, 13832147:13834875, 13834941:13835339, 13840047:13840703, 13843996:13844738, 13846957:13849500, 13850893:13852385, 13854047:13854637, 13859699:13861467, 13887640:13890341, 13891406:13893132, 13894249:13899084, 13914008:13915663, 13917809:13919374, 13922122:13922584, 13922659:13923060, 13923763:13926735, 13927556:13927945, 13928792:13938809, 13942016:13942543, 13951630:13951749, 13952052:13952160, 13953203:13957198, 13960612:13961010, 13969571:13969843, 13970877:13971282, 13971419:13972616, 13973412:13978429, 13980130:13980857, 13980954:13983225, 13989668:13992993, 13996221:13997362, 14002972:14004427, 14008484:14009870, 14011531:14012368, 14013629:14015753, 14018267:14018798, 14019759:14020360, 14021312:14027235, 14027588:14027779, 14029212:14029278, 14029364:14031146, 14031667:14034953, 14036057:14037150, 14037777:14041626, 14043478:14043530, 14045694:14046131, 14048022:14048093, 14051010:14052135, 14054154:14054592, 14056894:14058838, 14058910:14059311, 14062314:14062748, 14068803:14070555, 14070619:14075066, 14075163:14075836, 14082582:14083044, 14084207:14084511, 14085525:14085687, 14086685:14088086, 14089858:14090729, 14096348:14097491, 14101090:14101321, 14104470:14104514, 14113108:14113635, 14114085:14114193, 14118650:14118750, 14127202:14128590, 14142207:14150627, 14151067:14151840, 14152282:14153227, 14154195:14154768, 14155296:14155428, 14156336:14157610, 14158078:14158507, 14161664:14162845, 14162939:14163529, 14168307:14171053, 14179360:14182594, 14196295:14198610, 14199136:14200019, 14204687:14205716, 14208726:14209147, 14212501:14213448, 14213515:14213680, 14215850:14216700, 14220624:14220692, 14231360:14233034, 14235948:14236394, 14241181:14241897, 14244452:14245499, 14245555:14246227, 14247314:14251082, 14252416:14252956, 14254513:14255683, 14257016:14263296, 14271321:14271892, 14277653:14278780, 14278828:14282074, 14282140:14283469, 14285325:14285933, 14287089:14287825, 14295954:14297286, 14298325:14298754, 14299780:14305069, 14307220:14311756, 14312842:14316649, 14318522:14319113, 14319913:14320341, 14320592:14322236, 14328965:14331084, 14339839:14340790, 14343815:14347548, 14358326:14358561, 14367594:14367615, 14369773:14371137, 14376086:14376094, 14376115:14378385, 14379425:14381359, 14384737:14390332, 14400226:14400682, 14405649:14406027, 14417934:14418196, 14431913:14432790, 14433750:14434641, 14435706:14436809, 14451889:14461037, 14465145:14465379, 14468214:14468331, 14469103:14469499, 14486767:14486876, 14496560:14496815, 14498295:14499734, 14510552:14511430, 14518719:14529528, 14546870:14552029, 14565853:14566306, 14566783:14567168, 14616523:14620182, 14620637:14621168, 14631094:14634452, 14642568:14642713, 14646601:14647781, 14651288:14652115, 14657275:14657354, 14661495:14662523, 14663141:14664448, 14674841:14675599, 14682566:14688626, 14692144:14694153, 14700392:14701121, 14706522:14707058, 14707875:14708194, 14719027:14719166, 14720477:14720587, 14726131:14726280, 14730738:14732930, 14733669:14735222, 14737195:14737701, 14738796:14738881, 14742671:14742764, 14747789:14748617, 14753777:14755153, 14756412:14756903, 14759473:14759929, 14760626:14760637, 14768989:14770688, 14775007:14775183, 14776663:14777136, 14787804:14788124, 14789517:14790243, 14790934:14791039, 14797167:14797659, 14798705:14806328, 14806473:14807461, 14820235:14820404, 14821651:14822049, 14823996:14824492, 14825952:14826557, 14827779:14828199, 14829594:14829731, 14834841:14838916, 14839138:14839855, 14843128:14846475, 14855827:14855900, 14863482:14867923, 14871365:14872111, 14875436:14875822, 14900322:14906269, 14906399:14906791, 14913083:14913765, 14952226:14953489, 14955492:14955635, 14957097:14961496, 14965361:14967300, 14968540:14975418, 14980596:14984177, 14990016:14993823, 14993959:14994580, 15001691:15001852, 15013904:15014836, 15016595:15017282, 15022990:15023167, 15030708:15031140, 15033039:15034514, 15035467:15036008, 15041274:15041432, 15050700:15050796, 15052420:15053009, 15058994:15060298, 15061159:15062082, 15063836:15064300, 15064453:15064837, 15068355:15069279, 15071033:15072265, 15072426:15072434]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))
</pre></div>
</div>
</div>
</div>
<p>We can slice the sequence by this new Feature to return the complete intergenic sequence! However, because the automated naming produces a crazy long name, we reset that.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">intergenic</span><span class="o">.</span><span class="n">get_slice</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">not gene</td><td><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span></td></tr>
</table>
<p><i>DnaSequence, length=4,010,835 (truncated to 60)</i></p>
</div></div></div>
</div>
</section>
</section>
<section id="how-to-mask-annotated-regions">
<h3>How to mask annotated regions<a class="headerlink" href="#how-to-mask-annotated-regions" title="Link to this heading">#</a></h3>
<section id="masking-annotated-regions-on-a-sequence">
<h4>Masking annotated regions on a sequence<a class="headerlink" href="#masking-annotated-regions-on-a-sequence" title="Link to this heading">#</a></h4>
<p>We can mask a certain annotation using <code class="docutils literal notranslate"><span class="pre">with_masked_annotations()</span></code></p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span><span class="s2">&quot;data/C-elegans-chromosome-I.gb&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">no_cds</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;CDS&quot;</span><span class="p">)</span>
<span class="n">no_cds</span><span class="p">[</span><span class="mi">2575800</span><span class="p">:</span><span class="mi">2575900</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">I</td><td><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="C_dna">C</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="C_dna">C</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span></td></tr>
</table>
<p><i>DnaSequence, length=100 (truncated to 60)</i></p>
</div></div></div>
</div>
<p>The above sequence could then have positions filtered so no position with the ambiguous character ‘?’ was present.</p>
</section>
<section id="masking-annotated-regions-on-an-alignment">
<h4>Masking annotated regions on an Alignment<a class="headerlink" href="#masking-annotated-regions-on-an-alignment" title="Link to this heading">#</a></h4>
<p>We can mask exons on an alignment.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">make_aligned_seqs</span>

<span class="n">aln</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">([[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;C-CCCAAAAAGGGAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;-T----TTTTG-GTT&quot;</span><span class="p">]],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">exon</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span>
    <span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span>
    <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;exon-be-gone&quot;</span><span class="p">,</span>
    <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span>
    <span class="n">on_alignment</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">aln</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">mask_char</span><span class="o">=</span><span class="s2">&quot;?&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">-</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="A_dna">A</span><span class="A_dna">A</span></td></tr>
<tr><td class="label">y</td><td><span class="ambig_dna">-</span><span class="T_dna">T</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">.</span><span class="ambig_dna">-</span><span class="G_dna">.</span><span class="T_dna">T</span><span class="T_dna">T</span></td></tr>
</table>
<p><i>2 x 15 dna alignment</i></p>
</div></div></div>
</div>
<p>After a reverse complement operation</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rc</span> <span class="o">=</span> <span class="n">aln</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="n">rc</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="T_dna">T</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="ambig_dna">-</span><span class="G_dna">G</span></td></tr>
<tr><td class="label">y</td><td><span class="A_dna">A</span><span class="A_dna">A</span><span class="C_dna">.</span><span class="ambig_dna">-</span><span class="C_dna">.</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="A_dna">A</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 15 dna alignment</i></p>
</div></div></div>
</div>
<p>these persist.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rc</span><span class="o">.</span><span class="n">with_masked_annotations</span><span class="p">(</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">mask_char</span><span class="o">=</span><span class="s2">&quot;?&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="T_dna">T</span><span class="T_dna">T</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">?</span><span class="terminal_ambig_dna">-</span><span class="terminal_ambig_dna">?</span></td></tr>
<tr><td class="label">y</td><td><span class="A_dna">A</span><span class="A_dna">A</span><span class="C_dna">.</span><span class="ambig_dna">-</span><span class="C_dna">.</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="A_dna">A</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 15 dna alignment</i></p>
</div></div></div>
</div>
</section>
</section>
<section id="how-to-find-the-children-of-a-feature">
<h3>How to find the “children” of a Feature<a class="headerlink" href="#how-to-find-the-children-of-a-feature" title="Link to this heading">#</a></h3>
<p>To find the “children” of a feature, we can use the <code class="docutils literal notranslate"><span class="pre">get_children()</span></code> method. A “child” refers to a feature that is nested within or contained by another “parent” feature. For example, a child feature could be an exon contained within a gene or a CDS contained within a transcript.</p>
<p>This method returns a generator that yields all the child features of the specified feature.</p>
<p>For example, let’s find the children of the gene “WBGene00021661”:</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">cogent3</span><span class="w"> </span><span class="kn">import</span> <span class="n">load_seq</span>

<span class="n">seq</span> <span class="o">=</span> <span class="n">load_seq</span><span class="p">(</span>
    <span class="s2">&quot;data/C-elegans-chromosome-I.fa&quot;</span><span class="p">,</span>
    <span class="n">annotation_path</span><span class="o">=</span><span class="s2">&quot;data/C-elegans-chromosome-I.gff&quot;</span><span class="p">,</span>
    <span class="n">label_to_name</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">gene</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;gene:WBGene00022276&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;gene&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">children</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">get_children</span><span class="p">())</span>
<span class="n">children</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&#39;I&#39;, biotype=&#39;mRNA&#39;, name=&#39;transcript:Y74C9A.2a.3&#39;, map=[11494:16793]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;mRNA&#39;, name=&#39;transcript:Y74C9A.2a.1&#39;, map=[11494:16837]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;mRNA&#39;, name=&#39;transcript:Y74C9A.2a.2&#39;, map=[11617:16833]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434)),
 Feature(seqid=&#39;I&#39;, biotype=&#39;mRNA&#39;, name=&#39;transcript:Y74C9A.2b.1&#39;, map=[15102:16585]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))]
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-find-the-parent-of-a-feature">
<h3>How to find the “parent” of a Feature<a class="headerlink" href="#how-to-find-the-parent-of-a-feature" title="Link to this heading">#</a></h3>
<p>To find the “parent” of a feature, we can use the <code class="docutils literal notranslate"><span class="pre">get_parent()</span></code> method, which achieves the inverse of the above method.</p>
<p>For example, we can use the first “child” we returned above, <code class="docutils literal notranslate"><span class="pre">&quot;transcript:Y74C9A.2a.3&quot;</span></code>, to find the original parent gene!</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">child</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;transcript:Y74C9A.2a.3&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;mRNA&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">parent</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">child</span><span class="o">.</span><span class="n">get_parent</span><span class="p">())</span>
<span class="n">parent</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&#39;I&#39;, biotype=&#39;gene&#39;, name=&#39;gene:WBGene00022276&#39;, map=[11494:16837]/15072434, parent=DnaSequence(GCCTAAG... 15,072,434))]
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-copy-features">
<h3>How to copy features<a class="headerlink" href="#how-to-copy-features" title="Link to this heading">#</a></h3>
<p>We can copy features onto sequences with the same name. Note that the <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instance bound to the alignment and its member sequences is the <strong>same</strong>.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">aln2</span> <span class="o">=</span> <span class="n">make_aligned_seqs</span><span class="p">([[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;-AAAAAAAAA&quot;</span><span class="p">],</span> <span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTT--TTTT&quot;</span><span class="p">]],</span>
    <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">aln2</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;x&quot;</span><span class="p">),</span> <span class="n">aln2</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="s2">&quot;y&quot;</span><span class="p">)</span>
<span class="n">x</span><span class="o">.</span><span class="n">annotation_db</span> <span class="ow">is</span> <span class="n">y</span><span class="o">.</span><span class="n">annotation_db</span> <span class="ow">is</span> <span class="n">aln2</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>True
</pre></div>
</div>
</div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Despite this, it is possible for the attributes to get out-of-sync. So, any copy annotations should be done using <code class="docutils literal notranslate"><span class="pre">alignment.copy_annotations()</span></code>, <strong>not</strong> <code class="docutils literal notranslate"><span class="pre">alignment.get_seq(&quot;x&quot;).copy_annotations()</span></code>.</p>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">seq</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCCCCCCCCCCCCCCCC&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">match_exon</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">)])</span>
<span class="n">aln2</span><span class="o">.</span><span class="n">copy_annotations</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">annotation_db</span><span class="p">)</span>
<span class="n">aln2</span><span class="o">.</span><span class="n">annotation_db</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>BasicAnnotationDb(source=&#39;:memory:&#39;, total_records=1)
</pre></div>
</div>
</div>
</div>
<p>However, if the feature lies outside the sequence being copied to, you get a lost span</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">copied</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">aln2</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">seqid</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;exon&quot;</span><span class="p">))</span>
<span class="n">copied</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-none notranslate"><div class="highlight"><pre><span></span>[Feature(seqid=&quot;from &#39;x&#39;&quot;, biotype=&#39;exon&#39;, name=&#39;A&#39;, map=[4:9]/10, parent=2 x 10 dna alignment: x[-AAAAAAAAA], y[TTTT--TTTT])]
</pre></div>
</div>
</div>
</div>
</section>
<section id="how-to-get-the-positions-of-a-feature-as-one-span">
<h3>How to get the positions of a feature as one span<a class="headerlink" href="#how-to-get-the-positions-of-a-feature-as-one-span" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">as_one_span</span></code> unifies features with discontinuous alignment coordinates and returns positions spanned by a feature, including gaps.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">unified</span> <span class="o">=</span> <span class="n">aln_exons</span><span class="o">.</span><span class="n">as_one_span</span><span class="p">()</span>
<span class="n">aln3</span><span class="p">[</span><span class="n">unified</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3align table {margin: 10px 0;}
.c3align td { border: none !important; text-align: left !important; }
.c3align tr:not(.num_row) td span {margin: 0 2px;}
.c3align tr:nth-child(even) {background: #f7f7f7;}
.c3align .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3align .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; display: table-cell !important; font-weight: normal !important; }
.c3align .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3align .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3align .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3align .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3align .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3align .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3align">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">x</td><td><span class="C_dna">C</span><span class="ambig_dna">-</span><span class="C_dna">C</span><span class="C_dna">C</span><span class="C_dna">C</span></td></tr>
<tr><td class="label">y</td><td><span class="ambig_dna">-</span><span class="T_dna">T</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span><span class="ambig_dna">-</span></td></tr>
</table>
<p><i>2 x 5 dna alignment</i></p>
</div></div></div>
</div>
</section>
<section id="behaviour-of-annotations-on-nucleic-acid-sequences">
<h3>Behaviour of annotations on nucleic acid sequences<a class="headerlink" href="#behaviour-of-annotations-on-nucleic-acid-sequences" title="Link to this heading">#</a></h3>
<p>Reverse complementing a sequence <strong>does not</strong> reverse features. Features are considered to have strand specific meaning (e.g. CDS, exons) and so they retain the reference to the frame for which they were defined.</p>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plus</span> <span class="o">=</span> <span class="n">make_seq</span><span class="p">(</span><span class="s2">&quot;CCCCCAAAAAAAAAATTTTTTTTTTAAAGG&quot;</span><span class="p">,</span> <span class="n">moltype</span><span class="o">=</span><span class="s2">&quot;dna&quot;</span><span class="p">)</span>
<span class="n">plus_rpt</span> <span class="o">=</span> <span class="n">plus</span><span class="o">.</span><span class="n">add_feature</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;blah&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="n">spans</span><span class="o">=</span><span class="p">[(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">15</span><span class="p">),</span> <span class="p">(</span><span class="mi">25</span><span class="p">,</span> <span class="mi">28</span><span class="p">)])</span>
<span class="n">plus</span><span class="p">[</span><span class="n">plus_rpt</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">a</td><td><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span></td></tr>
</table>
<p><i>DnaSequence, length=13</i></p>
</div></div></div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">minus</span> <span class="o">=</span> <span class="n">plus</span><span class="o">.</span><span class="n">rc</span><span class="p">()</span>
<span class="n">minus</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">None</td><td><span class="C_dna">C</span><span class="C_dna">C</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="T_dna">T</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span><span class="G_dna">G</span></td></tr>
</table>
<p><i>DnaSequence, length=30</i></p>
</div></div></div>
</div>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">minus_rpt</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">minus</span><span class="o">.</span><span class="n">get_features</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s2">&quot;blah&quot;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">minus</span><span class="p">[</span><span class="n">minus_rpt</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style>
.c3seq table {margin: 10px 0;}
.c3seq td { border: none !important; text-align: left !important; }
.c3seq tr:not(.num_row) td span {margin: 0 2px;}
.c3seq tr:nth-child(even) {background: #f7f7f7;}
.c3seq .num_row {background-color:rgba(161, 195, 209, 0.5) !important; border-top: solid 1px black; }
.c3seq .label { font-size: 12pt ; text-align: right !important; color: black !important; padding: 0 4px; }}
.c3seq .T_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: blue; }
.c3seq .C_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: black; }
.c3seq .A_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: #FF0102; }
.c3seq .G_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: green; }
.c3seq .terminal_ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
.c3seq .ambig_dna{font-family: "Lucida Console",monospace !important; font-size: 12pt !important; color: gray; }
</style>
<div class="c3seq">
<table>
<tr class="num_row"><td></td><td><b>0</b></td></tr>
<tr><td class="label">a</td><td><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span><span class="A_dna">A</span></td></tr>
</table>
<p><i>DnaSequence, length=13</i></p>
</div></div></div>
</div>
</section>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="alignments.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Sequence Collections and Alignments</p>
      </div>
    </a>
    <a class="right-next"
       href="annotation_db.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Annotation Databases</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-create-a-custom-feature">How to create a custom <code class="docutils literal notranslate"><span class="pre">Feature</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#via-add-feature">Via <code class="docutils literal notranslate"><span class="pre">add_feature</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#on-a-sequence">On a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#on-a-sequence-within-an-alignment">On a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> within an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#on-an-alignment">On an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#via-an-annotationdb">Via an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-load-bulk-features-from-a-file">How to load bulk Features from a File</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#from-a-genbank-file">From a Genbank file</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#directly-using-the-genbank-parser-to-get-features-as-python-primitives">Directly using the genbank parser to get features as Python primitives</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-load-features-and-sequence-data">How to load features and sequence data</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#from-a-gff-file">From a GFF file</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">How to load features and sequence data</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-load-features-and-associate-them-with-an-existing-sequence">How to load features and associate them with an existing sequence</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-load-features-and-associate-them-with-sequences-in-an-existing-alignment">How to load features and associate them with sequences in an existing alignment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-query-a-sequence-or-alignment-for-features">How to query a Sequence or Alignment for Features</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-a-sequence-for-features">Querying a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> for Features</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-feature-name">Querying via Feature Name</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-feature-biotype">Querying via Feature Biotype</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-via-region-of-interest">Querying via region of interest</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-a-sequence-via-an-alignment-for-features">Querying a Sequence (via an Alignment) for Features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-an-alignment-for-features">Querying an Alignment for Features</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-for-features-on-any-sequence-in-an-alignment">Querying for features on any <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> in an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-for-features-on-an-alignment">Querying for features on an <code class="docutils literal notranslate"><span class="pre">Alignment</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#querying-features-that-span-gaps-in-alignments">Querying features that span gaps in alignments</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#examples-using-the-methods-available-on-features">Examples using the methods available on Features</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-slice-a-sequence-or-alignment-by-its-features">How to slice a <code class="docutils literal notranslate"><span class="pre">Sequence</span></code> or <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> by its features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-display-the-features-of-a-sequence">How to display the features of a Sequence</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-the-coordinates-of-a-feature">How to find the coordinates of a feature</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-take-the-union-of-features">How to take the union of features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-get-the-shadow-of-a-feature">How to get the shadow of a Feature</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-use-the-shadow-of-a-feature-to-return-the-intergenic-sequence">How to use the shadow of a Feature to return the intergenic sequence</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-mask-annotated-regions">How to mask annotated regions</a><ul class="visible nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#masking-annotated-regions-on-a-sequence">Masking annotated regions on a sequence</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#masking-annotated-regions-on-an-alignment">Masking annotated regions on an Alignment</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-the-children-of-a-feature">How to find the “children” of a Feature</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-find-the-parent-of-a-feature">How to find the “parent” of a Feature</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-copy-features">How to copy features</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#how-to-get-the-positions-of-a-feature-as-one-span">How to get the positions of a feature as one span</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#behaviour-of-annotations-on-nucleic-acid-sequences">Behaviour of annotations on nucleic acid sequences</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020-2026, cogent3 Team.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>