
<!DOCTYPE html>


<html lang="en" data-content_root="../../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>DnaSequence &#8212; cogent3 2023.12.12 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/jupyter-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/thebelab.css" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />
    <link rel="stylesheet" type="text/css" href="../../../../_static/custom.css?v=da3cec82" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script src="../../../../_static/documentation_options.js?v=d202a915"></script>
    <script src="../../../../_static/doctools.js?v=888ff710"></script>
    <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../../_static/thebelab-helper.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../../../_static/design-tabs.js?v=36754332"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'doc/api/sequence/classes/cogent3.core.sequence.DnaSequence';</script>
    <link rel="icon" href="../../../../_static/c3-square.svg"/>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="RnaSequence" href="cogent3.core.sequence.RnaSequence.html" />
    <link rel="prev" title="sequence" href="../sequence.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../../_static/c3-logo.png" class="logo__image only-light" alt="cogent3 2023.12.12 documentation - Home"/>
    <script>document.write(`<img src="../../../../_static/c3-logo.png" class="logo__image only-dark" alt="cogent3 2023.12.12 documentation - Home"/>`);</script>
  
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../draw/index.html">
                        Image Gallery
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../install.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        User Guides
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../community.html">
                        Community
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../licenses.html">
                        License
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../history.html">
                        Project History
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../pycogent.html">
                        How does cogent3 relate to PyCogent?
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../projects.html">
                        Projects using cogent3
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../data_file_links.html">
                        Data Files Used in the Documentation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../genindex.html">
                        Index
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../draw/index.html">
                        Image Gallery
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../install.html">
                        Installation
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../../../index.html">
                        User Guides
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../community.html">
                        Community
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../../../licenses.html">
                        License
                      </a>
                    </li>
                
            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-controls="pst-nav-more-links-2">
                    More
                </button>
                <ul id="pst-nav-more-links-2" class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../history.html">
                        Project History
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../pycogent.html">
                        How does cogent3 relate to PyCogent?
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../projects.html">
                        Projects using cogent3
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../data_file_links.html">
                        Data Files Used in the Documentation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link dropdown-item nav-internal" href="../../../../genindex.html">
                        Index
                      </a>
                    </li>
                
                </ul>
            </li>
            
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/cogent3/cogent3" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../app/index.html">apps</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-app.html">Overview of apps</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app-overview.html">Overview of using apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app-help.html">Getting help on an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/app-get.html">Getting an app</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/dstore.html">Data stores – collections of data records</a></li>

<li class="toctree-l3"><a class="reference internal" href="../../../app/not-completed.html">Tracking records that could not be processed</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-align.html">Sequence Alignment</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/align-nucleotide.html">Using a nucleotide model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/align-codon.html">Using a codon model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/align-protein.html">Using a protein model</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-evo-model.html">Modelling sequence evolution</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-ancestral-states.html">Reconstructing ancestral states</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-evo-natsel.html">Testing for natural selection</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../app/index-custom.html">Write your own Apps</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/user_function.html">Turn your functions into composable apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html">Manipulating biological data</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/loading_sequences.html">Loading a sequence from a file</a></li>





<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html">Sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/protein_sequences.html">Protein sequences</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/alignments.html">Sequence Collections and Alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/features.html">Features</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/annotation_db.html">Annotation Databases</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/moltypes.html">Molecular types</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/alphabet.html">Alphabets</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/what_codes.html">Available genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/genetic_code.html">Using genetic codes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/simple_trees.html">Trees</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/tables.html">Tabular data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/building_alignments.html">Building alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html">Genetic distance calculation</a></li>


<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/building_phylogenies.html">Building phylogenies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/useful_utilities.html">Useful Utilities</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../examples/index.html">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../../examples/period_estimation.html">Estimating periodic signals</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/simple.html">The simplest script</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/codon_models.html">Using codon models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/parallel.html">Parallel computations</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../../index.html">API</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.open_.html">open</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.get_model.html">get_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-14" name="toctree-checkbox-14" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-14"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</li>
<li class="toctree-l2 current active has-children"><a class="reference internal" href="../sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><input checked="" class="toctree-checkbox" id="toctree-checkbox-15" name="toctree-checkbox-15" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-15"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l3 current active"><a class="current reference internal" href="#">DnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="cogent3.core.sequence.ProteinSequence.html">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-16" name="toctree-checkbox-16" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-16"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-17" name="toctree-checkbox-17" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-17"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-18" name="toctree-checkbox-18" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-18"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-19" name="toctree-checkbox-19" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-19"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-20" name="toctree-checkbox-20" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-20"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../app/composable/define_app.html">define_app</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">





<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../../index.html" class="nav-link">User Guides</a></li>
    
    
    <li class="breadcrumb-item"><i class="fa-solid fa-ellipsis"></i></li>
    
    
    <li class="breadcrumb-item"><a href="../sequence.html" class="nav-link"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a></li>
    
    <li class="breadcrumb-item active" aria-current="page">DnaSequence</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="dnasequence">
<h1>DnaSequence<a class="headerlink" href="#dnasequence" title="Link to this heading">#</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DnaSequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_case</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gaps_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wildcards_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence" title="Link to this definition">#</a></dt>
<dd><p>Holds the standard DNA sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_db</strong></dt><dd></dd>
<dt><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.annotation_offset" title="cogent3.core.sequence.DnaSequence.annotation_offset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotation_offset</span></code></a></dt><dd><p>The offset between annotation coordinates and sequence coordinates.</p>
</dd>
<dt><strong>codon_alphabet</strong></dt><dd></dd>
<dt><strong>line_wrap</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.add_feature" title="cogent3.core.sequence.DnaSequence.add_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_feature</span></code></a>(*, biotype, name, spans[, ...])</p></td>
<td><p>add a feature to annotation_db</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.annotate_from_gff" title="cogent3.core.sequence.DnaSequence.annotate_from_gff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_from_gff</span></code></a>(f[, offset])</p></td>
<td><p>copies annotations from a gff file to self,</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.annotate_matches_to" title="cogent3.core.sequence.DnaSequence.annotate_matches_to"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_matches_to</span></code></a>(pattern, biotype, name)</p></td>
<td><p>Adds an annotation at sequence positions matching pattern.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.can_match" title="cogent3.core.sequence.DnaSequence.can_match"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_match</span></code></a>(other)</p></td>
<td><p>Returns True if every pos in self could match same pos in other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.can_mismatch" title="cogent3.core.sequence.DnaSequence.can_mismatch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_mismatch</span></code></a>(other)</p></td>
<td><p>Returns True if any position in self could mismatch with other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.can_mispair" title="cogent3.core.sequence.DnaSequence.can_mispair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_mispair</span></code></a>(other)</p></td>
<td><p>Returns True if any position in self could mispair with other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.can_pair" title="cogent3.core.sequence.DnaSequence.can_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_pair</span></code></a>(other)</p></td>
<td><p>Returns True if self and other could pair.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.complement" title="cogent3.core.sequence.DnaSequence.complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complement</span></code></a>()</p></td>
<td><p>Returns complement of self, using data from MolType.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.copy" title="cogent3.core.sequence.DnaSequence.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>([exclude_annotations])</p></td>
<td><p>returns a copy of self</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.copy_annotations" title="cogent3.core.sequence.DnaSequence.copy_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_annotations</span></code></a>(seq_db)</p></td>
<td><p>copy annotations into attached annotation db</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.count" title="cogent3.core.sequence.DnaSequence.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(item)</p></td>
<td><p>count() delegates to self._seq.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.count_degenerate" title="cogent3.core.sequence.DnaSequence.count_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_degenerate</span></code></a>()</p></td>
<td><p>Counts the degenerate bases in the specified sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.count_gaps" title="cogent3.core.sequence.DnaSequence.count_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_gaps</span></code></a>()</p></td>
<td><p>Counts the gaps in the specified sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.counts" title="cogent3.core.sequence.DnaSequence.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>([motif_length, include_ambiguity, ...])</p></td>
<td><p>returns dict of counts of motifs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.degap" title="cogent3.core.sequence.DnaSequence.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>()</p></td>
<td><p>Deletes all gap characters from sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.diff" title="cogent3.core.sequence.DnaSequence.diff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diff</span></code></a>(other)</p></td>
<td><p>Returns number of differences between self and other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.disambiguate" title="cogent3.core.sequence.DnaSequence.disambiguate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">disambiguate</span></code></a>([method])</p></td>
<td><p>Returns a non-degenerate sequence from a degenerate one.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.distance" title="cogent3.core.sequence.DnaSequence.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a>(other[, function])</p></td>
<td><p>Returns distance between self and other using function(i,j).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.first_degenerate" title="cogent3.core.sequence.DnaSequence.first_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_degenerate</span></code></a>()</p></td>
<td><p>Returns the index of first degenerate symbol in sequence, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.first_gap" title="cogent3.core.sequence.DnaSequence.first_gap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_gap</span></code></a>()</p></td>
<td><p>Returns the index of the first gap in the sequence, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.first_invalid" title="cogent3.core.sequence.DnaSequence.first_invalid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_invalid</span></code></a>()</p></td>
<td><p>Returns the index of first invalid symbol in sequence, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.first_non_strict" title="cogent3.core.sequence.DnaSequence.first_non_strict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_non_strict</span></code></a>()</p></td>
<td><p>Returns the index of first non-strict symbol in sequence, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_diff" title="cogent3.core.sequence.DnaSequence.frac_diff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other differ.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_diff_gaps" title="cogent3.core.sequence.DnaSequence.frac_diff_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff_gaps</span></code></a>(other)</p></td>
<td><p>Returns frac.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_diff_non_gaps" title="cogent3.core.sequence.DnaSequence.frac_diff_non_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff_non_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of non-gap positions where self differs from other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_same" title="cogent3.core.sequence.DnaSequence.frac_same"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other are the same.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_same_gaps" title="cogent3.core.sequence.DnaSequence.frac_same_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other share gap states.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_same_non_gaps" title="cogent3.core.sequence.DnaSequence.frac_same_non_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same_non_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of non-gap positions where self matches other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.frac_similar" title="cogent3.core.sequence.DnaSequence.frac_similar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_similar</span></code></a>(other, similar_pairs)</p></td>
<td><p>Returns fraction of positions where self[i] is similar to other[i].</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.gap_indices" title="cogent3.core.sequence.DnaSequence.gap_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_indices</span></code></a>()</p></td>
<td><p>Returns list of indices of all gaps in the sequence, or [].</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.gap_maps" title="cogent3.core.sequence.DnaSequence.gap_maps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_maps</span></code></a>()</p></td>
<td><p>Returns dicts mapping between gapped and ungapped positions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.gap_vector" title="cogent3.core.sequence.DnaSequence.gap_vector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_vector</span></code></a>()</p></td>
<td><p>Returns vector of True or False according to which pos are gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_drawable" title="cogent3.core.sequence.DnaSequence.get_drawable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawable</span></code></a>(*[, biotype, width, vertical])</p></td>
<td><p>make a figure from sequence features</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_drawables" title="cogent3.core.sequence.DnaSequence.get_drawables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawables</span></code></a>(*[, biotype])</p></td>
<td><p>returns a dict of drawables, keyed by type</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_features" title="cogent3.core.sequence.DnaSequence.get_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_features</span></code></a>(*[, biotype, name, start, ...])</p></td>
<td><p>yields Feature instances</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_in_motif_size" title="cogent3.core.sequence.DnaSequence.get_in_motif_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_in_motif_size</span></code></a>([motif_length, log_warnings])</p></td>
<td><p>returns sequence as list of non-overlapping motifs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_kmers" title="cogent3.core.sequence.DnaSequence.get_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kmers</span></code></a>(k[, strict])</p></td>
<td><p>return all overlapping k-mers</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_name" title="cogent3.core.sequence.DnaSequence.get_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_name</span></code></a>()</p></td>
<td><p>Return the sequence name -- should just use name instead.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_translation" title="cogent3.core.sequence.DnaSequence.get_translation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_translation</span></code></a>([gc, incomplete_ok, ...])</p></td>
<td><p>translate to amino acid sequence</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.get_type" title="cogent3.core.sequence.DnaSequence.get_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_type</span></code></a>()</p></td>
<td><p>Return the sequence type as moltype label.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.has_terminal_stop" title="cogent3.core.sequence.DnaSequence.has_terminal_stop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_terminal_stop</span></code></a>([gc, strict, allow_partial])</p></td>
<td><p>Return True if the sequence has a terminal stop codon.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.is_annotated" title="cogent3.core.sequence.DnaSequence.is_annotated"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_annotated</span></code></a>()</p></td>
<td><p>returns True if sequence has any annotations</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.is_degenerate" title="cogent3.core.sequence.DnaSequence.is_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_degenerate</span></code></a>()</p></td>
<td><p>Returns True if sequence contains degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.is_gap" title="cogent3.core.sequence.DnaSequence.is_gap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_gap</span></code></a>([char])</p></td>
<td><p>Returns True if char is a gap.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.is_gapped" title="cogent3.core.sequence.DnaSequence.is_gapped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_gapped</span></code></a>()</p></td>
<td><p>Returns True if sequence contains gaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.is_strict" title="cogent3.core.sequence.DnaSequence.is_strict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_strict</span></code></a>()</p></td>
<td><p>Returns True if sequence contains only monomers.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.is_valid" title="cogent3.core.sequence.DnaSequence.is_valid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_valid</span></code></a>()</p></td>
<td><p>Returns True if sequence contains no items absent from alphabet.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.iter_kmers" title="cogent3.core.sequence.DnaSequence.iter_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_kmers</span></code></a>(k[, strict])</p></td>
<td><p>generates all overlapping k-mers.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.make_feature" title="cogent3.core.sequence.DnaSequence.make_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_feature</span></code></a>(feature, *args)</p></td>
<td><p>return an Feature instance from feature data</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.matrix_distance" title="cogent3.core.sequence.DnaSequence.matrix_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matrix_distance</span></code></a>(other, matrix)</p></td>
<td><p>Returns distance between self and other using a score matrix.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.must_match" title="cogent3.core.sequence.DnaSequence.must_match"><code class="xref py py-obj docutils literal notranslate"><span class="pre">must_match</span></code></a>(other)</p></td>
<td><p>Returns True if all positions in self must match positions in other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.must_pair" title="cogent3.core.sequence.DnaSequence.must_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">must_pair</span></code></a>(other)</p></td>
<td><p>Returns True if all positions in self must pair with other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.mw" title="cogent3.core.sequence.DnaSequence.mw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mw</span></code></a>([method, delta])</p></td>
<td><p>Returns the molecular weight of (one strand of) the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.parse_out_gaps" title="cogent3.core.sequence.DnaSequence.parse_out_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_out_gaps</span></code></a>()</p></td>
<td><p>returns Map corresponding to gap locations and ungapped Sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.possibilities" title="cogent3.core.sequence.DnaSequence.possibilities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">possibilities</span></code></a>()</p></td>
<td><p>Counts number of possible sequences matching the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.rc" title="cogent3.core.sequence.DnaSequence.rc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rc</span></code></a>()</p></td>
<td><p>Converts a nucleic acid sequence to its reverse complement.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.replace" title="cogent3.core.sequence.DnaSequence.replace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code></a>(oldchar, newchar)</p></td>
<td><p>return new instance with oldchar replaced by newchar</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.replace_annotation_db" title="cogent3.core.sequence.DnaSequence.replace_annotation_db"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_annotation_db</span></code></a>(value[, check])</p></td>
<td><p>public interface to assigning the annotation_db</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.resolved_ambiguities" title="cogent3.core.sequence.DnaSequence.resolved_ambiguities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">resolved_ambiguities</span></code></a>()</p></td>
<td><p>Returns a list of tuples of strings.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.reverse_complement" title="cogent3.core.sequence.DnaSequence.reverse_complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_complement</span></code></a>()</p></td>
<td><p>Converts a nucleic acid sequence to its reverse complement.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.shuffle" title="cogent3.core.sequence.DnaSequence.shuffle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shuffle</span></code></a>()</p></td>
<td><p>returns a randomized copy of the Sequence object</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.sliding_windows" title="cogent3.core.sequence.DnaSequence.sliding_windows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sliding_windows</span></code></a>(window, step[, start, end])</p></td>
<td><p>Generator function that yield new sequence objects of a given length at a given interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.strand_symmetry" title="cogent3.core.sequence.DnaSequence.strand_symmetry"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strand_symmetry</span></code></a>([motif_length])</p></td>
<td><p>returns G-test for strand symmetry</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.strip_bad" title="cogent3.core.sequence.DnaSequence.strip_bad"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_bad</span></code></a>()</p></td>
<td><p>Removes any symbols not in the alphabet.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.strip_bad_and_gaps" title="cogent3.core.sequence.DnaSequence.strip_bad_and_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_bad_and_gaps</span></code></a>()</p></td>
<td><p>Removes any symbols not in the alphabet, and any gaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.strip_degenerate" title="cogent3.core.sequence.DnaSequence.strip_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_degenerate</span></code></a>()</p></td>
<td><p>Removes degenerate bases by stripping them out of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_dna" title="cogent3.core.sequence.DnaSequence.to_dna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dna</span></code></a>()</p></td>
<td><p>Returns copy of self as DNA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_fasta" title="cogent3.core.sequence.DnaSequence.to_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_fasta</span></code></a>([make_seqlabel, block_size])</p></td>
<td><p>Return string of self in FASTA format, no trailing newline</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_html" title="cogent3.core.sequence.DnaSequence.to_html"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_html</span></code></a>([wrap, limit, colors, font_size, ...])</p></td>
<td><p>returns html with embedded styles for sequence colouring</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_json" title="cogent3.core.sequence.DnaSequence.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns a json formatted string</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_moltype" title="cogent3.core.sequence.DnaSequence.to_moltype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_moltype</span></code></a>(moltype)</p></td>
<td><p>returns copy of self with moltype seq</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_rich_dict" title="cogent3.core.sequence.DnaSequence.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>([exclude_annotations])</p></td>
<td><p>returns {'name': name, 'seq': sequence, 'moltype': moltype.label}</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.to_rna" title="cogent3.core.sequence.DnaSequence.to_rna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rna</span></code></a>()</p></td>
<td><p>Returns copy of self as RNA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.trim_stop_codon" title="cogent3.core.sequence.DnaSequence.trim_stop_codon"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim_stop_codon</span></code></a>([gc, strict, allow_partial])</p></td>
<td><p>Removes a terminal stop codon from the sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.with_masked_annotations" title="cogent3.core.sequence.DnaSequence.with_masked_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_masked_annotations</span></code></a>(annot_types[, ...])</p></td>
<td><p>returns a sequence with annot_types regions replaced by mask_char if shadow is False, otherwise all other regions are masked.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.DnaSequence.with_termini_unknown" title="cogent3.core.sequence.DnaSequence.with_termini_unknown"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_termini_unknown</span></code></a>()</p></td>
<td><p>Returns copy of sequence with terminal gaps remapped as missing.</p></td>
</tr>
</tbody>
</table>
<table class="table">
<tbody>
<tr class="row-odd"><td><p><strong>gapped_by_map</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>gapped_by_map_motif_iter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gapped_by_map_segment_iter</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spans</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.add_feature" title="Link to this definition">#</a></dt>
<dd><p>add a feature to annotation_db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>biological type</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>spans</strong></dt><dd><p>coordinates for this sequence</p>
</dd>
<dt><strong>parent_id</strong></dt><dd><p>name of the feature parent</p>
</dd>
<dt><strong>strand</strong></dt><dd><p>‘+’ or ‘-’, defaults to ‘+’</p>
</dd>
<dt><strong>on_alignment</strong></dt><dd><p>whether the feature spans are alignment coordinates</p>
</dd>
<dt><strong>seqid</strong></dt><dd><p>ignored since the feature is added to this sequence</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Feature instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.annotate_from_gff">
<span class="sig-name descname"><span class="pre">annotate_from_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PathLike</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.annotate_from_gff" title="Link to this definition">#</a></dt>
<dd><p>copies annotations from a gff file to self,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>f</strong><span class="classifier">path to gff annotation file.</span></dt><dd></dd>
<dt><strong>offset</strong><span class="classifier">Optional, the offset between annotation coordinates and sequence coordinates.</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.annotate_matches_to">
<span class="sig-name descname"><span class="pre">annotate_matches_to</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_multiple</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.annotate_matches_to" title="Link to this definition">#</a></dt>
<dd><p>Adds an annotation at sequence positions matching pattern.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pattern</strong></dt><dd><p>The search string for which annotations are made. IUPAC ambiguities
are converted to regex on sequences with the appropriate MolType.</p>
</dd>
<dt><strong>biotype</strong></dt><dd><p>The type of the annotation (e.g. “domain”).</p>
</dd>
<dt><strong>name</strong></dt><dd><p>The name of the annotation.</p>
</dd>
<dt><strong>allow_multiple</strong></dt><dd><p>If True, allows multiple occurrences of the input pattern. Otherwise,
only the first match is used.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Returns a list of Feature instances.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.annotation_db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation_db</span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.annotation_db" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.annotation_offset">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation_offset</span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.annotation_offset" title="Link to this definition">#</a></dt>
<dd><p>The offset between annotation coordinates and sequence coordinates.</p>
<p>The offset can be used to adjust annotation coordinates to match the position
of the given Sequence within a larger genomic context. For example, if the
Annotations are with respect to a chromosome, and the sequence represents
a gene that is 100 bp from the start of a chromosome, the offset can be set to
100 to ensure that the gene’s annotations are aligned with the appropriate
genomic positions.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>int: The offset between annotation coordinates and sequence coordinates.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.can_match">
<span class="sig-name descname"><span class="pre">can_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.can_match" title="Link to this definition">#</a></dt>
<dd><p>Returns True if every pos in self could match same pos in other.</p>
<p>Truncates at length of shorter sequence.
gaps are only allowed to match other gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.can_mismatch">
<span class="sig-name descname"><span class="pre">can_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.can_mismatch" title="Link to this definition">#</a></dt>
<dd><p>Returns True if any position in self could mismatch with other.</p>
<p>Truncates at length of shorter sequence.
gaps are always counted as matches.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.can_mispair">
<span class="sig-name descname"><span class="pre">can_mispair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.can_mispair" title="Link to this definition">#</a></dt>
<dd><p>Returns True if any position in self could mispair with other.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
<p>Truncates at length of shorter sequence.
gaps are always counted as possible mispairs, as are weak pairs like GU.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.can_pair">
<span class="sig-name descname"><span class="pre">can_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.can_pair" title="Link to this definition">#</a></dt>
<dd><p>Returns True if self and other could pair.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
<p>Truncates at length of shorter sequence.
gaps are only allowed to pair with other gaps, and are counted as ‘weak’
(same category as GU and degenerate pairs).</p>
<p>NOTE: second must be able to be reverse</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.codon_alphabet">
<span class="sig-name descname"><span class="pre">codon_alphabet</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.codon_alphabet" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.complement">
<span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.complement" title="Link to this definition">#</a></dt>
<dd><p>Returns complement of self, using data from MolType.</p>
<p>Always tries to return same type as item: if item looks like a dict,
will return list of keys.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exclude_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.copy" title="Link to this definition">#</a></dt>
<dd><p>returns a copy of self</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.copy_annotations">
<span class="sig-name descname"><span class="pre">copy_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SupportsFeatures</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.copy_annotations" title="Link to this definition">#</a></dt>
<dd><p>copy annotations into attached annotation db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_db</strong></dt><dd><p>compatible annotation db</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Only copies annotations for records with seqid equal to self.name</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.count">
<span class="sig-name descname"><span class="pre">count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.count" title="Link to this definition">#</a></dt>
<dd><p>count() delegates to self._seq.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.count_degenerate">
<span class="sig-name descname"><span class="pre">count_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.count_degenerate" title="Link to this definition">#</a></dt>
<dd><p>Counts the degenerate bases in the specified sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.count_gaps">
<span class="sig-name descname"><span class="pre">count_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.count_gaps" title="Link to this definition">#</a></dt>
<dd><p>Counts the gaps in the specified sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.counts" title="Link to this definition">#</a></dt>
<dd><p>returns dict of counts of motifs</p>
<p>only non-overlapping motifs are counted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.degap">
<span class="sig-name descname"><span class="pre">degap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.degap" title="Link to this definition">#</a></dt>
<dd><p>Deletes all gap characters from sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.diff">
<span class="sig-name descname"><span class="pre">diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.diff" title="Link to this definition">#</a></dt>
<dd><p>Returns number of differences between self and other.</p>
<p>NOTE: truncates at the length of the shorter sequence. Case-sensitive.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.disambiguate">
<span class="sig-name descname"><span class="pre">disambiguate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strip'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.disambiguate" title="Link to this definition">#</a></dt>
<dd><p>Returns a non-degenerate sequence from a degenerate one.</p>
<p>method can be ‘strip’ (deletes any characters not in monomers or gaps)
or ‘random’(assigns the possibilities at random, using equal
frequencies).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">function</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.distance" title="Link to this definition">#</a></dt>
<dd><p>Returns distance between self and other using function(i,j).</p>
<p>other must be a sequence.</p>
<p>function should be a function that takes two items and returns a
number. To turn a 2D matrix into a function, use
cogent3.util.miscs.DistanceFromMatrix(matrix).</p>
<p>NOTE: Truncates at the length of the shorter sequence.</p>
<p>Note that the function acts on two _elements_ of the sequences, not
the two sequences themselves (i.e. the behavior will be the same for
every position in the sequences, such as identity scoring or a function
derived from a distance matrix as suggested above). One limitation of
this approach is that the distance function cannot use properties of
the sequences themselves: for example, it cannot use the lengths of the
sequences to normalize the scores as percent similarities or percent
differences.</p>
<p>If you want functions that act on the two sequences themselves, there
is no particular advantage in making these functions methods of the
first sequences by passing them in as parameters like the function
in this method. It makes more sense to use them as standalone functions.
The factory function cogent3.util.transform.for_seq is useful for
converting per-element functions into per-sequence functions, since it
takes as parameters a per-element scoring function, a score aggregation
function, and a normalization function (which itself takes the two
sequences as parameters), returning a single function that combines
these functions and that acts on two complete sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.first_degenerate">
<span class="sig-name descname"><span class="pre">first_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.first_degenerate" title="Link to this definition">#</a></dt>
<dd><p>Returns the index of first degenerate symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.first_gap">
<span class="sig-name descname"><span class="pre">first_gap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.first_gap" title="Link to this definition">#</a></dt>
<dd><p>Returns the index of the first gap in the sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.first_invalid">
<span class="sig-name descname"><span class="pre">first_invalid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.first_invalid" title="Link to this definition">#</a></dt>
<dd><p>Returns the index of first invalid symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.first_non_strict">
<span class="sig-name descname"><span class="pre">first_non_strict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.first_non_strict" title="Link to this definition">#</a></dt>
<dd><p>Returns the index of first non-strict symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_diff">
<span class="sig-name descname"><span class="pre">frac_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_diff" title="Link to this definition">#</a></dt>
<dd><p>Returns fraction of positions where self and other differ.</p>
<p>Truncates at length of shorter sequence.
Note that frac_same and frac_diff are both 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_diff_gaps">
<span class="sig-name descname"><span class="pre">frac_diff_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_diff_gaps" title="Link to this definition">#</a></dt>
<dd><p>Returns frac. of positions where self and other’s gap states differ.</p>
<p>In other words, if self and other are both all gaps, or both all
non-gaps, or both have gaps in the same places, frac_diff_gaps will
return 0.0. If self is all gaps and other has no gaps, frac_diff_gaps
will return 1.0.</p>
<p>Returns 0 if one sequence is empty.</p>
<p>Uses self’s gap characters for both sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_diff_non_gaps">
<span class="sig-name descname"><span class="pre">frac_diff_non_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_diff_non_gaps" title="Link to this definition">#</a></dt>
<dd><p>Returns fraction of non-gap positions where self differs from other.</p>
<p>Doesn’t count any position where self or other has a gap.
Truncates at the length of the shorter sequence.</p>
<p>Returns 0 if one sequence is empty. Note that this means that
frac_diff_non_gaps is _not_ the same as 1 - frac_same_non_gaps, since both
return 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_same">
<span class="sig-name descname"><span class="pre">frac_same</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_same" title="Link to this definition">#</a></dt>
<dd><p>Returns fraction of positions where self and other are the same.</p>
<p>Truncates at length of shorter sequence.
Note that frac_same and frac_diff are both 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_same_gaps">
<span class="sig-name descname"><span class="pre">frac_same_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_same_gaps" title="Link to this definition">#</a></dt>
<dd><p>Returns fraction of positions where self and other share gap states.</p>
<p>In other words, if self and other are both all gaps, or both all
non-gaps, or both have gaps in the same places, frac_same_gaps will
return 1.0. If self is all gaps and other has no gaps, frac_same_gaps
will return 0.0. Returns 0 if one sequence is empty.</p>
<p>Uses self’s gap characters for both sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_same_non_gaps">
<span class="sig-name descname"><span class="pre">frac_same_non_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_same_non_gaps" title="Link to this definition">#</a></dt>
<dd><p>Returns fraction of non-gap positions where self matches other.</p>
<p>Doesn’t count any position where self or other has a gap.
Truncates at the length of the shorter sequence.</p>
<p>Returns 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.frac_similar">
<span class="sig-name descname"><span class="pre">frac_similar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">similar_pairs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.frac_similar" title="Link to this definition">#</a></dt>
<dd><p>Returns fraction of positions where self[i] is similar to other[i].</p>
<p>similar_pairs must be a dict such that d[(i,j)] exists if i and j are
to be counted as similar. Use PairsFromGroups in cogent3.util.misc to
construct such a dict from a list of lists of similar residues.</p>
<p>Truncates at the length of the shorter sequence.</p>
<p>Note: current implementation re-creates the distance function each
time, so may be expensive compared to creating the distance function
using for_seq separately.</p>
<p>Returns 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.gap_indices">
<span class="sig-name descname"><span class="pre">gap_indices</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.gap_indices" title="Link to this definition">#</a></dt>
<dd><p>Returns list of indices of all gaps in the sequence, or [].</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.gap_maps">
<span class="sig-name descname"><span class="pre">gap_maps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.gap_maps" title="Link to this definition">#</a></dt>
<dd><p>Returns dicts mapping between gapped and ungapped positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.gap_vector">
<span class="sig-name descname"><span class="pre">gap_vector</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.gap_vector" title="Link to this definition">#</a></dt>
<dd><p>Returns vector of True or False according to which pos are gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.gapped_by_map">
<span class="sig-name descname"><span class="pre">gapped_by_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.gapped_by_map" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.gapped_by_map_motif_iter">
<span class="sig-name descname"><span class="pre">gapped_by_map_motif_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.gapped_by_map_motif_iter" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.gapped_by_map_segment_iter">
<span class="sig-name descname"><span class="pre">gapped_by_map_segment_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.gapped_by_map_segment_iter" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_drawable">
<span class="sig-name descname"><span class="pre">get_drawable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_drawable" title="Link to this definition">#</a></dt>
<dd><p>make a figure from sequence features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>passed to get_features(biotype). Can be a single biotype or
series. Only features matching this will be included.</p>
</dd>
<dt><strong>width</strong></dt><dd><p>width in pixels</p>
</dd>
<dt><strong>vertical</strong></dt><dd><p>rotates the drawable</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>a Drawable instance</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If provided, the biotype is used for plot order.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_drawables">
<span class="sig-name descname"><span class="pre">get_drawables</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_drawables" title="Link to this definition">#</a></dt>
<dd><p>returns a dict of drawables, keyed by type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>passed to get_features(biotype). Can be a single biotype or
series. Only features matching this will be included.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_features">
<span class="sig-name descname"><span class="pre">get_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_features" title="Link to this definition">#</a></dt>
<dd><p>yields Feature instances</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>biotype of the feature</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>start, stop</strong></dt><dd><p>start, end positions to search between, relative to offset
of this sequence. If not provided, entire span of sequence is used.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>When dealing with a nucleic acid moltype, the returned features will
yield a sequence segment that is consistently oriented irrespective
of strand of the current instance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_in_motif_size">
<span class="sig-name descname"><span class="pre">get_in_motif_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_warnings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_in_motif_size" title="Link to this definition">#</a></dt>
<dd><p>returns sequence as list of non-overlapping motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>length of the motifs</p>
</dd>
<dt><strong>log_warnings</strong></dt><dd><p>whether to notify of an incomplete terminal motif</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_kmers">
<span class="sig-name descname"><span class="pre">get_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_kmers" title="Link to this definition">#</a></dt>
<dd><p>return all overlapping k-mers</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_name">
<span class="sig-name descname"><span class="pre">get_name</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_name" title="Link to this definition">#</a></dt>
<dd><p>Return the sequence name – should just use name instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_translation">
<span class="sig-name descname"><span class="pre">get_translation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">incomplete_ok</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim_stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_translation" title="Link to this definition">#</a></dt>
<dd><p>translate to amino acid sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>incomplete_ok</strong></dt><dd><p>codons that are mixes of nucleotide and gaps converted to ‘?’.
raises a ValueError if False</p>
</dd>
<dt><strong>include_stop</strong></dt><dd><p>allows stop codons in translation</p>
</dd>
<dt><strong>trim_stop</strong></dt><dd><p>trims a terminal stop codon if it exists</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>sequence of PROTEIN moltype</dt><dd></dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>AlphabetError if include_stop is False and a stop codon occurs</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.get_type" title="Link to this definition">#</a></dt>
<dd><p>Return the sequence type as moltype label.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.has_terminal_stop">
<span class="sig-name descname"><span class="pre">has_terminal_stop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.has_terminal_stop" title="Link to this definition">#</a></dt>
<dd><p>Return True if the sequence has a terminal stop codon.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>strict</strong></dt><dd><p>If True, raises an exception if length not divisible by 3</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.is_annotated">
<span class="sig-name descname"><span class="pre">is_annotated</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.is_annotated" title="Link to this definition">#</a></dt>
<dd><p>returns True if sequence has any annotations</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.is_degenerate">
<span class="sig-name descname"><span class="pre">is_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.is_degenerate" title="Link to this definition">#</a></dt>
<dd><p>Returns True if sequence contains degenerate characters.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.is_gap">
<span class="sig-name descname"><span class="pre">is_gap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.is_gap" title="Link to this definition">#</a></dt>
<dd><p>Returns True if char is a gap.</p>
<p>If char is not supplied, tests whether self is gaps only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.is_gapped">
<span class="sig-name descname"><span class="pre">is_gapped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.is_gapped" title="Link to this definition">#</a></dt>
<dd><p>Returns True if sequence contains gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.is_strict">
<span class="sig-name descname"><span class="pre">is_strict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.is_strict" title="Link to this definition">#</a></dt>
<dd><p>Returns True if sequence contains only monomers.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.is_valid">
<span class="sig-name descname"><span class="pre">is_valid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.is_valid" title="Link to this definition">#</a></dt>
<dd><p>Returns True if sequence contains no items absent from alphabet.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.iter_kmers">
<span class="sig-name descname"><span class="pre">iter_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Generator</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.iter_kmers" title="Link to this definition">#</a></dt>
<dd><p>generates all overlapping k-mers.
When strict is True, the characters in the k-mer must be
a subset of the canonical characters for the moltype</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.line_wrap">
<span class="sig-name descname"><span class="pre">line_wrap</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.line_wrap" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.make_feature">
<span class="sig-name descname"><span class="pre">make_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">FeatureDataType</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.make_feature" title="Link to this definition">#</a></dt>
<dd><p>return an Feature instance from feature data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>feature</strong></dt><dd><p>dict of key data to make an Feature instance</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Unlike add_feature(), this method does not add the feature to the
database.
We assume that spans represent the coordinates for this instance!</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.matrix_distance">
<span class="sig-name descname"><span class="pre">matrix_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.matrix_distance" title="Link to this definition">#</a></dt>
<dd><p>Returns distance between self and other using a score matrix.</p>
<p>WARNING: the matrix must explicitly contain scores for the case where
a position is the same in self and other (e.g. for a distance matrix,
an identity between U and U might have a score of 0). The reason the
scores for the ‘diagonals’ need to be passed explicitly is that for
some kinds of distance matrices, e.g. log-odds matrices, the ‘diagonal’
scores differ from each other. If these elements are missing, this
function will raise a KeyError at the first position that the two
sequences are identical.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.moltype">
<span class="sig-name descname"><span class="pre">moltype</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">MolType(('T',</span> <span class="pre">'C',</span> <span class="pre">'A',</span> <span class="pre">'G'))</span></em><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.moltype" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.must_match">
<span class="sig-name descname"><span class="pre">must_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.must_match" title="Link to this definition">#</a></dt>
<dd><p>Returns True if all positions in self must match positions in other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.must_pair">
<span class="sig-name descname"><span class="pre">must_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.must_pair" title="Link to this definition">#</a></dt>
<dd><p>Returns True if all positions in self must pair with other.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.mw">
<span class="sig-name descname"><span class="pre">mw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.mw" title="Link to this definition">#</a></dt>
<dd><p>Returns the molecular weight of (one strand of) the sequence.</p>
<p>If the sequence is ambiguous, uses method (random or strip) to
disambiguate the sequence.</p>
<p>If delta is passed in, adds delta per strand (default is None, which
uses the alphabet default. Typically, this adds 18 Da for terminal
water. However, note that the default nucleic acid weight assumes
5’ monophosphate and 3’ OH: pass in delta=18.0 if you want 5’ OH as
well.</p>
<p>Note that this method only calculates the MW of the coding strand. If
you want the MW of the reverse strand, add self.rc().mw(). DO NOT
just multiply the MW by 2: the results may not be accurate due to
strand bias, e.g. in mitochondrial genomes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.parse_out_gaps">
<span class="sig-name descname"><span class="pre">parse_out_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.parse_out_gaps" title="Link to this definition">#</a></dt>
<dd><p>returns Map corresponding to gap locations and ungapped Sequence</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.possibilities">
<span class="sig-name descname"><span class="pre">possibilities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.possibilities" title="Link to this definition">#</a></dt>
<dd><p>Counts number of possible sequences matching the sequence.</p>
<p>Uses self.degenerates to decide how many possibilities there are at
each position in the sequence.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.protein">
<span class="sig-name descname"><span class="pre">protein</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">MolType(('A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'))</span></em><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.protein" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.rc" title="Link to this definition">#</a></dt>
<dd><p>Converts a nucleic acid sequence to its reverse complement.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">oldchar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newchar</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.replace" title="Link to this definition">#</a></dt>
<dd><p>return new instance with oldchar replaced by newchar</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.replace_annotation_db">
<span class="sig-name descname"><span class="pre">replace_annotation_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">SupportsFeatures</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.replace_annotation_db" title="Link to this definition">#</a></dt>
<dd><p>public interface to assigning the annotation_db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>value</strong></dt><dd><p>the annotation db instance</p>
</dd>
<dt><strong>check</strong></dt><dd><p>whether to check value supports the feature interface</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The check can be very expensive, so if you’re confident set it to False</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.resolved_ambiguities">
<span class="sig-name descname"><span class="pre">resolved_ambiguities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.resolved_ambiguities" title="Link to this definition">#</a></dt>
<dd><p>Returns a list of tuples of strings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.reverse_complement">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.reverse_complement" title="Link to this definition">#</a></dt>
<dd><p>Converts a nucleic acid sequence to its reverse complement.
Synonymn for rc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.shuffle">
<span class="sig-name descname"><span class="pre">shuffle</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.shuffle" title="Link to this definition">#</a></dt>
<dd><p>returns a randomized copy of the Sequence object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.sliding_windows">
<span class="sig-name descname"><span class="pre">sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.sliding_windows" title="Link to this definition">#</a></dt>
<dd><p>Generator function that yield new sequence objects
of a given length at a given interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>window</strong></dt><dd><p>The length of the returned sequence</p>
</dd>
<dt><strong>step</strong></dt><dd><p>The interval between the start of the returned
sequence objects</p>
</dd>
<dt><strong>start</strong></dt><dd><p>first window start position</p>
</dd>
<dt><strong>end</strong></dt><dd><p>last window start position</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.strand_symmetry">
<span class="sig-name descname"><span class="pre">strand_symmetry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.strand_symmetry" title="Link to this definition">#</a></dt>
<dd><p>returns G-test for strand symmetry</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.strip_bad">
<span class="sig-name descname"><span class="pre">strip_bad</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.strip_bad" title="Link to this definition">#</a></dt>
<dd><p>Removes any symbols not in the alphabet.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.strip_bad_and_gaps">
<span class="sig-name descname"><span class="pre">strip_bad_and_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.strip_bad_and_gaps" title="Link to this definition">#</a></dt>
<dd><p>Removes any symbols not in the alphabet, and any gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.strip_degenerate">
<span class="sig-name descname"><span class="pre">strip_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.strip_degenerate" title="Link to this definition">#</a></dt>
<dd><p>Removes degenerate bases by stripping them out of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_dna">
<span class="sig-name descname"><span class="pre">to_dna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_dna" title="Link to this definition">#</a></dt>
<dd><p>Returns copy of self as DNA.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_fasta">
<span class="sig-name descname"><span class="pre">to_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">make_seqlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_fasta" title="Link to this definition">#</a></dt>
<dd><p>Return string of self in FASTA format, no trailing newline</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>make_seqlabel</strong></dt><dd><p>callback function that takes the seq object and
returns a label str</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_html">
<span class="sig-name descname"><span class="pre">to_html</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_family</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Lucida</span> <span class="pre">Console'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_html" title="Link to this definition">#</a></dt>
<dd><p>returns html with embedded styles for sequence colouring</p>
<dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl>
<dt><strong>wrap</strong></dt><dd><p>maximum number of printed bases, defaults to
alignment length</p>
</dd>
<dt><strong>limit</strong></dt><dd><p>truncate alignment to this length</p>
</dd>
<dt><strong>colors</strong></dt><dd><p>{character
moltype.</p>
</dd>
<dt><strong>font_size</strong></dt><dd><p>in points. Affects labels and sequence and line spacing
(proportional to value)</p>
</dd>
<dt><strong>font_family</strong></dt><dd><p>string denoting font family</p>
</dd>
<dt><strong>To display in jupyter notebook:</strong></dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">IPython.core.display</span> <span class="kn">import</span> <span class="n">HTML</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HTML</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">to_html</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_json">
<span class="sig-name descname"><span class="pre">to_json</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_json" title="Link to this definition">#</a></dt>
<dd><p>returns a json formatted string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_moltype">
<span class="sig-name descname"><span class="pre">to_moltype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">moltype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_moltype" title="Link to this definition">#</a></dt>
<dd><p>returns copy of self with moltype seq</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>moltype</strong><span class="classifier">str</span></dt><dd><p>molecular type</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_rich_dict">
<span class="sig-name descname"><span class="pre">to_rich_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exclude_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_rich_dict" title="Link to this definition">#</a></dt>
<dd><p>returns {‘name’: name, ‘seq’: sequence, ‘moltype’: moltype.label}</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.to_rna">
<span class="sig-name descname"><span class="pre">to_rna</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.to_rna" title="Link to this definition">#</a></dt>
<dd><p>Returns copy of self as RNA.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.trim_stop_codon">
<span class="sig-name descname"><span class="pre">trim_stop_codon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.trim_stop_codon" title="Link to this definition">#</a></dt>
<dd><p>Removes a terminal stop codon from the sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gc</strong></dt><dd><p>valid input to cogent3.get_code(), a genetic code object, number
or name</p>
</dd>
<dt><strong>strict</strong></dt><dd><p>If True, raises an exception if length not divisible by 3</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If sequence contains gap characters, the result preserves the sequence
length by adding gap characters at the end.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.with_masked_annotations">
<span class="sig-name descname"><span class="pre">with_masked_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annot_types</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shadow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.with_masked_annotations" title="Link to this definition">#</a></dt>
<dd><p>returns a sequence with annot_types regions replaced by mask_char
if shadow is False, otherwise all other regions are masked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annot_types</strong></dt><dd><p>annotation type(s)</p>
</dd>
<dt><strong>mask_char</strong></dt><dd><p>must be a character valid for the seq MolType. The
default value is the most ambiguous character, eg. ‘?’ for DNA</p>
</dd>
<dt><strong>shadow</strong></dt><dd><p>whether to mask the annotated regions, or everything but
the annotated regions</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.DnaSequence.with_termini_unknown">
<span class="sig-name descname"><span class="pre">with_termini_unknown</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.DnaSequence.with_termini_unknown" title="Link to this definition">#</a></dt>
<dd><p>Returns copy of sequence with terminal gaps remapped as missing.</p>
</dd></dl>

</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../sequence.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></p>
      </div>
    </a>
    <a class="right-next"
       href="cogent3.core.sequence.RnaSequence.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">RnaSequence</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">

  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence"><code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code></a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.add_feature"><code class="docutils literal notranslate"><span class="pre">DnaSequence.add_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.annotate_from_gff"><code class="docutils literal notranslate"><span class="pre">DnaSequence.annotate_from_gff()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.annotate_matches_to"><code class="docutils literal notranslate"><span class="pre">DnaSequence.annotate_matches_to()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.annotation_db"><code class="docutils literal notranslate"><span class="pre">DnaSequence.annotation_db</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.annotation_offset"><code class="docutils literal notranslate"><span class="pre">DnaSequence.annotation_offset</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.can_match"><code class="docutils literal notranslate"><span class="pre">DnaSequence.can_match()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.can_mismatch"><code class="docutils literal notranslate"><span class="pre">DnaSequence.can_mismatch()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.can_mispair"><code class="docutils literal notranslate"><span class="pre">DnaSequence.can_mispair()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.can_pair"><code class="docutils literal notranslate"><span class="pre">DnaSequence.can_pair()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.codon_alphabet"><code class="docutils literal notranslate"><span class="pre">DnaSequence.codon_alphabet</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.complement"><code class="docutils literal notranslate"><span class="pre">DnaSequence.complement()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.copy"><code class="docutils literal notranslate"><span class="pre">DnaSequence.copy()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.copy_annotations"><code class="docutils literal notranslate"><span class="pre">DnaSequence.copy_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.count"><code class="docutils literal notranslate"><span class="pre">DnaSequence.count()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.count_degenerate"><code class="docutils literal notranslate"><span class="pre">DnaSequence.count_degenerate()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.count_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.count_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.counts"><code class="docutils literal notranslate"><span class="pre">DnaSequence.counts()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.degap"><code class="docutils literal notranslate"><span class="pre">DnaSequence.degap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.diff"><code class="docutils literal notranslate"><span class="pre">DnaSequence.diff()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.disambiguate"><code class="docutils literal notranslate"><span class="pre">DnaSequence.disambiguate()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.distance"><code class="docutils literal notranslate"><span class="pre">DnaSequence.distance()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.first_degenerate"><code class="docutils literal notranslate"><span class="pre">DnaSequence.first_degenerate()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.first_gap"><code class="docutils literal notranslate"><span class="pre">DnaSequence.first_gap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.first_invalid"><code class="docutils literal notranslate"><span class="pre">DnaSequence.first_invalid()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.first_non_strict"><code class="docutils literal notranslate"><span class="pre">DnaSequence.first_non_strict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_diff"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_diff()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_diff_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_diff_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_diff_non_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_diff_non_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_same"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_same()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_same_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_same_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_same_non_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_same_non_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.frac_similar"><code class="docutils literal notranslate"><span class="pre">DnaSequence.frac_similar()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.gap_indices"><code class="docutils literal notranslate"><span class="pre">DnaSequence.gap_indices()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.gap_maps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.gap_maps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.gap_vector"><code class="docutils literal notranslate"><span class="pre">DnaSequence.gap_vector()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.gapped_by_map"><code class="docutils literal notranslate"><span class="pre">DnaSequence.gapped_by_map()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.gapped_by_map_motif_iter"><code class="docutils literal notranslate"><span class="pre">DnaSequence.gapped_by_map_motif_iter()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.gapped_by_map_segment_iter"><code class="docutils literal notranslate"><span class="pre">DnaSequence.gapped_by_map_segment_iter()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_drawable"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_drawable()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_drawables"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_drawables()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_features"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_features()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_in_motif_size"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_in_motif_size()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_kmers"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_kmers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_name"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_name()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_translation"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_translation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.get_type"><code class="docutils literal notranslate"><span class="pre">DnaSequence.get_type()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.has_terminal_stop"><code class="docutils literal notranslate"><span class="pre">DnaSequence.has_terminal_stop()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.is_annotated"><code class="docutils literal notranslate"><span class="pre">DnaSequence.is_annotated()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.is_degenerate"><code class="docutils literal notranslate"><span class="pre">DnaSequence.is_degenerate()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.is_gap"><code class="docutils literal notranslate"><span class="pre">DnaSequence.is_gap()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.is_gapped"><code class="docutils literal notranslate"><span class="pre">DnaSequence.is_gapped()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.is_strict"><code class="docutils literal notranslate"><span class="pre">DnaSequence.is_strict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.is_valid"><code class="docutils literal notranslate"><span class="pre">DnaSequence.is_valid()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.iter_kmers"><code class="docutils literal notranslate"><span class="pre">DnaSequence.iter_kmers()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.line_wrap"><code class="docutils literal notranslate"><span class="pre">DnaSequence.line_wrap</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.make_feature"><code class="docutils literal notranslate"><span class="pre">DnaSequence.make_feature()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.matrix_distance"><code class="docutils literal notranslate"><span class="pre">DnaSequence.matrix_distance()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.moltype"><code class="docutils literal notranslate"><span class="pre">DnaSequence.moltype</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.must_match"><code class="docutils literal notranslate"><span class="pre">DnaSequence.must_match()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.must_pair"><code class="docutils literal notranslate"><span class="pre">DnaSequence.must_pair()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.mw"><code class="docutils literal notranslate"><span class="pre">DnaSequence.mw()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.parse_out_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.parse_out_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.possibilities"><code class="docutils literal notranslate"><span class="pre">DnaSequence.possibilities()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.protein"><code class="docutils literal notranslate"><span class="pre">DnaSequence.protein</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.rc"><code class="docutils literal notranslate"><span class="pre">DnaSequence.rc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.replace"><code class="docutils literal notranslate"><span class="pre">DnaSequence.replace()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.replace_annotation_db"><code class="docutils literal notranslate"><span class="pre">DnaSequence.replace_annotation_db()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.resolved_ambiguities"><code class="docutils literal notranslate"><span class="pre">DnaSequence.resolved_ambiguities()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.reverse_complement"><code class="docutils literal notranslate"><span class="pre">DnaSequence.reverse_complement()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.shuffle"><code class="docutils literal notranslate"><span class="pre">DnaSequence.shuffle()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.sliding_windows"><code class="docutils literal notranslate"><span class="pre">DnaSequence.sliding_windows()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.strand_symmetry"><code class="docutils literal notranslate"><span class="pre">DnaSequence.strand_symmetry()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.strip_bad"><code class="docutils literal notranslate"><span class="pre">DnaSequence.strip_bad()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.strip_bad_and_gaps"><code class="docutils literal notranslate"><span class="pre">DnaSequence.strip_bad_and_gaps()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.strip_degenerate"><code class="docutils literal notranslate"><span class="pre">DnaSequence.strip_degenerate()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_dna"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_dna()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_fasta"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_fasta()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_html"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_html()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_json"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_json()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_moltype"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_moltype()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_rich_dict"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_rich_dict()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.to_rna"><code class="docutils literal notranslate"><span class="pre">DnaSequence.to_rna()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.trim_stop_codon"><code class="docutils literal notranslate"><span class="pre">DnaSequence.trim_stop_codon()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.with_masked_annotations"><code class="docutils literal notranslate"><span class="pre">DnaSequence.with_masked_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#cogent3.core.sequence.DnaSequence.with_termini_unknown"><code class="docutils literal notranslate"><span class="pre">DnaSequence.with_termini_unknown()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2020-2023, cogent3 Team.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.2.6.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.14.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>