<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

<!-- IE Compatibility tag -->
<meta http-equiv="X-UA-Compatible" content="IE=edge">

<title>cogent3 2023.07.18 documentation</title>

<link href="../../../.././_static/favicon.ico" rel="shortcut icon" type="image/x-icon"/>


<!-- Metadata -->
<meta name="description"
      content="cogent3 documentation site">



<!-- Facebook OpenGraph -->
<meta property="og:type" content="website"/>
<meta property="og:site" content="cogent3 documentation site"/>
<meta property="og:title" content="cogent3"/>
<meta property="og:description"
      content="Check the documentation site for cogent3"/>



<!-- HTML5 Shiv. For IE6-8 support of HTML5 elements -->
<!--[if lt IE 9]>
<script src="../../../.././_static/lib/html5shiv/html5shiv.min.js"></script>
<![endif]-->
<!-- CSS -->
<link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/css/docs.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/jupyter-sphinx.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/thebelab.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/sg_gallery.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
<link rel="stylesheet" href="../../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />

<link rel="stylesheet" href="../../../.././_static/lib/bootswatch/litera/bootstrap.min.css">
<link rel="stylesheet" href="../../../.././_static/lib/font-awesome/css/font-awesome.min.css">
<link rel="stylesheet" href="../../../.././_static/lib/highlight/styles/default.css">

<!-- Docs Template style -->
<link rel="stylesheet" href="../../../.././_static/pygments.css">
<link rel="stylesheet" href="../../../.././_static/css/common.css">
<link rel="stylesheet" href="../../../.././_static/css/link.css">
<link rel="stylesheet" href="../../../.././_static/css/custom.css">

<!-- JavaScript -->

<script type="text/javascript" id="documentation_options" data-url_root="../../../../" src="../../../../_static/documentation_options.js"></script>
<script src="../../../.././_static/lib/jquery/jquery.min.js"></script>
<script src="../../../.././_static/lib/bootstrap/js/bootstrap.min.js"></script>
<script src="../../../.././_static/lib/highlight/highlight.pack.js"></script>
<script src="../../../.././_static/js/initializeHighlight.js"></script>
<script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
<script src="../../../../_static/jquery.js"></script>
<script src="../../../../_static/underscore.js"></script>
<script src="../../../../_static/doctools.js"></script>
<script src="../../../../_static/thebelab-helper.js"></script>
<script src="../../../../_static/plotly-latest.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
<script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="../../../../_static/require.js"></script>
<script src="../../../.././_static/js/custom.js"></script>

</head>
<body class="d-flex flex-column">
<header>
    <nav id="navbar-main" class="navbar navbar-expand navbar-light bg-light">
    <a class="navbar-brand" href="../../../.././index.html">cogent3</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-main-menu" aria-controls="navbar-main" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar-main-menu" class="collapse navbar-collapse">
        <small class="navbar-text"><span id="navbar-version"></span></time></small>
        <div class="ml-auto d-flex align-items-center row">
            
            
            <div class="col-sm">
                <ul class="navbar-nav">
                    
                    
                    <li class="nav-item">
                        <a href="../../../install.html" class="nav-link" title="Install">Install</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../../../draw/index.html" class="nav-link" title="Gallery">Gallery</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="../../../index.html" class="nav-link" title="Docs">Docs</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3/discussions" class="nav-link" title="Forum">Forum</a>
                    </li>
                    
                    
                    
                    <li class="nav-item">
                        <a href="https://github.com/cogent3/cogent3" class="nav-link" title="GitHub">GitHub</a>
                    </li>
                    
                    
                    
                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" id="examplesDropdown" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Site</a>
                        <div class="dropdown-menu dropdown-menu-right" role="menu" aria-labelledby="examplesDropdown">
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Posting Bugs">Posting Bugs</a>
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Citation">Citation</a>
                            
                            <a href="../../../../general.html" class="dropdown-item" title="Support">Support</a>
                            
                            <a href="../../../../history.html" class="dropdown-item" title="History">History</a>
                            
                            <a href="../../../../projects.html" class="dropdown-item" title="Projects">Projects</a>
                            
                            <a href="../../../licenses.html" class="dropdown-item" title="License">License</a>
                            
                        </div>
                    </li>
                    
                    
                </ul>
            </div>
            
            <div role="search">
                <form id="Search-form" class="wy-form" action="../../../../search.html" method="get">
                    <input type="text" name="q" placeholder="Search docs" />
                    <input type="hidden" name="check_keywords" value="yes" />
                    <input type="hidden" name="area" value="default" />
                </form>
            </div>
        </div>
    </div>
</nav>
</header>

<nav class="side-nav">
    <div class="side-nav-container">
        
        <div class="side-nav-menu hidden">
        
            
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#minimal-installation">Minimal installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#install-with-graphing-tools">Install with graphing tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#install-with-developer-tools">Install with developer tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#installing-the-development-version">Installing the development version</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#installing-using-conda-mamba">Installing using <code class="docutils literal notranslate"><span class="pre">conda</span></code> / <code class="docutils literal notranslate"><span class="pre">mamba</span></code></a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../../../index.html">User Guides</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../../app/index.html">apps</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index-app.html">Overview of apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app-overview.html">Overview of using apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#types-of-apps">Types of apps</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#composability">Composability</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#composability-rules">Composability rules</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#an-example">An example</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#using-apps-sequentially-like-functions">Using apps sequentially like functions</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#composing-a-multi-step-process-from-several-apps">Composing a multi-step process from several apps</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#applying-a-process-to-multiple-data-records">Applying a process to multiple data records</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/app-overview.html#other-important-features">Other important features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#the-settings-and-data-analysed-will-be-logged">The settings and data analysed will be logged</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#failures-are-recorded">Failures are recorded</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-track-progress">You can track progress</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#you-can-do-parallel-computation">You can do parallel computation</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/app-overview.html#all-of-the-above">All of the above</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app-help.html">Getting help on an app</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/app-get.html">Getting an app</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/available-apps.html">Displaying installed apps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/dstore.html">Data stores – collections of data records</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#how-do-i-use-a-data-store">How do I use a data store?</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#types-of-data-store">Types of data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#the-structure-of-data-stores">The structure of data stores</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#supported-operations-on-a-data-store">Supported operations on a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#opening-a-data-store">Opening a data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#opening-a-read-only-data-store">Opening a read only data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#summarising-the-data-store">Summarising the data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#data-store-members">Data store “members”</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/dstore.html#get-one-member">Get one member</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/dstore.html#looping-over-a-data-store">Looping over a data store</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/dstore.html#members-can-read-their-own-data">Members can read their own data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#making-a-writeable-data-store">Making a writeable data store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#datastoresqlite-stores-serialised-data"><code class="docutils literal notranslate"><span class="pre">DataStoreSqlite</span></code> stores serialised data</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#interrogating-run-logs">Interrogating run logs</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/dstore.html#pulling-it-all-together">Pulling it all together</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#defining-the-data-stores-for-analysis">Defining the data stores for analysis</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#create-an-app-and-apply-it">Create an app and apply it</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/dstore.html#inspecting-the-outcome">Inspecting the outcome</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/not-completed.html">Tracking records that could not be processed</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#the-notcompleted-object">The <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> object</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/not-completed.html#notcompleted-false-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> FALSE type</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/not-completed.html#notcompleted-error-type"><code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> ERROR type</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/not-completed.html#composed-functions-propagate-notcompleted-results">Composed functions propagate <code class="docutils literal notranslate"><span class="pre">NotCompleted</span></code> results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index-align.html">Sequence Alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-nucleotide.html">Using a nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#nucleotide-alignment-with-default-settings">Nucleotide alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#specifying-the-substitution-model">Specifying the substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-nucleotide.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-codon.html">Using a codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#codon-alignment-with-default-settings">Codon alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#providing-a-guide-tree">Providing a guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specifying-the-gap-parameters">Specifying the gap parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#specifying-the-substitution-model-and-parameters">Specifying the substitution model and parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-codon.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/align-protein.html">Using a protein model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#protein-alignment-with-default-settings">Protein alignment with default settings</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#specify-a-different-distance-measure-for-estimating-the-guide-tree">Specify a different distance measure for estimating the guide tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/align-protein.html#alignment-settings-and-file-provenance-are-recorded-in-the-info-attribute">Alignment settings and file provenance are recorded in the <code class="docutils literal notranslate"><span class="pre">info</span></code> attribute</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index-evo-model.html">Modelling sequence evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model.html">Apply a non-stationary nucleotide model to an alignment with 3 sequences</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model.html#model-result"><code class="docutils literal notranslate"><span class="pre">model_result</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model.html#fitting-a-separate-nucleotide-model-to-each-codon-position">Fitting a separate nucleotide model to each codon position</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model-with-tree.html">Apply a non-stationary nucleotide model to an alignment with a tree</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#specify-the-tree-via-a-tree-instance">Specify the tree via a tree instance</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#specify-the-tree-via-a-path">Specify the tree via a path.</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-with-tree.html#apply-the-model-to-an-alignment">Apply the model to an alignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-model-timehet.html">Specifying a non-stationary model with time-heterogeneous parameters</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-model-timehet.html#specify-a-distinct-rate-matrix-for-the-great-apes">Specify a distinct rate matrix for the great apes</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-extract-model-stats.html">Extracting maximum likelihood estimates from a <code class="docutils literal notranslate"><span class="pre">model_result</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#create-and-apply-tabulate-stats-app">Create and apply <code class="docutils literal notranslate"><span class="pre">tabulate_stats</span></code> app</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#edge-parameters">Edge parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#global-parameters">Global parameters</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-extract-model-stats.html#motif-parameters">Motif parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-ancestral-states.html">Reconstructing ancestral states</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-ancestral-states.html#define-the-ancestral-states-app">Define the <code class="docutils literal notranslate"><span class="pre">ancestral_states</span></code> app</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html">Applying a discrete-time, non-stationary nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html#get-a-tree-with-branch-lengths-as-paralinear">Get a tree with branch lengths as paralinear</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-dt-nuc-model.html#getting-parameter-estimates">Getting parameter estimates</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html">Applying a time-reversible nucleotide model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-tr-nuc-model.html#using-the-gtr-model">Using the GTR model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html">Applying GNC, a non-stationary codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-nstat-codon-model.html#we-can-obtain-the-tree-with-branch-lengths-as-ens">We can obtain the tree with branch lengths as ENS</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-tr-codon-model.html">Applying a time-reversible codon model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-tr-codon-model.html#using-the-conditional-nucleotide-form-codon-model">Using the conditional nucleotide form codon model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index-evo-hypothesis.html">Testing hypotheses about sequence evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-hypothesis.html">Testing a hypothesis – non-stationary or time-reversible</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-hypothesis.html#hypothesis-result-has-attributes-and-keys"><code class="docutils literal notranslate"><span class="pre">hypothesis_result</span></code> has attributes and keys</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#accessing-the-test-statistics">Accessing the test statistics</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#the-null-hypothesis">The null hypothesis</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../app/evo-hypothesis.html#the-alternate-hypothesis">The alternate hypothesis</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-hypothesis.html#saving-hypothesis-results">Saving hypothesis results</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index-evo-natsel.html">Testing for natural selection</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_neutral.html"><code class="docutils literal notranslate"><span class="pre">natsel_neutral</span></code> – a test for selective neutrality</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_timehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_timehet</span></code> – a test of branch heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html"><code class="docutils literal notranslate"><span class="pre">natsel_sitehet</span></code> – a test of site heterogeneity</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_sitehet.html#getting-the-individual-site-posterior-probabilities">Getting the individual site posterior probabilities</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../app/evo-natsel_zhang.html"><code class="docutils literal notranslate"><span class="pre">natsel_zhang</span></code> – a branch-site test</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_zhang.html#getting-the-posterior-probabilities-of-site-class-membership">Getting the posterior probabilities of site-class membership</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/evo-natsel_zhang.html#getting-all-the-statistics-in-tabular-form">Getting all the statistics in tabular form</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../app/index-custom.html">Write your own Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../app/user_function.html">Turn your functions into composable apps</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#a-simple-example">A simple example</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#renaming-sequences">Renaming sequences</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../app/user_function.html#a-user-app-with-a-different-output-type">A user app with a different output type</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../cookbook/index.html">Cookbook</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html">Manipulating biological data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#sequences-and-alignments">Sequences and Alignments</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html">Loading a sequence from a file</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-an-alignment-from-a-file-or-url">Loading an alignment from a file or url</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-aligned-sequences">Loading aligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-unaligned-sequences">Loading unaligned sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-from-a-url">Loading from a url</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#specifying-the-file-format">Specifying the file format</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#specifying-the-sequence-molecular-type">Specifying the sequence molecular type</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#making-an-alignment-from-standard-python-objects">Making an alignment from standard python objects</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#from-a-dict-of-strings">From a dict of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#from-a-series-of-strings">From a series of strings</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#stripping-label-characters-on-loading">Stripping label characters on loading</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#making-a-sequence-collection-from-standard-python-objects">Making a sequence collection from standard python objects</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-sequences-using-format-parsers">Loading sequences using format parsers</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#loading-fasta-sequences-from-an-open-file-or-list-of-lines">Loading FASTA sequences from an open file or list of lines</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/loading_sequences.html#handling-overloaded-fasta-sequence-labels">Handling overloaded FASTA sequence labels</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html">Sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html#generic-molecular-types">Generic molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/DNA_and_RNA_sequences.html#dna-and-rna-sequences">DNA and RNA sequences</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/protein_sequences.html">Protein sequences</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#creating-a-proteinsequence-with-a-name">Creating a ProteinSequence with a name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#converting-a-dna-sequence-string-to-protein-sequence-string">Converting a DNA sequence string to protein sequence string</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#converting-a-nucleic-acid-sequence-object-to-protein">Converting a nucleic acid sequence object to protein</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/protein_sequences.html#loading-protein-sequences-from-a-phylip-file">Loading protein sequences from a Phylip file</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/alignments.html">Sequence Collections and Alignments</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#what-s-the-difference-between-alignment-and-arrayalignment">What’s the difference between <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> and <code class="docutils literal notranslate"><span class="pre">ArrayAlignment</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#basic-collection-objects">Basic Collection objects</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#adding-new-sequences-to-an-existing-collection-or-alignment">Adding new sequences to an existing collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#the-elements-of-a-collection-or-alignment">The elements of a collection or alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#alignments">Alignments</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#slicing-an-alignment">Slicing an alignment</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alignments.html#computing-alignment-statistics">Computing alignment statistics</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/features.html">Features</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/features.html#how-to-create-a-custom-feature">How to create a custom <code class="docutils literal notranslate"><span class="pre">Feature</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/features.html#how-to-load-bulk-features-from-a-file">How to load bulk Features from a File</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/features.html#how-to-query-a-sequence-or-alignment-for-features">How to query a Sequence or Alignment for Features</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/features.html#examples-using-the-methods-available-on-features">Examples using the methods available on Features</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/annotation_db.html">Annotation Databases</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#what-are-the-different-types-of-annotationdb">What are the different types of <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code>?</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-create-a-standalone-basicannotationdb">How to create a standalone <code class="docutils literal notranslate"><span class="pre">BasicAnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-load-an-standalone-annotationdb-from-a-data-file">How to load an standalone <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> from a data file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-generate-a-summary-of-an-annotationdb">How to generate a summary of an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-add-custom-features-to-an-annotationdb">How to add custom features to an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-write-an-annotationdb-to-disk-for-efficient-re-loading">How to write an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to disk for efficient re-loading</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-query-an-annotationdb">How to query an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-interrogate-an-annotationdb">How to interrogate an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-combine-two-annotationdb-instances">How to combine two <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> instances</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/annotation_db.html#how-to-assign-an-annotationdb-to-a-sequence">How to assign an <code class="docutils literal notranslate"><span class="pre">AnnotationDb</span></code> to a sequence</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/moltypes.html">Molecular types</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#available-molecular-types">Available molecular types</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#getting-a-moltype">Getting a <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#using-a-moltype-to-get-ambiguity-codes">Using a <code class="docutils literal notranslate"><span class="pre">MolType</span></code> to get ambiguity codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#moltype-definition-of-degenerate-codes"><code class="docutils literal notranslate"><span class="pre">MolType</span></code> definition of degenerate codes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#nucleic-acid-moltype-and-complementing">Nucleic acid <code class="docutils literal notranslate"><span class="pre">MolType</span></code> and complementing</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#making-sequences">Making sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#verify-sequences">Verify sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/moltypes.html#making-a-custom-moltype">Making a custom <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/alphabet.html">Alphabets</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#alphabet-and-moltype"><code class="docutils literal notranslate"><span class="pre">Alphabet</span></code> and <code class="docutils literal notranslate"><span class="pre">MolType</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#creating-tuple-alphabets">Creating tuple alphabets</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#convert-a-sequence-into-integers">Convert a sequence into integers</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/alphabet.html#convert-integers-to-a-sequence">Convert integers to a sequence</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#genetic-code">Genetic code</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/what_codes.html">Available genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/what_codes.html#getting-a-genetic-code-with-get-code">Getting a genetic code with <code class="docutils literal notranslate"><span class="pre">get_code()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/genetic_code.html">Using genetic codes</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#selecting-codes-in-methods-that-support-them">Selecting codes in methods that support them</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-dna-sequences">Translate DNA sequences</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-all-six-frames">Translate all six frames</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#find-out-how-many-stops-in-a-frame">Find out how many stops in a frame</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#translate-a-codon">Translate a codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#look-up-the-amino-acid-corresponding-to-a-single-codon">Look up the amino acid corresponding to a single codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#get-all-the-codons-for-one-amino-acid">Get all the codons for one amino acid</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#get-all-the-codons-for-a-group-of-amino-acids">Get all the codons for a group of amino acids</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#converting-the-codonalphabet-to-codon-series">Converting the <code class="docutils literal notranslate"><span class="pre">CodonAlphabet</span></code> to codon series</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#obtaining-the-codons-from-a-dnasequence-object">Obtaining the codons from a <code class="docutils literal notranslate"><span class="pre">DnaSequence</span></code> object</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#translating-a-dna-sequence">Translating a DNA sequence</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/genetic_code.html#translating-a-dna-sequence-containing-stop-codons">Translating a DNA sequence containing stop codons</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/manipulating_biological_data.html#trees">Trees</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/simple_trees.html">Trees</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#loading-a-tree-from-a-file-and-visualizing-it-with-ascii-art">Loading a tree from a file and visualizing it with <code class="docutils literal notranslate"><span class="pre">ascii_art()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#writing-a-tree-to-a-file">Writing a tree to a file</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-individual-nodes-of-a-tree-by-name">Getting the individual nodes of a tree by name</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-name-of-a-node-or-a-tree">Getting the name of a node (or a tree)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#the-object-type-of-a-tree-and-its-nodes-is-the-same">The object type of a tree and its nodes is the same</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#working-with-the-nodes-of-a-tree">Working with the nodes of a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-path-between-two-tips-or-edges-connecting-edges">Getting the path between two tips or edges (connecting edges)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-distance-between-two-nodes">Getting the distance between two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-last-common-ancestor-lca-for-two-nodes">Getting the last common ancestor (LCA) for two nodes</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-ancestors-for-a-node">Getting all the ancestors for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-children-for-a-node">Getting all the children for a node</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-all-the-distances-for-a-tree">Getting all the distances for a tree</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#getting-the-two-nodes-that-are-farthest-apart">Getting the two nodes that are farthest apart</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#get-the-nodes-within-a-given-distance">Get the nodes within a given distance</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#rerooting-trees">Rerooting trees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-representations">Tree representations</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-traversal">Tree traversal</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#selecting-subtrees">Selecting subtrees</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/simple_trees.html#tree-manipulation-methods">Tree manipulation methods</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/tables.html">Tabular data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-a-csv-file">Loading a csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-from-a-url">Loading from a url</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-delimited-specifying-the-format">Loading delimited specifying the format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#loading-delimited-data-without-a-header-line">Loading delimited data without a header line</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/tables.html#using-load-delimited">Using <code class="docutils literal notranslate"><span class="pre">load_delimited()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/tables.html#using-filteringparser">Using <code class="docutils literal notranslate"><span class="pre">FilteringParser</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#selectively-loading-parts-of-a-big-file">Selectively loading parts of a big file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/tables.html#loading-a-set-number-of-lines-from-a-file">Loading a set number of lines from a file</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/tables.html#loading-only-some-rows">Loading only some rows</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/tables.html#loading-only-some-columns">Loading only some columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/tables.html#load-raw-data-as-a-list-of-lists-of-strings">Load raw data as a list of lists of strings</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#make-a-table-from-header-and-rows">Make a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#make-a-table-from-a-dict">Make a table from a <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-the-column-order-when-creating-from-a-dict">Specify the column order when creating from a <code class="docutils literal notranslate"><span class="pre">dict</span></code>.</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-the-table-with-an-index">Create the table with an index</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-a-pandas-dataframe">Create a table from a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-header-and-rows">Create a table from header and rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-dict">Create a table from dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-from-a-2d-dict">Create a table from a 2D dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-table-that-has-complex-python-objects-as-elements">Create a table that has complex python objects as elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-an-empty-table">Create an empty table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#adding-a-new-column">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#add-a-title-and-a-legend-to-a-table">Add a title and a legend to a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#iterating-over-table-rows">Iterating over table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#how-many-rows-are-there">How many rows are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#how-many-columns-are-there">How many columns are there?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#iterating-over-table-columns">Iterating over table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#table-slicing-using-column-names">Table slicing using column names</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#table-slicing-using-indices">Table slicing using indices</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#changing-displayed-numerical-precision">Changing displayed numerical precision</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#change-digits-or-column-spacing">Change digits or column spacing</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#wrapping-tables-for-display">Wrapping tables for display</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-the-top-of-a-table-using-head">Display the top of a table using <code class="docutils literal notranslate"><span class="pre">head()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-the-bottom-of-a-table-using-tail">Display the bottom of a table using <code class="docutils literal notranslate"><span class="pre">tail()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#display-random-rows-from-a-table">Display random rows from a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#change-the-number-of-rows-displayed-by-repr">Change the number of rows displayed by <code class="docutils literal notranslate"><span class="pre">repr()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#changing-column-headings">Changing column headings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#id1">Adding a new column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#create-a-new-column-from-existing-ones">Create a new column from existing ones</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-table-data-as-a-numpy-array">Get table data as a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-column-as-a-list">Get a table column as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-multiple-table-columns-as-a-list">Get multiple table columns as a list</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-row-oriented-dict">Get the table as a row oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-column-oriented-dict">Get the table as a column oriented <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-the-table-as-a-pandas-dataframe">Get the table as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-of-counts-as-a-contingency-table">Get a table of counts as a contingency table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#appending-tables">Appending tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-a-single-column">Summing a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-multiple-columns-or-rows-strictly-numerical-data">Summing multiple columns or rows - strictly numerical data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#summing-multiple-columns-or-rows-with-mixed-non-numeric-numeric-data">Summing multiple columns or rows with mixed non-numeric/numeric data</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#filtering-table-rows">Filtering table rows</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#filtering-table-columns">Filtering table columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#standard-sorting">Standard sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#reverse-sorting">Reverse sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#sorting-involving-multiple-columns-one-reversed">Sorting involving multiple columns, one reversed</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-raw-data-for-a-single-column">Getting raw data for a single column</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-raw-data-for-multiple-columns">Getting raw data for multiple columns</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-distinct-values">Getting distinct values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#counting-occurrences-of-values">Counting occurrences of values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#counting-unique-values">Counting unique values</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#joining-or-merging-tables">Joining or merging tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#transpose-a-table">Transpose a table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-markdown-as-the-str-format">Specify markdown as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#specify-latex-as-the-str-format">Specify latex as the <code class="docutils literal notranslate"><span class="pre">str()</span></code> format</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-markdown-formatted-string">Get a table as a markdown formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-latex-formatted-string">Get a table as a latex formatted string</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-restructured-text-csv-table">Get a table as a restructured text csv-table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#get-a-table-as-a-restructured-text-grid-table">Get a table as a restructured text grid table</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-latex-format-table-with-to-string">Getting a latex format table with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-bedgraph-format-with-to-string">Getting a bedGraph format with <code class="docutils literal notranslate"><span class="pre">to_string()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#getting-a-table-as-html">Getting a table as html</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#what-formats-can-be-written">What formats can be written?</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#writing-a-latex-formmated-file">Writing a latex formmated file</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/tables.html#writing-delimited-formats">Writing delimited formats</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/building_alignments.html">Building alignments</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-nucleotides">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for nucleotides</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#using-a-cogent3-progressive-aligner-for-codons">Using a <code class="docutils literal notranslate"><span class="pre">cogent3</span></code> progressive aligner for codons</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_alignments.html#converting-gaps-from-aa-seq-alignment-to-nuc-seq-alignment">Converting gaps from aa-seq alignment to nuc seq alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html">Genetic distance calculation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#fast-pairwise-distance-estimation">Fast pairwise distance estimation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#computing-genetic-distances-using-the-alignment-object">Computing genetic distances using the <code class="docutils literal notranslate"><span class="pre">Alignment</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#using-the-distance-calculator-directly">Using the distance calculator directly</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/calc_genetic_distance.html#likelihood-based-pairwise-distance-estimation">Likelihood based pairwise distance estimation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/building_phylogenies.html">Building phylogenies</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#building-a-phylogenetic-tree-from-pairwise-distances">Building A Phylogenetic Tree From Pairwise Distances</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#directly-via-alignment-quick-tree">Directly via <code class="docutils literal notranslate"><span class="pre">alignment.quick_tree()</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#using-the-distancematrix-object">Using the <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#explicitly-via-distancematrix-and-cogent3-phylo-nj-nj">Explicitly via <code class="docutils literal notranslate"><span class="pre">DistanceMatrix</span></code> and <code class="docutils literal notranslate"><span class="pre">cogent3.phylo.nj.nj()`</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#directly-from-a-pairwise-distance-dict">Directly from a pairwise distance <code class="docutils literal notranslate"><span class="pre">dict</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#by-least-squares">By Least-squares</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/building_phylogenies.html#by-ml">By ML</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/evo_modelling.html">Evolutionary Analysis Using Likelihood</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-substitution-models">Specifying substitution models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#the-available-pre-defined-substitution-models">The available pre-defined substitution models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#getting-a-substitution-model-with-get-model">Getting a substitution model with <code class="docutils literal notranslate"><span class="pre">get_model()</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#rate-heterogeneity-models">Rate heterogeneity models</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-nucleotide">For nucleotide</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-codon">For codon</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#for-protein">For protein</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/evo_modelling.html#making-a-likelihood-function">Making a likelihood function</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#providing-an-alignment-to-a-likelihood-function">Providing an alignment to a likelihood function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#scoping-parameters-on-trees-time-heterogeneous-models">Scoping parameters on trees – time heterogeneous models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-a-parameter-as-constant">Specifying a parameter as constant</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#providing-a-starting-value-for-a-parameter">Providing a starting value for a parameter</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#setting-parameter-bounds-for-optimisation">Setting parameter bounds for optimisation</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#setting-an-upper-bound-for-branch-length">Setting an upper bound for branch length</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-rate-heterogeneity-functions">Specifying rate heterogeneity functions</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#specifying-phylo-hmms">Specifying Phylo-HMMs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#fitting-likelihood-functions-choice-of-optimisers">Fitting likelihood functions - Choice of optimisers</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#how-to-check-your-optimisation-was-successful">How to check your optimisation was successful</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#overview-of-the-fitted-likelihood-function">Overview of the fitted likelihood function</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#log-likelihood-and-number-of-free-parameters">Log likelihood and number of free parameters</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#aikake-information-criterion">Aikake Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#bayesian-information-criterion">Bayesian Information Criterion</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#getting-maximum-likelihood-estimates">Getting maximum likelihood estimates</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#testing-hypotheses-using-likelihood-ratio-tests">Testing Hypotheses - Using Likelihood Ratio Tests</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../cookbook/evo_modelling.html#testing-hypotheses-by-parametric-bootstrapping">Testing Hypotheses - By parametric bootstrapping</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#determining-confidence-intervals-on-mles">Determining confidence intervals on MLEs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#saving-results">Saving results</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/evo_modelling.html#reconstructing-ancestral-sequences">Reconstructing ancestral sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../cookbook/useful_utilities.html">Useful Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#uniondict-a-dict-with-set-like-operations-and-keys-as-attributes"><code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> – a dict with set like operations and keys as attributes</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#accessing-elements-of-a-uniondict">Accessing elements of a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#updating-a-uniondict">Updating a <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#accessing-a-non-existent-uniondict-key">Accessing a non-existent <code class="docutils literal notranslate"><span class="pre">UnionDict</span></code> key</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#using-cogent3-s-optimisers-for-your-own-functions">Using Cogent3’s optimisers for your own functions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../cookbook/useful_utilities.html#miscellaneous-functions">Miscellaneous functions</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#force-a-variable-to-be-iterable">Force a variable to be iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#curry-a-function">Curry a function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-see-if-an-object-is-iterable">Test to see if an object is iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-see-if-an-object-is-a-single-char">Test to see if an object is a single char</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#flatten-a-deeply-nested-iterable">Flatten a deeply nested iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#test-to-determine-if-list-of-tuple">Test to determine if <code class="docutils literal notranslate"><span class="pre">list</span></code> of <code class="docutils literal notranslate"><span class="pre">tuple</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#create-a-case-insensitive-iterable">Create a case-insensitive iterable</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#construct-a-distance-matrix-lookup-function">Construct a distance matrix lookup function</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#check-class-types">Check class types</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#delegate-to-a-separate-object">Delegate to a separate object</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#wrap-a-function-to-hide-from-a-class">Wrap a function to hide from a class</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../cookbook/useful_utilities.html#construct-a-constrained-container">Construct a constrained container</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../examples/index.html">Tutorials</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#data-manipulation">Data manipulation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/align_codons_to_protein.html">Map protein alignment gaps to DNA alignment gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html">Manipulation of Tree Node Objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/manipulating_tree_nodes.html#an-example-of-conditional-tree-node-modifications">An Example of Conditional Tree Node Modifications</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#general-data-analysis">General data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/period_estimation.html">Estimating periodic signals</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#estimating-the-periods-in-a-signal">Estimating the periods in a signal</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#for-numerical-continuous-data">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#for-symbolic-data">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#estimating-power-for-specified-period">Estimating power for specified period</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id1">For numerical (continuous) data</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id2">For symbolic data</a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/period_estimation.html#measuring-statistical-significance-of-periodic-signals">Measuring statistical significance of periodic signals</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id3">For numerical (continuous data)</a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/period_estimation.html#id4">For symbolic data</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#modelling-evolution">Modelling Evolution</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/simple.html">The simplest script</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/relative_rate.html">Performing a relative rate test</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/neutral_test.html">A test of the neutral theory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html">Allowing substitution model parameters to differ between branches</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-a-clade">Specifying a clade</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-a-stem">Specifying a stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#specifying-clade-and-stem">Specifying clade and stem</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#alternate-arguments-for-specifying-edges">Alternate arguments for specifying edges</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/scope_model_params_on_trees.html#applications-of-scoped-parameters">Applications of scoped parameters</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/codon_models.html">Using codon models</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#constructing-the-codon-substitution-model">Constructing the codon substitution model</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#standard-test-of-neutrality">Standard test of neutrality</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#rate-heterogeneity-model-variants">Rate-heterogeneity model variants</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/codon_models.html#mixing-branch-and-site-heterogeneity">Mixing branch and site-heterogeneity</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/empirical_protein_models.html">Use an empirical protein substitution model</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/testing_multi_loci.html">Likelihood analysis of multiple loci</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/simulate_alignment.html">Simulate an alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/parametric_bootstrap.html">Performing a parametric bootstrap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/coevolution.html">Perform a coevolutionary analysis on biological sequence alignments</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/rate_heterogeneity.html">Analysis of rate heterogeneity</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html">Evaluate process heterogeneity using a Hidden Markov Model</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ho-no-rate-heterogeneity">Model Ho: no rate heterogeneity</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ha-1-two-classes-of-gamma-distributed-but-independent-sites">Model Ha(1): two classes of gamma distributed but independent sites</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#model-ha-2-fast-and-slowly-evolving-sites-are-auto-correlated">Model Ha(2): fast and slowly evolving sites are auto-correlated</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/hmm_par_heterogeneity.html#a-model-with-patches-of-kappa">A model with patches of <code class="docutils literal notranslate"><span class="pre">kappa</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#phylogenetic-reconstruction">Phylogenetic Reconstruction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_pairwise_distances.html">Calculate pairwise distances between sequences</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_neigbourjoining_tree.html">Make a neighbor joining tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/calculate_UPGMA_cluster.html">Make a UPGMA cluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/phylo_by_ls.html">Phylogenetic reconstruction by least squares</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#look-for-the-single-best-tree">Look for the single best tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#a-more-rigorous-tree-space-search">A more rigorous tree space search</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#assessing-the-fit-for-a-pre-specified-tree-topology">Assessing the fit for a pre-specified tree topology</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/phylo_by_ls.html#using-maximum-likelihood-for-measuring-tree-fit">Using maximum likelihood for measuring tree fit</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../examples/index.html#utilities">Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../examples/parallel.html">Parallel computations</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/parallel.html#parallel-computation-on-a-single-computer">Parallel computation on a single computer</a><ul>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/parallel.html#using-app-apply-to">Using <code class="docutils literal notranslate"><span class="pre">app.apply_to()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="../../../examples/parallel.html#directly-using-cogent3-util-parallel-map">Directly using <code class="docutils literal notranslate"><span class="pre">cogent3.util.parallel.map()</span></code></a></li>
</ul>
</li>
<li class="toctree-l5"><a class="reference internal" href="../../../examples/parallel.html#parallel-computation-on-multiple-computers">Parallel computation on multiple computers</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="../../index.html">API</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../index.html#utility-functions-for">Utility Functions For</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#loading-data-from-file">Loading data from file</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_seq.html">load_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_aligned_seqs.html">load_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_unaligned_seqs.html">load_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_delimited.html">load_delimited</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_table.html">load_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_tree.html">load_tree</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.load_annotations.html">load_annotations</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.open_data_store.html">open_data_store</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.open_.html">open</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#making-cogent3-types-from-standard-python-types">Making cogent3 types from standard Python types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_seq.html">make_seq</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_aligned_seqs.html">make_aligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_unaligned_seqs.html">make_unaligned_seqs</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_table.html">make_table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.make_tree.html">make_tree</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#getting-commonly-used-cogent3-types">Getting commonly used cogent3 types</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_code.html">get_code</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_moltype.html">get_moltype</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.get_model.html">get_model</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../index.html#displaying-cogent3-builtins">Displaying cogent3 builtins</a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_codes.html">available_codes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_moltypes.html">available_moltypes</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_models.html">available_models</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../__init__/cogent3.__init__.available_apps.html">available_apps</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3 current"><a class="reference internal" href="../../index.html#the-major-cogent3-data-types">The Major cogent3 Data Types</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../../alignment/alignment.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alignment</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.SequenceCollection.html">SequenceCollection</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.Alignment.html">Alignment</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../alignment/classes/cogent3.core.alignment.ArrayAlignment.html">ArrayAlignment</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../annotation/annotation.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html">Feature</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../annotation_db/annotation_db.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">annotation_db</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.BasicAnnotationDb.html">BasicAnnotationDb</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GffAnnotationDb.html">GffAnnotationDb</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../annotation_db/classes/cogent3.core.annotation_db.GenbankAnnotationDb.html">GenbankAnnotationDb</a></li>
</ul>
</li>
<li class="toctree-l4 current"><a class="reference internal" href="../sequence.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">sequence</span></code></a><ul class="current">
<li class="toctree-l5"><a class="reference internal" href="cogent3.core.sequence.DnaSequence.html">DnaSequence</a></li>
<li class="toctree-l5"><a class="reference internal" href="cogent3.core.sequence.RnaSequence.html">RnaSequence</a></li>
<li class="toctree-l5 current"><a class="current reference internal" href="#">ProteinSequence</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../genetic_code/genetic_code.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">genetic_code</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../genetic_code/classes/cogent3.core.genetic_code.GeneticCode.html">GeneticCode</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../moltype/moltype.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">moltype</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../moltype/classes/cogent3.core.moltype.MolType.html">MolType</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../alphabet/alphabet.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">alphabet</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../alphabet/classes/cogent3.core.alphabet.Alphabet.html">Alphabet</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../table/table.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">table</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.util.table.Table.html">Table</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.util.table.Columns.html">Columns</a></li>
<li class="toctree-l5"><a class="reference internal" href="../../table/classes/cogent3.parse.table.FilteringParser.html">FilteringParser</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../../tree/tree.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">tree</span></code></a><ul>
<li class="toctree-l5"><a class="reference internal" href="../../tree/classes/cogent3.core.tree.PhyloNode.html">PhyloNode</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../index.html#composable-apps">Composable Apps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../app/composable/define_app.html">define_app</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../index.html#datastore">Datastore</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../app/io/classes/cogent3.app.io.register_datastore_reader.html">register_datastore_reader</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataMember.html">DataMember</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.DataStoreDirectory.html">DataStoreDirectory</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../app/data_store/classes/cogent3.app.data_store.ReadOnlyDataStoreZipped.html">ReadOnlyDataStoreZipped</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../app/sqlite_data_store/classes/cogent3.app.sqlite_data_store.DataStoreSqlite.html">DataStoreSqlite</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../index.html#deserialising">Deserialising</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../util/deserialise/deserialise_object.html">deserialise_object</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../util/deserialise/classes/cogent3.util.deserialise.register_deserialiser.html">register_deserialiser</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../draw/index.html">Image Gallery</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../draw/index.html#alignments-sequences">Alignments &amp; Sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_seq-features.html">Sequence Features</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_seq-features.html#drawing-all-features-on-a-sequence-segment">Drawing all features on a sequence segment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_seq-features.html#drawing-selected-feature-biotypes">Drawing selected feature biotypes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html">Sequence logos</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#drawing-logo-for-a-tfbs">Drawing logo for a TFBS</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#drawing-a-sequence-logo-from-a-multiple-sequence-alignment">Drawing a sequence logo from a multiple sequence alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-seqlogo.html#sequence-logo-of-protein-alignment">Sequence logo of protein alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html">Information analysis of an alignment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html#illustrated-with-a-simple-example">Illustrated with a simple example</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-info-plot.html#on-a-sample-data-set">On a sample data set</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html">Counting gaps per sequence</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#plotting-counts-of-unique-gaps">Plotting counts of unique gaps</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-bar-chart">Displaying unique gaps as a bar chart</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-violin-plot">Displaying unique gaps as a violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-gaps-per-seq.html#displaying-unique-gaps-as-a-box-plot">Displaying unique gaps as a box plot</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html">Dotplot with annotated sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html#reloading-from-json">Reloading from json</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-2.html#removing-annotation-tracks">Removing annotation tracks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html">Dotplot basics</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#modifying-the-matching-parameters">Modifying the matching parameters</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#displaying-dotplot-for-the-reverse-complement">Displaying dotplot for the reverse complement</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-dotplot-1.html#setting-plot-attributes">Setting plot attributes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html">Coevolution analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-as-a-heatmap">Display coevolution as a heatmap</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-violin-plot">Display coevolution scores as a Violin plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/aln/plot_aln-coevolution.html#display-coevolution-scores-as-a-boxplot">Display coevolution scores as a Boxplot</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../draw/index.html#phylogenetic-trees">Phylogenetic Trees</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-support.html">Showing Bootstrap Support</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-support.html#change-the-placement-of-support-text">Change the placement of support text</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html">Square Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-square.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-radial.html">Radial Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-radial.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html">Circular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html#colouring-a-set-of-edges">Colouring a set of edges</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-circular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../draw/tree/plot_tree-angular.html">Angular Dendrogram Style</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../draw/tree/plot_tree-angular.html#with-contemporaneous-tips">With Contemporaneous Tips</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html">Posting Bugs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html#citation">Citation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../general.html#support">Support</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../history.html">Project History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../../history.html#cogent3-no-longer-includes-module-x-what-do-i-do"><code class="docutils literal notranslate"><span class="pre">cogent3</span></code> no longer includes module <code class="docutils literal notranslate"><span class="pre">x</span></code>, what do I do?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../projects.html">Projects using <code class="docutils literal notranslate"><span class="pre">cogent3</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../licenses.html">License</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../licenses.html#cogent3-license">Cogent3 License</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../../genindex.html">Index</a></li>
</ul>

        
        </div>
        
    </div>
</nav>

<section id="main-section" class="container flex-grow-1 has-nav">
    
  <section id="proteinsequence">
<h1>ProteinSequence<a class="headerlink" href="#proteinsequence" title="Permalink to this headline">¶</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ProteinSequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_case</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gaps_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wildcards_allowed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Holds the standard Protein sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_db</strong></dt><dd></dd>
<dt><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.annotation_offset" title="cogent3.core.sequence.ProteinSequence.annotation_offset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotation_offset</span></code></a></dt><dd><p>The offset between annotation coordinates and sequence coordinates.</p>
</dd>
<dt><strong>line_wrap</strong></dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.add_feature" title="cogent3.core.sequence.ProteinSequence.add_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_feature</span></code></a>(*, biotype, name, spans[, ...])</p></td>
<td><p>add a feature to annotation_db</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.annotate_from_gff" title="cogent3.core.sequence.ProteinSequence.annotate_from_gff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_from_gff</span></code></a>(f[, offset])</p></td>
<td><p>copies annotations from a gff file to self,</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.annotate_matches_to" title="cogent3.core.sequence.ProteinSequence.annotate_matches_to"><code class="xref py py-obj docutils literal notranslate"><span class="pre">annotate_matches_to</span></code></a>(pattern, biotype, name)</p></td>
<td><p>Adds an annotation at sequence positions matching pattern.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_match" title="cogent3.core.sequence.ProteinSequence.can_match"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_match</span></code></a>(other)</p></td>
<td><p>Returns True if every pos in self could match same pos in other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_mismatch" title="cogent3.core.sequence.ProteinSequence.can_mismatch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_mismatch</span></code></a>(other)</p></td>
<td><p>Returns True if any position in self could mismatch with other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_mispair" title="cogent3.core.sequence.ProteinSequence.can_mispair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_mispair</span></code></a>(other)</p></td>
<td><p>Returns True if any position in self could mispair with other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.can_pair" title="cogent3.core.sequence.ProteinSequence.can_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">can_pair</span></code></a>(other)</p></td>
<td><p>Returns True if self and other could pair.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.complement" title="cogent3.core.sequence.ProteinSequence.complement"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complement</span></code></a>()</p></td>
<td><p>Returns complement of self, using data from MolType.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.copy" title="cogent3.core.sequence.ProteinSequence.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>([exclude_annotations])</p></td>
<td><p>returns a copy of self</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.copy_annotations" title="cogent3.core.sequence.ProteinSequence.copy_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_annotations</span></code></a>(seq_db)</p></td>
<td><p>copy annotations into attached annotation db</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.count" title="cogent3.core.sequence.ProteinSequence.count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count</span></code></a>(item)</p></td>
<td><p>count() delegates to self._seq.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.count_degenerate" title="cogent3.core.sequence.ProteinSequence.count_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_degenerate</span></code></a>()</p></td>
<td><p>Counts the degenerate bases in the specified sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.count_gaps" title="cogent3.core.sequence.ProteinSequence.count_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">count_gaps</span></code></a>()</p></td>
<td><p>Counts the gaps in the specified sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.counts" title="cogent3.core.sequence.ProteinSequence.counts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">counts</span></code></a>([motif_length, include_ambiguity, ...])</p></td>
<td><p>returns dict of counts of motifs</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.degap" title="cogent3.core.sequence.ProteinSequence.degap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">degap</span></code></a>()</p></td>
<td><p>Deletes all gap characters from sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.diff" title="cogent3.core.sequence.ProteinSequence.diff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">diff</span></code></a>(other)</p></td>
<td><p>Returns number of differences between self and other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.disambiguate" title="cogent3.core.sequence.ProteinSequence.disambiguate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">disambiguate</span></code></a>([method])</p></td>
<td><p>Returns a non-degenerate sequence from a degenerate one.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.distance" title="cogent3.core.sequence.ProteinSequence.distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code></a>(other[, function])</p></td>
<td><p>Returns distance between self and other using function(i,j).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_degenerate" title="cogent3.core.sequence.ProteinSequence.first_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_degenerate</span></code></a>()</p></td>
<td><p>Returns the index of first degenerate symbol in sequence, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_gap" title="cogent3.core.sequence.ProteinSequence.first_gap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_gap</span></code></a>()</p></td>
<td><p>Returns the index of the first gap in the sequence, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_invalid" title="cogent3.core.sequence.ProteinSequence.first_invalid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_invalid</span></code></a>()</p></td>
<td><p>Returns the index of first invalid symbol in sequence, or None.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.first_non_strict" title="cogent3.core.sequence.ProteinSequence.first_non_strict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">first_non_strict</span></code></a>()</p></td>
<td><p>Returns the index of first non-strict symbol in sequence, or None.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_diff" title="cogent3.core.sequence.ProteinSequence.frac_diff"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other differ.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_diff_gaps" title="cogent3.core.sequence.ProteinSequence.frac_diff_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff_gaps</span></code></a>(other)</p></td>
<td><p>Returns frac.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps" title="cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_diff_non_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of non-gap positions where self differs from other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_same" title="cogent3.core.sequence.ProteinSequence.frac_same"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other are the same.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_same_gaps" title="cogent3.core.sequence.ProteinSequence.frac_same_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of positions where self and other share gap states.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_same_non_gaps" title="cogent3.core.sequence.ProteinSequence.frac_same_non_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_same_non_gaps</span></code></a>(other)</p></td>
<td><p>Returns fraction of non-gap positions where self matches other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.frac_similar" title="cogent3.core.sequence.ProteinSequence.frac_similar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">frac_similar</span></code></a>(other, similar_pairs)</p></td>
<td><p>Returns fraction of positions where self[i] is similar to other[i].</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gap_indices" title="cogent3.core.sequence.ProteinSequence.gap_indices"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_indices</span></code></a>()</p></td>
<td><p>Returns list of indices of all gaps in the sequence, or [].</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gap_maps" title="cogent3.core.sequence.ProteinSequence.gap_maps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_maps</span></code></a>()</p></td>
<td><p>Returns dicts mapping between gapped and ungapped positions.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gap_vector" title="cogent3.core.sequence.ProteinSequence.gap_vector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gap_vector</span></code></a>()</p></td>
<td><p>Returns vector of True or False according to which pos are gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_drawable" title="cogent3.core.sequence.ProteinSequence.get_drawable"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawable</span></code></a>(*[, biotype, width, vertical])</p></td>
<td><p>make a figure from sequence features</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_drawables" title="cogent3.core.sequence.ProteinSequence.get_drawables"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_drawables</span></code></a>(*[, biotype])</p></td>
<td><p>returns a dict of drawables, keyed by type</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_features" title="cogent3.core.sequence.ProteinSequence.get_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_features</span></code></a>(*[, biotype, name, start, ...])</p></td>
<td><p>yields Feature instances</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_features_matching" title="cogent3.core.sequence.ProteinSequence.get_features_matching"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_features_matching</span></code></a>(**kwargs)</p></td>
<td><p>use .get_features()</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_in_motif_size" title="cogent3.core.sequence.ProteinSequence.get_in_motif_size"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_in_motif_size</span></code></a>([motif_length, log_warnings])</p></td>
<td><p>returns sequence as list of non-overlapping motifs</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_kmers" title="cogent3.core.sequence.ProteinSequence.get_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kmers</span></code></a>(k[, strict])</p></td>
<td><p>return all overlapping k-mers</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_name" title="cogent3.core.sequence.ProteinSequence.get_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_name</span></code></a>()</p></td>
<td><p>Return the sequence name -- should just use name instead.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.get_type" title="cogent3.core.sequence.ProteinSequence.get_type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_type</span></code></a>()</p></td>
<td><p>Return the sequence type as moltype label.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.gettype" title="cogent3.core.sequence.ProteinSequence.gettype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gettype</span></code></a>()</p></td>
<td><p>Return the sequence type.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_annotated" title="cogent3.core.sequence.ProteinSequence.is_annotated"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_annotated</span></code></a>()</p></td>
<td><p>returns True if sequence has any annotations</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_degenerate" title="cogent3.core.sequence.ProteinSequence.is_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_degenerate</span></code></a>()</p></td>
<td><p>Returns True if sequence contains degenerate characters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_gap" title="cogent3.core.sequence.ProteinSequence.is_gap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_gap</span></code></a>([char])</p></td>
<td><p>Returns True if char is a gap.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_gapped" title="cogent3.core.sequence.ProteinSequence.is_gapped"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_gapped</span></code></a>()</p></td>
<td><p>Returns True if sequence contains gaps.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_strict" title="cogent3.core.sequence.ProteinSequence.is_strict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_strict</span></code></a>()</p></td>
<td><p>Returns True if sequence contains only monomers.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.is_valid" title="cogent3.core.sequence.ProteinSequence.is_valid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_valid</span></code></a>()</p></td>
<td><p>Returns True if sequence contains no items absent from alphabet.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.iter_kmers" title="cogent3.core.sequence.ProteinSequence.iter_kmers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">iter_kmers</span></code></a>(k[, strict])</p></td>
<td><p>generates all overlapping k-mers.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.make_feature" title="cogent3.core.sequence.ProteinSequence.make_feature"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_feature</span></code></a>(feature, *args)</p></td>
<td><p>return an Feature instance from feature data</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.matrix_distance" title="cogent3.core.sequence.ProteinSequence.matrix_distance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">matrix_distance</span></code></a>(other, matrix)</p></td>
<td><p>Returns distance between self and other using a score matrix.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.must_match" title="cogent3.core.sequence.ProteinSequence.must_match"><code class="xref py py-obj docutils literal notranslate"><span class="pre">must_match</span></code></a>(other)</p></td>
<td><p>Returns True if all positions in self must match positions in other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.must_pair" title="cogent3.core.sequence.ProteinSequence.must_pair"><code class="xref py py-obj docutils literal notranslate"><span class="pre">must_pair</span></code></a>(other)</p></td>
<td><p>Returns True if all positions in self must pair with other.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.mw" title="cogent3.core.sequence.ProteinSequence.mw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mw</span></code></a>([method, delta])</p></td>
<td><p>Returns the molecular weight of (one strand of) the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.parse_out_gaps" title="cogent3.core.sequence.ProteinSequence.parse_out_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_out_gaps</span></code></a>()</p></td>
<td><p>returns Map corresponding to gap locations and ungapped Sequence</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.possibilities" title="cogent3.core.sequence.ProteinSequence.possibilities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">possibilities</span></code></a>()</p></td>
<td><p>Counts number of possible sequences matching the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.rc" title="cogent3.core.sequence.ProteinSequence.rc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rc</span></code></a>()</p></td>
<td><p>Returns reverse complement of self w/ data from MolType.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.replace" title="cogent3.core.sequence.ProteinSequence.replace"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code></a>(oldchar, newchar)</p></td>
<td><p>return new instance with oldchar replaced by newchar</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.replace_annotation_db" title="cogent3.core.sequence.ProteinSequence.replace_annotation_db"><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_annotation_db</span></code></a>(value[, check])</p></td>
<td><p>public interface to assigning the annotation_db</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.resolveambiguities" title="cogent3.core.sequence.ProteinSequence.resolveambiguities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">resolveambiguities</span></code></a>()</p></td>
<td><p>Returns a list of tuples of strings.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.resolved_ambiguities" title="cogent3.core.sequence.ProteinSequence.resolved_ambiguities"><code class="xref py py-obj docutils literal notranslate"><span class="pre">resolved_ambiguities</span></code></a>()</p></td>
<td><p>Returns a list of tuples of strings.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.shuffle" title="cogent3.core.sequence.ProteinSequence.shuffle"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shuffle</span></code></a>()</p></td>
<td><p>returns a randomized copy of the Sequence object</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.sliding_windows" title="cogent3.core.sequence.ProteinSequence.sliding_windows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sliding_windows</span></code></a>(window, step[, start, end])</p></td>
<td><p>Generator function that yield new sequence objects of a given length at a given interval.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.strip_bad" title="cogent3.core.sequence.ProteinSequence.strip_bad"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_bad</span></code></a>()</p></td>
<td><p>Removes any symbols not in the alphabet.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps" title="cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_bad_and_gaps</span></code></a>()</p></td>
<td><p>Removes any symbols not in the alphabet, and any gaps.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.strip_degenerate" title="cogent3.core.sequence.ProteinSequence.strip_degenerate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip_degenerate</span></code></a>()</p></td>
<td><p>Removes degenerate bases by stripping them out of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_fasta" title="cogent3.core.sequence.ProteinSequence.to_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_fasta</span></code></a>([make_seqlabel, block_size])</p></td>
<td><p>Return string of self in FASTA format, no trailing newline</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_html" title="cogent3.core.sequence.ProteinSequence.to_html"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_html</span></code></a>([wrap, limit, colors, font_size, ...])</p></td>
<td><p>returns html with embedded styles for sequence colouring</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_json" title="cogent3.core.sequence.ProteinSequence.to_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_json</span></code></a>()</p></td>
<td><p>returns a json formatted string</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_moltype" title="cogent3.core.sequence.ProteinSequence.to_moltype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_moltype</span></code></a>(moltype)</p></td>
<td><p>returns copy of self with moltype seq</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.to_rich_dict" title="cogent3.core.sequence.ProteinSequence.to_rich_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_rich_dict</span></code></a>([exclude_annotations])</p></td>
<td><p>returns {'name': name, 'seq': sequence, 'moltype': moltype.label}</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.translate" title="cogent3.core.sequence.ProteinSequence.translate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate</span></code></a>(*args, **kwargs)</p></td>
<td><p>returns the result of call str.translate</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.with_masked_annotations" title="cogent3.core.sequence.ProteinSequence.with_masked_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_masked_annotations</span></code></a>(annot_types[, ...])</p></td>
<td><p>returns a sequence with annot_types regions replaced by mask_char if shadow is False, otherwise all other regions are masked.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#cogent3.core.sequence.ProteinSequence.with_termini_unknown" title="cogent3.core.sequence.ProteinSequence.with_termini_unknown"><code class="xref py py-obj docutils literal notranslate"><span class="pre">with_termini_unknown</span></code></a>()</p></td>
<td><p>Returns copy of sequence with terminal gaps remapped as missing.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 75%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>gapped_by_map</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>gapped_by_map_motif_iter</strong></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p><strong>gapped_by_map_segment_iter</strong></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p><strong>strand_symmetry</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.add_feature">
<span class="sig-name descname"><span class="pre">add_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spans</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parent_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">on_alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seqid</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">cogent3.core.annotation.Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.add_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>add a feature to annotation_db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>biological type</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>spans</strong></dt><dd><p>coordinates for this sequence</p>
</dd>
<dt><strong>parent_id</strong></dt><dd><p>name of the feature parent</p>
</dd>
<dt><strong>strand</strong></dt><dd><p>‘+’ or ‘-’, defaults to ‘+’</p>
</dd>
<dt><strong>on_alignment</strong></dt><dd><p>whether the feature spans are alignment coordinates</p>
</dd>
<dt><strong>seqid</strong></dt><dd><p>ignored since the feature is added to this sequence</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Feature instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotate_from_gff">
<span class="sig-name descname"><span class="pre">annotate_from_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">os.PathLike</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotate_from_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>copies annotations from a gff file to self,</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>f</strong><span class="classifier">path to gff annotation file.</span></dt><dd></dd>
<dt><strong>offset</strong><span class="classifier">Optional, the offset between annotation coordinates and sequence coordinates.</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotate_matches_to">
<span class="sig-name descname"><span class="pre">annotate_matches_to</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_multiple</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotate_matches_to" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds an annotation at sequence positions matching pattern.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>pattern</strong></dt><dd><p>The search string for which annotations are made. IUPAC ambiguities
are converted to regex on sequences with the appropriate MolType.</p>
</dd>
<dt><strong>biotype</strong></dt><dd><p>The type of the annotation (e.g. “domain”).</p>
</dd>
<dt><strong>name</strong></dt><dd><p>The name of the annotation.</p>
</dd>
<dt><strong>allow_multiple</strong></dt><dd><p>If True, allows multiple occurrences of the input pattern. Otherwise,
only the first match is used.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Returns a list of Feature instances.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotation_db">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation_db</span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotation_db" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.annotation_offset">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotation_offset</span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.annotation_offset" title="Permalink to this definition">¶</a></dt>
<dd><p>The offset between annotation coordinates and sequence coordinates.</p>
<p>The offset can be used to adjust annotation coordinates to match the position
of the given Sequence within a larger genomic context. For example, if the
Annotations are with respect to a chromosome, and the sequence represents
a gene that is 100 bp from the start of a chromosome, the offset can be set to
100 to ensure that the gene’s annotations are aligned with the appropriate
genomic positions.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>int: The offset between annotation coordinates and sequence coordinates.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_match">
<span class="sig-name descname"><span class="pre">can_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if every pos in self could match same pos in other.</p>
<p>Truncates at length of shorter sequence.
gaps are only allowed to match other gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_mismatch">
<span class="sig-name descname"><span class="pre">can_mismatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_mismatch" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if any position in self could mismatch with other.</p>
<p>Truncates at length of shorter sequence.
gaps are always counted as matches.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_mispair">
<span class="sig-name descname"><span class="pre">can_mispair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_mispair" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if any position in self could mispair with other.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
<p>Truncates at length of shorter sequence.
gaps are always counted as possible mispairs, as are weak pairs like GU.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.can_pair">
<span class="sig-name descname"><span class="pre">can_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.can_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if self and other could pair.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
<p>Truncates at length of shorter sequence.
gaps are only allowed to pair with other gaps, and are counted as ‘weak’
(same category as GU and degenerate pairs).</p>
<p>NOTE: second must be able to be reverse</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.complement">
<span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.complement" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns complement of self, using data from MolType.</p>
<p>Always tries to return same type as item: if item looks like a dict,
will return list of keys.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exclude_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a copy of self</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.copy_annotations">
<span class="sig-name descname"><span class="pre">copy_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cogent3.core.annotation_db.SupportsFeatures</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.copy_annotations" title="Permalink to this definition">¶</a></dt>
<dd><p>copy annotations into attached annotation db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_db</strong></dt><dd><p>compatible annotation db</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Only copies annotations for records with seqid equal to self.name</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.count">
<span class="sig-name descname"><span class="pre">count</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.count" title="Permalink to this definition">¶</a></dt>
<dd><p>count() delegates to self._seq.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.count_degenerate">
<span class="sig-name descname"><span class="pre">count_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.count_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts the degenerate bases in the specified sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.count_gaps">
<span class="sig-name descname"><span class="pre">count_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.count_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts the gaps in the specified sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.counts">
<span class="sig-name descname"><span class="pre">counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ambiguity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unobserved</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.counts" title="Permalink to this definition">¶</a></dt>
<dd><p>returns dict of counts of motifs</p>
<p>only non-overlapping motifs are counted.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>number of elements per character.</p>
</dd>
<dt><strong>include_ambiguity</strong></dt><dd><p>if True, motifs containing ambiguous characters
from the seq moltype are included. No expansion of those is attempted.</p>
</dd>
<dt><strong>allow_gaps</strong></dt><dd><p>if True, motifs containing a gap character are included.</p>
</dd>
<dt><strong>exclude_unobserved</strong></dt><dd><p>if True, unobserved motif combinations are excluded.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.degap">
<span class="sig-name descname"><span class="pre">degap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.degap" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes all gap characters from sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.diff">
<span class="sig-name descname"><span class="pre">diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of differences between self and other.</p>
<p>NOTE: truncates at the length of the shorter sequence. Case-sensitive.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.disambiguate">
<span class="sig-name descname"><span class="pre">disambiguate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strip'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.disambiguate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a non-degenerate sequence from a degenerate one.</p>
<p>method can be ‘strip’ (deletes any characters not in monomers or gaps)
or ‘random’(assigns the possibilities at random, using equal
frequencies).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.distance">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">function</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns distance between self and other using function(i,j).</p>
<p>other must be a sequence.</p>
<p>function should be a function that takes two items and returns a
number. To turn a 2D matrix into a function, use
cogent3.util.miscs.DistanceFromMatrix(matrix).</p>
<p>NOTE: Truncates at the length of the shorter sequence.</p>
<p>Note that the function acts on two _elements_ of the sequences, not
the two sequences themselves (i.e. the behavior will be the same for
every position in the sequences, such as identity scoring or a function
derived from a distance matrix as suggested above). One limitation of
this approach is that the distance function cannot use properties of
the sequences themselves: for example, it cannot use the lengths of the
sequences to normalize the scores as percent similarities or percent
differences.</p>
<p>If you want functions that act on the two sequences themselves, there
is no particular advantage in making these functions methods of the
first sequences by passing them in as parameters like the function
in this method. It makes more sense to use them as standalone functions.
The factory function cogent3.util.transform.for_seq is useful for
converting per-element functions into per-sequence functions, since it
takes as parameters a per-element scoring function, a score aggregation
function, and a normalization function (which itself takes the two
sequences as parameters), returning a single function that combines
these functions and that acts on two complete sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_degenerate">
<span class="sig-name descname"><span class="pre">first_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of first degenerate symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_gap">
<span class="sig-name descname"><span class="pre">first_gap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_gap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of the first gap in the sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_invalid">
<span class="sig-name descname"><span class="pre">first_invalid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_invalid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of first invalid symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.first_non_strict">
<span class="sig-name descname"><span class="pre">first_non_strict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.first_non_strict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the index of first non-strict symbol in sequence, or None.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_diff">
<span class="sig-name descname"><span class="pre">frac_diff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_diff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self and other differ.</p>
<p>Truncates at length of shorter sequence.
Note that frac_same and frac_diff are both 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_diff_gaps">
<span class="sig-name descname"><span class="pre">frac_diff_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_diff_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns frac. of positions where self and other’s gap states differ.</p>
<p>In other words, if self and other are both all gaps, or both all
non-gaps, or both have gaps in the same places, frac_diff_gaps will
return 0.0. If self is all gaps and other has no gaps, frac_diff_gaps
will return 1.0.</p>
<p>Returns 0 if one sequence is empty.</p>
<p>Uses self’s gap characters for both sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps">
<span class="sig-name descname"><span class="pre">frac_diff_non_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_diff_non_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of non-gap positions where self differs from other.</p>
<p>Doesn’t count any position where self or other has a gap.
Truncates at the length of the shorter sequence.</p>
<p>Returns 0 if one sequence is empty. Note that this means that
frac_diff_non_gaps is _not_ the same as 1 - frac_same_non_gaps, since both
return 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_same">
<span class="sig-name descname"><span class="pre">frac_same</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_same" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self and other are the same.</p>
<p>Truncates at length of shorter sequence.
Note that frac_same and frac_diff are both 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_same_gaps">
<span class="sig-name descname"><span class="pre">frac_same_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_same_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self and other share gap states.</p>
<p>In other words, if self and other are both all gaps, or both all
non-gaps, or both have gaps in the same places, frac_same_gaps will
return 1.0. If self is all gaps and other has no gaps, frac_same_gaps
will return 0.0. Returns 0 if one sequence is empty.</p>
<p>Uses self’s gap characters for both sequences.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_same_non_gaps">
<span class="sig-name descname"><span class="pre">frac_same_non_gaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_same_non_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of non-gap positions where self matches other.</p>
<p>Doesn’t count any position where self or other has a gap.
Truncates at the length of the shorter sequence.</p>
<p>Returns 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.frac_similar">
<span class="sig-name descname"><span class="pre">frac_similar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">similar_pairs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.frac_similar" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns fraction of positions where self[i] is similar to other[i].</p>
<p>similar_pairs must be a dict such that d[(i,j)] exists if i and j are
to be counted as similar. Use PairsFromGroups in cogent3.util.misc to
construct such a dict from a list of lists of similar residues.</p>
<p>Truncates at the length of the shorter sequence.</p>
<p>Note: current implementation re-creates the distance function each
time, so may be expensive compared to creating the distance function
using for_seq separately.</p>
<p>Returns 0 if one sequence is empty.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gap_indices">
<span class="sig-name descname"><span class="pre">gap_indices</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gap_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of indices of all gaps in the sequence, or [].</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gap_maps">
<span class="sig-name descname"><span class="pre">gap_maps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gap_maps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dicts mapping between gapped and ungapped positions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gap_vector">
<span class="sig-name descname"><span class="pre">gap_vector</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gap_vector" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns vector of True or False according to which pos are gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gapped_by_map">
<span class="sig-name descname"><span class="pre">gapped_by_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gapped_by_map" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gapped_by_map_motif_iter">
<span class="sig-name descname"><span class="pre">gapped_by_map_motif_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gapped_by_map_motif_iter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gapped_by_map_segment_iter">
<span class="sig-name descname"><span class="pre">gapped_by_map_segment_iter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recode_gaps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gapped_by_map_segment_iter" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_drawable">
<span class="sig-name descname"><span class="pre">get_drawable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">600</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_drawable" title="Permalink to this definition">¶</a></dt>
<dd><p>make a figure from sequence features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>passed to get_features(biotype). Can be a single biotype or
series. Only features matching this will be included.</p>
</dd>
<dt><strong>width</strong></dt><dd><p>width in pixels</p>
</dd>
<dt><strong>vertical</strong></dt><dd><p>rotates the drawable</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>a Drawable instance</dt><dd></dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>If provided, the biotype is used for plot order.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_drawables">
<span class="sig-name descname"><span class="pre">get_drawables</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Iterable</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_drawables" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a dict of drawables, keyed by type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>passed to get_features(biotype). Can be a single biotype or
series. Only features matching this will be included.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_features">
<span class="sig-name descname"><span class="pre">get_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_partial</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_features" title="Permalink to this definition">¶</a></dt>
<dd><p>yields Feature instances</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>biotype</strong></dt><dd><p>biotype of the feature</p>
</dd>
<dt><strong>name</strong></dt><dd><p>name of the feature</p>
</dd>
<dt><strong>start, stop</strong></dt><dd><p>start, end positions to search between, relative to offset
of this sequence. If not provided, entire span of sequence is used.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>When dealing with a nucleic acid moltype, the returned features will
yield a sequence segment that is consistently oriented irrespective
of strand of the current instance.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_features_matching">
<span class="sig-name descname"><span class="pre">get_features_matching</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_features_matching" title="Permalink to this definition">¶</a></dt>
<dd><p>use .get_features()</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_in_motif_size">
<span class="sig-name descname"><span class="pre">get_in_motif_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_warnings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_in_motif_size" title="Permalink to this definition">¶</a></dt>
<dd><p>returns sequence as list of non-overlapping motifs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>motif_length</strong></dt><dd><p>length of the motifs</p>
</dd>
<dt><strong>log_warnings</strong></dt><dd><p>whether to notify of an incomplete terminal motif</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_kmers">
<span class="sig-name descname"><span class="pre">get_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>return all overlapping k-mers</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_name">
<span class="sig-name descname"><span class="pre">get_name</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sequence name – should just use name instead.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.get_type">
<span class="sig-name descname"><span class="pre">get_type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.get_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sequence type as moltype label.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.gettype">
<span class="sig-name descname"><span class="pre">gettype</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.gettype" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sequence type.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_annotated">
<span class="sig-name descname"><span class="pre">is_annotated</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_annotated" title="Permalink to this definition">¶</a></dt>
<dd><p>returns True if sequence has any annotations</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_degenerate">
<span class="sig-name descname"><span class="pre">is_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains degenerate characters.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_gap">
<span class="sig-name descname"><span class="pre">is_gap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_gap" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if char is a gap.</p>
<p>If char is not supplied, tests whether self is gaps only.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_gapped">
<span class="sig-name descname"><span class="pre">is_gapped</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_gapped" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_strict">
<span class="sig-name descname"><span class="pre">is_strict</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_strict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains only monomers.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.is_valid">
<span class="sig-name descname"><span class="pre">is_valid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.is_valid" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if sequence contains no items absent from alphabet.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.iter_kmers">
<span class="sig-name descname"><span class="pre">iter_kmers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Generator</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.iter_kmers" title="Permalink to this definition">¶</a></dt>
<dd><p>generates all overlapping k-mers.
When strict is True, the characters in the k-mer must be
a subset of the canonical characters for the moltype</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.line_wrap">
<span class="sig-name descname"><span class="pre">line_wrap</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">None</span></em><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.line_wrap" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.make_feature">
<span class="sig-name descname"><span class="pre">make_feature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feature</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cogent3.core.annotation_db.FeatureDataType</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="../../annotation/classes/cogent3.core.annotation.Feature.html#cogent3.core.annotation.Feature" title="cogent3.core.annotation.Feature"><span class="pre">cogent3.core.annotation.Feature</span></a></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.make_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>return an Feature instance from feature data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>feature</strong></dt><dd><p>dict of key data to make an Feature instance</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Unlike add_feature(), this method does not add the feature to the
database.
We assume that spans represent the coordinates for this instance!</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.matrix_distance">
<span class="sig-name descname"><span class="pre">matrix_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.matrix_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns distance between self and other using a score matrix.</p>
<p>WARNING: the matrix must explicitly contain scores for the case where
a position is the same in self and other (e.g. for a distance matrix,
an identity between U and U might have a score of 0). The reason the
scores for the ‘diagonals’ need to be passed explicitly is that for
some kinds of distance matrices, e.g. log-odds matrices, the ‘diagonal’
scores differ from each other. If these elements are missing, this
function will raise a KeyError at the first position that the two
sequences are identical.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.moltype">
<span class="sig-name descname"><span class="pre">moltype</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">MolType(('A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'))</span></em><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.moltype" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.must_match">
<span class="sig-name descname"><span class="pre">must_match</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.must_match" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if all positions in self must match positions in other.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.must_pair">
<span class="sig-name descname"><span class="pre">must_pair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.must_pair" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if all positions in self must pair with other.</p>
<p>Pairing occurs in reverse order, i.e. last position of other with
first position of self, etc.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.mw">
<span class="sig-name descname"><span class="pre">mw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'random'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.mw" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the molecular weight of (one strand of) the sequence.</p>
<p>If the sequence is ambiguous, uses method (random or strip) to
disambiguate the sequence.</p>
<p>If delta is passed in, adds delta per strand (default is None, which
uses the alphabet default. Typically, this adds 18 Da for terminal
water. However, note that the default nucleic acid weight assumes
5’ monophosphate and 3’ OH: pass in delta=18.0 if you want 5’ OH as
well.</p>
<p>Note that this method only calculates the MW of the coding strand. If
you want the MW of the reverse strand, add self.rc().mw(). DO NOT
just multiply the MW by 2: the results may not be accurate due to
strand bias, e.g. in mitochondrial genomes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.parse_out_gaps">
<span class="sig-name descname"><span class="pre">parse_out_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.parse_out_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>returns Map corresponding to gap locations and ungapped Sequence</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.possibilities">
<span class="sig-name descname"><span class="pre">possibilities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.possibilities" title="Permalink to this definition">¶</a></dt>
<dd><p>Counts number of possible sequences matching the sequence.</p>
<p>Uses self.degenerates to decide how many possibilities there are at
each position in the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.rc">
<span class="sig-name descname"><span class="pre">rc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.rc" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns reverse complement of self w/ data from MolType.</p>
<p>Always returns same type self.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.replace">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">oldchar</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">newchar</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.replace" title="Permalink to this definition">¶</a></dt>
<dd><p>return new instance with oldchar replaced by newchar</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.replace_annotation_db">
<span class="sig-name descname"><span class="pre">replace_annotation_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">cogent3.core.annotation_db.SupportsFeatures</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.replace_annotation_db" title="Permalink to this definition">¶</a></dt>
<dd><p>public interface to assigning the annotation_db</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>value</strong></dt><dd><p>the annotation db instance</p>
</dd>
<dt><strong>check</strong></dt><dd><p>whether to check value supports the feature interface</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The check can be very expensive, so if you’re confident set it to False</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.resolveambiguities">
<span class="sig-name descname"><span class="pre">resolveambiguities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.resolveambiguities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of tuples of strings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.resolved_ambiguities">
<span class="sig-name descname"><span class="pre">resolved_ambiguities</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.resolved_ambiguities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of tuples of strings.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.shuffle">
<span class="sig-name descname"><span class="pre">shuffle</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.shuffle" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a randomized copy of the Sequence object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.sliding_windows">
<span class="sig-name descname"><span class="pre">sliding_windows</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">window</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.sliding_windows" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator function that yield new sequence objects
of a given length at a given interval.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>window</strong></dt><dd><p>The length of the returned sequence</p>
</dd>
<dt><strong>step</strong></dt><dd><p>The interval between the start of the returned
sequence objects</p>
</dd>
<dt><strong>start</strong></dt><dd><p>first window start position</p>
</dd>
<dt><strong>end</strong></dt><dd><p>last window start position</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strand_symmetry">
<span class="sig-name descname"><span class="pre">strand_symmetry</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strand_symmetry" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strip_bad">
<span class="sig-name descname"><span class="pre">strip_bad</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strip_bad" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes any symbols not in the alphabet.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps">
<span class="sig-name descname"><span class="pre">strip_bad_and_gaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strip_bad_and_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes any symbols not in the alphabet, and any gaps.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.strip_degenerate">
<span class="sig-name descname"><span class="pre">strip_degenerate</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.strip_degenerate" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes degenerate bases by stripping them out of the sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_fasta">
<span class="sig-name descname"><span class="pre">to_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">make_seqlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Return string of self in FASTA format, no trailing newline</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>make_seqlabel</strong></dt><dd><p>callback function that takes the seq object and
returns a label str</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_html">
<span class="sig-name descname"><span class="pre">to_html</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">wrap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_family</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Lucida</span> <span class="pre">Console'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_html" title="Permalink to this definition">¶</a></dt>
<dd><p>returns html with embedded styles for sequence colouring</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>wrap</strong></dt><dd><p>maximum number of printed bases, defaults to
alignment length</p>
</dd>
<dt><strong>limit</strong></dt><dd><p>truncate alignment to this length</p>
</dd>
<dt><strong>colors</strong></dt><dd><p>{character
moltype.</p>
</dd>
<dt><strong>font_size</strong></dt><dd><p>in points. Affects labels and sequence and line spacing
(proportional to value)</p>
</dd>
<dt><strong>font_family</strong></dt><dd><p>string denoting font family</p>
</dd>
<dt><strong>To display in jupyter notebook:</strong></dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">IPython.core.display</span> <span class="kn">import</span> <span class="n">HTML</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">HTML</span><span class="p">(</span><span class="n">aln</span><span class="o">.</span><span class="n">to_html</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_json">
<span class="sig-name descname"><span class="pre">to_json</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a json formatted string</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_moltype">
<span class="sig-name descname"><span class="pre">to_moltype</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">moltype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_moltype" title="Permalink to this definition">¶</a></dt>
<dd><p>returns copy of self with moltype seq</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>moltype</strong><span class="classifier">str</span></dt><dd><p>molecular type</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.to_rich_dict">
<span class="sig-name descname"><span class="pre">to_rich_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">exclude_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.to_rich_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>returns {‘name’: name, ‘seq’: sequence, ‘moltype’: moltype.label}</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.translate" title="Permalink to this definition">¶</a></dt>
<dd><p>returns the result of call str.translate</p>
<p class="rubric">Notes</p>
<p>This is a string method, nothing to do with translating into a
protein sequence.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.with_masked_annotations">
<span class="sig-name descname"><span class="pre">with_masked_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annot_types</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_char</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shadow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend_query</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.with_masked_annotations" title="Permalink to this definition">¶</a></dt>
<dd><p>returns a sequence with annot_types regions replaced by mask_char
if shadow is False, otherwise all other regions are masked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annot_types</strong></dt><dd><p>annotation type(s)</p>
</dd>
<dt><strong>mask_char</strong></dt><dd><p>must be a character valid for the seq MolType. The
default value is the most ambiguous character, eg. ‘?’ for DNA</p>
</dd>
<dt><strong>shadow</strong></dt><dd><p>whether to mask the annotated regions, or everything but
the annotated regions</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="cogent3.core.sequence.ProteinSequence.with_termini_unknown">
<span class="sig-name descname"><span class="pre">with_termini_unknown</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#cogent3.core.sequence.ProteinSequence.with_termini_unknown" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns copy of sequence with terminal gaps remapped as missing.</p>
</dd></dl>

</dd></dl>

</section>


</section>

<footer class="footer w-100 pt-2 border-top bg-light">
    <div id="footer-nav" class="bg-light container">
        <div id="navbar-footer" class="row">
            
            <div class="col">
                <dl class="text-center mb-1">
                    
                    
                    
                </dl>
            </div>
            
        </div>
    </div>
    <div id="footer-info" class="bg-light pl-2 pb-1 w-100">
        <div><span class="fa fa-copyright"></span>2020-2023, cogent3 Team
        <a id="author-link" href=""></a> </div>
        <div>Powered by <a href="http://sphinx-doc.org/">Sphinx</a> 4.5.0 and the <a href="https://github.com/Bernardo-MG/sphinx-docs-theme">Sphinx Docs Theme</a></div>
    </div>
</footer>

</body>
</html>